{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/anaconda3/lib/python3.7/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'process_jaspar_without_strand' from '/home/louiscam/projects/gpcr/code/JASPAR_processing/process_jaspar_without_strand.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import itertools\n",
    "from itertools import groupby\n",
    "import os.path\n",
    "import math\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import MOODS.parsers\n",
    "import MOODS.tools\n",
    "import MOODS.scan\n",
    "import subprocess\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import process_jaspar_without_strand as pj\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(pj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "prom_hg19_seq_dir = dir_genome+'prom_hg19_seq_dir/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data_dir/'\n",
    "# Directory for storing preliminary results\n",
    "prelim_results_dir = '/home/louiscam/projects/gpcr/save/prelim_results_dir/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "# Directory of tf data\n",
    "tf_dir = '/home/louiscam/projects/gpcr/data/tf_data/'\n",
    "jaspar_dir = tf_dir+'jaspar_data/'\n",
    "pfm_dir = tf_dir+'pfm_data/'\n",
    "fantom5_dir = '/home/louiscam/projects/gpcr/data/fantom5/'\n",
    "moods_out_dir = tf_dir+'moods_cage_outdir_without_strand/'\n",
    "cage_fasta_dir = dir_genome+'cage_hg19_seq_dir_without_strand/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Select genes of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genes of interest (active adhesome genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes of interest = 219\n"
     ]
    }
   ],
   "source": [
    "# Genes of interest and location\n",
    "selected_genes = pickle.load(open(saving_dir+'active_adhesome_genes.pkl', 'rb'))\n",
    "print('Number of genes of interest = '+str(len(selected_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping UCSC to HGNC\n",
    "ucsc_to_hgnc = pj.create_ucsc_hgnc_dict(dir_genome+'chrom_hg19.name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load human TFs list\n",
    "from http://humantfs.ccbr.utoronto.ca/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of TFs = 1639\n"
     ]
    }
   ],
   "source": [
    "# Load TFs\n",
    "tf_df = pd.read_csv(dir_genome+'human_tf.database', sep='\\t', header=0, index_col=0)\n",
    "tf_df = tf_df[tf_df['Is TF?']=='Yes'].sort_values(by=['HGNC symbol'])\n",
    "all_tfs = tf_df['HGNC symbol'].values\n",
    "print('Total number of TFs = '+str(len(all_tfs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene location dataframe\n",
    "gene_id_filename = dir_genome+'chrom_hg19.name'\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc_canonical'\n",
    "df_loc = lu.get_gene_locations(gene_locations_filename, gene_id_filename)\n",
    "df_loc.columns = ['transcript','gene','chrom','gene_start','gene_end','clusterId','protein']\n",
    "df_loc['gene_length'] = df_loc['gene_end']-df_loc['gene_start']\n",
    "df_loc['gene_center'] = df_loc['gene_start']+df_loc['gene_length']//2\n",
    "# Restrict to selected genes\n",
    "df_loc = df_loc[df_loc['gene'].isin(selected_genes)]\n",
    "# Add TF information\n",
    "df_loc['isTF'] = [df_loc['gene'].iloc[i] in all_tfs for i in range(df_loc.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adhesome gene lengths quartile = [ 21507.   49385.  114977.5]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeoklEQVR4nO3deZgdVbnv8e+PBAIIh4BpkAChGSPzFBEEFAE9ERE8RxDiQQ2gUbwyKIpBvBfk4hXn4ShiVAhHEHBAiaAIKgjImEAIEOTIIQGCQBKGMCmQ8N4/1mopNr27K91de2d3/T7Ps5+uca23qnu/vfaq2qsUEZiZWX2s1O4AzMystZz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJv80kzZe0X5N1e0m6p9UxrUiUnCPpCUk3D0F50yWdnqf3lrRg8FFakaSrJX2oTXWHpM3bUXcnceJfgUXEtRExvr/tJJ0q6bxWxNQGewJvAzaMiF3bHYytONr5D6bTOfFbnySNbHMIGwPzI+LZNsdhNmw48a8YdpQ0R9ISSRdJWhVe3RUh6TOSHpL0tKR7JO0raSLwWeBQSc9Iuj1vO1bSDEmPS7pX0ocL5awm6dzcfXK3pBMb6pmf65oDPCtppKSpkv4n1z1X0r8Vtp8s6c+SviHpSUn3SXpTXv6gpIWSPtjs4JvFKuko4IfA7vnYPt/LvptJ+qOkxyQtlnS+pNGF9TtJujXHfRGwai9lnJBjfFjSEYXloyR9VdIDkh6VdJak1fK6MZIuzcf7uKRrJa2U122VW6NPSrpL0oGFMqdLOlPSb/Mx/VnS6yR9M/8+/iJpp4Zz8wtJiyTNk3RsH+fxtZJ+LekpSbdIOl3SdYX1r5d0ZY73HknvbYjru5Iuy+fqJkmbldm3P5KOzH9nT0j6naSNC+tC0kcl/TWfr+9KUl43QtLX8u91nqSP5+1HSvoCsBfwnXwev1Oocr8m5W0u6U9K77PF+e+hniLCrza+gPnAzcBYYB3gbuCjed3ewII8PR54EBib57uBzfL0qcB5DeVeA5xJSnQ7AouAffK6M4A/AWsDGwJzeuopxDQb2AhYLS87JMe4EnAo8Cywfl43GVgKHAGMAE4HHgC+C4wC3g48DazR5Bz0Fetk4Lo+zt/mpK6gUUBXLuubed0qwP3AJ4CVgYOBF4HTC+d3KXBaXr8/8Bywdl7/DWBG/r2sCfwa+GJe90XgrLzfyqQkpDx9L+mf8SrAPvnYx+f9pgOLgV3y8f4RmAd8oHDursrbrgTMAv5PLmtT4D7gX5uciwvza3Vga9Lfy3V53Wvy/BHASGCnHMfWhbgeA3bN688HLiyzby9xXA18KE8flM/HVnnfzwHXF7YN4FJgNDAu/+4n5nUfBeaS/kbXBn6ftx/ZWE/J8i4ATs7ndVVgz3a//9uWd9odQN1fpCR7eGH+y8BZeXpvXk78mwMLgf2AlRvKOJVC4icl7GXAmoVlXwSm5+lXJA/gQ7w68R/ZT9yzgYPy9GTgr4V12+U34HqFZY8BO/ZSTn+xTqaPxN9Lee8GbsvTbwb+Bqiw/npemfj/3pNI8rKFwG6kJP4s+Z9rXrc7MC9PnwZcAmzeUP9ewCPASoVlFwCn5unpwA8K644B7m44d0/m6TcCDzSUfxJwTi/HPYL0T218YdnpvJz4DwWubdjn+8Aphbh+WFi3P/CXMvv2EsvVvJz4fwscVVi3Eumf68Z5PigkYOCnwNQ8/UfgI4V1+1Eu8Tcr77+AaaTrRW1/77fz5a6eFcMjhenngDUaN4iIe4HjSUl+oaQLJY1tUt5Y4PGIeLqw7H5gg8L6BwvritO9LpP0AUmz88fnJ4FtgTGFTR4tTP89x9y47FXHVSLWPklaL5+LhyQ9BZxXiGss8FDkd32h7KLHImJpYb7n/HeRWs6zCsd8eV4O8BVSS/YKpa6tqYU6H4yIl/o4nsbz0uw8bQyM7ak/x/BZYL1eTkUXqUXd7Pe6MfDGhrL+A3hdYZtmf4dl9m1mY+Bbhf0eJ/1TLZ6PZvWW+TvtTbPyTsx135y74I4sWd6w48TfQSLiJxGxJ+nNFMCXelY1bPo3YB1JaxaWjQMeytMPkz4+99iot+p6JnKf7A+AjwOvjYjRwJ2kN9Fg9Rdrf/5fjnW7iPgX4PBCXA8DG/T08RbKLmMxKQlvExGj82utiFgDICKejogTImJT4EDgk5L2zcezUU9//wCOp+hB0ieM0YXXmhGxfy/bLiJ1WzX7vT4I/KmhrDUi4uiScQxm34807LtaRFxfYt/+/k6Xa2jhiHgkIj4cEWOBjwBnqqa3fjrxdwhJ4yXtI2kU8A9SUuppVT4KdPckm4h4kNSl8UVJq0raHjiK1BqG9PH3JElrS9qAlND78hrSm2xRjuUIUot/0ErE2p81gWeAJflYPl1YdwMpGR4raWVJ/07qwy4T10ukf3bfkLQugKQNJP1rnj4gXywUsITUXfUScBOplXlirnNv4F2kvvfldTPwtNKF9tXyxc5tJb2hl3iXARcDp0paXdLrSdcNelwKbCnp/TmulSW9QdJWJeIYzL5nkf7WtgGQtJakQ0rsB+nv9Lh83kcDn2lY/yjpukcpkg6R1POP5AnS3/RLfewybDnxd45RpIuyi0kfZdcl9fcC/Cz/fEzSrXl6EukC8N+AX5L6Y3+f150GLCBdVPw98HPg+WYVR8Rc4GukRPooqR/6z0NxUCVi7c/ngZ1JyfcyUvIDICJeAP6ddJ3gcVJf9cWvLqKpz5C6c27M3Ui/J11kB9gizz9DOi9nRsRVuc53Ae8g/a7OBD4QEX9Zjnp74l8GHEC64D0vl/dDYK0mu3w8r3sE+DHp2sLzuaynSRfZDyOd50dInxhHlYhjMPv+Mm97YT6Hd5LOTRk/AK4g3XxwG/Ab0j/yZXn9t4CD891C3y5R3huAmyQ9Q7pof1xE3FcylmFFr+z+tDqSdDRwWES8pd2x2NCR9CXgdRHR9FbaTiLpHaQbHzbud2Prk1v8NSRpfUl7SFpJ0njgBFJL2zqY0r322yvZldRl1rG/19y9tX++b38D4BQ6+HhWJE789bQK6Xa8p0m3zF1C6pKwzrYmqSvrWeAiUvfcJW2NaHBE6sp7gtTVczfpOw02SO7qMTOrGbf4zcxqpt0DcJUyZsyY6O7ubncYZmYdZdasWYsjoqtxeUck/u7ubmbOnNnuMMzMOoqkxm+qA+7qMTOrHSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmqks8Us6W+k5pnc2LD9G6bmid0n6clX1m5lZ76ps8U8HJhYXSHor6RmcO0TENsBXK6zfzMx6UVnij4hrSGOgFx0NnBERPWOEL6yqfjMz612r+/i3BPaSdJOkP/X2JKEekqZImilp5qJFiwZdcffUywZdhpnZcNDqxD8SWAfYjfSIvJ82PA/1nyJiWkRMiIgJXV2vGmrCzMwGqNWJfwFwcSQ3k553OabFMZiZ1VqrE/+vgLcCSNqS9ECQxS2Owcys1iobnVPSBcDewBhJC0iPTTsbODvf4vkC8MHwk2DMzFqqssQfEZOarDq8qjrNzKx//uaumVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1UxliV/S2ZIW5qdtNa47QVJI8vN2zcxarMoW/3RgYuNCSRsBbwceqLBuMzNrorLEHxHXAI/3suobwImAn7VrZtYGLe3jl3QQ8FBE3N7Kes3M7GWVPWy9kaTVgc+SunnKbD8FmAIwbty4AdfbPfWyAe9rZjYctbLFvxmwCXC7pPnAhsCtkl7X28YRMS0iJkTEhK6urhaGaWY2vLWsxR8RdwDr9szn5D8hIha3KgYzM6v2ds4LgBuA8ZIWSDqqqrrMzKy8ylr8ETGpn/XdVdVtZmbN+Zu7ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzLRuyYUVWHMht/hnvbGMkZmbVc4vfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGqmykcvni1poaQ7C8u+IukvkuZI+qWk0VXVb2ZmvauyxT8dmNiw7Epg24jYHvhv4KQK6zczs15Ulvgj4hrg8YZlV0TE0jx7I7BhVfWbmVnv2tnHfyTw22YrJU2RNFPSzEWLFg1pxcWxeczM6qYtiV/SycBS4Pxm20TEtIiYEBETurq6Whecmdkw1/LROSVNBg4A9o2IaHX9ZmZ119LEL2kicCLwloh4rpV1m5lZUuXtnBcANwDjJS2QdBTwHWBN4EpJsyWdVVX9ZmbWu8pa/BExqZfFP6qqPjMzK6ffFr+kzSSNytN7SzrWX7wyM+tcZbp6fgEsk7Q5MA3YCPhJpVGZmVllyiT+l/KXrv4N+M+I+DSwfrVhmZlZVcok/hclTQI+CFyal61cXUhmZlalMon/CGB34AsRMU/SJsCPqw3LzMyq0u9dPRExV9JngHF5fh7wpaoDMzOzapS5q+ddwGzg8jy/o6QZFcdlZmYVKXMf/6nArsDVABExW9KmFcZUmd4GZ/OAbWZWN6Uu7kbEkoZlL1URjJmZVa9Mi/8uSe8DRkjaAjgWuL7asMzMrCplWvzHANsAzwMXAE8Bx1cYk5mZVajMXT3PASfnl5mZdbh+E7+kLYFPAd3F7SNin+rCMjOzqpTp4/8ZcBbwQ2BZteGYmVnVyiT+pRHxvcojMTOzlihzcffXkj4maX1J6/S8Ko/MzMwqUabF/8H889OFZQF05Je4zMzqrsxdPZsMpGBJZ5Meqr4wIrbNy9YBLiJdKJ4PvDcinhhI+WZmNjBlxupZXdLnJE3L81tIOqBE2dOBiQ3LpgJ/iIgtgD/keTMza6EyffznAC8Ab8rzDwGn97dTRFwDPN6w+CDg3Dx9LvDuUlGamdmQKdPHv1lEHJofxkJEPCdJA6xvvYh4OE8/AqzXbENJU4ApAOPGjRtgdc15cDYzq6syLf4XJK1GuqCLpM1IwzcMSkRET5lN1k+LiAkRMaGrq2uw1ZmZWVamxX8KaSz+jSSdD+wBTB5gfY9KWj8iHpa0PrBwgOWYmdkAlbmr50pJtwK7AQKOi4jFA6xvBun20DPyz0sGWI6ZmQ1QmbF6ds6TPX3z4yStBdwfEUv72O8CYG9gjKQFpE8OZwA/lXQUcD/w3kHEbmZmA1Cmq+dMYGdgDqnFvy1wF7CWpKMj4oredoqISU3K23cggZqZ2dAoc3H3b8BO+ULrLsBOwH3A24AvVxmcmZkNvTKJf8uIuKtnJiLmAq+PiPuqC8vMzKpS9tGL3wMuzPOHAnMljQJerCwyMzOrRJkW/2TgXtLjFo8ndfNMJiX9t1YTlpmZVaXM7Zx/B76WX42eGfKIzMysUmVa/GZmNow48ZuZ1UzTxC/px/nnca0LZ8XhQdzMbLjqq8W/i6SxwJGS1i4+dtGPXjQz61x9Xdw9i/SwlE2BWaRv7fbwoxfNzDpU0xZ/RHw7IrYCzo6ITSNik8LLSd/MrEOVuZ3zaEk7AHvlRddExJxqwzIzs6qUeebuscD5wLr5db6kY6oOzMzMqlFmyIYPAW+MiGcBJH0JuAH4zyoDMzOzapS5j1/AssL8Ml55odfMzDpImRb/OcBNkn6Z598N/KiyiMzMrFJlLu5+XdLVwJ550RERcVulUZmZWWXKtPiJiFuBW4eqUkmfIF07COAO0j+TfwxV+WZm1lzLx+qRtAFwLDAhIrYFRgCHtToOM7O6atcgbSOB1SSNBFYnPd7RzMxaoM/EL2mEpKuGssKIeAj4KvAA8DCwpLcHtkuaImmmpJmLFi0ayhCWW/fUyzxom5kNG30m/ohYBrwkaa2hqlDS2sBBwCbAWOA1kg7vpe5p+QHvE7q6uoaqejOz2itzcfcZ4A5JVwLP9iyMiGMHWOd+wLyIWAQg6WLgTcB5AyzPzMyWQ5nEf3F+DZUHgN0krQ78HdgXmDmE5ZuZWR/K3Md/rqTVgHERcc9gK4yImyT9nHR76FLgNmDaYMs1M7NyygzS9i5gNnB5nt9R0ozBVBoRp0TE6yNi24h4f0Q8P5jyzMysvDK3c54K7Ao8CRARs/FDWMzMOlaZxP9iRCxpWPZSFcGYmVn1ylzcvUvS+4ARkrYgfev2+mrDMjOzqpRp8R8DbAM8D1wAPAUcX2FMZmZWoTJ39TwHnJwfwBIR8XT1YZmZWVXK3NXzBkl3AHNIX+S6XdIu1YdmZmZVKNPH/yPgYxFxLYCkPUkPZ9m+ysDMzKwaZfr4l/UkfYCIuI70xSvLPICbmXWSpi1+STvnyT9J+j7pwm4AhwJXVx+amZlVoa+unq81zJ9SmI4KYjEzsxZomvgj4q2tDMTMzFqj34u7kkYDHwC6i9sPYlhmMzNrozJ39fwGuJH0UHQP1WBm1uHKJP5VI+KTlUdiZmYtUeZ2zh9L+rCk9SWt0/OqPDIzM6tEmRb/C8BXgJN5+W6ewEMzm5l1pDKJ/wRg84hYXHUwZmZWvTJdPfcCzw1lpZJGS/q5pL9IulvS7kNZvpmZNVemxf8sMFvSVaShmYFB3875LeDyiDhY0irA6oMoy8zMlkOZxP+r/BoSktYC3gxMBoiIF0jXEczMrAXKjMd/7hDXuQmwCDhH0g7ALOC4iHi2uJGkKcAUgHHjxg1xCGZm9VVmPP55ku5rfA2izpHAzsD3ImInUlfS1MaNImJaREyIiAldXV2DqM7MzIrKdPVMKEyvChwCDOY+/gXAgoi4Kc//nF4Sv5mZVaPfFn9EPFZ4PRQR3wTeOdAKI+IR4EFJ4/OifYG5Ay3PzMyWT5lB2nYuzK5E+gRQ5pNCX44Bzs939NwHHDHI8szMrKQyCbw4Lv9SYD7w3sFUGhGzeWUXkpmZtUiZu3o8Lr+Z2TBSpqtnFPAeXj0e/2nVhWVmZlUp09VzCbCEdL/98/1sa2ZmK7gyiX/DiJhYeSRmZtYSZQZpu17SdpVHYmZmLVGmxb8nMFnSPFJXj4CIiO0rjczMzCpRJvG/o/IozMysZcp8c/f+3l6tCG5F1j31slLLu6de1nRbM7N2KNPHb2Zmw4gTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTNsSv6QRkm6TdGm7YjAzq6N2tviPA+5uY/1mZrXUlsQvaUPgncAP21G/mVmdtavF/03gROClZhtImiJppqSZixYtallgRYMdYK3MQG4ewM3MWq3liV/SAcDCiJjV13YRMS0iJkTEhK6urhZFZ2Y2/LWjxb8HcKCk+cCFwD6SzmtDHGZmtdTyxB8RJ0XEhhHRDRwG/DEiDm91HGZmdeX7+M3MaqbMoxcrExFXA1e3MwYzs7pxi9/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGbaeh9/J2o2wFp/g60t72BsPdvPP+Ody7WfmVl/3OI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczq5mWJ35JG0m6StJcSXdJOq7VMZiZ1Vk7hmxYCpwQEbdKWhOYJenKiJjbhljMzGqn5S3+iHg4Im7N008DdwMbtDoOM7O6ausgbZK6gZ2Am3pZNwWYAjBu3LjWBtZE2YHWymzX1wBvxYHZuqde9s/5xuni9sN1ULfG4xyK8obbOTJbXm27uCtpDeAXwPER8VTj+oiYFhETImJCV1dX6wM0Mxum2pL4Ja1MSvrnR8TF7YjBzKyu2nFXj4AfAXdHxNdbXb+ZWd21o8W/B/B+YB9Js/Nr/zbEYWZWSy2/uBsR1wFqdb1mZpb4m7tmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxN+ge+plpQdja5ViTP3F17O+r+0btxloLEOx7WDOd+O+y1tOs/MyUM32XdH+nhp1atwrkk47V078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXTroetT5R0j6R7JU1tRwxmZnXVjoetjwC+C7wD2BqYJGnrVsdhZlZX7Wjx7wrcGxH3RcQLwIXAQW2Iw8yslhQRra1QOhiYGBEfyvPvB94YER9v2G4KMCXPjgfuGWCVY4DFA9y3U9XxmKGex+1jroeBHvPGEdHVuHDk4OOpRkRMA6YNthxJMyNiwhCE1DHqeMxQz+P2MdfDUB9zO7p6HgI2KsxvmJeZmVkLtCPx3wJsIWkTSasAhwEz2hCHmVkttbyrJyKWSvo48DtgBHB2RNxVYZWD7i7qQHU8ZqjncfuY62FIj7nlF3fNzKy9/M1dM7OaceI3M6uZYZP4+xsGQtIoSRfl9TdJ6m5DmEOqxDF/UtJcSXMk/UHSxu2IcyiVHe5D0nskhaSOv+2vzDFLem/+Xd8l6SetjrEKJf6+x0m6StJt+W98/3bEOVQknS1poaQ7m6yXpG/n8zFH0s4DriwiOv5Fukj8P8CmwCrA7cDWDdt8DDgrTx8GXNTuuFtwzG8FVs/TR9fhmPN2awLXADcCE9oddwt+z1sAtwFr5/l12x13i457GnB0nt4amN/uuAd5zG8GdgbubLJ+f+C3gIDdgJsGWtdwafGXGQbiIODcPP1zYF9JamGMQ63fY46IqyLiuTx7I+k7E52s7HAf/xf4EvCPVgZXkTLH/GHguxHxBEBELGxxjFUoc9wB/EueXgv4WwvjG3IRcQ3weB+bHAT8VyQ3AqMlrT+QuoZL4t8AeLAwvyAv63WbiFgKLAFe25LoqlHmmIuOIrUWOlm/x5w//m4UEZe1MrAKlfk9bwlsKenPkm6UNLFl0VWnzHGfChwuaQHwG+CY1oTWNsv7nm9qhR2ywYaOpMOBCcBb2h1LlSStBHwdmNzmUFptJKm7Z2/Sp7prJG0XEU+2M6gWmARMj4ivSdod+LGkbSPipXYHtqIbLi3+MsNA/HMbSSNJHw0fa0l01Sg19IWk/YCTgQMj4vkWxVaV/o55TWBb4GpJ80n9oDM6/AJvmd/zAmBGRLwYEfOA/yb9I+hkZY77KOCnABFxA7AqaTCz4WrIhrsZLom/zDAQM4AP5umDgT9GvmLSofo9Zkk7Ad8nJf3h0O/b5zFHxJKIGBMR3RHRTbqucWBEzGxPuEOizN/2r0itfSSNIXX93NfCGKtQ5rgfAPYFkLQVKfEvammUrTUD+EC+u2c3YElEPDyQgoZFV080GQZC0mnAzIiYAfyI9FHwXtIFlMPaF/HglTzmrwBrAD/L17EfiIgD2xb0IJU85mGl5DH/Dni7pLnAMuDTEdHJn2bLHvcJwA8kfYJ0oXdyJzfmJF1A+gc+Jl+3OAVYGSAiziJdx9gfuBd4DjhiwHV18HkyM7MBGC5dPWZmVpITv5lZzTjxm5nVjBO/mVnNOPGbma1g+huwrZftl2uQPid+swGSND/fNz+UZXZLel9hfrKk7wxlHdYRpgOlht6QtAVwErBHRGwDHN/fPk78ZiuWbuB9/W1kw1tvA7ZJ2kzS5ZJmSbpW0uvzquUepM+J34YNSf87j99+naQLJH0qL+/1DSNpeh7f/HpJ90k6uFDWpyXdksc9/3yJug+XdLOk2ZK+L2lEXv6MpC9Iuj0PoLZeIaYbJd0h6XRJz+SizgD2yuV8Ii8bm+P/q6QvD+Eps84yDTgmInYBPgWcmZcv9yB9Tvw2LEh6A/AeYAfgHaRB6Xo0e8MArA/sCRxASrpIejtprJtdgR2BXSS9uY+6twIOJX3U3pH07dn/yKtfA9wYETuQnhHw4bz8W8C3ImI70lg7PaYC10bEjhHxjbxsx1z+dsChkorjtVgNSFoDeBPpW/izSUOx9AzJXBykbxLp28yj+ypvWAzZYAbsAVwSEf8A/iHp1/CqN0zPtqMK+/0qj+Y4t6c1Drw9v27L82uQ3ljXNKl7X2AX4JZcx2pAz8ftF4BL8/Qs4G15enfg3Xn6J8BX+zi2P0TEknw8c4GNeeXwvDb8rQQ8mRsWjRaQHsryIjBPUs8gfbc0K8yJ34a7vt4wAMURS1X4+cWI+H7JOgScGxEn9bLuxcL4McsY2HuuGONAy7AOFhFPSZon6ZCI+JlSC2P7iLidNEjfJOCcsoP0uavHhos/A++StGpu5R8A6Q1DagUdAv98bukO/ZT1O+DIXA6SNpC0bh/b/wE4uGcbSeuo/+cb30jqmoJXDhj4NGl4aauxPGDbDcB4SQskHUXqPjxK0u3AXbz8RLLfAY/lT4NXUWKQPrccbFiIiFskzQDmAI8Cd5CesgbpDfM9SZ8jjXZ4IekZrs3KuiL329+Qu26eAQ7n5e6bxu3n5rKvUHoYzIvA/wLu7yPk44HzJJ0MXF6IdQ6wLL+5pwNP9H3kNhxFxKQmq1514TZ/ovxkfpXi0Tlt2JC0RkQ8I2l1Un/8lIi4td1x9SbH+PeICEmHAZMiorfnB5sNObf4bTiZJmlr0gM5zl1Rk362C/Cd3Ff7JHBke8OxOnGL38ysZnxx18ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGb+PxoH2CpMdXS9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize distribution of gene lengths\n",
    "print('adhesome gene lengths quartile = '+str(np.quantile(df_loc['gene_length'].values, [0.25,0.5,0.75])))\n",
    "plt.figure()\n",
    "plt.hist(df_loc['gene_length'], bins=np.arange(0,1000000,5000))\n",
    "plt.xlabel('gene length')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of adhesome gene lengths')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine gene windows (1 locus centered at the gene)\n",
    "df_loc['gene_window_start'] = df_loc['gene_center']-250000//2\n",
    "df_loc['gene_window_end'] = df_loc['gene_center']+250000//2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>clusterId</th>\n",
       "      <th>protein</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>gene_center</th>\n",
       "      <th>isTF</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3334</th>\n",
       "      <td>uc001isx.3</td>\n",
       "      <td>ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>3074</td>\n",
       "      <td>uc001isx.3</td>\n",
       "      <td>114492</td>\n",
       "      <td>27092770</td>\n",
       "      <td>False</td>\n",
       "      <td>26967770</td>\n",
       "      <td>27217770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15824</th>\n",
       "      <td>uc002uzz.3</td>\n",
       "      <td>ABI2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>17484</td>\n",
       "      <td>uc002uzz.3</td>\n",
       "      <td>103890</td>\n",
       "      <td>204244947</td>\n",
       "      <td>False</td>\n",
       "      <td>204119947</td>\n",
       "      <td>204369947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11717</th>\n",
       "      <td>uc002iop.1</td>\n",
       "      <td>ABI3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>13058</td>\n",
       "      <td>uc002iop.1</td>\n",
       "      <td>12999</td>\n",
       "      <td>47294087</td>\n",
       "      <td>False</td>\n",
       "      <td>47169087</td>\n",
       "      <td>47419087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27778</th>\n",
       "      <td>uc004bzv.3</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>30188</td>\n",
       "      <td>uc004bzv.3</td>\n",
       "      <td>173795</td>\n",
       "      <td>133676164</td>\n",
       "      <td>False</td>\n",
       "      <td>133551164</td>\n",
       "      <td>133801164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25716</th>\n",
       "      <td>uc003sot.4</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>26639</td>\n",
       "      <td>uc003sot.4</td>\n",
       "      <td>3454</td>\n",
       "      <td>5568505</td>\n",
       "      <td>False</td>\n",
       "      <td>5443505</td>\n",
       "      <td>5693505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript  gene  chrom  gene_start   gene_end  clusterId     protein  \\\n",
       "3334   uc001isx.3  ABI1  chr10    27035524   27150016       3074  uc001isx.3   \n",
       "15824  uc002uzz.3  ABI2   chr2   204193002  204296892      17484  uc002uzz.3   \n",
       "11717  uc002iop.1  ABI3  chr17    47287588   47300587      13058  uc002iop.1   \n",
       "27778  uc004bzv.3  ABL1   chr9   133589267  133763062      30188  uc004bzv.3   \n",
       "25716  uc003sot.4  ACTB   chr7     5566778    5570232      26639  uc003sot.4   \n",
       "\n",
       "       gene_length  gene_center   isTF  gene_window_start  gene_window_end  \n",
       "3334        114492     27092770  False           26967770         27217770  \n",
       "15824       103890    204244947  False          204119947        204369947  \n",
       "11717        12999     47294087  False           47169087         47419087  \n",
       "27778       173795    133676164  False          133551164        133801164  \n",
       "25716         3454      5568505  False            5443505          5693505  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Select CAGE peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process robust CAGE peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique CAGE peaks = 201802\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95563</th>\n",
       "      <td>p1@MTND1P23</td>\n",
       "      <td>129bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564571</td>\n",
       "      <td>564600</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95564</th>\n",
       "      <td>p3@MTND1P23</td>\n",
       "      <td>197bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564639</td>\n",
       "      <td>564649</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95565</th>\n",
       "      <td>p3@MTND2P28</td>\n",
       "      <td>246bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565266</td>\n",
       "      <td>565278</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95566</th>\n",
       "      <td>p4@MTND2P28</td>\n",
       "      <td>458bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565478</td>\n",
       "      <td>565483</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95567</th>\n",
       "      <td>p1@MTND2P28</td>\n",
       "      <td>489bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565509</td>\n",
       "      <td>565541</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      short_description    association_with_transcript chrom  cage_start  \\\n",
       "95563       p1@MTND1P23  129bp_to_ENST00000416931_5end  chr1      564571   \n",
       "95564       p3@MTND1P23  197bp_to_ENST00000416931_5end  chr1      564639   \n",
       "95565       p3@MTND2P28  246bp_to_ENST00000457540_5end  chr1      565266   \n",
       "95566       p4@MTND2P28  458bp_to_ENST00000457540_5end  chr1      565478   \n",
       "95567       p1@MTND2P28  489bp_to_ENST00000457540_5end  chr1      565509   \n",
       "\n",
       "       cage_end strand peak_id  \n",
       "95563    564600      +       1  \n",
       "95564    564649      +       3  \n",
       "95565    565278      +       3  \n",
       "95566    565483      +       4  \n",
       "95567    565541      +       1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load annotated FANTOM5 CAGE peaks\n",
    "cage_df = pd.read_csv(fantom5_dir+'hg19.cage_peak_phase1and2combined_ann.txt', sep='\\t', skiprows=7, header=0)\n",
    "# Define new columns\n",
    "cage_df['chrom'] = cage_df['00Annotation'].str.split(':', expand=True)[0]\n",
    "cage_df['cage_start'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[0].astype(int)\n",
    "cage_df['cage_end'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[2].astype(int)\n",
    "cage_df['strand'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[1]\n",
    "cage_df['peak_id'] = cage_df['short_description'].str.split('@', expand=True)[0].str.strip('p')\n",
    "# Reformat dataframe\n",
    "cage_df = cage_df[['short_description','association_with_transcript','chrom','cage_start','cage_end',\n",
    "                   'strand','peak_id']]\n",
    "cage_df = cage_df.sort_values(by=['chrom','cage_start'])\n",
    "print('Number of unique CAGE peaks = '+str(len(np.unique(cage_df['short_description'].values))))\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CAGE peaks within selected gene windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bed file for regions upstream of genes of interest\n",
    "gene_window_df = df_loc[['chrom','gene_window_start','gene_window_end','gene']]\n",
    "gene_window_df.columns = ['chrom','start','stop','name']\n",
    "gene_window_bed = pybedtools.BedTool.from_dataframe(gene_window_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a bed file for robust CAGE peaks\n",
    "cage_peak_df = cage_df[['chrom','cage_start','cage_end','short_description']]\n",
    "cage_peak_df.columns = ['chrom','start','end','name']\n",
    "cage_bed = pybedtools.BedTool.from_dataframe(cage_peak_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get CAGE peaks for each gene of interest\n",
    "out = pybedtools.bedtool.BedTool.map(gene_window_bed, cage_bed, c = 4, o = 'distinct')\n",
    "out_df = out.to_dataframe()\n",
    "out_df.columns=['chrom','gene_window_start','gene_window_end','gene','peaks']\n",
    "out_df['peaks'] = out_df['peaks'].str.replace(',p',';p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat matching results\n",
    "peaks_per_gene_df = pd.merge(df_loc, out_df, on=['chrom','gene_window_start','gene_window_end','gene'])\n",
    "peaks_per_gene_df = peaks_per_gene_df[peaks_per_gene_df['peaks'] != '.']\n",
    "peaks_per_gene_df['peaks'] = peaks_per_gene_df['peaks'].str.split(';')\n",
    "peaks_per_gene_df['n_peaks'] = [len(peaks) for peaks in peaks_per_gene_df['peaks']]\n",
    "peaks_per_gene_df = peaks_per_gene_df[['gene', 'isTF', 'chrom', 'gene_start', 'gene_end', 'gene_length',\n",
    "                                       'gene_window_start', 'gene_window_end','peaks','n_peaks']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isTF</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "      <th>peaks</th>\n",
       "      <th>n_peaks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>26967770</td>\n",
       "      <td>27217770</td>\n",
       "      <td>[p10@PDSS1, p11@PDSS1, p12@PDSS1, p1@ABI1, p1@...</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>204119947</td>\n",
       "      <td>204369947</td>\n",
       "      <td>[p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>47169087</td>\n",
       "      <td>47419087</td>\n",
       "      <td>[p1@ABI3, p1@B4GALNT2, p1@FLJ40194, p1@GNGT2, ...</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr9</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>133551164</td>\n",
       "      <td>133801164</td>\n",
       "      <td>[p1@ABL1, p1@AF533988, p1@EXOSC2, p1@M14755, p...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>False</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>5443505</td>\n",
       "      <td>5693505</td>\n",
       "      <td>[p10@ACTB, p10@TNRC18, p11@ACTB, p12@ACTB, p13...</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene   isTF  chrom  gene_start   gene_end  gene_length  gene_window_start  \\\n",
       "0  ABI1  False  chr10    27035524   27150016       114492           26967770   \n",
       "1  ABI2  False   chr2   204193002  204296892       103890          204119947   \n",
       "2  ABI3  False  chr17    47287588   47300587        12999           47169087   \n",
       "3  ABL1  False   chr9   133589267  133763062       173795          133551164   \n",
       "4  ACTB  False   chr7     5566778    5570232         3454            5443505   \n",
       "\n",
       "   gene_window_end                                              peaks  n_peaks  \n",
       "0         27217770  [p10@PDSS1, p11@PDSS1, p12@PDSS1, p1@ABI1, p1@...       35  \n",
       "1        204369947  [p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...       17  \n",
       "2         47419087  [p1@ABI3, p1@B4GALNT2, p1@FLJ40194, p1@GNGT2, ...       35  \n",
       "3        133801164  [p1@ABL1, p1@AF533988, p1@EXOSC2, p1@M14755, p...       21  \n",
       "4          5693505  [p10@ACTB, p10@TNRC18, p11@ACTB, p12@ACTB, p13...       79  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "peaks_per_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXgAAAEWCAYAAABsY4yMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdcElEQVR4nO3deZgdVZ3/8feHEEgIS4A0DFvTYTEuiCytoqLDsDgsjugjDqAIQbRHlE1EJ/x0hmVwBAVFZxSIrApG2UGiCCLIooBJjIEkMCIJkhBWCYSAQML390edDpWmu29131t9+1Y+r+epJ7We861blW/XPVX3lCICMzOrntWaHYCZmZXDCd7MrKKc4M3MKsoJ3sysopzgzcwqygnezKyinOAbSNJ8SXv2sez9kh4c6piGE2UukvSspHubHMtESXc2sf4jJT0h6QVJGw5hvRdLOm2o6rPmWr3ZAawqIuIOYEKt9SSdDGwTEYeUHtTQ2xXYC9g8IpY2O5hmkTQS+DawS0T8qdnxWHX5Cn4VIqnZf9C3BOZXLbkP4nPdGBgFzC4hnJYyDM7JSnOCb7wdJM2S9Jykn0kaBSBpN0kLuleS9O+SFkpaIulBSXtI2hv4f8CB6av7n9K6m0q6XtLfJD0k6bO5ckZLuiQ1e8yV9JUe9cxPdc0ClkpaXdIkSX9Jdc+R9NHc+hMl3SXpO5IWS3pY0nvT/EclPSnpsL52vq9YJR0BnA+8J+3bKb1sO1HSnZLOTPszT9I+PfZlz9z0yZIuTeMdkkLS4SnOZyV9TtI70/FYLOl/31il/jcdqwck7ZFbsJ6kCyQtSsfpNEkjevmMngFO7mVf1pR0tqTH0nB2mvcmoLupbrGk3/Sybfe+dKVtF0k6Ibd8tdwxfEbS5ZI2yC2/QtLjab9ul/S2Po7VOpJulfQ9ZfZN58OStM8n9LFd9/6X9dnVOqc3lXSVpKfSOXJMbtnJ6fP4UdqP2ZI6i2xbSRHhoUEDMB+4F9gU2ACYC3wuLdsNWJDGJwCPApum6Q5g6zR+MnBpj3JvB35AdtW3A/AUsHtadjrwW2B9YHNgVnc9uZhmAlsAo9O8j6cYVwMOBJYCm6RlE4FlwOHACOA04K/A94E1gQ8CS4C1+/gM+ot1InBnP5/fROBV4LOp7iOBxwDl9mXP3PorPqv0GQZwbqr7g8DfgWuBjYDNgCeBf+yxn18ERqbP4Tlgg7T8GuA8YEza/l7g33psezRZM+foXvblVODutG0b8Dvgv3rEunofn0P38imp/renz3HPtPzYVPbm6ZicB0zJbf9pYJ207GxgZm7ZxemYbpj26bTcskXA+9P4+sBO/RynMj+7Ps9psnN2OvCfwBrAVsDDwD/nzom/A/uSnUPfAO4usm0Vh6YHUKWBLAEdkpv+JnBuGt8td5JuQ5Zs9gRG9ijjZHIJniwxLwfWyc37BnBxGl/pBAU+wxsT/KdrxD0T2D+NTwT+nFv2drJks3Fu3jPADr2UUyvWidRO8A/lptdKdf9Dbl9qJfjNesR5YG76KuC4XF0r/nikefcCnyJrQnmZXPIBDgZuzW371xqf6V+AfXPT/0zWPJWPtVaCf3OPc+mCND4X2CO3bBOyP4xvKA8Ym8paL01fDFwI3A98uce6fwX+DVi3xr6V/dn1eU4D7+65PXAicFHunPh1btlbgZeKbFvFwe1fjfd4bvxFsivllUTEQ5KOIzsZ3ybpV8DxEfFYL+VtCvwtIpbk5j0CdOaWP5pblh/vdZ6kQ4HjyRIJwNrAuNwqT+TGX0ox95y39iBiLWLF5xcRL0rqjq+onnH2F/fCSP/Lk0fI9mFLsivTRal+yK7+an3OeZum8nqWPRD5Oh4h+2NLiu8aSa/lli8HNpb0OPB1sm9pbUD3OuPIrrIB9gNeIPu2k/cx4GvA6cqa9CZFxO/7iK3sz66v9bcENpW0ODdvBHBHbrrn/8FRytr6i2xbKW6Db5KI+ElE7Ep20gVwRveiHqs+BmwgaZ3cvHZgYRpfRPY1ttsWvVXXPSJpS+CHwFHAhhExluxqTr1sN1C1Yq3XUrKr+m7/UGd5mymXhchifYwsobwMjIuIsWlYNyLybdm1umF9jOzY9ix7IPLHMr/9o8A+udjGRsSoiFgIfALYn+zb4Xq8/kc8v58/BG4EfiFpTPfMiPhDROxP1qxyLXB5P7GV+dn1d04/Cszrse/rRMS+Ncqsd9uW5ATfBJImSNpd0ppk7YUv8fqV1hNAh6TVACLiUbL2229IGiVpe+AI4NK0/uXAiZLWl7QZWeLuzxiy/2BPpVgOB7ZrxH4ViLVeM4GDJI1MN84OqLO8jYBjUnkfB94C/CIiFgE3AWdJWjfd1Nxa0j8OoOwpwNcktUkaR9buO9DP4T8krZVukh4O/CzNPxf4evpjTapj/7RsHbIE+wzZH8P/7qPso8hu9v483dRcQ9InJa0XEa8Cz/P6OdmbMj+7/s7pe4Elyh4cGC1phKTtJL2zQLn1bNuSnOCbY02yG0lPk32d3IisLRDgivTvM5JmpPGDya7EHiO7gXVSRPw6LTsVWADMA34NXEn2H7xXETEHOAv4Pdkfk7cDdzVipwrEWq//ALYGngVOAX5SZ3n3ANuSHYevAwdExDNp2aFkN+LmpPquJGvrLuo0YBrZDcL7gBlp3kD8FngIuAU4MyJuSvO/C1wP3CRpCdkN13enZT8iay5ZmGK/u7eCU/NKF9m5cx3ZjelPAfMlPQ98DvhkP7GV+dn1eU5HxHLgQ2Q38Oel+s8n+7bSr3q2bVXdTydYRUg6EjgoIgZyxWTDiKQOsgQ0MiKWNTmcN5A0EfhMamIcivp8Tg+Sr+BbnKRNJL0vfRWeAHyJ7MrZrCX5nG4cP0XT+tYge+Z4PLAY+CnZc+hmrcrndIO4icbMrKLcRGNmVlHDqolm3Lhx0dHR0ewwzMxaxvTp05+OiLbelg2rBN/R0cG0adOaHYaZWcuQ9Ehfy9xEY2ZWUU7wZmYV5QRvZlZRTvBmZhXlBG9mVlFO8GZmFVVqgpf0xfROxPslTVF6P6mZmZWvtASf+nE+BuiMiO3I3pxyUFn1mZnZyspuolkdGJ1el7UWA3+jjZmZDVJpv2SNiIWSziR7ke9LwE25FxasIKmL7MUDtLe3lxXOsNYxaeqK8fmn79fESMysSspsolmf7N2Q48leojtG0iE914uIyRHRGRGdbW29dqdgZmaDUGYTzZ5kL7h9Kr3j8WrgvSXWZ2ZmOWUm+L8Cu6SXBgvYA5hbYn1mZpZTWoKPiHvIXpY7g+ylw6sBk8uqz8zMVlZqd8ERcRJwUpl1mJlZ7/xLVjOzinKCNzOrKCd4M7OKcoI3M6soJ3gzs4pygjczqygneDOzinKCNzOrKCd4M7OKcoI3M6soJ3gzs4pygjczqygneDOzinKCNzOrKCd4M7OKcoI3M6uoMl+6PUHSzNzwvKTjyqrPzMxWVtobnSLiQWAHAEkjgIXANWXVZ2ZmKxuqJpo9gL9ExCNDVJ+Z2SpvqBL8QcCUIarLzMwYggQvaQ3gw8AVfSzvkjRN0rSnnnqq7HCsZB2TptIxaWqzwzAzhuYKfh9gRkQ80dvCiJgcEZ0R0dnW1jYE4ZiZrRqGIsEfjJtnzMyGXKkJXtIYYC/g6jLrMTOzNyrtMUmAiFgKbFhmHWZm1jv/ktXMrKKc4M3MKsoJ3sysopzgzcwqygnezKyinODNzCrKCd7MrKKc4M3MKsoJ3sysopzgzcwqygnezKyinODNzCrKCd7MrKKc4M3MKsoJ3sysopzgzcwqygnezKyiyn5l31hJV0p6QNJcSe8psz4zM3tdqa/sA74L3BgRB0haA1ir5PrMzCwpLcFLWg/4ADARICJeAV4pqz4zM1tZmVfw44GngIskvQOYDhybXsS9gqQuoAugvb29xHBWbR2Tpq4Yn3/6fpWrb7jHYdYMZbbBrw7sBJwTETsCS4FJPVeKiMkR0RkRnW1tbSWGY2a2aikzwS8AFkTEPWn6SrKEb2ZmQ6C0BB8RjwOPSpqQZu0BzCmrPjMzW1nZT9EcDVyWnqB5GDi85PrMzCwpNcFHxEygs8w6zMysdzWbaCRtLWnNNL6bpGMkjS09MjMzq0uRNvirgOWStgEmA1sAPyk1KjMzq1uRBP9aRCwDPgr8T0R8Gdik3LDMzKxeRRL8q5IOBg4DbkjzRpYXkpmZNUKRBH848B7g6xExT9J44MflhmVmZvWq+RRNRMyR9O9Ae5qeB5xRdmBmZlafIk/R/AswE7gxTe8g6fqS4zIzszoVaaI5GXgXsBhWPNu+VWkRmZlZQxS6yRoRz/WY91oZwZiZWeMU+SXrbEmfAEZI2hY4BvhduWGZmVm9ilzBHw28DXgZmAI8DxxXYkxmZtYARZ6ieRH4ahrMzKxF1Ezwkt4EnAB05NePiN3LC8vMzOpVpA3+CuBc4HxgebnhmJlZoxRJ8Msi4pzSIzEzs4YqcpP155I+L2kTSRt0D6VHZmZmdSlyBX9Y+vfLuXmBf+xkZjasFXmKZvxgC5c0H1hC1na/LCL8diczsyFS5CmatYDjgfaI6Eo/dpoQETfU2LTbP0XE0/UEaWZmA1ekDf4i4BXgvWl6IXBaaRGZmVlDFGmD3zoiDkwv/SAiXpSkguUHcJOkAM6LiMk9V5DUBXQBtLe3Fyx21dUxaeqK8fmn7zeg9VvNQPe17DiaGYPZYBS5gn9F0miyZI2krcm6LShi14jYCdgH+IKkD/RcISImR0RnRHS2tbUVjdvMzGookuBPIusLfgtJlwG3AF8pUnhELEz/PglcQ9btsJmZDYEiT9HcLGkGsAsg4NgiN00ljQFWi4glafyDwKn1BmxmZsUUeYpmpzS6KP3bLmk94JGIWNbPphsD16Tm+tWBn0TEjfUEa2ZmxRW5yfoDYCdgFtkV/HbAbGA9SUdGxE29bRQRDwPvaFSgZmY2MEXa4B8Ddkw3QncGdgQeBvYCvllmcGZmNnhFEvybImJ290REzAHenK7QzcxsmCr6yr5zgJ+m6QOBOZLWBF4tLTIzM6tLkSv4icBDZK/pO46seWYiWXL/p3LCMjOzehV5TPIl4Kw09PRCwyMyM7OGKHIFb2ZmLcgJ3sysovpM8JJ+nP49dujCMTOzRunvCn5nSZsCn5a0fv51fX5ln5nZ8NffTdZzyToW2wqYTvYr1m5+ZZ+Z2TDX5xV8RHwvIt4CXBgRW0XE+Nzg5G5mNswVeUzySEnvAN6fZt0eEbPKDcvMzOpV8ykaSccAlwEbpeEySUeXHZiZmdWnSFcFnwHeHRFLASSdAfwe+J8yAzMzs/oUeQ5ewPLc9HJWvuFqZmbDUJEr+IuAeyRdk6Y/AlxQWkRmZtYQRW6yflvSbcCuadbhEfHHUqMyM7O6FbmCJyJmADMGU4GkEcA0YGFEfGgwZZiZ2cANRV80xwJzh6AeMzPLKTXBS9oc2A84v8x6zMzsjfptoknNK7+OiMG+2ONs4CvAOv3U0QV0AbS3tw+ymuGhY9LUFePzT9+v5eqrVd5g6xvIdvl1h1ojPs+hPgfM+tPvFXxELAdek7TeQAuW9CHgyYiYXqOOyemF3p1tbW0DrcbMzPpQ5CbrC8B9km4GlnbPjIhjamz3PuDDkvYFRgHrSro0Ig4ZdLRmZlZYkQR/dRoGJCJOBE4EkLQbcIKTu5nZ0CnyHPwlkkYD7RHx4BDEZGZmDVCks7F/AWYCN6bpHSRdP5BKIuI2PwNvZja0ijwmeTLwLmAxQETMxC/7MDMb9ook+Fcj4rke814rIxgzM2ucIjdZZ0v6BDBC0rbAMcDvyg3LzMzqVeQK/mjgbcDLwBTgeeC4EmMyM7MGKPIUzYvAV9OLPiIilpQflpmZ1avIUzTvlHQfMIvsB09/krRz+aGZmVk9irTBXwB8PiLuAJC0K9lLQLYvMzAzM6tPkTb45d3JHSAi7gSWlReSmZk1Qp9X8JJ2SqO/lXQe2Q3WAA4Ebis/NDMzq0d/TTRn9Zg+KTceJcRiZmYN1GeCr6MPeDMzGwZq3mSVNBY4FOjIr1+gu2AzM2uiIk/R/AK4G7gPd1FgZtYyiiT4URFxfOmRmJlZQxV5TPLHkj4raRNJG3QPpUdmZmZ1KXIF/wrwLeCrvP70TOAug83MhrUiCf5LwDYR8XTZwZiZWeMUaaJ5CHhxoAVLGiXp3tR3zWxJpww8PDMzG6wiV/BLgZmSbiXrMhgo9Jjky8DuEfGCpJHAnZJ+GRF3Dz5cMzMrqkiCvzYNAxIRAbyQJkemwb+ANTMbIkX6g79ksIVLGgFMB7YBvh8R9/SyThfQBdDe3j7YqlpGx6SpAMw/fb/C6xZZfyDrFjHYOBu5blka/Vk1U2/HqUr7162s87tR5Q1XRfqDnyfp4Z5DkcIjYnlE7ABsDrxL0na9rDM5IjojorOtrW3AO2BmZr0r0kTTmRsfBXwcGNBz8BGxOLXh7w3cP5BtzcxscGpewUfEM7lhYUScDdT8TiOpLfVjg6TRwF7AA3XGa2ZmBRXpbGyn3ORqZFf0Ra78NwEuSe3wqwGXR8QNg4rSzMwGrEiizvcLvwyYD/xrrY0iYhaw4+DCMjOzehV5isb9wpuZtaAiTTRrAh/jjf3Bn1peWGZmVq8iTTTXAc+RPc/+co11zcxsmCiS4DePiL1Lj8TMzBqqSGdjv5P09tIjMTOzhipyBb8rMFHSPLImGpF1NbN9qZGZmVldiiT4fUqPwszMGq7IY5KPDEUgZmbWWEXa4M3MrAU5wZuZVZQTvJlZRTnBm5lVlBO8mVlFOcGbmVWUE7yZWUU5wZuZVZQTvJlZRZWW4CVtIelWSXMkzZZ0bFl1mZnZGxXpi2awlgFfiogZktYBpku6OSLmlFinmZklpV3BR8SiiJiRxpcAc4HNyqrPzMxWVuYV/AqSOshewH1PL8u6gC6A9vb2oQin4TomTR1WZQ2kjFrrNnLfGqW3mOafvl+/y/uaP5B1BxJPPbrLy+9TrfqKxDCQz63I/Fp19LZuX3HWKneoDWSfB7vdYOsYiNJvskpaG7gKOC4inu+5PCImR0RnRHS2tbWVHY6Z2Sqj1AQvaSRZcr8sIq4usy4zM1tZmU/RCLgAmBsR3y6rHjMz612ZV/DvAz4F7C5pZhr2LbE+MzPLKe0ma0TcSfb+VjMzawL/ktXMrKKc4M3MKsoJ3sysopzgzcwqygnezKyinODNzCrKCd7MrKKc4M3MKsoJ3sysopzgzcwqygnezKyinODNzCrKCd7MrKKc4M3MKsoJ3sysopzgzcwqygnezKyiynwn64WSnpR0f1l1mJlZ38q8gr8Y2LvE8s3MrB+lJfiIuB34W1nlm5lZ/0p76XZRkrqALoD29vaGlNkxaSoA80/fr/C6fcmXkV+3Vtm1yh2oRpbX6NiGg+G4T72dhwOJcyj2qa86BjK/Ef/PBmsg5RaJs1YdAymjr9gGG8dgNP0ma0RMjojOiOhsa2trdjhmZpXR9ARvZmblcII3M6uoMh+TnAL8HpggaYGkI8qqy8zM3qi0m6wRcXBZZZuZWW1uojEzqygneDOzinKCNzOrKCd4M7OKcoI3M6soJ3gzs4pygjczqygneDOzinKCNzOrKCd4M7OKcoI3M6soJ3gzs4pygjczqygneDOzinKCNzOrKCd4M7OKcoI3M6uoUhO8pL0lPSjpIUmTyqzLzMxWVuY7WUcA3wf2Ad4KHCzprWXVZ2ZmKyvzCv5dwEMR8XBEvAL8FNi/xPrMzCxHEVFOwdIBwN4R8Zk0/Sng3RFxVI/1uoCuNDkBeHCQVY4Dnh7ktq2gyvtX5X0D71+rG+77t2VEtPW2YPWhjqSniJgMTK63HEnTIqKzASENS1XevyrvG3j/Wl0r71+ZTTQLgS1y05uneWZmNgTKTPB/ALaVNF7SGsBBwPUl1mdmZjmlNdFExDJJRwG/AkYAF0bE7LLqowHNPMNclfevyvsG3r9W17L7V9pNVjMzay7/ktXMrKKc4M3MKqrlE3zVukOQtIWkWyXNkTRb0rFp/gaSbpb05/Tv+s2OtR6SRkj6o6Qb0vR4Sfek4/izdGO+JUkaK+lKSQ9ImivpPVU6fpK+mM7N+yVNkTSqlY+fpAslPSnp/ty8Xo+XMt9L+zlL0k7Ni7y2lk7wFe0OYRnwpYh4K7AL8IW0T5OAWyJiW+CWNN3KjgXm5qbPAL4TEdsAzwJHNCWqxvgucGNEvBl4B9l+VuL4SdoMOAbojIjtyB6gOIjWPn4XA3v3mNfX8doH2DYNXcA5QxTjoLR0gqeC3SFExKKImJHGl5Alh83I9uuStNolwEeaEmADSNoc2A84P00L2B24Mq3SsvsnaT3gA8AFABHxSkQspkLHj+zpu9GSVgfWAhbRwscvIm4H/tZjdl/Ha3/gR5G5GxgraZMhCXQQWj3BbwY8mptekOZVgqQOYEfgHmDjiFiUFj0ObNysuBrgbOArwGtpekNgcUQsS9OtfBzHA08BF6UmqPMljaEixy8iFgJnAn8lS+zPAdOpzvHr1tfxaqmc0+oJvrIkrQ1cBRwXEc/nl0X2bGtLPt8q6UPAkxExvdmxlGR1YCfgnIjYEVhKj+aYFj9+65NdxY4HNgXG8MbmjUpp5ePV6gm+kt0hSBpJltwvi4ir0+wnur8Kpn+fbFZ8dXof8GFJ88ma1HYna7Mem77yQ2sfxwXAgoi4J01fSZbwq3L89gTmRcRTEfEqcDXZMa3K8evW1/FqqZzT6gm+ct0hpPboC4C5EfHt3KLrgcPS+GHAdUMdWyNExIkRsXlEdJAdr99ExCeBW4ED0mqtvH+PA49KmpBm7QHMoSLHj6xpZhdJa6VztXv/KnH8cvo6XtcDh6anaXYBnss15Qw/EdHSA7Av8H/AX4CvNjueBuzPrmRfB2cBM9OwL1k79S3An4FfAxs0O9YG7OtuwA1pfCvgXuAh4ApgzWbHV8d+7QBMS8fwWmD9Kh0/4BTgAeB+4MfAmq18/IApZPcTXiX7BnZEX8cLENmTe38B7iN7mqjp+9DX4K4KzMwqqtWbaMzMrA9O8GZmFeUEb2ZWUU7wZmYV5QRvZlZRTvBWWZJuk1T6y5IlHZN6jbysxDrmSxpXVvlWTaW9ss+slUlaPV7vW6WWzwN7RsSCMmMyGyhfwVtTSepIV78/TH2M3yRpdFq24gpc0rjUvQGSJkq6NvXTPV/SUZKOT5173S1pg1wVn5I0M/Vd/q60/ZjUB/i9aZv9c+VeL+k3ZD9y6Rnr8amc+yUdl+adS/Yjn19K+mKP9SdKui7tx58lnZRbdkiqf6ak81LX10g6R9K09Fmc0ksMoyX9UtJn035MlfSnFNOBgz8SVknN/qWVh1V7ADrI+sDfIU1fDhySxm8j/VIQGAfMT+MTyX4xuQ7QRtaj4efSsu+QddDWvf0P0/gHgPvT+H/n6hhL9kvoMancBfTyK1NgZ7JfLo4B1gZmAzumZfOBcb1sM5HsF5IbAqPJfvnZCbwF+DkwMq33A+DQNN79i8kRKf7tc3V0kP2qsnvdj3XvX5per9nH08PwGtxEY8PBvIiYmcankyWyWm6NrL/8JZKeI0uYkCXh7XPrTYGsz29J60oaC3yQrMOzE9I6o4D2NH5zRPTsGxyyLiSuiYilAJKuBt4P/LFGnDdHxDO5bXYl+4O2M/CHrDsXRvN6Z1b/KqmLrPl0E7IX2cxKy64DvhkR3W399wFnSTqDrMuHO2rEYqsYJ3gbDl7OjS8nS3iQJcLuZsRR/WzzWm76NVY+r3v2xRFk/Yl8LCIezC+Q9G6y7n0bqa/6L4mIE3vUPx44AXhnRDwr6WJW3u+7gL0l/SQy/5deGbcvcJqkWyLi1AbHby3MbfA2nM0nu9KF13sqHKgDASTtStbz33PAr4CjU2+ISNqxQDl3AB9JvSiOAT6a5tWyl7L3e44meyvQXWTt+wdI2ijVv4GkLYF1yf7APCdpY7LXw+X9J9nr8L6fttsUeDEiLgW+RdYtsdkKvoK34exM4PLUZDF1kGX8XdIfgZHAp9O8/yJ7q9QsSasB84AP9VdIRMxIV9T3plnnR0St5hnS+leR9Rt+aURMA5D0NeCmVP+rwBci4u4U6wNkbw26q5fyjgUulPRNsj8U35L0WirjyALx2CrEvUmalUTSRLKbxEc1OxZbNbmJxsysonwFb2ZWUb6CNzOrKCd4M7OKcoI3M6soJ3gzs4pygjczq6j/D7nosfY8FkxqAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Histogram of number of peaks per gene\n",
    "plt.figure()\n",
    "plt.hist(peaks_per_gene_df['n_peaks'], bins=np.arange(0,110,1))\n",
    "plt.xlabel('number of peaks')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of number of peaks per gene')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Associate CAGE peaks with genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Dataframe of genes for each peak\n",
    "genes_per_peak = peaks_per_gene_df[['gene','peaks']]\n",
    "genes_per_peak = lu.unnesting(genes_per_peak,['peaks'])\n",
    "genes_per_peak = genes_per_peak.groupby('peaks')['gene'].apply(list).reset_index(name='genes')\n",
    "genes_per_peak['n_genes'] = [len(genes) for genes in genes_per_peak['genes']]\n",
    "genes_per_peak.columns = ['short_description','genes','n_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add peak location\n",
    "genes_per_peak = pd.merge(genes_per_peak, cage_df, on=['short_description'])\n",
    "genes_per_peak = genes_per_peak[['short_description','chrom','cage_start','cage_end','strand','genes']]\n",
    "genes_per_peak = genes_per_peak.sort_values(by=['chrom'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consider a window around each peak\n",
    "genes_per_peak['cage_window_start'] = genes_per_peak['cage_start']-400\n",
    "genes_per_peak['cage_window_end'] = genes_per_peak['cage_end']+400"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genes</th>\n",
       "      <th>cage_window_start</th>\n",
       "      <th>cage_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3103</th>\n",
       "      <td>p2@TPM3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154164569</td>\n",
       "      <td>154164584</td>\n",
       "      <td>-</td>\n",
       "      <td>[HAX1]</td>\n",
       "      <td>154164169</td>\n",
       "      <td>154164984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4130</th>\n",
       "      <td>p4@DCST2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155006254</td>\n",
       "      <td>155006259</td>\n",
       "      <td>-</td>\n",
       "      <td>[SHC1]</td>\n",
       "      <td>155005854</td>\n",
       "      <td>155006659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>p1@ADAM15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155023757</td>\n",
       "      <td>155023833</td>\n",
       "      <td>+</td>\n",
       "      <td>[SHC1]</td>\n",
       "      <td>155023357</td>\n",
       "      <td>155024233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>p1@SERINC2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31886034</td>\n",
       "      <td>31886099</td>\n",
       "      <td>+</td>\n",
       "      <td>[FABP3]</td>\n",
       "      <td>31885634</td>\n",
       "      <td>31886499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4134</th>\n",
       "      <td>p4@DNAJB4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>78444874</td>\n",
       "      <td>78444905</td>\n",
       "      <td>+</td>\n",
       "      <td>[NEXN]</td>\n",
       "      <td>78444474</td>\n",
       "      <td>78445305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     short_description chrom  cage_start   cage_end strand    genes  \\\n",
       "3103           p2@TPM3  chr1   154164569  154164584      -   [HAX1]   \n",
       "4130          p4@DCST2  chr1   155006254  155006259      -   [SHC1]   \n",
       "781          p1@ADAM15  chr1   155023757  155023833      +   [SHC1]   \n",
       "1763        p1@SERINC2  chr1    31886034   31886099      +  [FABP3]   \n",
       "4134         p4@DNAJB4  chr1    78444874   78444905      +   [NEXN]   \n",
       "\n",
       "      cage_window_start  cage_window_end  \n",
       "3103          154164169        154164984  \n",
       "4130          155005854        155006659  \n",
       "781           155023357        155024233  \n",
       "1763           31885634         31886499  \n",
       "4134           78444474         78445305  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize\n",
    "genes_per_peak.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Create sequence files for all CAGE windows"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over all CAGE windows to generate sequence files\n",
    "for i in tqdm(range(genes_per_peak.shape[0])):\n",
    "    time.sleep(.01)\n",
    "    # Peak characteristics\n",
    "    peak = genes_per_peak['short_description'].iloc[i]\n",
    "    chrom = genes_per_peak['chrom'].iloc[i]\n",
    "    start = genes_per_peak['cage_window_start'].iloc[i]\n",
    "    end = genes_per_peak['cage_window_end'].iloc[i]\n",
    "    # Generate window sequence\n",
    "    result = subprocess.run(cage_fasta_dir+'twoBitToFa http://hgdownload.cse.ucsc.edu/gbdb/hg19/hg19.2bit '+\n",
    "                            cage_fasta_dir+peak+'.fa'+\n",
    "                            ' -seq='+chrom+' -start='+str(start)+' -end='+str(end)+' -noMask',\n",
    "                            shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Search for TFBS in all CAGE windows\n",
    "See process_JASPAR_without_strand.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Find TFs binding each CAGE window"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Useful data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load JASPAR TFBS metadata\n",
    "jaspar_df = pd.read_csv(tf_dir+'JASPAR-HomoSapiens.csv', \n",
    "                        header=0, usecols=['ID','Name','Species','Class','Family'])\n",
    "# Mapping JASPAR ID to HGNC\n",
    "jasparid_to_hgnc = {jaspar_df.iloc[i,0]:jaspar_df.iloc[i,1] for i in range(jaspar_df.shape[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique TFs = 639\n",
      "Number of unique TF motifs = 810\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Class</th>\n",
       "      <th>Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0002.1</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Runt domain factors</td>\n",
       "      <td>Runt-related factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0003.1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0003.2</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0003.3</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID    Name       Species                                  Class  \\\n",
       "0  MA0002.1   RUNX1  Homo sapiens                    Runt domain factors   \n",
       "1  MA0003.1  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "2  MA0003.2  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "3  MA0003.3  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "4  MA0003.4  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "\n",
       "                 Family  \n",
       "0  Runt-related factors  \n",
       "1                  AP-2  \n",
       "2                  AP-2  \n",
       "3                  AP-2  \n",
       "4                  AP-2  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize data\n",
    "print('Number of unique TFs = '+str(len(np.unique(jaspar_df['Name']))))\n",
    "print('Number of unique TF motifs = '+str(len(np.unique(jaspar_df['ID']))))\n",
    "jaspar_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load MOODS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary mapping CAGE peaks to the corresponding TFs\n",
    "tfs_associated_to_cage_peaks = {}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over files in moods_out_dir and summarize results\n",
    "for peak_moods_file in os.listdir(moods_out_dir):\n",
    "    # load results dataframe for that gene\n",
    "    cage_peak = peak_moods_file.strip('.txt')\n",
    "    res_df = pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                         usecols=[0,1,2,3,4,5], \n",
    "                         names=['peak_id','tfbs_file','motif_loc','strand','score','motif'])\n",
    "    # Add columns corresponding to the target gene\n",
    "    res_df['cage_peak'] = cage_peak\n",
    "    res_df['chrom'] = res_df['peak_id'].str.split(':', expand=True)[0]\n",
    "    location_cols = res_df['peak_id'].str.split(':', expand=True)[1]\n",
    "    res_df[['start','stop']] = location_cols.str.split('-', expand=True)\n",
    "    # Add columns corresponding to the TFBS\n",
    "    res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "    res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "    # Select relevant columns\n",
    "    res_df = res_df[['cage_peak', 'chrom','start','stop','motif_loc',\n",
    "                     'jaspar_tfbs','TF','score','strand','motif']]\n",
    "    # Store list of TFs associated with the current CAGE peak in dict\n",
    "    tfs_associated_to_cage_peaks[cage_peak] = np.sort(np.unique(res_df['TF'].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_moods_file = os.listdir(moods_out_dir)[0]\n",
    "# load results dataframe for that gene\n",
    "cage_peak = peak_moods_file.strip('.txt')\n",
    "res_df = pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                     usecols=[0,1,2,3,4,5], \n",
    "                     names=['peak_id','tfbs_file','motif_loc','strand','score','motif'])\n",
    "# Add columns corresponding to the target gene\n",
    "res_df['cage_peak'] = cage_peak\n",
    "res_df['chrom'] = res_df['peak_id'].str.split(':', expand=True)[0]\n",
    "location_cols = res_df['peak_id'].str.split(':', expand=True)[1]\n",
    "res_df[['start','stop']] = location_cols.str.split('-', expand=True)\n",
    "# Add columns corresponding to the TFBS\n",
    "res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "# Select relevant columns\n",
    "res_df = res_df[['cage_peak', 'chrom','start','stop','motif_loc',\n",
    "                 'jaspar_tfbs','TF','score','strand','motif']]\n",
    "# Store list of TFs associated with the current CAGE peak in dict\n",
    "tfs_associated_to_cage_peaks[cage_peak] = np.sort(np.unique(res_df['TF'].values))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Find TFs targeting genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Construct dataframe with gene-tf pairs\n",
    "tf_and_genes_df = genes_per_peak.copy()\n",
    "tf_and_genes_df['tfs'] = [tfs_associated_to_cage_peaks[cage_peak] if cage_peak in tfs_associated_to_cage_peaks.keys()\n",
    "                          else ['ToDo']\n",
    "                          for cage_peak in tf_and_genes_df['short_description'].values]\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['genes'])\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['tfs'])\n",
    "tf_and_genes_df = tf_and_genes_df[['genes','tfs','short_description','chrom','cage_start','cage_end','strand']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ucsc_genes</th>\n",
       "      <th>hgnc_gene</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uc002lga.3</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>ToDo</td>\n",
       "      <td>p101@TCF4</td>\n",
       "      <td>chr18</td>\n",
       "      <td>53303180</td>\n",
       "      <td>53303190</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uc021ukp.1</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>ToDo</td>\n",
       "      <td>p101@TCF4</td>\n",
       "      <td>chr18</td>\n",
       "      <td>53303180</td>\n",
       "      <td>53303190</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uc002lga.3</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>ToDo</td>\n",
       "      <td>p101@TCF4</td>\n",
       "      <td>chr18</td>\n",
       "      <td>53303180</td>\n",
       "      <td>53303190</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uc021ukp.1</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>ToDo</td>\n",
       "      <td>p101@TCF4</td>\n",
       "      <td>chr18</td>\n",
       "      <td>53303180</td>\n",
       "      <td>53303190</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uc011avz.1</td>\n",
       "      <td>ANKRD28</td>\n",
       "      <td>ToDo</td>\n",
       "      <td>p10@ANKRD28</td>\n",
       "      <td>chr3</td>\n",
       "      <td>15839693</td>\n",
       "      <td>15839707</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ucsc_genes hgnc_gene   tfs short_description  chrom  cage_start  cage_end  \\\n",
       "0  uc002lga.3      TCF4  ToDo         p101@TCF4  chr18    53303180  53303190   \n",
       "0  uc021ukp.1      TCF4  ToDo         p101@TCF4  chr18    53303180  53303190   \n",
       "0  uc002lga.3      TCF4  ToDo         p101@TCF4  chr18    53303180  53303190   \n",
       "0  uc021ukp.1      TCF4  ToDo         p101@TCF4  chr18    53303180  53303190   \n",
       "1  uc011avz.1   ANKRD28  ToDo       p10@ANKRD28   chr3    15839693  15839707   \n",
       "\n",
       "  strand  \n",
       "0      -  \n",
       "0      -  \n",
       "0      -  \n",
       "0      -  \n",
       "1      -  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "tf_and_genes_df.sort_values('short_description').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
