{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'process_jaspar_with_strand' from '/home/louiscam/projects/gpcr/code/JASPAR_processing/process_jaspar_with_strand.py'>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import itertools\n",
    "from itertools import groupby\n",
    "import os.path\n",
    "import math\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import MOODS.parsers\n",
    "import MOODS.tools\n",
    "import MOODS.scan\n",
    "import subprocess\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import process_jaspar_with_strand as pj\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(pj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "prom_hg19_seq_dir = dir_genome+'prom_hg19_seq_dir/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data_dir/'\n",
    "# Directory for storing preliminary results\n",
    "prelim_results_dir = '/home/louiscam/projects/gpcr/save/prelim_results_dir/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "# Directory of tf data\n",
    "tf_dir = '/home/louiscam/projects/gpcr/data/tf_data/'\n",
    "jaspar_dir = tf_dir+'jaspar_data/'\n",
    "pfm_dir = tf_dir+'pfm_data/'\n",
    "fantom5_dir = '/home/louiscam/projects/gpcr/data/fantom5/'\n",
    "moods_out_dir = tf_dir+'moods_cage_outdir_with_strand/'\n",
    "cage_fasta_dir = dir_genome+'cage_hg19_seq_dir_with_strand/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Select genes of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genes of interest (active adhesome genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes of interest = 219\n"
     ]
    }
   ],
   "source": [
    "# Genes of interest and location\n",
    "selected_genes = pickle.load(open(saving_dir+'active_adhesome_genes.pkl', 'rb'))\n",
    "print('Number of genes of interest = '+str(len(selected_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping UCSC to HGNC\n",
    "ucsc_to_hgnc = pj.create_ucsc_hgnc_dict(dir_genome+'chrom_hg19.name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load human TFs list\n",
    "from http://humantfs.ccbr.utoronto.ca/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of TFs = 1639\n"
     ]
    }
   ],
   "source": [
    "# Load TFs\n",
    "tf_df = pd.read_csv(dir_genome+'human_tf.database', sep='\\t', header=0, index_col=0)\n",
    "tf_df = tf_df[tf_df['Is TF?']=='Yes'].sort_values(by=['HGNC symbol'])\n",
    "all_tfs = tf_df['HGNC symbol'].values\n",
    "print('Total number of TFs = '+str(len(all_tfs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>isTF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46268</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13550</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70810</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41777</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33033</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  chrom strand    txStart      txEnd  txLength   isTF\n",
       "46268  ABI1  chr10      -   27035524   27150016    114492  False\n",
       "13550  ABI2   chr2      +  204193002  204296892    103890  False\n",
       "70810  ABI3  chr17      +   47287588   47300587     12999  False\n",
       "41777  ABL1   chr9      +  133589267  133763062    173795  False\n",
       "33033  ACTB   chr7      -    5566778    5570232      3454  False"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load gene locations\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc'\n",
    "df_loc = pd.read_csv(gene_locations_filename, sep = '\\t', header = 0)\n",
    "df_loc.columns = ['transcript','chrom','strand','txStart','txEnd','cdsStart','cdsEnd',\n",
    "                  'exonCount','exonStarts','exonEnds','proteinID','alignID']\n",
    "df_loc['gene'] = [ucsc_to_hgnc[t] for t in df_loc['transcript']]\n",
    "# Pick largest extent of each gene\n",
    "min_extent_gene = df_loc.groupby(['gene'], sort=False)['txStart'].min().rename('tx_minStart')\n",
    "max_extent_gene = df_loc.groupby(['gene'], sort=False)['txEnd'].max().rename('tx_maxEnd')\n",
    "df_loc = pd.merge(df_loc, min_extent_gene, on='gene')\n",
    "df_loc = pd.merge(df_loc, max_extent_gene, on='gene')\n",
    "df_loc = df_loc[['transcript','gene','chrom','strand',\n",
    "                 'tx_minStart','tx_maxEnd']]\n",
    "# Determine the strand of the gene (in case of labeling error)\n",
    "df_loc = pd.merge(df_loc, \n",
    "                  df_loc.groupby(['gene','chrom','tx_minStart','tx_maxEnd'])['strand'].count().rename('strand_vote'),\n",
    "                  on=['gene','chrom','tx_minStart','tx_maxEnd'])\n",
    "df_loc = df_loc.sort_values(by=['gene','strand_vote'])\n",
    "df_loc = df_loc.drop_duplicates(subset=['gene','chrom','tx_minStart','tx_maxEnd'], keep='first')\n",
    "df_loc['txLength'] = df_loc['tx_maxEnd']-df_loc['tx_minStart']\n",
    "df_loc = df_loc[['gene','chrom','strand','tx_minStart','tx_maxEnd','txLength']]\n",
    "df_loc.columns = ['gene','chrom','strand','txStart','txEnd','txLength']\n",
    "# Restrict to genes of interest\n",
    "df_loc = df_loc[df_loc['gene'].isin(selected_genes)]\n",
    "df_loc['isTF'] = [df_loc.iloc[i,0] in all_tfs for i in range(df_loc.shape[0])]\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adhesome gene lengths quartile = [ 26394.  66588. 149211.]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdzklEQVR4nO3deZwdVZn/8c+XAAGEIWBaJEDSCBhBdhoUQUVAJzAIzghCHNSwGMVhU1zC4AzID8ddxAUxLoQBBEZEiaAsIsiaCIEQIMDAkAhBlrAnoJCE5/fHOS3FpW93pbvr3nTX9/163VfXes5zqvs+XfdU3VOKCMzMrD5WancAZmbWWk78ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePE32aS5kvas8m6d0q6t9UxrUiUnCnpaUl/GoTypkk6JU/vJmnBwKO0IknXSDq8TXWHpE3bUfdQ4sS/AouI6yJifF/bSTpJ0jmtiKkNdgXeC2wYETu1OxhbcbTzH8xQ58RvvZK0cptDGAfMj4jn2xyH2bDhxL9i2FbSHEnPSrpA0mrw2q4ISV+Q9LCkRZLulbSHpAnAvwMHSlos6fa87RhJ0yU9Jel+SR8vlLO6pLNy98ndkj7fUM/8XNcc4HlJK0uaIun/ct1zJf1zYftJkm6QdKqkZyQ9IOkdeflDkh6X9LFmjW8Wq6TDgJ8AO+e2famHfTeR9AdJT0p6QtK5kkYV1m8n6dYc9wXAaj2UcVyO8RFJhxSWj5T0TUkPSnpM0hmSVs/rRku6JLf3KUnXSVopr9s8n40+I+kuSfsWypwm6XRJv8ttukHSGyV9J/8+7pG0XcOx+aWkhZLmSTq6l+P4ekm/kfScpJslnSLp+sL6t0i6Msd7r6QPNcT1A0mX5mM1U9ImZfbti6RD89/Z05IulzSusC4kfVLSffl4/UCS8roRkr6Vf6/zJB2Zt19Z0peBdwLfz8fx+4Uq92xS3qaS/qj0Pnsi/z3UU0T41cYXMB/4EzAGWBe4G/hkXrcbsCBPjwceAsbk+U5gkzx9EnBOQ7nXAqeTEt22wEJg97zuq8AfgXWADYE53fUUYpoNbASsnpcdkGNcCTgQeB5YP6+bBCwFDgFGAKcADwI/AEYC7wMWAWs2OQa9xToJuL6X47cpqStoJNCRy/pOXrcq8Gfg08AqwP7AEuCUwvFdCpyc1+8NvACsk9efCkzPv5e1gN8AX8nrvgKckfdbhZSElKfvJ/0zXhXYPbd9fN5vGvAEsENu7x+AecBHC8fu6rztSsAs4D9zWW8CHgD+scmxOD+/1gC2IP29XJ/XvS7PHwKsDGyX49iiENeTwE55/bnA+WX27SGOa4DD8/R++Xhsnvf9InBjYdsALgFGAWPz735CXvdJYC7pb3Qd4Pd5+5Ub6ylZ3nnACfm4rgbs2u73f9vyTrsDqPuLlGQPLsx/HTgjT+/GK4l/U+BxYE9glYYyTqKQ+EkJexmwVmHZV4BpefpVyQM4nNcm/kP7iHs2sF+engTcV1i3VX4DrldY9iSwbQ/l9BXrJHpJ/D2U9wHgtjz9LuAvgArrb+TVif+v3YkkL3sceDspiT9P/uea1+0MzMvTJwMXA5s21P9O4FFgpcKy84CT8vQ04MeFdUcBdzccu2fy9NuABxvKPx44s4d2jyD9UxtfWHYKryT+A4HrGvb5EXBiIa6fFNbtDdxTZt8eYrmGVxL/74DDCutWIv1zHZfng0ICBv4HmJKn/wB8orBuT8ol/mbl/TcwlXS9qO3v/Xa+3NWzYni0MP0CsGbjBhFxP3AsKck/Lul8SWOalDcGeCoiFhWW/RnYoLD+ocK64nSPyyR9VNLs/PH5GWBLYHRhk8cK03/NMTcue027SsTaK0nr5WPxsKTngHMKcY0BHo78ri+UXfRkRCwtzHcf/w7SmfOsQpsvy8sBvkE6k71CqWtrSqHOhyLi5V7a03hcmh2nccCY7vpzDP8OrNfDoeggnVE3+72OA97WUNa/Am8sbNPs77DMvs2MA04r7PcU6Z9q8Xg0q7fM32lPmpX3+Vz3n3IX3KElyxt2nPiHkIj4eUTsSnozBfC17lUNm/4FWFfSWoVlY4GH8/QjpI/P3Tbqqbruidwn+2PgSOD1ETEKuJP0JhqovmLty3/lWLeKiH8ADi7E9QiwQXcfb6HsMp4gJeG3RsSo/Fo7ItYEiIhFEXFcRLwJ2Bf4jKQ9cns26u7v70d7ih4ifcIYVXitFRF797DtQlK3VbPf60PAHxvKWjMijigZx0D2/UTDvqtHxI0l9u3r73S5hhaOiEcj4uMRMQb4BHC6anrrpxP/ECFpvKTdJY0E/kZKSt1nlY8Bnd3JJiIeInVpfEXSapK2Bg4jnQ1D+vh7vKR1JG1ASui9eR3pTbYwx3II6Yx/wErE2pe1gMXAs7ktnyusu4mUDI+WtIqkfyH1YZeJ62XSP7tTJb0BQNIGkv4xT++TLxYKeJbUXfUyMJN0lvn5XOduwPtJfe/L60/AIqUL7avni51bStqxh3iXARcBJ0laQ9JbSNcNul0CvFnSR3Jcq0jaUdLmJeIYyL5nkP7W3gogaW1JB5TYD9Lf6TH5uI8CvtCw/jHSdY9SJB0gqfsfydOkv+mXe9ll2HLiHzpGki7KPkH6KPsGUn8vwC/yzycl3ZqnJ5IuAP8F+BWpP/b3ed3JwALSRcXfAxcCLzarOCLmAt8iJdLHSP3QNwxGo0rE2pcvAduTku+lpOQHQES8BPwL6TrBU6S+6oteW0RTXyB158zI3Ui/J11kB9gszy8mHZfTI+LqXOf7gb1Iv6vTgY9GxD3LUW93/MuAfUgXvOfl8n4CrN1klyPzukeBs0nXFl7MZS0iXWQ/iHScHyV9YhxZIo6B7PurvO35+RjeSTo2ZfwYuIJ088FtwG9J/8iX5fWnAfvnu4W+W6K8HYGZkhaTLtofExEPlIxlWNGruz+tjiQdARwUEe9udyw2eCR9DXhjRDS9lXYokbQX6caHcX1ubL3yGX8NSVpf0i6SVpI0HjiOdKZtQ5jSvfZbK9mJ1GU2ZH+vuXtr73zf/gbAiQzh9qxInPjraVXS7XiLSLfMXUzqkrChbS1SV9bzwAWk7rmL2xrRwIjUlfc0qavnbtJ3GmyA3NVjZlYzPuM3M6uZdg/AVcro0aOjs7Oz3WGYmQ0ps2bNeiIiOhqXD4nE39nZyS233NLuMMzMhhRJjd9UB9zVY2ZWO078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNVNZ4pf0M6XnmN7Zw7rjlJ6dObqnfc3MrDpVnvFPAyY0LpS0EWmI1wcrrNvMzJqoLPFHxLWkMdAbnUp6BJoHCTIza4OW9vFL2o/0DNTbS2w7WdItkm5ZuHDhgOvunHJpj9NmZnXTssQvaQ3Sg6JLDasaEVMjoisiujo6XjPUhJmZ9VMrz/g3ATYGbpc0n/QQ5VslvbGFMZiZ1V7LBmmLiDtIz4kFICf/roh4olUxmJlZtbdznkd6CPV4SQskHVZVXWZmVl5lZ/wRMbGP9Z1V1W1mZs35m7tmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjOVJX5JP5P0uKQ7C8u+IekeSXMk/UrSqKrqNzOznlV5xj8NmNCw7Epgy4jYGvhf4PgK6zczsx5Ulvgj4lrgqYZlV0TE0jw7A9iwqvrNzKxn7ezjPxT4XbOVkiZLukXSLQsXLmxhWGZmw1tbEr+kE4ClwLnNtomIqRHRFRFdHR0drQvOzGyYW7nVFUqaBOwD7BER0er6zczqrqWJX9IE4PPAuyPihVbWbWZmSZW3c54H3ASMl7RA0mHA94G1gCslzZZ0RlX1m5lZzyo744+IiT0s/mlV9ZmZWTn+5q6ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzdQ+8XdOubTHaTOz4ar2id/MrG6c+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmKkv8kn4m6XFJdxaWrSvpSkn35Z/rVFW/mZn1rM/EL2kTSSPz9G6SjpY0qkTZ04AJDcumAFdFxGbAVXnezMxaqMwZ/y+BZZI2BaYCGwE/72uniLgWeKph8X7AWXn6LOADpSM1M7NBsXKJbV6OiKWS/hn4XkR8T9Jt/axvvYh4JE8/CqzXbENJk4HJAGPHju1ndeWGWvZwzGZWJ2XO+JdImgh8DLgkL1tloBVHRADRy/qpEdEVEV0dHR0Drc7MzLIyif8QYGfgyxExT9LGwNn9rO8xSesD5J+P97McMzPrpz4Tf0TMBb4A3Jrn50XE1/pZ33TSJwfyz4v7WY6ZmfVTmbt63g/MBi7L89tKml5iv/OAm4DxkhZIOgz4KvBeSfcBe+Z5MzNroTIXd08CdgKuAYiI2ZLe1NdOETGxyao9ygZnZmaDr9TF3Yh4tmHZy1UEY2Zm1Stzxn+XpA8DIyRtBhwN3FhtWGZmVpUyZ/xHAW8FXgTOA54Djq0wJjMzq1CfZ/wR8QJwQn6ZmdkQ12fil/Rm4LNAZ3H7iNi9urDMzKwqZfr4fwGcAfwEWFZtOGZmVrUyiX9pRPyw8kjMzKwlylzc/Y2kT0laP4+nv66kdSuPzMzMKlHmjL97iIXPFZYF0OeXuMzMbMVT5q6ejVsRSKt5KGYzq6syY/WsIemLkqbm+c0k7VN9aGZmVoUyffxnAi8B78jzDwOnVBaRmZlVqkzi3yQivg4sgb9/oUuVRmVmZpUpk/hfkrQ6+WlZkjYhDd9gZmZDUJm7ek4kjcW/kaRzgV2ASVUGZWZm1SlzV8+Vkm4F3k7q4jkmIp6oPDIzM6tEmbF6ts+Tj+SfYyWtDfw5IpZWFpmZmVWiTFfP6cD2wBzSGf+WwF3A2pKOiIgrKozPzMwGWZmLu38BtouIrojYAdgOeAB4L/D1KoMzM7PBVybxvzki7uqeiYi5wFsi4oHqwjIzs6qUffTiD4Hz8/yBwFxJI8n39i8vSZ8GDifdInoHcEhE/K0/ZZmZ2fIpc8Y/Cbif9LjFY0ndPJNISf89y1uhpA1Iz+3tiogtgRHAQctbjpmZ9U+Z2zn/CnwrvxotHkC9q0taAqxBuo5gZmYtUKarZ1BFxMOSvgk8CPwVuKKnO4MkTQYmA4wdO3ZQ6vaInGZm5bp6BpWkdYD9gI2BMcDrJB3cuF1ETM13EnV1dHS0Okwzs2GraeKXdHb+ecwg17knMC8iFkbEEuAiXhn508zMKtbbGf8OksYAh0pap/jYxQE+evFB4O15nH8BewB3D6A8MzNbDr318Z8BXEV6xOIsXj0Uc78fvRgRMyVdCNwKLAVuA6b2pywzM1t+TRN/RHwX+K6kH0bEEYNZaUScSBr108zMWqzM7ZxHSNoGeGdedG1EzKk2LDMzq0qZZ+4eDZwLvCG/zpV0VNWBmZlZNcrcx3848LaIeB5A0teAm4DvVRmYmZlVo8x9/AKWFeaX4WfumpkNWWXO+M8EZkr6VZ7/APDTyiIyM7NKlbm4+21J1wC75kWHRMRtlUZlZmaVKTVWT0TcSrrv3szMhriWj9VjZmbt5cRvZlYzvSZ+SSMkXd2qYFYkHsLZzIarXhN/RCwDXpa0doviMTOzipW5uLsYuEPSlcDz3Qsj4ujKojIzs8qUSfwX5ZeZmQ0DZe7jP0vS6sDYiLi3BTGZmVmFygzS9n5gNnBZnt9W0vSK4zIzs4qUuZ3zJGAn4BmAiJhNPx/CYmZm7Vcm8S+JiGcblr1cRTBmZla9Mhd375L0YWCEpM2Ao4Ebqw3LzMyqUuaM/yjgrcCLwHnAc8CxFcZkZmYVKnNXzwvACfkBLBERi6oPy8zMqlLmrp4dJd0BzCF9ket2STtUH5qZmVWhTFfPT4FPRURnRHQC/0Z6OEu/SRol6UJJ90i6W9LOAynPzMzKK3Nxd1lEXNc9ExHXS1o6wHpPAy6LiP0lrQqsMcDyzMyspKaJX9L2efKPkn5EurAbwIHANf2tMA/49i5gEkBEvAS81N/yzMxs+fR2xv+thvkTC9MxgDo3BhYCZ0raBpgFHBMRzxc3kjQZmAwwduzYAVTXf41DM8//6j+9al33fHG6WRnN1puZtVrTxB8R76mwzu2BoyJipqTTgCnAfzTUPxWYCtDV1TWQfzRmZlbQZx+/pFHAR4HO4vYDGJZ5AbAgImbm+QtJid/MzFqgzMXd3wIzgDsYhKEaIuJRSQ9JGp9H+9wDmDvQcs3MrJwyiX+1iPjMINd7FHBuvqPnAeCQQS7fzMyaKJP4z5b0ceAS0rANAETEU/2tNI/w2dXf/c3MrP/KJP6XgG8AJ/DK3TyBh2Y2MxuSyiT+44BNI+KJqoMxM7PqlRmy4X7ghaoDMTOz1ihzxv88MFvS1by6j7+/t3OamVkblUn8v84vMzMbBsqMx39WKwIxM7PWKPPN3Xn0MDZPRPiuHjOzIahMV0/xfvvVgAOAdasJx8zMqtbnXT0R8WTh9XBEfAfwUJNmZkNUmUcvbl94dUn6JOU+KdRScSjnxmGd+1puZtYKZRJ4cVz+pcB84EOVRGNmZpUrc1dPVePym5lZG5S5q2ck8EFeOx7/ydWFZWZmVSnT1XMx8CzpEYkv9rGtmZmt4Mok/g0jYkLlkZiZWUuUGaTtRklbVR6JmZm1RJkz/l2BSfkbvC8CAiIitq40MjMzq0SZxL9X5VGYmVnLlLmd88+tCMTMzFqjTB+/mZkNI21L/JJGSLpN0iXtisHMrI7aecZ/DHB3G+s3M6ultiR+SRuSRvj8STvqNzOrs3ad8X8H+DzwcpvqNzOrrZYPryxpH+DxiJglabdetpsMTAYYO3Zsa4JbTmWGYB5o2fO/+k+llpuZldWOM/5dgH0lzQfOB3aXdE7jRhExNSK6IqKro6Oj1TGamQ1bLU/8EXF8RGwYEZ3AQcAfIuLgVsdhZlZXvo/fzKxm2voIxYi4BrimnTGYmdWNz/jNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3Maqat39xdEfU2yuZgj8A5mDxqp5mV5TN+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmml54pe0kaSrJc2VdJekY1odg5lZnbVjkLalwHERcauktYBZkq6MiLltiMXMrHZafsYfEY9ExK15ehFwN7BBq+MwM6urtg7LLKkT2A6Y2cO6ycBkgLFjx7Y2sAHqafjmzimXvmrI5OI2jUMpF9ct71DQjeUO9nDNrR7+ubfj1N/yPHS11V3bLu5KWhP4JXBsRDzXuD4ipkZEV0R0dXR0tD5AM7Nhqi2JX9IqpKR/bkRc1I4YzMzqqh139Qj4KXB3RHy71fWbmdVdO874dwE+AuwuaXZ+7d2GOMzMaqnlF3cj4npAra7XzMwSf3PXzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+Fuoc8qlTYdsLr4GWl7Z/Rr37au8/mzfV3z9bUOzfZenvIHsW6a8vpb3t7zBVnU9Az2uNvic+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGqmLYlf0gRJ90q6X9KUdsRgZlZXLU/8kkYAPwD2ArYAJkraotVxmJnVVTvO+HcC7o+IByLiJeB8YL82xGFmVkuKiNZWKO0PTIiIw/P8R4C3RcSRDdtNBibn2fHAvf2scjTwRD/3Harq2GaoZ7vd5nrob5vHRURH48KVBx5PNSJiKjB1oOVIuiUiugYhpCGjjm2Gerbbba6HwW5zO7p6HgY2KsxvmJeZmVkLtCPx3wxsJmljSasCBwHT2xCHmVkttbyrJyKWSjoSuBwYAfwsIu6qsMoBdxcNQXVsM9Sz3W5zPQxqm1t+cdfMzNrL39w1M6sZJ34zs5oZNom/r2EgJI2UdEFeP1NSZxvCHFQl2vwZSXMlzZF0laRx7YhzMJUd7kPSByWFpCF/21+ZNkv6UP5d3yXp562OsQol/r7HSrpa0m35b3zvdsQ5WCT9TNLjku5ssl6SvpuPxxxJ2/e7sogY8i/SReL/A94ErArcDmzRsM2ngDPy9EHABe2OuwVtfg+wRp4+og5tztutBVwLzAC62h13C37PmwG3Aevk+Te0O+4WtXsqcESe3gKY3+64B9jmdwHbA3c2Wb838DtAwNuBmf2ta7ic8ZcZBmI/4Kw8fSGwhyS1MMbB1mebI+LqiHghz84gfWdiKCs73Mf/A74G/K2VwVWkTJs/DvwgIp4GiIjHWxxjFcq0O4B/yNNrA39pYXyDLiKuBZ7qZZP9gP+OZAYwStL6/alruCT+DYCHCvML8rIet4mIpcCzwOtbEl01yrS56DDS2cJQ1meb88ffjSLi0lYGVqEyv+c3A2+WdIOkGZImtCy66pRp90nAwZIWAL8FjmpNaG2zvO/5plbYIRts8Eg6GOgC3t3uWKokaSXg28CkNofSaiuTunt2I32qu1bSVhHxTDuDaoGJwLSI+JaknYGzJW0ZES+3O7AV3XA54y8zDMTft5G0Mumj4ZMtia4apYa+kLQncAKwb0S82KLYqtJXm9cCtgSukTSf1A86fYhf4C3ze14ATI+IJRExD/hf0j+CoaxMuw8D/gcgIm4CViMNZjZcDdpwN8Ml8ZcZBmI68LE8vT/wh8hXTIaoPtssaTvgR6SkPxz6fXttc0Q8GxGjI6IzIjpJ1zX2jYhb2hPuoCjzt/1r0tk+kkaTun4eaGGMVSjT7geBPQAkbU5K/AtbGmVrTQc+mu/ueTvwbEQ80p+ChkVXTzQZBkLSycAtETEd+Cnpo+D9pAsoB7Uv4oEr2eZvAGsCv8jXsR+MiH3bFvQAlWzzsFKyzZcD75M0F1gGfC4ihvKn2bLtPg74saRPky70ThrKJ3OSziP9Ax+dr1ucCKwCEBFnkK5j7A3cD7wAHNLvuobwcTIzs34YLl09ZmZWkhO/mVnNOPGbmdWME7+ZWc048ZuZrWD6GrCth+2Xa5A+J36zfpI0P983P5hldkr6cGF+kqTvD2YdNiRMA0oNvSFpM+B4YJeIeCtwbF/7OPGbrVg6gQ/3tZENbz0N2CZpE0mXSZol6TpJb8mrlnuQPid+GzYk/Ucev/16SedJ+mxe3uMbRtK0PL75jZIekLR/oazPSbo5j3v+pRJ1HyzpT5JmS/qRpBF5+WJJX5Z0ex5Abb1CTDMk3SHpFEmLc1FfBd6Zy/l0XjYmx3+fpK8P4iGzoWUqcFRE7AB8Fjg9L1/uQfqc+G1YkLQj8EFgG2Av0qB03Zq9YQDWB3YF9iElXSS9jzTWzU7AtsAOkt7VS92bAweSPmpvS/r27L/m1a8DZkTENqRnBHw8Lz8NOC0itiKNtdNtCnBdRGwbEafmZdvm8rcCDpRUHK/FakDSmsA7SN/Cn00aiqV7SObiIH0TSd9mHtVbecNiyAYzYBfg4oj4G/A3Sb+B17xhurcdWdjv13k0x7ndZ+PA+/Lrtjy/JumNdW2TuvcAdgBuznWsDnR/3H4JuCRPzwLem6d3Bj6Qp38OfLOXtl0VEc/m9swFxvHq4Xlt+FsJeCafWDRaQHooyxJgnqTuQfpublaYE78Nd729YQCKI5aq8PMrEfGjknUIOCsiju9h3ZLC+DHL6N97rhhjf8uwISwinpM0T9IBEfELpTOMrSPidtIgfROBM8sO0ueuHhsubgDeL2m1fJa/D6Q3DOks6AD4+3NLt+mjrMuBQ3M5SNpA0ht62f4qYP/ubSStq76fbzyD1DUFrx4wcBFpeGmrsTxg203AeEkLJB1G6j48TNLtwF288kSyy4En86fBqykxSJ/PHGxYiIibJU0H5gCPAXeQnrIG6Q3zQ0lfJI12eD7pGa7Nyroi99vflLtuFgMH80r3TeP2c3PZVyg9DGYJ8G/An3sJ+VjgHEknAJcVYp0DLMtv7mnA07233IajiJjYZNVrLtzmT5Sfya9SPDqnDRuS1oyIxZLWIPXHT46IW9sdV09yjH+NiJB0EDAxInp6frDZoPMZvw0nUyVtQXogx1kratLPdgC+n/tqnwEObW84Vic+4zczqxlf3DUzqxknfjOzmnHiNzOrGSd+M7OaceI3M6uZ/w9b/nA9VAkLFwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize distribution of gene lengths\n",
    "print('adhesome gene lengths quartile = '+str(np.quantile(df_loc['txLength'].values, [0.25,0.5,0.75])))\n",
    "plt.figure()\n",
    "plt.hist(df_loc['txLength'], bins=np.arange(0,1000000,5000))\n",
    "plt.xlabel('gene length')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of adhesome gene lengths')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define gene window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define region 1kb upstream of transcripts\n",
    "df_loc['gene_window_start'] = [df_loc['txStart'].iloc[i]-1000 if (df_loc['strand'].iloc[i]=='+') \n",
    "                             else max(df_loc['txStart'].iloc[i],df_loc['txStart'].iloc[i]-5000) for i in range(df_loc.shape[0])]\n",
    "df_loc['gene_window_end'] = [min(df_loc['txEnd'].iloc[i], df_loc['txStart'].iloc[i]+5000) \n",
    "                             if (df_loc['strand'].iloc[i]=='+') \n",
    "                             else \n",
    "                             df_loc['txEnd'].iloc[i]+1000 \n",
    "                             for i in range(df_loc.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>isTF</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46268</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>False</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27151016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13550</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>False</td>\n",
       "      <td>204192002</td>\n",
       "      <td>204198002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70810</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>False</td>\n",
       "      <td>47286588</td>\n",
       "      <td>47292588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41777</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>False</td>\n",
       "      <td>133588267</td>\n",
       "      <td>133594267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33033</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>False</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5571232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  chrom strand    txStart      txEnd  txLength   isTF  \\\n",
       "46268  ABI1  chr10      -   27035524   27150016    114492  False   \n",
       "13550  ABI2   chr2      +  204193002  204296892    103890  False   \n",
       "70810  ABI3  chr17      +   47287588   47300587     12999  False   \n",
       "41777  ABL1   chr9      +  133589267  133763062    173795  False   \n",
       "33033  ACTB   chr7      -    5566778    5570232      3454  False   \n",
       "\n",
       "       gene_window_start  gene_window_end  \n",
       "46268           27035524         27151016  \n",
       "13550          204192002        204198002  \n",
       "70810           47286588         47292588  \n",
       "41777          133588267        133594267  \n",
       "33033            5566778          5571232  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Select CAGE peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process robust CAGE peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique CAGE peaks = 201802\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>gene</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95563</th>\n",
       "      <td>p1@MTND1P23</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>129bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564571</td>\n",
       "      <td>564600</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95564</th>\n",
       "      <td>p3@MTND1P23</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>197bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564639</td>\n",
       "      <td>564649</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95565</th>\n",
       "      <td>p3@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>246bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565266</td>\n",
       "      <td>565278</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95566</th>\n",
       "      <td>p4@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>458bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565478</td>\n",
       "      <td>565483</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95567</th>\n",
       "      <td>p1@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>489bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565509</td>\n",
       "      <td>565541</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      short_description      gene    association_with_transcript chrom  \\\n",
       "95563       p1@MTND1P23  MTND1P23  129bp_to_ENST00000416931_5end  chr1   \n",
       "95564       p3@MTND1P23  MTND1P23  197bp_to_ENST00000416931_5end  chr1   \n",
       "95565       p3@MTND2P28  MTND2P28  246bp_to_ENST00000457540_5end  chr1   \n",
       "95566       p4@MTND2P28  MTND2P28  458bp_to_ENST00000457540_5end  chr1   \n",
       "95567       p1@MTND2P28  MTND2P28  489bp_to_ENST00000457540_5end  chr1   \n",
       "\n",
       "       cage_start  cage_end strand peak_id  \n",
       "95563      564571    564600      +       1  \n",
       "95564      564639    564649      +       3  \n",
       "95565      565266    565278      +       3  \n",
       "95566      565478    565483      +       4  \n",
       "95567      565509    565541      +       1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load annotated FANTOM5 CAGE peaks\n",
    "cage_df = pd.read_csv(fantom5_dir+'hg19.cage_peak_phase1and2combined_ann.txt', sep='\\t', skiprows=7, header=0)\n",
    "# Define new columns\n",
    "cage_df['gene'] = cage_df['short_description'].str.split('@',expand=True)[1]\n",
    "cage_df['chrom'] = cage_df['00Annotation'].str.split(':', expand=True)[0]\n",
    "cage_df['cage_start'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[0].astype(int)\n",
    "cage_df['cage_end'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[2].astype(int)\n",
    "cage_df['strand'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[1]\n",
    "cage_df['peak_id'] = cage_df['short_description'].str.split('@', expand=True)[0].str.strip('p')\n",
    "# Reformat dataframe\n",
    "cage_df = cage_df[['short_description', 'gene','association_with_transcript',\n",
    "                   'chrom','cage_start','cage_end','strand','peak_id']]\n",
    "cage_df = cage_df.sort_values(by=['chrom','cage_start'])\n",
    "print('Number of unique CAGE peaks = '+str(len(np.unique(cage_df['short_description'].values))))\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>strand_gene</th>\n",
       "      <th>strand_prom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>CALR</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene strand_gene strand_prom\n",
       "279  CALR           +           -"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check which genes are located on the opposite strand of their promoter (mislabeling error)\n",
    "# The only problematic error was on PDPK1, which has been taken care of before\n",
    "merge_df = pd.merge(df_loc[['gene','strand']],cage_df[['gene','strand']], on=['gene'], suffixes=['_gene','_prom'])\n",
    "merge_df[merge_df['strand_gene'] != merge_df['strand_prom']].drop_duplicates(subset=['gene'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CAGE peaks within selected gene windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bed file for regions upstream of genes of interest\n",
    "gene_window_df = df_loc[['chrom','gene_window_start','gene_window_end','gene','strand']]\n",
    "gene_window_df['score'] = 0\n",
    "gene_window_df = gene_window_df[['chrom','gene_window_start','gene_window_end','gene','score','strand']]\n",
    "gene_window_bed = pybedtools.BedTool.from_dataframe(gene_window_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a bed file for robust CAGE peaks\n",
    "cage_peak_df = cage_df[['chrom','cage_start','cage_end','short_description','peak_id','strand']]\n",
    "cage_bed = pybedtools.BedTool.from_dataframe(cage_peak_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get CAGE peaks for each gene of interest\n",
    "out = pybedtools.bedtool.BedTool.map(gene_window_bed, cage_bed, c=4, o='distinct', s=True)\n",
    "out_df = out.to_dataframe()\n",
    "out_df.columns=['chrom','gene_window_start','gene_window_end','gene','score','strand','peaks']\n",
    "out_df = out_df[['chrom','gene_window_start','gene_window_end','gene','strand','peaks']]\n",
    "out_df['peaks'] = out_df['peaks'].str.replace(',p',';p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat matching results\n",
    "peaks_per_gene_df = pd.merge(df_loc, out_df, on=['chrom','gene_window_start','gene_window_end','gene','strand'])\n",
    "peaks_per_gene_df = peaks_per_gene_df[peaks_per_gene_df['peaks'] != '.']\n",
    "peaks_per_gene_df['peaks'] = peaks_per_gene_df['peaks'].str.split(';')\n",
    "peaks_per_gene_df['n_peaks'] = [len(peaks) for peaks in peaks_per_gene_df['peaks']]\n",
    "peaks_per_gene_df\n",
    "peaks_per_gene_df = peaks_per_gene_df[['gene', 'isTF', 'chrom', 'strand', 'txStart', 'txEnd', 'txLength',\n",
    "                                       'gene_window_start', 'gene_window_end','peaks','n_peaks']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isTF</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "      <th>peaks</th>\n",
       "      <th>n_peaks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27151016</td>\n",
       "      <td>[p1@ABI1, p2@ABI1, p3@ABI1, p4@ABI1, p@chr10:2...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>204192002</td>\n",
       "      <td>204198002</td>\n",
       "      <td>[p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>47286588</td>\n",
       "      <td>47292588</td>\n",
       "      <td>[p1@ABI3, p2@ABI3, p3@ABI3, p4@ABI3, p5@ABI3]</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>133588267</td>\n",
       "      <td>133594267</td>\n",
       "      <td>[p@chr9:133588280..133588291,+, p@chr9:1335884...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>False</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5571232</td>\n",
       "      <td>[p10@ACTB, p11@ACTB, p12@ACTB, p13@ACTB, p14@A...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene   isTF  chrom strand    txStart      txEnd  txLength  \\\n",
       "0  ABI1  False  chr10      -   27035524   27150016    114492   \n",
       "1  ABI2  False   chr2      +  204193002  204296892    103890   \n",
       "2  ABI3  False  chr17      +   47287588   47300587     12999   \n",
       "3  ABL1  False   chr9      +  133589267  133763062    173795   \n",
       "4  ACTB  False   chr7      -    5566778    5570232      3454   \n",
       "\n",
       "   gene_window_start  gene_window_end  \\\n",
       "0           27035524         27151016   \n",
       "1          204192002        204198002   \n",
       "2           47286588         47292588   \n",
       "3          133588267        133594267   \n",
       "4            5566778          5571232   \n",
       "\n",
       "                                               peaks  n_peaks  \n",
       "0  [p1@ABI1, p2@ABI1, p3@ABI1, p4@ABI1, p@chr10:2...       10  \n",
       "1  [p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...        6  \n",
       "2      [p1@ABI3, p2@ABI3, p3@ABI3, p4@ABI3, p5@ABI3]        5  \n",
       "3  [p@chr9:133588280..133588291,+, p@chr9:1335884...        3  \n",
       "4  [p10@ACTB, p11@ACTB, p12@ACTB, p13@ACTB, p14@A...       28  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "peaks_per_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAci0lEQVR4nO3debgdVZnv8e+PgCQMEmJCOgzhICAKiAHCYIOXUZtBL3BFERUIIlEUAQHbKHaLNlxBAWlbBcOsQBQZo9AIIsqggAlGpkCDEIQQSEACARRIePuPtQ4UJ2eonHNq77N3/T7Ps59UrZre2lV5z9qrqlYpIjAzs/pYrtkBmJlZYznxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTfwNImiNp1x6mvU/SA42OaShRcp6kZyXd0eRYJkm6pYnbP0zSU5JekPS2Bm73fEknNGp71lzLNzuAuouIm4GN+ppP0vHABhHxycqDarztgfcDa0fEi80OplkkrQCcBmwbEX9udjzWvlzjNyQ1uwKwLjCn3ZJ+P77XscBw4N4KwmkpQ+CcbGtO/I0zQdJdkp6T9DNJwwEk7Sjp8c6ZJH1Z0lxJiyQ9IGkXSbsBXwX2y00Af87zrilpuqS/SXpI0qGF9YyQdEFuPpkt6V+7bGdO3tZdwIuSlpc0RdJf8rbvk7RPYf5Jkm6V9F1JCyU9LOmfc/ljkuZLOqinne8pVkmHAGcD78379o1ulp0k6RZJp+T9eUTS7l32ZdfC+PGSLszDHZJC0sE5zmclfVbSVvl4LJT0/aU3qe/nY3W/pF0KE1aTdI6kefk4nSBpWDff0TPA8d3sy4qSTpf0RP6cnsveAXQ2+S2U9Jtulu3cl8l52XmSji1MX65wDJ+RdImkUYXpP5f0ZN6vmyRt0sOxWlXSjZK+p2SPfD4syvt8bA/Lde5/Vd9dX+f0mpIuk7QgnyNHFKYdn7+PH+f9uFfSxDLLtqWI8KfiDzAHuANYExgFzAY+m6ftCDyehzcCHgPWzOMdwPp5+Hjgwi7rvQn4IamWOAFYAOycp50E/A5YHVgbuKtzO4WYZgHrACNy2UdyjMsB+wEvAuPytEnAYuBgYBhwAvBX4AfAisAHgEXAKj18B73FOgm4pZfvbxLwKnBo3vZhwBOACvuya2H+17+r/B0GcGbe9geAfwBXAmsAawHzgR267OcXgRXy9/AcMCpPvwL4EbByXv4O4DNdlv0CqRl1RDf78k3gtrzsGOD3wH90iXX5Hr6HzunT8vbfnb/HXfP0I/O6187H5EfAtMLynwJWzdNOB2YVpp2fj+nb8j6dUJg2D3hfHl4d2KKX41Tld9fjOU06Z2cC/w68BXg78DDwL4Vz4h/AHqRz6FvAbWWWbcdP0wOow4eUmD5ZGP82cGYe3rFw8m5ASkK7Ait0WcfxFBI/KWEvAVYtlH0LOD8Pv+nEBT7N0on/U33EPQvYKw9PAh4sTHs3KQmNLZQ9A0zoZj19xTqJvhP/Q4XxlfK2/6mwL30l/rW6xLlfYfwy4KjCtl7/o5LL7gAOIDXFvEwhKQH7AzcWlv1rH9/pX4A9CuP/QmrmKsbaV+J/Z5dz6Zw8PBvYpTBtHOkP5lLrA0bmda2Wx88HzgXuAb7UZd6/Ap8B3trHvlX93fV4TgPbdF0e+ApwXuGc+HVh2sbA38ss244ft6M1zpOF4ZdINes3iYiHJB1FOkk3kfQr4OiIeKKb9a0J/C0iFhXKHgUmFqY/VphWHO62TNKBwNGkBAOwCjC6MMtTheG/55i7lq3Sj1jLeP37i4iXJHXGV1bXOHuLe27k//3Zo6R9WJdUk52Xtw+pttjX91y0Zl5f13Uvi+I2HiX9ESbHd4Wk1wrTlwBjJT0JnEj6VTcG6JxnNKlWDrAn8ALp11HRh4GvAScpNQ1OiYg/9BBb1d9dT/OvC6wpaWGhbBhwc2G86//B4UrXEsos21bcxj/ERMTFEbE96WQM4OTOSV1mfQIYJWnVQtl4YG4enkf6Odxpne421zkgaV3gLOBw4G0RMZJU+1M3yy2rvmIdqBdJvwI6/dMA17eWCtmJFOsTpETzMjA6Ikbmz1sjothW3ld3t0+Qjm3XdS+L4rEsLv8YsHshtpERMTwi5gIfB/Yi/ZpcjTf+uBf38yzgWuAaSSt3FkbEHyNiL1LzzJXAJb3EVuV319s5/RjwSJd9XzUi9uhjnQNdtiU58Q8hkjaStLOkFUntkX/njZrZU0CHpOUAIuIxUvvwtyQNl7QZcAhwYZ7/EuArklaXtBYpofdmZdJ/vAU5loOBTQdjv0rEOlCzgI9JWiFfsNt3gOtbAzgir+8jwLuAayJiHnAdcKqkt+aLqetL2mEZ1j0N+JqkMZJGk9qVl/V7+DdJK+WLswcDP8vlZwIn5j/i5G3slaetSkq8z5D+SP7/HtZ9OOki8y/yxdS3SPqEpNUi4lXged44J7tT5XfX2zl9B7BI6YaFEZKGSdpU0lYl1juQZVuSE//QsiLpAtbTpJ+la5DaGgF+nv99RtKdeXh/Us3tCdKFs69HxK/ztG8CjwOPAL8GLiX9x+9WRNwHnAr8gfRH5t3ArYOxUyViHah/A9YHngW+AVw8wPXdDmxIOg4nAvtGxDN52oGkC4D35e1dSmpLL+sEYAbpwuTdwJ25bFn8DngIuAE4JSKuy+X/CUwHrpO0iHShd5s87cekZpe5OfbbultxbqaZTDp3riJdED8AmCPpeeCzwCd6ia3K767HczoilgAfJN048Eje/tmkXze9Gsiyrarzrghrc5IOAz4WEctSw7IhRFIHKTGtEBGLmxzOUiRNAj6dmyobsT2f0/3kGn+bkjRO0nb5J/VGwDGkmrZZS/I5PXh8V0/7egvpnun1gIXAT0n30Zu1Kp/Tg8RNPWZmNVNZU4+kdZQe+74vPx59ZC4/Pj+uPSt/2vaWKTOzoaiyGr+kcaTH/e/M92/PBPYGPgq8EBGnlF3X6NGjo6Ojo5I4zcza1cyZM5+OiDFdyytr48/37c7Lw4skzSb1i7LMOjo6mDFjxmCGZ2bW9iQ92l15Q+7qybehbU66xxfgcKWeEc+VtHoPy0yWNEPSjAULFjQiTDOzWqg88UtahTc6wXoeOIP0sM0E0i+CU7tbLiKmRsTEiJg4ZsxSv1TMzKyfKk38Sm8Uugy4KCIuh9SpV0QsiYjXSH2DbF1lDGZm9mZV3tUj4BxgdkScVigvPqK9D6kjMDMza5AqH+DajtTHx92SZuWyrwL7S5pA6hBsDqmfbzMza5Aq7+q5he679L2mqm2amVnf3FePmVnNOPGbmdWME7+ZWc24d86CjilXvz4856Q9mxiJmVl1XOM3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6uZ5ZsdwFDQMeXqZodgZtYwrvGbmdWME7+ZWc048ZuZ1YwTv5lZzVSW+CWtI+lGSfdJulfSkbl8lKTrJT2Y/129qhjMzGxpVdb4FwPHRMTGwLbA5yVtDEwBboiIDYEb8riZmTVIZYk/IuZFxJ15eBEwG1gL2Au4IM92AbB3VTGYmdnSGtLGL6kD2By4HRgbEfPypCeBsT0sM1nSDEkzFixY0IgwzcxqofLEL2kV4DLgqIh4vjgtIgKI7paLiKkRMTEiJo4ZM6bqMM3MaqPSxC9pBVLSvygiLs/FT0kal6ePA+ZXGYOZmb1ZlXf1CDgHmB0RpxUmTQcOysMHAVdVFYOZmS2tyr56tgMOAO6WNCuXfRU4CbhE0iHAo8BHK4zBzMy6qCzxR8QtgHqYvEtV2zUzs975yV0zs5px4jczqxknfjOzmvGLWEoovqhlzkl7NjESM7OBc43fzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxm+kz8ktaXtGIe3lHSEZJGVh6ZmZlVokyN/zJgiaQNgKnAOsDFlUZlZmaVKZP4X4uIxcA+wH9FxJeAcdWGZWZmVSnzIpZXJe0PHAR8KJetUF1IjVF8uYqZWZ2UqfEfDLwXODEiHpG0HvCTasMyM7Oq9Fnjj4j7JH0ZGJ/HHwFOrjowMzOrRpm7ej4EzAKuzeMTJE2vOC4zM6tImaae44GtgYUAETELeHtlEZmZWaXKJP5XI+K5LmWvVRGMmZlVr8xdPfdK+jgwTNKGwBHA76sNy8zMqlKmxv8FYBPgZWAa8DxwVIUxmZlZhcrc1fMScFz+mJlZi+sz8Ut6B3As0FGcPyJ2ri4sMzOrSpk2/p8DZwJnA0uqDcfMzKpWJvEvjogzKo/EzMwaoszF3V9I+pykcZJGdX4qj8zMzCpRpsZ/UP73S4WywA9xmZm1pD5r/BGxXjefPpO+pHMlzZd0T6HseElzJc3Knz0GugNmZrZsyvTVs5Kkr0mamsc3lPTBEus+H9itm/LvRsSE/Llm2cI1M7OBKtPGfx7wCvDPeXwucEJfC0XETcDf+h+amZlVoUwb//oRsV9+GQsR8ZIkDWCbh0s6EJgBHBMRz3Y3k6TJwGSA8ePHD2Bzg6v4Apc5J+3ZxEjMzPqnTI3/FUkjSBd0kbQ+qfuG/jgDWB+YAMwDTu1pxoiYGhETI2LimDFj+rk5MzPrqkyN/+ukvvjXkXQRsB0wqT8bi4inOoclnQX8sj/rMTOz/ivTV8/1ku4EtgUEHBkRT/dnY5LGRcS8PLoPcE9v85uZ2eAr01fPFnmwM2GPl7Qa8GhELO5luWnAjsBoSY+TfjnsKGkCqdloDvCZfkduZmb9Uqap54fAFsBdpBr/psC9wGqSDouI67pbKCL276b4nP4GamZmg6PMxd0ngM3zhdYtgc2Bh4H3A9+uMjgzMxt8ZRL/OyLi3s6RiLgPeGdEPFxdWGZmVpWyr148A/hpHt8PuE/SisCrlUVmZmaVKFPjnwQ8RHrd4lGkZp5JpKS/UzVhmZlZVcrczvl30oNW3T1s9cKgR2RmZpUqU+M3M7M24sRvZlYzPSZ+ST/J/x7ZuHDMzKxqvdX4t5S0JvApSasXX7voVy+ambWu3i7ungncQHrF4kzSU7ud/OpFM7MW1WONPyK+FxHvAs6NiLcv66sXzcxsaCpzO+dhkt4DvC8X3RQRd1UblpmZVaVM75xHkN6EdXkuukjS1Ij4r0oja7Lim7aWZV6/lcvMhroyXTZ8GtgmIl4EkHQy8AegrRO/mVm7KnMfv4AlhfElvPlCr5mZtZAyNf7zgNslXZHH98b96puZtawyF3dPk/RbYPtcdHBE/KnSqMzMrDJlavxExJ3AnRXHYmZmDeC+eszMasaJ38ysZnpN/JKGSbqxUcGYmVn1ek38EbEEeE3Sag2Kx8zMKlbm4u4LwN2Srgde7CyMiCMqi8rMzCpTJvFfzhvdNZiZWYsrcx//BZJGAOMj4oEGxGRmZhXq864eSR8CZgHX5vEJkqZXHJeZmVWkzO2cxwNbAwsBImIWfgmLmVnLKpP4X42I57qUvVZFMGZmVr0yF3fvlfRxYJikDYEjgN9XG5aZmVWlTI3/C8AmwMvANOB54KgKYzIzswqVuavnJeC4/AKWiIhF1YdlZmZVKXNXz1aS7gbuIj3I9WdJW1YfmpmZVaFMG/85wOci4mYASduTXs6yWZWBmZlZNcq08S/pTPoAEXELsLi6kMzMrEo9Jn5JW0jaAvidpB9J2lHSDpJ+CPy2rxVLOlfSfEn3FMpGSbpe0oP539UHZS/MzKy03pp6Tu0y/vXCcJRY9/nA94EfF8qmADdExEmSpuTxL5dYl5mZDZIeE39E7DSQFUfETZI6uhTvBeyYhy8g/XJw4jcza6A+L+5KGgkcCHQU5+9nt8xjI2JeHn4SGNuPdZiZ2QCUuavnGuA24G4GsauGiAhJPTYZSZoMTAYYP378YG3WzKz2yiT+4RFx9CBt7ylJ4yJinqRxwPyeZoyIqcBUgIkTJ5a5pmBmZiWUuZ3zJ5IOlTQu35UzStKofm5vOnBQHj4IuKqf6zEzs34qU+N/BfgOcBxv3M0T9NE1s6RppAu5oyU9Tror6CTgEkmHAI8CH+1f2GZm1l9lEv8xwAYR8fSyrDgi9u9h0i7Lsh4zMxtcZZp6HgJeqjoQMzNrjDI1/heBWZJuJHXNDPT7dk4zM2uyMon/yvwxM7M2UKY//gsaEYiZmTVGmSd3H6Gbvnkiwi9cNzNrQWWaeiYWhocDHwH6ex+/mZk1WZ939UTEM4XP3Ig4Hdiz+tDMzKwKZZp6tiiMLkf6BVDml4KZmQ1BZRJ4sV/+xcAc/MStmVnLKnNXz4D65Tczs6GlTFPPisCHWbo//m9WF5aZmVWlTFPPVcBzwEwKT+6amVlrKpP4146I3SqPxMzMGqJMJ22/l/TuyiMxM7OGKFPj3x6YlJ/gfRkQ6c2Jm1UamZmZVaJM4t+98ijMzKxhytzO+WgjAjEzs8Yo08ZvZmZtxInfzKxmnPjNzGrGna01UceUq18fnnOSOzw1s8Zwjd/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MaqZWD3AVH5hqxPr8gJaZDUWu8ZuZ1YwTv5lZzTjxm5nVjBO/mVnNNOXirqQ5wCJgCbA4IiY2Iw4zszpq5l09O0XE003cvplZLbmpx8ysZpqV+AO4TtJMSZObFIOZWS01q6ln+4iYK2kN4HpJ90fETcUZ8h+EyQDjx49vRoxDQudDYFU+AOYHzczqpSk1/oiYm/+dD1wBbN3NPFMjYmJETBwzZkyjQzQza1sNT/ySVpa0aucw8AHgnkbHYWZWV81o6hkLXCGpc/sXR8S1TYjDzKyWGp74I+Jh4D2N3q6ZmSW+ndPMrGac+M3MasaJ38ysZmr1IpZGWJaXsyzLPGXute9uOd+Xb2ZducZvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjB7gapMyDW0OJX85i1r5c4zczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmvEDXEPEsjzg1d+HwQb7IbLuHvLyg19mQ59r/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNeP7+NvcYNy7Pxj35ncXR3FdPW2jv8stSzyNWK7MOjrLe1qvn5EYWlr5eLjGb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNNSfySdpP0gKSHJE1pRgxmZnXV8MQvaRjwA2B3YGNgf0kbNzoOM7O6akaNf2vgoYh4OCJeAX4K7NWEOMzMakkR0dgNSvsCu0XEp/P4AcA2EXF4l/kmA5Pz6EbAA/3c5Gjg6X4u2wq8f63N+9e6WmHf1o2IMV0Lh+yTuxExFZg60PVImhEREwchpCHJ+9favH+tq5X3rRlNPXOBdQrja+cyMzNrgGYk/j8CG0paT9JbgI8B05sQh5lZLTW8qSciFks6HPgVMAw4NyLurXCTA24uGuK8f63N+9e6WnbfGn5x18zMmstP7pqZ1YwTv5lZzbR14m+nriEkrSPpRkn3SbpX0pG5fJSk6yU9mP9dvdmxDoSkYZL+JOmXeXw9SbfnY/izfENAS5I0UtKlku6XNFvSe9vp+En6Yj4375E0TdLwVj5+ks6VNF/SPYWybo+Xku/l/bxL0hbNi7xvbZv427BriMXAMRGxMbAt8Pm8P1OAGyJiQ+CGPN7KjgRmF8ZPBr4bERsAzwKHNCWqwfGfwLUR8U7gPaT9bIvjJ2kt4AhgYkRsSrpx42O09vE7H9itS1lPx2t3YMP8mQyc0aAY+6VtEz9t1jVERMyLiDvz8CJS0liLtE8X5NkuAPZuSoCDQNLawJ7A2XlcwM7ApXmWlt0/SasB/wc4ByAiXomIhbTR8SPdJThC0vLASsA8Wvj4RcRNwN+6FPd0vPYCfhzJbcBISeMaEmg/tHPiXwt4rDD+eC5reZI6gM2B24GxETEvT3oSGNusuAbB6cC/Aq/l8bcBCyNicR5v5WO4HrAAOC83ZZ0taWXa5PhFxFzgFOCvpIT/HDCT9jl+nXo6Xi2Vb9o58bclSasAlwFHRcTzxWmR7s1tyftzJX0QmB8RM5sdS0WWB7YAzoiIzYEX6dKs0+LHb3VSrXc9YE1gZZZuJmkrrXy82jnxt13XEJJWICX9iyLi8lz8VOdPyvzv/GbFN0DbAf9X0hxSs9zOpDbxkbnpAFr7GD4OPB4Rt+fxS0l/CNrl+O0KPBIRCyLiVeBy0jFtl+PXqafj1VL5pp0Tf1t1DZHbu88BZkfEaYVJ04GD8vBBwFWNjm0wRMRXImLtiOggHavfRMQngBuBffNsrbx/TwKPSdooF+0C3EebHD9SE8+2klbK52rn/rXF8Svo6XhNBw7Md/dsCzxXaBIaeiKibT/AHsD/AH8Bjmt2PAPcl+1JPyvvAmblzx6kdvAbgAeBXwOjmh3rIOzrjsAv8/DbgTuAh4CfAys2O74B7NcEYEY+hlcCq7fT8QO+AdwP3AP8BFixlY8fMI10veJV0i+2Q3o6XoBIdxH+BbibdHdT0/ehp4+7bDAzq5l2buoxM7NuOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxW+1I+q2kyl+SLemI3AvnRRVuY46k0VWt39pTw1+9aNbKJC0fb/Q905fPAbtGxONVxmS2rFzjtyFJUkeuLZ+V+3i/TtKIPO31Gruk0bmbByRNknRl7id9jqTDJR2dO0W7TdKowiYOkDQr9x2/dV5+5dwH+x15mb0K650u6Tekh3e6xnp0Xs89ko7KZWeSHl76b0lf7DL/JElX5f14UNLXC9M+mbc/S9KPcvfiSDpD0oz8XXyjmxhGSPpvSYfm/bha0p9zTPv1/0hYO3Lit6FsQ+AHEbEJsBD4cIllNgX+H7AVcCLwUqRO0f4AHFiYb6WImECqlZ+by44jdRWxNbAT8J3cgyakfnX2jYgdihuTtCVwMLAN6T0Jh0raPCI+CzwB7BQR3+0mzq3z/mwGfETSREnvAvYDtsuxLQE+0RlbREzM8+8gabPCulYBfgFMi4izSJ2jPRER74nUN/61Jb43qxE39dhQ9khEzMrDM4GOEsvcGOl9BYskPUdKiJAeoy8my2mQ+lyX9FZJI4EPkDqKOzbPMxwYn4evj4iufbND6krjioh4EUDS5cD7gD/1Eef1EfFMYZntSS/b2RL4Y+ruhhG80QnYRyVNJv2fHUd6udBdedpVwLcjovNawt3AqZJOJnV9cXMfsVjNOPHbUPZyYXgJKRFCSpCdv1aH97LMa4Xx13jz+d61r5Ig9bfy4Yh4oDhB0jakbpQHU0/bvyAivtJl++sBxwJbRcSzks7nzft9K7CbpIsj+R+lV//tAZwg6YaI+OYgx28tzE091ormkGrG8EbPj8tqPwBJ25N6UnwO+BXwhdy7JJI2L7Gem4G9c6+UKwP75LK+vF/p/a0jSG9xupV0/WBfSWvk7Y+StC7wVtIfnuckjSW95q/o30mvNfxBXm5NUhPXhcB3SM1UZq9zjd9a0SnAJbnp4+p+ruMfkv4ErAB8Kpf9B+ktYHdJWg54BPhgbyuJiDtzDfyOXHR2RPTVzEOe/zJSv+0XRsQMAElfA67L238V+HxE3JZjvZ/0lqdbu1nfkcC5kr5N+gPyHUmv5XUcViIeqxH3zmnWYJImkbrtPbzZsVg9uanHzKxmXOM3M6sZ1/jNzGrGid/MrGac+M3MasaJ38ysZpz4zcxq5n8B0nzfr5ApdgAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Histogram of number of peaks per gene\n",
    "plt.figure()\n",
    "plt.hist(peaks_per_gene_df['n_peaks'], bins=np.arange(0,110,1))\n",
    "plt.xlabel('number of peaks')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of number of peaks per gene')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Associate CAGE peaks with genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Dataframe of genes for each peak\n",
    "genes_per_peak = peaks_per_gene_df[['gene','peaks']]\n",
    "genes_per_peak = lu.unnesting(genes_per_peak,['peaks'])\n",
    "genes_per_peak = genes_per_peak.groupby('peaks')['gene'].apply(list).reset_index(name='genes')\n",
    "genes_per_peak['n_genes'] = [len(genes) for genes in genes_per_peak['genes']]\n",
    "genes_per_peak.columns = ['short_description','genes','n_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add peak location\n",
    "genes_per_peak = pd.merge(genes_per_peak, cage_df, on=['short_description'])\n",
    "genes_per_peak = genes_per_peak[['short_description','chrom','cage_start','cage_end','strand','genes']]\n",
    "genes_per_peak = genes_per_peak.sort_values(by=['chrom'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consider a window around each peak\n",
    "genes_per_peak['cage_window_start'] = [genes_per_peak['cage_start'].iloc[i]-400 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                       else genes_per_peak['cage_start'].iloc[i]-50 for i in range(genes_per_peak.shape[0])]\n",
    "genes_per_peak['cage_window_end'] = [genes_per_peak['cage_end'].iloc[i]+50 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                     else genes_per_peak['cage_end'].iloc[i]+400 for i in range(genes_per_peak.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genes</th>\n",
       "      <th>cage_window_start</th>\n",
       "      <th>cage_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>p2@PYGO2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154936109</td>\n",
       "      <td>154936127</td>\n",
       "      <td>-</td>\n",
       "      <td>[SHC1]</td>\n",
       "      <td>154936059</td>\n",
       "      <td>154936527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>p2@VAV3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108231101</td>\n",
       "      <td>108231133</td>\n",
       "      <td>-</td>\n",
       "      <td>[VAV3]</td>\n",
       "      <td>108231051</td>\n",
       "      <td>108231533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>p2@uc001dqd.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>94240877</td>\n",
       "      <td>94240925</td>\n",
       "      <td>-</td>\n",
       "      <td>[BCAR3]</td>\n",
       "      <td>94240827</td>\n",
       "      <td>94241325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>p30@ARHGEF2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155940369</td>\n",
       "      <td>155940382</td>\n",
       "      <td>-</td>\n",
       "      <td>[ARHGEF2]</td>\n",
       "      <td>155940319</td>\n",
       "      <td>155940782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>p35@ARHGEF2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155942148</td>\n",
       "      <td>155942156</td>\n",
       "      <td>-</td>\n",
       "      <td>[ARHGEF2]</td>\n",
       "      <td>155942098</td>\n",
       "      <td>155942556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    short_description chrom  cage_start   cage_end strand      genes  \\\n",
       "757          p2@PYGO2  chr1   154936109  154936127      -     [SHC1]   \n",
       "795           p2@VAV3  chr1   108231101  108231133      -     [VAV3]   \n",
       "799     p2@uc001dqd.1  chr1    94240877   94240925      -    [BCAR3]   \n",
       "801       p30@ARHGEF2  chr1   155940369  155940382      -  [ARHGEF2]   \n",
       "830       p35@ARHGEF2  chr1   155942148  155942156      -  [ARHGEF2]   \n",
       "\n",
       "     cage_window_start  cage_window_end  \n",
       "757          154936059        154936527  \n",
       "795          108231051        108231533  \n",
       "799           94240827         94241325  \n",
       "801          155940319        155940782  \n",
       "830          155942098        155942556  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize\n",
    "genes_per_peak.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Create sequence files for all CAGE windows"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over all CAGE windows to generate sequence files\n",
    "for i in tqdm(range(genes_per_peak.shape[0])):\n",
    "    time.sleep(.01)\n",
    "    # Peak characteristics\n",
    "    peak = genes_per_peak['short_description'].iloc[i]\n",
    "    chrom = genes_per_peak['chrom'].iloc[i]\n",
    "    start = genes_per_peak['cage_window_start'].iloc[i]\n",
    "    end = genes_per_peak['cage_window_end'].iloc[i]\n",
    "    # Generate window sequence (sequence on the + strand)\n",
    "    result = subprocess.run(cage_fasta_dir+'twoBitToFa http://hgdownload.cse.ucsc.edu/gbdb/hg19/hg19.2bit '+\n",
    "                            cage_fasta_dir+peak+'.fa'+\n",
    "                            ' -seq='+chrom+' -start='+str(start)+' -end='+str(end)+' -noMask',\n",
    "                            shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Search for TFBS in all CAGE windows\n",
    "See process_JASPAR_with_strand.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Find TFs binding each CAGE window"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Useful data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load JASPAR TFBS metadata\n",
    "jaspar_df = pd.read_csv(tf_dir+'JASPAR-HomoSapiens.csv', \n",
    "                        header=0, usecols=['ID','Name','Species','Class','Family'])\n",
    "jaspar_tfs = np.unique(jaspar_df['Name'])\n",
    "# Mapping JASPAR ID to HGNC\n",
    "jasparid_to_hgnc = {jaspar_df.iloc[i,0]:jaspar_df.iloc[i,1] for i in range(jaspar_df.shape[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique TFs = 639\n",
      "Number of unique TF motifs = 810\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Class</th>\n",
       "      <th>Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0002.1</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Runt domain factors</td>\n",
       "      <td>Runt-related factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0003.1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0003.2</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0003.3</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID    Name       Species                                  Class  \\\n",
       "0  MA0002.1   RUNX1  Homo sapiens                    Runt domain factors   \n",
       "1  MA0003.1  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "2  MA0003.2  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "3  MA0003.3  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "4  MA0003.4  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "\n",
       "                 Family  \n",
       "0  Runt-related factors  \n",
       "1                  AP-2  \n",
       "2                  AP-2  \n",
       "3                  AP-2  \n",
       "4                  AP-2  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize data\n",
    "print('Number of unique TFs = '+str(len(np.unique(jaspar_df['Name']))))\n",
    "print('Number of unique TF motifs = '+str(len(np.unique(jaspar_df['ID']))))\n",
    "jaspar_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load MOODS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary mapping CAGE peaks to the corresponding TFs\n",
    "tfs_associated_to_cage_peaks = {}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over files in moods_out_dir and summarize results\n",
    "for peak_moods_file in tqdm(os.listdir(moods_out_dir)):\n",
    "    time.sleep(.01)\n",
    "    # load results dataframe for that gene\n",
    "    cage_peak = peak_moods_file.strip('.txt')\n",
    "    res_df = pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                         usecols=[0,1,2,3,4,5], \n",
    "                         names=['peak_id','tfbs_file','motif_loc','match_strand','match_score','motif'])\n",
    "    if (res_df.shape[0]>0):\n",
    "        # Add columns corresponding to the target peak\n",
    "        res_df['cage_peak'] = cage_peak\n",
    "        res_df['chrom'] = res_df['peak_id'].str.split(':', expand=True)[0]\n",
    "        location_cols = res_df['peak_id'].str.split(':', expand=True)[1]\n",
    "        res_df[['start','stop']] = location_cols.str.split('-', expand=True)\n",
    "        # Add columns corresponding to the TFBS\n",
    "        res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "        res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "        # Only keep TFs matching on the peak strand\n",
    "        peak_strand = cage_df[cage_df['short_description']==cage_peak]['strand'].values[0]\n",
    "        res_df = res_df[res_df['match_strand']==peak_strand]\n",
    "        # Select relevant columns\n",
    "        res_df = res_df[['cage_peak', 'chrom','start','stop','motif_loc',\n",
    "                         'jaspar_tfbs','TF','match_score','match_strand','motif']]\n",
    "        # Store list of TFs associated with the current CAGE peak in dict\n",
    "        tfs_associated_to_cage_peaks[cage_peak] = np.sort(np.unique(res_df['TF'].values))\n",
    "pickle.dump(tfs_associated_to_cage_peaks, open(saving_dir+'tfs_cage_with_strand.pkl', 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Find TFs targeting genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mapping between cage peaks and corresponding TFs\n",
    "tfs_associated_to_cage_peaks = pickle.load(open(saving_dir+'tfs_cage_with_strand.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Construct dataframe with gene-tf pairs\n",
    "tf_and_genes_df = genes_per_peak.copy()\n",
    "tf_and_genes_df['tfs'] = [tfs_associated_to_cage_peaks[cage_peak] if cage_peak in tfs_associated_to_cage_peaks.keys()\n",
    "                          else ['missing']\n",
    "                          for cage_peak in tf_and_genes_df['short_description'].values]\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['genes'])\n",
    "gene_to_tfs = {tf_and_genes_df.iloc[i,0]:tf_and_genes_df.iloc[i,8] for i in range(tf_and_genes_df.shape[0])}\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['tfs'])\n",
    "tf_and_genes_df = tf_and_genes_df[['genes','tfs','short_description','chrom','cage_start','cage_end','strand']]\n",
    "tf_and_genes_df = tf_and_genes_df.sort_values(by=['genes','tfs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ALX3</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ARGFX</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1639</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>BACH2</td>\n",
       "      <td>p@chr10:27149068..27149079,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149068</td>\n",
       "      <td>27149079</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1637</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>p@chr10:27132379..27132382,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27132379</td>\n",
       "      <td>27132382</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>p@chr10:27037479..27037487,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27037479</td>\n",
       "      <td>27037487</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genes    tfs             short_description  chrom  cage_start  cage_end  \\\n",
       "853   ABI1   ALX3                       p3@ABI1  chr10    27149904  27149920   \n",
       "853   ABI1  ARGFX                       p3@ABI1  chr10    27149904  27149920   \n",
       "1639  ABI1  BACH2  p@chr10:27149068..27149079,-  chr10    27149068  27149079   \n",
       "1637  ABI1   E2F6  p@chr10:27132379..27132382,-  chr10    27132379  27132382   \n",
       "1634  ABI1   EBF1  p@chr10:27037479..27037487,-  chr10    27037479  27037487   \n",
       "\n",
       "     strand  \n",
       "853       -  \n",
       "853       -  \n",
       "1639      -  \n",
       "1637      -  \n",
       "1634      -  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "tf_and_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SHC1': ['missing'],\n",
       " 'VAV3': array(['ESR1', 'ESR2', 'SMAD2::SMAD3::SMAD4'], dtype=object),\n",
       " 'BCAR3': ['missing'],\n",
       " 'ARHGEF2': array(['RREB1', 'ZBTB12'], dtype=object),\n",
       " 'ITGA10': array(['CTCF', 'CTCFL'], dtype=object),\n",
       " 'CSRP1': array(['ZNF263'], dtype=object),\n",
       " 'HAX1': array(['ZNF384'], dtype=object),\n",
       " 'FABP3': array(['EWSR1-FLI1', 'NR1I2', 'NR5A1', 'SPI1', 'ZKSCAN1', 'ZKSCAN5',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'FBLIM1': ['missing'],\n",
       " 'NEXN': array(['GRHL2', 'MAZ', 'NFATC4'], dtype=object),\n",
       " 'PTPRF': array(['ZNF740'], dtype=object),\n",
       " 'TSPAN1': array(['NFATC4', 'ZNF354C'], dtype=object),\n",
       " 'MACF1': ['missing'],\n",
       " 'ENAH': array(['ASCL1(var.2)', 'GLI2', 'GLI3', 'INSM1', 'NFKB1', 'TCF12(var.2)',\n",
       "        'TCF3', 'TCF4', 'TCF7L1', 'ZEB1', 'ZNF135', 'ZNF24', 'ZNF354C',\n",
       "        'ZNF460'], dtype=object),\n",
       " 'ARF1': array(['ATF6', 'KLF15', 'KLF5'], dtype=object),\n",
       " 'ASAP3': array(['ZNF341'], dtype=object),\n",
       " 'ITGB3BP': array(['EBF3', 'SPI1'], dtype=object),\n",
       " 'SLC9A1': array(['ESRRA', 'ESRRB', 'GRHL2', 'NR2F2', 'NR4A2', 'NR5A1', 'PKNOX1'],\n",
       "       dtype=object),\n",
       " 'CAPN2': ['missing'],\n",
       " 'RHOU': ['missing'],\n",
       " 'VIM': array(['FOS::JUN', 'FOS::JUNB', 'FOS::JUND', 'FOSL2::JUN', 'MZF1(var.2)',\n",
       "        'ZNF282'], dtype=object),\n",
       " 'ITGA8': array(['CDX2', 'ZNF384'], dtype=object),\n",
       " 'NRP1': array(['RBPJ', 'STAT1'], dtype=object),\n",
       " 'LDB3': array(['EBF1', 'EBF3', 'FOS', 'FOS::JUN', 'FOS::JUNB', 'FOS::JUND',\n",
       "        'FOSB::JUNB', 'FOSL2::JUN', 'FOSL2::JUNB', 'SPIB', 'STAT1::STAT2'],\n",
       "       dtype=object),\n",
       " 'PTEN': array(['TFAP2B(var.2)'], dtype=object),\n",
       " 'ADAM12': array(['EHF', 'ELF3', 'EWSR1-FLI1', 'GABPA', 'PAX1', 'PAX9', 'RFX2',\n",
       "        'RFX4', 'RFX5'], dtype=object),\n",
       " 'ITGB1': array(['EHF', 'ELF3', 'GABPA', 'PRDM4', 'SPI1', 'SPIB'], dtype=object),\n",
       " 'SVIL': ['missing'],\n",
       " 'SORBS1': array(['HOXB13', 'MAFF', 'MYBL1', 'NR2F1', 'NR2F2', 'NR4A1', 'OTX2',\n",
       "        'RORA', 'TFE3', 'TFEC', 'ZNF317'], dtype=object),\n",
       " 'VCL': array(['SMAD2::SMAD3::SMAD4', 'SRF', 'ZNF143', 'ZNF682'], dtype=object),\n",
       " 'ABI1': array(['MAZ', 'RREB1', 'THAP11', 'ZNF148', 'ZNF263', 'ZNF384', 'ZNF410'],\n",
       "       dtype=object),\n",
       " 'MAPK8': ['missing'],\n",
       " 'DOCK1': array(['NFATC2', 'NFATC4', 'RELA', 'SPI1', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'THY1': array(['IKZF1', 'MAFK', 'MAZ', 'ZNF148'], dtype=object),\n",
       " 'PARVA': array(['EWSR1-FLI1', 'NR2C2'], dtype=object),\n",
       " 'ARHGEF12': array(['CDX1', 'CDX2', 'CDX4', 'HOXA10', 'HOXD9', 'IRF3', 'IRF9'],\n",
       "       dtype=object),\n",
       " 'RDX': array(['CTCF', 'NFATC3', 'STAT1::STAT2'], dtype=object),\n",
       " 'LAYN': array(['PPARG'], dtype=object),\n",
       " 'PAK1': array(['IRF2', 'NFATC2', 'SPI1'], dtype=object),\n",
       " 'ILK': array(['RFX1'], dtype=object),\n",
       " 'CAPN1': array(['EHF', 'GATA3', 'MAZ', 'POU6F1', 'RREB1', 'ZBTB7B', 'ZNF148'],\n",
       "       dtype=object),\n",
       " 'ARHGAP32': array(['ELF1', 'ELF5', 'ETV1', 'ETV4', 'GBX2', 'SMAD2::SMAD3::SMAD4',\n",
       "        'ZBTB7A'], dtype=object),\n",
       " 'CBL': array(['CTCF', 'IRF3', 'KLF5', 'ZNF136'], dtype=object),\n",
       " 'CD151': array([], dtype=object),\n",
       " 'CFL1': array(['RREB1', 'SP3', 'SP8', 'SP9'], dtype=object),\n",
       " 'CORO1B': array(['EHF', 'ELF3', 'GABPA', 'KLF15', 'KLF5', 'SP2', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'INPPL1': array(['FOXF2', 'MYF5', 'SMAD2::SMAD3::SMAD4'], dtype=object),\n",
       " 'LPXN': array(['ELF1', 'ETV6', 'SMAD2::SMAD3::SMAD4', 'SPI1', 'ZNF140'],\n",
       "       dtype=object),\n",
       " 'CTTN': array(['EGR1', 'ELF2', 'ELF4', 'ELF5', 'ELK1', 'ELK3', 'ERF', 'ERG',\n",
       "        'ETS1', 'ETS2', 'ETV1', 'ETV2', 'ETV3', 'FEV', 'FLI1', 'HES2',\n",
       "        'IRF4', 'IRF8', 'KLF15', 'KLF5', 'SPDEF', 'SPIC'], dtype=object),\n",
       " 'HRAS': array(['ZNF449', 'ZNF460'], dtype=object),\n",
       " 'FERMT3': array(['E2F7', 'EWSR1-FLI1', 'ZBTB12', 'ZNF136'], dtype=object),\n",
       " 'SLC3A2': array([], dtype=object),\n",
       " 'PPFIA1': array(['EOMES', 'GFI1', 'KLF15', 'KLF5', 'NFYA', 'NFYC', 'TBR1', 'TBX2',\n",
       "        'TBX21', 'TBX6'], dtype=object),\n",
       " 'PTPRO': array(['ZSCAN4'], dtype=object),\n",
       " 'ITGA7': ['missing'],\n",
       " 'TUBA1B': array(['HIC2', 'IRF3', 'KLF15', 'KLF5', 'MYC', 'SP1', 'SP4', 'ZNF24'],\n",
       "       dtype=object),\n",
       " 'SSH1': array(['KLF4', 'NFATC2', 'NR1H2::RXRA', 'NR2F1(var.2)', 'NR2F2', 'NR4A1',\n",
       "        'PPARA::RXRA', 'RREB1', 'ZNF740'], dtype=object),\n",
       " 'TENC1': array(['MYOG', 'NR4A2::RXRA', 'POU1F1', 'POU2F3', 'POU5F1B', 'REST',\n",
       "        'ZNF354C', 'ZNF528'], dtype=object),\n",
       " 'ITGA5': array(['MZF1(var.2)', 'NFIC(var.2)', 'NFIX(var.2)', 'ZNF740'],\n",
       "       dtype=object),\n",
       " 'LRP1': array(['BCL6B', 'NR2C2', 'NRF1', 'ZNF263'], dtype=object),\n",
       " 'PXN': ['missing'],\n",
       " 'AGAP2': array([], dtype=object),\n",
       " 'PTPN6': array(['GABPA', 'TBX19', 'TBXT'], dtype=object),\n",
       " 'PTPN11': array(['ATF3', 'CREM', 'ELF1', 'ETV1', 'ETV4', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUND(var.2)', 'HOXD3', 'IKZF1', 'ONECUT2', 'ONECUT3',\n",
       "        'ZKSCAN1'], dtype=object),\n",
       " 'GIT2': array(['BHLHE22(var.2)', 'FERD3L', 'MITF', 'TFAP2A', 'YY1'], dtype=object),\n",
       " 'ITGB7': array(['EHF', 'ELF3', 'KLF4', 'KLF5', 'RREB1', 'SP2', 'SPIB', 'ZNF682'],\n",
       "       dtype=object),\n",
       " 'STARD13': array(['NRL'], dtype=object),\n",
       " 'ARHGEF7': array(['ATF2', 'ATF3', 'ATF7', 'CDX2', 'CREB1', 'CREM', 'FOSB::JUN',\n",
       "        'FOSB::JUNB(var.2)', 'FOSL1::JUN(var.2)', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'FOSL2::JUND(var.2)', 'FOXC1', 'FOXC2',\n",
       "        'FOXD2', 'FOXE1', 'FOXF2', 'FOXG1', 'FOXK1', 'FOXK2', 'FOXN3',\n",
       "        'FOXO3', 'FOXO6', 'FOXP1', 'HOXB13', 'IRF1', 'JUN',\n",
       "        'JUN::JUNB(var.2)', 'JUND(var.2)', 'MEF2A', 'MEF2C', 'SOX10',\n",
       "        'SOX4', 'SPIB', 'SRY', 'STAT1', 'STAT1::STAT2', 'STAT3', 'ZNF384'],\n",
       "       dtype=object),\n",
       " 'MMP14': array(['BCL6B', 'IRF1', 'NR1D1'], dtype=object),\n",
       " 'FERMT2': array(['BHLHA15(var.2)', 'FOXN3', 'ZNF140'], dtype=object),\n",
       " 'ACTN1': array(['LEF1', 'NR5A1', 'ZNF354C'], dtype=object),\n",
       " 'ZFYVE21': array(['NFYA', 'NFYC', 'PBX1', 'VENTX'], dtype=object),\n",
       " 'HSPA2': array(['FOXN3', 'NFYB', 'POU2F2'], dtype=object),\n",
       " 'KTN1': array(['ARNT::HIF1A', 'KLF5'], dtype=object),\n",
       " 'ARHGAP5': array(['MAZ', 'ZNF148'], dtype=object),\n",
       " 'AKT1': array(['MYBL1'], dtype=object),\n",
       " 'SYNM': array(['CTCFL', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D', 'NFATC3',\n",
       "        'STAT1::STAT2', 'THAP1', 'ZNF354C'], dtype=object),\n",
       " 'ITGA11': array(['ELF3', 'GABPA', 'IRF3', 'MAZ', 'SPIB', 'ZNF354C'], dtype=object),\n",
       " 'CSK': array(['CTCF', 'CTCFL', 'KLF10', 'POU4F1', 'ZIC3', 'ZIC4', 'ZIC5'],\n",
       "       dtype=object),\n",
       " 'PEAK1': array(['EWSR1-FLI1', 'KLF5', 'NFYA', 'NFYC', 'SP1', 'SP2', 'SP4'],\n",
       "       dtype=object),\n",
       " 'TRPM7': array(['BCL6B', 'MLX', 'MLXIPL', 'NFIL3', 'STAT1', 'TFE3'], dtype=object),\n",
       " 'CIB2': array(['E2F6', 'ELF3', 'ELK4', 'ETV5', 'NR2C2', 'NR2F1(var.2)', 'PPARD',\n",
       "        'RXRB', 'RXRG', 'TFDP1', 'THRB', 'ZNF354C'], dtype=object),\n",
       " 'CIB1': array(['NR2C2(var.2)', 'RARA(var.2)', 'RARA::RXRG', 'RARB(var.3)',\n",
       "        'RXRB(var.2)', 'RXRG(var.2)', 'SNAI3', 'SOX21'], dtype=object),\n",
       " 'BCAR1': array(['BCL6B', 'SREBF2', 'TFAP2C(var.2)'], dtype=object),\n",
       " 'MAPK8IP3': array(['ATF3', 'ATF7', 'CREB1', 'CREB3L4(var.2)', 'CREM', 'ELK4', 'ETS2',\n",
       "        'ETV5', 'FOSB::JUN', 'FOSB::JUNB(var.2)', 'FOSL1::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'FOSL2::JUND(var.2)', 'JDP2(var.2)',\n",
       "        'JUN::JUNB(var.2)', 'JUNB(var.2)', 'MYOG', 'NHLH1', 'NR2C1',\n",
       "        'NR4A2', 'PBX3', 'PKNOX1', 'RORA', 'RORB', 'SIX1'], dtype=object),\n",
       " 'TGFB1I1': array([], dtype=object),\n",
       " 'PDPK1': ['missing'],\n",
       " 'ITGAL': array(['ELF1', 'ETV1', 'HNF4A(var.2)', 'IKZF1', 'NR1H2::RXRA',\n",
       "        'NR2F1(var.2)', 'PLAG1', 'PPARA::RXRA', 'PPARD', 'RXRB', 'RXRG',\n",
       "        'THRB'], dtype=object),\n",
       " 'ITGAX': array(['ESRRA', 'ESRRB', 'FOS', 'NR2F1', 'NR2F1(var.2)', 'NR2F2', 'NR5A1',\n",
       "        'NR6A1', 'RORA', 'SPDEF', 'THRB'], dtype=object),\n",
       " 'SH2B1': ['missing'],\n",
       " 'PKD1': array(['ATF2', 'ATF3', 'CREB1', 'FOSB::JUN', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'JDP2(var.2)', 'JUN', 'JUNB(var.2)',\n",
       "        'JUND(var.2)'], dtype=object),\n",
       " 'ITGAM': array(['ARGFX', 'EWSR1-FLI1', 'ISX', 'LHX9', 'MEF2A', 'MIXL1', 'MNX1',\n",
       "        'NKX6-2', 'POU6F2', 'PRRX1', 'PRRX2', 'RAX', 'UNCX', 'ZNF135',\n",
       "        'ZNF263', 'ZNF460'], dtype=object),\n",
       " 'NUDT16L1': array(['MTF1', 'NFIA', 'RUNX2', 'ZBTB6'], dtype=object),\n",
       " 'ITGAD': array(['EWSR1-FLI1', 'KLF13', 'POU4F2', 'POU4F3', 'SP1', 'SP4', 'SP9',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'SRCIN1': array(['CDX4', 'HOXA13', 'HOXC13'], dtype=object),\n",
       " 'STAT3': array(['BACH2(var.2)', 'MAX', 'MAX::MYC', 'POU4F1', 'POU4F3', 'PRDM1',\n",
       "        'ZNF341'], dtype=object),\n",
       " 'GIT1': array(['BACH2', 'BATF', 'BATF3', 'BATF::JUN', 'ELF3', 'FOS', 'FOS::JUN',\n",
       "        'FOS::JUNB', 'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'FOSL1::JUNB',\n",
       "        'FOSL1::JUND', 'FOSL2', 'FOSL2::JUN', 'FOSL2::JUNB', 'FOSL2::JUND',\n",
       "        'GSC', 'GSC2', 'IRF2', 'IRF7', 'JDP2', 'JUN(var.2)', 'JUN::JUNB',\n",
       "        'JUNB', 'JUND', 'MZF1', 'NFE2', 'OTX1', 'PITX1', 'PITX2', 'PITX3',\n",
       "        'PRDM1'], dtype=object),\n",
       " 'ITGB4': array(['BACH1', 'BATF', 'BATF3', 'BATF::JUN', 'CDX4', 'FOS', 'FOS::JUNB',\n",
       "        'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'HOXA13', 'HOXD13', 'JUNB',\n",
       "        'JUND', 'MAFK', 'SNAI3'], dtype=object),\n",
       " 'PRKCA': array(['ETV4', 'MZF1', 'SOX9', 'SPI1', 'ZNF384'], dtype=object),\n",
       " 'PFN1': array(['ELF1', 'ELF5', 'ETV1', 'ETV4', 'HOXB13', 'IKZF1', 'KLF15', 'KLF5',\n",
       "        'SP1', 'SP2', 'SP4', 'SPI1', 'ZBTB7A', 'ZNF263'], dtype=object),\n",
       " 'NDEL1': array(['ATF6', 'KLF15', 'KLF16', 'KLF4', 'KLF5', 'NR1I3', 'NRF1'],\n",
       "       dtype=object),\n",
       " 'GRB2': array(['EWSR1-FLI1', 'NFYB', 'PKNOX2', 'SPI1', 'SPIC', 'TGIF2'],\n",
       "       dtype=object),\n",
       " 'GRB7': array(['SP4', 'SRY'], dtype=object),\n",
       " 'ITGA3': array(['IRF1', 'STAT1::STAT2'], dtype=object),\n",
       " 'CRK': array(['MYB', 'NEUROD1', 'NEUROG2(var.2)', 'NFATC2', 'NR2C1', 'NRL',\n",
       "        'TGIF1'], dtype=object),\n",
       " 'ITGAE': array(['ASCL1', 'FERD3L', 'MYOG', 'SP2', 'ZSCAN4'], dtype=object),\n",
       " 'ABI3': array(['EHF', 'EWSR1-FLI1', 'GFI1', 'GRHL2', 'IRF2', 'NFATC4', 'SPIB',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'SLC16A3': array(['GRHL2', 'NFATC3', 'POU1F1'], dtype=object),\n",
       " 'ITGB3': array(['EOMES', 'MAZ', 'NRL', 'PAX3', 'PAX7', 'PPARG', 'PROP1', 'TBR1',\n",
       "        'TBX2', 'TBX21', 'ZNF148'], dtype=object),\n",
       " 'LASP1': array(['TWIST1', 'ZBTB18', 'ZNF75D'], dtype=object),\n",
       " 'PTPN2': array(['ARNT::HIF1A', 'EWSR1-FLI1'], dtype=object),\n",
       " 'MYOM1': array(['ARNT::HIF1A', 'ZBTB14', 'ZNF460'], dtype=object),\n",
       " 'ROCK1': array(['ASCL1', 'ETV1', 'ETV4', 'IKZF1', 'KLF4', 'KLF5', 'MYOG', 'SP1',\n",
       "        'SP4', 'SPI1', 'THAP1'], dtype=object),\n",
       " 'INSR': array(['FOXA1', 'FOXB1', 'FOXC1', 'FOXC2', 'FOXD2', 'FOXE1', 'FOXF2',\n",
       "        'IRF3', 'IRF4', 'IRF9', 'ZNF384'], dtype=object),\n",
       " 'CALR': array(['INSM1', 'MEIS1', 'RXRB', 'STAT1::STAT2'], dtype=object),\n",
       " 'PLAUR': array(['E2F6', 'MEF2A', 'MEF2D', 'RUNX3', 'RXRB(var.2)', 'RXRG(var.2)',\n",
       "        'ZBTB12'], dtype=object),\n",
       " 'PRKACA': array(['GLIS3', 'IRF1', 'STAT1::STAT2'], dtype=object),\n",
       " 'DNM2': array(['CEBPA', 'CEBPD', 'EBF1', 'MEIS1', 'MZF1', 'PBX3', 'PKNOX1',\n",
       "        'SPIB'], dtype=object),\n",
       " 'KEAP1': array(['FOS', 'GLIS3', 'MZF1', 'ZBTB12'], dtype=object),\n",
       " 'PVR': array(['IKZF1', 'NFKB2', 'TFDP1'], dtype=object),\n",
       " 'VASP': array(['MEIS1', 'ZBTB7B', 'ZNF143'], dtype=object),\n",
       " 'CYTH2': array(['SP1', 'SP2', 'SP4', 'ZNF263'], dtype=object),\n",
       " 'PTPRH': array(['EHF', 'ELF3', 'GABPA', 'IRF1', 'PRDM1', 'SPIB'], dtype=object),\n",
       " 'VAV1': array(['ELF1', 'RREB1', 'SPI1'], dtype=object),\n",
       " 'PIP5K1C': array(['FOS::JUN', 'FOS::JUND', 'FOSL2::JUN', 'HES7', 'KLF15', 'KLF16',\n",
       "        'KLF5', 'NFE2', 'SOX21', 'SP1', 'SP4'], dtype=object),\n",
       " 'RAVER1': array(['E2F6', 'GATA1', 'HOXA13', 'HOXB13', 'OSR2', 'TFDP1', 'YY1', 'YY2',\n",
       "        'ZFP42', 'ZNF354C', 'ZSCAN29'], dtype=object),\n",
       " 'TNS1': array(['SOHLH2'], dtype=object),\n",
       " 'SOS1': array(['ASCL1', 'MYOG', 'THAP11'], dtype=object),\n",
       " 'LIMS2': array(['ASCL1', 'BHLHE22(var.2)', 'IRF1', 'NHLH1', 'NR5A1', 'PRDM1',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'FHL2': array(['LEF1', 'TCF7L1', 'TCF7L2', 'ZNF384'], dtype=object),\n",
       " 'IRS1': array(['CTCF', 'EBF1'], dtype=object),\n",
       " 'INPP5D': array(['ELF1', 'ETV1', 'ETV4', 'IKZF1', 'IRF1', 'POU6F1(var.2)', 'PRDM1',\n",
       "        'SPI1', 'SPIC', 'STAT1::STAT2', 'TBXT'], dtype=object),\n",
       " 'ITGA4': array(['GATA1::TAL1', 'GATA6', 'HES6', 'SOHLH2'], dtype=object),\n",
       " 'NRP2': array(['LMX1B', 'PAX6', 'SPI1'], dtype=object),\n",
       " 'ITGA6': array(['FIGLA', 'FOXE1', 'HNF1A'], dtype=object),\n",
       " 'ITGAV': array(['EHF', 'ELF3', 'ELF5', 'EWSR1-FLI1', 'SPIB'], dtype=object),\n",
       " 'ITGB6': ['missing'],\n",
       " 'ITGB1BP1': array(['MAX::MYC', 'MSANTD3', 'NR2F1(var.3)', 'PKNOX2', 'TGIF2',\n",
       "        'THRB(var.2)', 'ZNF317'], dtype=object),\n",
       " 'ASAP2': array(['KLF5', 'SOX13', 'SOX18', 'SOX2', 'SOX4', 'SOX8', 'SRY'],\n",
       "       dtype=object),\n",
       " 'ARPC2': array(['ETV6'], dtype=object),\n",
       " 'LIMS1': ['missing'],\n",
       " 'ILKAP': array(['FOXC1', 'ZNF384'], dtype=object),\n",
       " 'NCK2': ['missing'],\n",
       " 'ABI2': array(['ELF1', 'ELF4', 'ELK1', 'ELK3', 'ERF', 'ERG', 'ETS1', 'ETV1',\n",
       "        'ETV3', 'ETV4', 'ETV6', 'EWSR1-FLI1', 'FEV', 'FLI1', 'HSF4',\n",
       "        'KLF15', 'KLF5', 'MYBL1', 'PRDM4', 'SOX10', 'SOX4', 'SOX8', 'SRY',\n",
       "        'ZBTB7A', 'ZNF652'], dtype=object),\n",
       " 'CASP8': ['missing'],\n",
       " 'SDC4': array(['ASCL1(var.2)', 'MEF2B', 'MEF2D', 'NHLH1', 'TCF12(var.2)', 'TCF4',\n",
       "        'ZNF341'], dtype=object),\n",
       " 'FERMT1': array(['SPI1'], dtype=object),\n",
       " 'SIRPA': array(['FOXC1', 'FOXC2', 'FOXD2', 'FOXE1', 'GRHL1', 'GRHL2', 'ZNF652'],\n",
       "       dtype=object),\n",
       " 'PLCG1': array(['EGR1', 'HNF4A', 'HNF4G', 'KLF15', 'KLF5', 'PROP1'], dtype=object),\n",
       " 'PTPN1': array(['BATF', 'BATF3', 'BATF::JUN', 'ETV6', 'MSANTD3', 'NR1I2', 'RFX5',\n",
       "        'SPIB'], dtype=object),\n",
       " 'PRNP': array(['NR3C2'], dtype=object),\n",
       " 'CASS4': array(['EWSR1-FLI1', 'NR1H2::RXRA', 'PBX3', 'PKNOX1', 'RXRB', 'TCF7L1',\n",
       "        'TCF7L2'], dtype=object),\n",
       " 'SRC': ['missing'],\n",
       " 'PTPRA': ['missing'],\n",
       " 'ITGB2': array(['ELF1', 'ELF3', 'ELF5', 'ERF', 'ETV1', 'ETV2', 'ETV4', 'IKZF1',\n",
       "        'NR2C2', 'OSR1', 'SPI1', 'TBR1'], dtype=object),\n",
       " 'TIAM1': array(['E2F8', 'NFATC2', 'PRDM1'], dtype=object),\n",
       " 'MYH9': array(['RARA::RXRA'], dtype=object),\n",
       " 'NF2': array(['NFATC3', 'NKX2-2', 'NKX2-5', 'THRB(var.2)'], dtype=object),\n",
       " 'CRKL': array(['ATF3', 'ATF7', 'DBP', 'FOSB::JUN', 'FOSB::JUNB(var.2)',\n",
       "        'FOSL1::JUN(var.2)', 'FOSL1::JUND(var.2)', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'GMEB2', 'GRHL2', 'MZF1(var.2)', 'OSR2',\n",
       "        'TEF', 'TFAP2A'], dtype=object),\n",
       " 'PPM1F': array(['FOXC1', 'FOXC2', 'GABPA', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D',\n",
       "        'MEIS1', 'MYF5', 'ONECUT1', 'SOX21', 'SPIB'], dtype=object),\n",
       " 'MAPK1': array(['EWSR1-FLI1', 'KLF15', 'KLF5', 'YY1', 'YY2', 'ZFP42'], dtype=object),\n",
       " 'RNF185': array(['MITF', 'NR2C2', 'THAP1', 'USF1', 'USF2', 'ZBTB7A', 'ZNF135',\n",
       "        'ZNF460'], dtype=object),\n",
       " 'ANKRD28': array(['GLIS2', 'RFX2', 'RFX4', 'RFX5'], dtype=object),\n",
       " 'CD47': array(['ETV6', 'NEUROG1', 'NEUROG2', 'NFATC2', 'NFATC4', 'OTX2', 'TGIF1',\n",
       "        'TGIF2', 'TGIF2LX', 'TGIF2LY'], dtype=object),\n",
       " 'LPP': ['missing'],\n",
       " 'PLD1': array(['PKNOX1', 'ZNF354C'], dtype=object),\n",
       " 'PPM1M': array(['PBX3', 'PKNOX1', 'ZKSCAN5'], dtype=object),\n",
       " 'ITGB5': array(['KLF5', 'SPI1', 'ZNF460'], dtype=object),\n",
       " 'RHOA': array(['SNAI2'], dtype=object),\n",
       " 'NISCH': array(['CEBPG', 'DBP', 'HLF', 'IRF1', 'IRF7', 'NR4A2::RXRA', 'SP1', 'SP4',\n",
       "        'SP8', 'SP9', 'YY1', 'YY2', 'ZFP42'], dtype=object),\n",
       " 'PIK3CA': array(['FOXN3', 'HOXA13', 'HOXB13', 'IRF1', 'SPIB', 'ZNF384'],\n",
       "       dtype=object),\n",
       " 'ITGA9': array(['CTCF', 'GATA1', 'GATA2', 'GATA4', 'GATA6', 'IRF3', 'SOX10'],\n",
       "       dtype=object),\n",
       " 'SORBS2': array(['GATA3', 'MEF2C', 'NFYB', 'TFAP2A(var.3)', 'TFAP2B(var.3)',\n",
       "        'TFAP2C(var.3)'], dtype=object),\n",
       " 'GAB1': array(['OSR2'], dtype=object),\n",
       " 'PALLD': array(['SIX2'], dtype=object),\n",
       " 'ARHGAP24': array(['NKX2-8', 'SIX2'], dtype=object),\n",
       " 'PDE4D': array(['FOXC2', 'FOXE1', 'KLF14'], dtype=object),\n",
       " 'ARHGAP26': array(['DUX4', 'FOXC2', 'FOXE1'], dtype=object),\n",
       " 'GNB2L1': array(['ATF4', 'CEBPG(var.2)', 'MZF1', 'NEUROD1', 'NEUROG2(var.2)',\n",
       "        'PKNOX2', 'TGIF1', 'TGIF2'], dtype=object),\n",
       " 'ITGA1': array(['TBX18', 'TBX20'], dtype=object),\n",
       " 'RASA1': array(['EMX1', 'EMX2', 'EN1', 'HOXB3', 'KLF4', 'KLF5', 'MEIS1', 'MSGN1',\n",
       "        'NEUROD2', 'NOTO', 'PAX4', 'SREBF2', 'TCF21(var.2)', 'TLX2',\n",
       "        'VSX2', 'ZNF135', 'ZNF460'], dtype=object),\n",
       " 'TRIO': ['missing'],\n",
       " 'ITGA2': array(['HIC2', 'MZF1'], dtype=object),\n",
       " 'EZR': array(['HOXC4', 'SPIB', 'ZNF354C'], dtype=object),\n",
       " 'MARCKS': array(['KLF14', 'KLF15', 'KLF16', 'LMX1A', 'LMX1B', 'MAZ', 'MEIS2(var.2)',\n",
       "        'PAX3', 'SOX9', 'SP1', 'SP3', 'SP4', 'SP8', 'SP9', 'ZNF148'],\n",
       "       dtype=object),\n",
       " 'FYN': array(['HNF4A(var.2)', 'SNAI1', 'SNAI3', 'TFAP2A(var.3)', 'TFAP2B(var.3)',\n",
       "        'TFAP2C(var.3)'], dtype=object),\n",
       " 'NEDD9': ['missing'],\n",
       " 'DEF6': array(['SIX2', 'TFCP2', 'ZBTB26', 'ZNF16'], dtype=object),\n",
       " 'KCNH2': array(['MAFK', 'MEIS1(var.2)', 'NKX6-3', 'PBX2'], dtype=object),\n",
       " 'ITGB8': array(['RARA'], dtype=object),\n",
       " 'ACTB': array(['ZNF354C'], dtype=object),\n",
       " 'RAC1': array(['BCL6B', 'E2F3', 'EHF', 'ELF3', 'GABPA', 'IRF1', 'KLF2', 'KLF3',\n",
       "        'KLF6', 'NFATC2', 'PRDM4', 'SPIB', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'HSPB1': array(['FOXC2', 'FOXN3', 'HSF1', 'HSF2', 'HSF4', 'NFIB', 'NFIC(var.2)',\n",
       "        'NFIX(var.2)', 'ZNF384'], dtype=object),\n",
       " 'TES': array(['HOXB13', 'NFATC4'], dtype=object),\n",
       " 'LIMK1': array(['EWSR1-FLI1', 'KLF15', 'KLF5', 'ONECUT1', 'POU4F1', 'POU4F2',\n",
       "        'POU4F3'], dtype=object),\n",
       " 'PTPN12': array(['SOX15', 'SOX9', 'ZBTB7B'], dtype=object),\n",
       " 'TRIP6': array(['SPI1'], dtype=object),\n",
       " 'PLEC': array(['ATF2', 'CREB1', 'JUN'], dtype=object),\n",
       " 'LYN': array(['ELF1', 'ELF5', 'POU1F1', 'POU2F1', 'POU2F3', 'POU3F1', 'POU3F2',\n",
       "        'POU3F3', 'POU3F4', 'POU5F1', 'POU5F1B', 'PRDM1', 'RARA', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'PABPC1': array(['ELF2', 'ELF4', 'ELF5', 'EWSR1-FLI1', 'IRF1', 'KLF5', 'NRF1',\n",
       "        'SP2', 'SP4', 'SPIB'], dtype=object),\n",
       " 'SHARPIN': array(['SMAD2::SMAD3::SMAD4', 'ZNF460'], dtype=object),\n",
       " 'DLC1': array(['ARGFX', 'ISX', 'LHX9', 'NKX6-2', 'PRRX1', 'PRRX2', 'UNCX'],\n",
       "       dtype=object),\n",
       " 'PTK2': array(['FOXI1', 'FOXL1', 'FOXO4', 'FOXO6', 'FOXP3', 'NFATC2', 'PRDM1',\n",
       "        'ZBTB26'], dtype=object),\n",
       " 'SORBS3': array(['IRF3', 'IRF4', 'IRF5', 'IRF8', 'SIX1', 'STAT3'], dtype=object),\n",
       " 'PTK2B': ['missing'],\n",
       " 'SDCBP': array(['RFX2', 'RFX5', 'SREBF1', 'TCF7'], dtype=object),\n",
       " 'TLN1': array(['IKZF1', 'MEF2C', 'POU1F1', 'POU2F1', 'POU2F3', 'POU3F1', 'POU3F2',\n",
       "        'POU3F3', 'POU3F4', 'POU5F1', 'POU5F1B', 'SPI1'], dtype=object),\n",
       " 'RAPGEF1': array(['ASCL1', 'EWSR1-FLI1', 'MYOG', 'NFIC'], dtype=object),\n",
       " 'ABL1': array(['MEIS1(var.2)', 'THAP11', 'ZFP42'], dtype=object),\n",
       " 'VAV2': array(['ELF1', 'ETV1', 'ETV4', 'HNF4A', 'HNF4G', 'HOXA9', 'KLF5', 'SP2'],\n",
       "       dtype=object),\n",
       " 'ENG': ['missing'],\n",
       " 'TESK1': array(['ATF3', 'FOS::JUN(var.2)', 'FOSL2::JUNB(var.2)',\n",
       "        'FOSL2::JUND(var.2)', 'JDP2(var.2)'], dtype=object),\n",
       " 'OSTF1': array(['ELF1', 'ETV1', 'HOXC4', 'NR1D1', 'NR1D2', 'PAX5', 'RXRB(var.2)',\n",
       "        'RXRG(var.2)', 'ZNF384'], dtype=object),\n",
       " 'SPTLC1': array(['ESR1', 'ESR2', 'FOXI1', 'FOXL1', 'FOXO4', 'FOXO6', 'FOXP3',\n",
       "        'GSX2', 'HOXA2', 'HOXB2', 'IRF4', 'IRF8', 'MSANTD3', 'NFIL3',\n",
       "        'PDX1', 'PPARG', 'ZNF528'], dtype=object),\n",
       " 'CORO2A': array(['FOXE1', 'NFATC2', 'NKX2-8', 'NR2F1(var.3)', 'THRB(var.3)',\n",
       "        'ZBTB12', 'ZNF135', 'ZNF460'], dtype=object)}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save dictionary mapping selected genes to their potential TFs to pickle\n",
    "pickle.dump(gene_to_tfs, open(saving_dir+'gene_to_tfs_with_strand.pkl', 'wb'))\n",
    "gene_to_tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p1@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2588012</td>\n",
       "      <td>2588031</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p2@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2587980</td>\n",
       "      <td>2587994</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p3@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2587999</td>\n",
       "      <td>2588011</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p4@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2588131</td>\n",
       "      <td>2588143</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p@chr16:2589098..2589102,+</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2589098</td>\n",
       "      <td>2589102</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes      tfs           short_description  chrom  cage_start  cage_end  \\\n",
       "430   PDPK1  missing                    p1@PDPK1  chr16     2588012   2588031   \n",
       "736   PDPK1  missing                    p2@PDPK1  chr16     2587980   2587994   \n",
       "952   PDPK1  missing                    p3@PDPK1  chr16     2587999   2588011   \n",
       "1123  PDPK1  missing                    p4@PDPK1  chr16     2588131   2588143   \n",
       "1876  PDPK1  missing  p@chr16:2589098..2589102,+  chr16     2589098   2589102   \n",
       "\n",
       "     strand  \n",
       "430       +  \n",
       "736       +  \n",
       "952       +  \n",
       "1123      +  \n",
       "1876      +  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_and_genes_df[tf_and_genes_df['genes']=='PDPK1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
