{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/anaconda3/lib/python3.7/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'process_jaspar_with_strand' from '/home/louiscam/projects/gpcr/code/JASPAR_processing/process_jaspar_with_strand.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import itertools\n",
    "from itertools import groupby\n",
    "import os.path\n",
    "import math\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import MOODS.parsers\n",
    "import MOODS.tools\n",
    "import MOODS.scan\n",
    "import subprocess\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import process_jaspar_with_strand as pj\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(pj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "prom_hg19_seq_dir = dir_genome+'prom_hg19_seq_dir/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data_dir/'\n",
    "# Directory for storing preliminary results\n",
    "prelim_results_dir = '/home/louiscam/projects/gpcr/save/prelim_results_dir/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "# Directory of tf data\n",
    "tf_dir = '/home/louiscam/projects/gpcr/data/tf_data/'\n",
    "jaspar_dir = tf_dir+'jaspar_data/'\n",
    "pfm_dir = tf_dir+'pfm_data/'\n",
    "fantom5_dir = '/home/louiscam/projects/gpcr/data/fantom5/'\n",
    "moods_out_dir = tf_dir+'moods_cage_outdir_with_strand/'\n",
    "cage_fasta_dir = dir_genome+'cage_hg19_seq_dir_with_strand/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Select genes of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genes of interest (active adhesome genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes of interest = 219\n"
     ]
    }
   ],
   "source": [
    "# Genes of interest and location\n",
    "selected_genes = pickle.load(open(saving_dir+'active_adhesome_genes.pkl', 'rb'))\n",
    "print('Number of genes of interest = '+str(len(selected_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping UCSC to HGNC\n",
    "ucsc_to_hgnc = pj.create_ucsc_hgnc_dict(dir_genome+'chrom_hg19.name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load human TFs list\n",
    "from http://humantfs.ccbr.utoronto.ca/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of TFs = 1639\n"
     ]
    }
   ],
   "source": [
    "# Load TFs\n",
    "tf_df = pd.read_csv(dir_genome+'human_tf.database', sep='\\t', header=0, index_col=0)\n",
    "tf_df = tf_df[tf_df['Is TF?']=='Yes'].sort_values(by=['HGNC symbol'])\n",
    "all_tfs = tf_df['HGNC symbol'].values\n",
    "print('Total number of TFs = '+str(len(all_tfs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>5S_RRNA</td>\n",
       "      <td>chr18</td>\n",
       "      <td>-</td>\n",
       "      <td>327954</td>\n",
       "      <td>242524490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>5S_RRNA</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>327954</td>\n",
       "      <td>242524490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>5S_RRNA</td>\n",
       "      <td>chr21</td>\n",
       "      <td>+</td>\n",
       "      <td>327954</td>\n",
       "      <td>242524490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>5S_RRNA</td>\n",
       "      <td>chr22</td>\n",
       "      <td>-</td>\n",
       "      <td>327954</td>\n",
       "      <td>242524490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>5S_RRNA</td>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>327954</td>\n",
       "      <td>242524490</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  chrom strand  txStart      txEnd\n",
       "817  5S_RRNA  chr18      -   327954  242524490\n",
       "818  5S_RRNA  chr19      +   327954  242524490\n",
       "823  5S_RRNA  chr21      +   327954  242524490\n",
       "825  5S_RRNA  chr22      -   327954  242524490\n",
       "793  5S_RRNA  chr12      +   327954  242524490"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load gene locations\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc'\n",
    "df_loc = pd.read_csv(gene_locations_filename, sep = '\\t', header = 0)\n",
    "df_loc.columns = ['transcript','chrom','strand','txStart','txEnd','cdsStart','cdsEnd',\n",
    "                  'exonCount','exonStarts','exonEnds','proteinID','alignID']\n",
    "df_loc['gene'] = [ucsc_to_hgnc[t] for t in df_loc['transcript']]\n",
    "# Pick largest extent of each gene\n",
    "min_extent_gene = df_loc.groupby(['gene'], sort=False)['txStart'].min().rename('tx_minStart')\n",
    "max_extent_gene = df_loc.groupby(['gene'], sort=False)['txEnd'].max().rename('tx_maxEnd')\n",
    "df_loc = pd.merge(df_loc, min_extent_gene, on='gene')\n",
    "df_loc = pd.merge(df_loc, max_extent_gene, on='gene')\n",
    "df_loc = df_loc[['transcript','gene','chrom','strand',\n",
    "                 'tx_minStart','tx_maxEnd']]\n",
    "# Determine the strand of the gene (in case of labeling error)\n",
    "df_loc = pd.merge(df_loc, \n",
    "                  df_loc.groupby(['gene','chrom','tx_minStart','tx_maxEnd'])['strand'].count().rename('strand_vote'),\n",
    "                  on=['gene','chrom','tx_minStart','tx_maxEnd'])\n",
    "df_loc = df_loc.sort_values(by=['gene','strand_vote'])\n",
    "df_loc = df_loc.drop_duplicates(subset=['gene','chrom','tx_minStart','tx_maxEnd'], keep='first')\n",
    "df_loc = df_loc[['gene','chrom','strand','tx_minStart','tx_maxEnd']]\n",
    "df_loc.columns = ['gene','chrom','strand','txStart','txEnd']\n",
    "# Only keep transcripts matching largest extent\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Canonical transcript  for each gene\n",
    "gene_id_filename = dir_genome+'chrom_hg19.name'\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc_canonical'\n",
    "canonical_transcripts_df = lu.get_gene_locations(gene_locations_filename, gene_id_filename)\n",
    "canonical_transcripts_df.columns = ['transcript','gene','chrom','start','end','clusterId','protein']\n",
    "canonical_transcripts_df = canonical_transcripts_df[['transcript','gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10708</th>\n",
       "      <td>uc002cqz.1</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript   gene\n",
       "10708  uc002cqz.1  PDPK1"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "canonical_transcripts_df[canonical_transcripts_df['gene']=='PDPK1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Canonical gene location and metadata\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc'\n",
    "df_loc = pd.read_csv(gene_locations_filename, sep = '\\t', header = 0)\n",
    "df_loc.columns = ['transcript','chrom','strand','txStart','txEnd','cdsStart','cdsEnd',\n",
    "                  'exonCount','exonStarts','exonEnds','proteinID','alignID']\n",
    "df_loc = pd.merge(canonical_transcripts_df, df_loc, how='inner', on=['transcript'])\n",
    "df_loc['txLength'] = df_loc['txEnd']-df_loc['txStart']\n",
    "df_loc['isTF'] = [gene in all_tfs for gene in df_loc['gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Restrict to genes of interest\n",
    "df_loc = df_loc[df_loc['gene'].isin(selected_genes)]\n",
    "df_loc.columns = ['transcript','gene','chrom','strand','txStart','txEnd','cdsStart','cdsEnd','exonCount',\n",
    "                  'exonStarts','exonEnds','proteinID','alignID','txLength','isTF']\n",
    "df_loc = df_loc[['transcript','gene','chrom','strand','txStart','txEnd','txLength','isTF']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct mistake at PDPK1\n",
    "df_loc.at[df_loc[df_loc['gene']=='PDPK1'].index[0],'strand'] = '+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>proteinID</th>\n",
       "      <th>alignID</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60757</th>\n",
       "      <td>uc010uwe.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2587964</td>\n",
       "      <td>2612228</td>\n",
       "      <td>2588113</td>\n",
       "      <td>2611618</td>\n",
       "      <td>3</td>\n",
       "      <td>2587964,2607703,2611480,</td>\n",
       "      <td>2588137,2607964,2612228,</td>\n",
       "      <td>B4DF26</td>\n",
       "      <td>uc010uwe.2</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60759</th>\n",
       "      <td>uc002cqv.1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2606896</td>\n",
       "      <td>2634065</td>\n",
       "      <td>2607760</td>\n",
       "      <td>2633733</td>\n",
       "      <td>9</td>\n",
       "      <td>2606896,2611480,2611771,2615553,2616356,262742...</td>\n",
       "      <td>2607964,2611523,2611909,2615698,2616454,262750...</td>\n",
       "      <td>Q6A1A2</td>\n",
       "      <td>uc002cqv.1</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60761</th>\n",
       "      <td>uc002cqs.4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2587964</td>\n",
       "      <td>2653191</td>\n",
       "      <td>2588113</td>\n",
       "      <td>2647768</td>\n",
       "      <td>14</td>\n",
       "      <td>2587964,2607703,2611480,2611771,2615553,261635...</td>\n",
       "      <td>2588137,2607964,2611523,2611909,2615698,261645...</td>\n",
       "      <td>O15530</td>\n",
       "      <td>uc002cqs.4</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60762</th>\n",
       "      <td>uc010bsn.4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2587964</td>\n",
       "      <td>2653191</td>\n",
       "      <td>2588113</td>\n",
       "      <td>2647673</td>\n",
       "      <td>12</td>\n",
       "      <td>2587964,2607703,2611480,2611771,2615553,261635...</td>\n",
       "      <td>2588137,2607964,2611523,2611909,2615698,261645...</td>\n",
       "      <td>H0Y4Z0</td>\n",
       "      <td>uc010bsn.4</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60763</th>\n",
       "      <td>uc002cqt.4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2587964</td>\n",
       "      <td>2653191</td>\n",
       "      <td>2588113</td>\n",
       "      <td>2647768</td>\n",
       "      <td>11</td>\n",
       "      <td>2587964,2607703,2611480,2627425,2631295,263160...</td>\n",
       "      <td>2588137,2607964,2611523,2627501,2631364,263170...</td>\n",
       "      <td>O15530-4</td>\n",
       "      <td>uc002cqt.4</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60764</th>\n",
       "      <td>uc002cqu.4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>2588164</td>\n",
       "      <td>2653191</td>\n",
       "      <td>2607760</td>\n",
       "      <td>2647768</td>\n",
       "      <td>14</td>\n",
       "      <td>2588164,2607703,2611480,2611771,2615553,261635...</td>\n",
       "      <td>2588772,2607964,2611523,2611909,2615698,261645...</td>\n",
       "      <td>O15530</td>\n",
       "      <td>uc002cqu.4</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60768</th>\n",
       "      <td>uc002cqz.1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>2666121</td>\n",
       "      <td>2693297</td>\n",
       "      <td>2666453</td>\n",
       "      <td>2692433</td>\n",
       "      <td>9</td>\n",
       "      <td>2666121,2668482,2668822,2672690,2683739,268449...</td>\n",
       "      <td>2666774,2668579,2668891,2672766,2683837,268464...</td>\n",
       "      <td>Q6A1A2</td>\n",
       "      <td>uc002cqz.1</td>\n",
       "      <td>PDPK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            #name  chrom strand  txStart    txEnd  cdsStart   cdsEnd  \\\n",
       "60757  uc010uwe.2  chr16      +  2587964  2612228   2588113  2611618   \n",
       "60759  uc002cqv.1  chr16      +  2606896  2634065   2607760  2633733   \n",
       "60761  uc002cqs.4  chr16      +  2587964  2653191   2588113  2647768   \n",
       "60762  uc010bsn.4  chr16      +  2587964  2653191   2588113  2647673   \n",
       "60763  uc002cqt.4  chr16      +  2587964  2653191   2588113  2647768   \n",
       "60764  uc002cqu.4  chr16      +  2588164  2653191   2607760  2647768   \n",
       "60768  uc002cqz.1  chr16      -  2666121  2693297   2666453  2692433   \n",
       "\n",
       "       exonCount                                         exonStarts  \\\n",
       "60757          3                           2587964,2607703,2611480,   \n",
       "60759          9  2606896,2611480,2611771,2615553,2616356,262742...   \n",
       "60761         14  2587964,2607703,2611480,2611771,2615553,261635...   \n",
       "60762         12  2587964,2607703,2611480,2611771,2615553,261635...   \n",
       "60763         11  2587964,2607703,2611480,2627425,2631295,263160...   \n",
       "60764         14  2588164,2607703,2611480,2611771,2615553,261635...   \n",
       "60768          9  2666121,2668482,2668822,2672690,2683739,268449...   \n",
       "\n",
       "                                                exonEnds proteinID  \\\n",
       "60757                           2588137,2607964,2612228,    B4DF26   \n",
       "60759  2607964,2611523,2611909,2615698,2616454,262750...    Q6A1A2   \n",
       "60761  2588137,2607964,2611523,2611909,2615698,261645...    O15530   \n",
       "60762  2588137,2607964,2611523,2611909,2615698,261645...    H0Y4Z0   \n",
       "60763  2588137,2607964,2611523,2627501,2631364,263170...  O15530-4   \n",
       "60764  2588772,2607964,2611523,2611909,2615698,261645...    O15530   \n",
       "60768  2666774,2668579,2668891,2672766,2683837,268464...    Q6A1A2   \n",
       "\n",
       "          alignID   gene  \n",
       "60757  uc010uwe.2  PDPK1  \n",
       "60759  uc002cqv.1  PDPK1  \n",
       "60761  uc002cqs.4  PDPK1  \n",
       "60762  uc010bsn.4  PDPK1  \n",
       "60763  uc002cqt.4  PDPK1  \n",
       "60764  uc002cqu.4  PDPK1  \n",
       "60768  uc002cqz.1  PDPK1  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_locations_filename = dir_genome+'chrom_hg19.loc'\n",
    "df_loc = pd.read_csv(gene_locations_filename, sep = '\\t', header = 0)\n",
    "df_loc['gene'] = [ucsc_to_hgnc[name] for name in df_loc['#name']]\n",
    "df_loc[df_loc['gene']=='PDPK1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adhesome gene lengths quartile = [ 21507.   49385.  114977.5]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeoklEQVR4nO3deZgdVbnv8e+PBAIIh4BpkAChGSPzFBEEFAE9ERE8RxDiQQ2gUbwyKIpBvBfk4hXn4ShiVAhHEHBAiaAIKgjImEAIEOTIIQGCQBKGMCmQ8N4/1mopNr27K91de2d3/T7Ps5+uca23qnu/vfaq2qsUEZiZWX2s1O4AzMystZz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJv80kzZe0X5N1e0m6p9UxrUiUnCPpCUk3D0F50yWdnqf3lrRg8FFakaSrJX2oTXWHpM3bUXcnceJfgUXEtRExvr/tJJ0q6bxWxNQGewJvAzaMiF3bHYytONr5D6bTOfFbnySNbHMIGwPzI+LZNsdhNmw48a8YdpQ0R9ISSRdJWhVe3RUh6TOSHpL0tKR7JO0raSLwWeBQSc9Iuj1vO1bSDEmPS7pX0ocL5awm6dzcfXK3pBMb6pmf65oDPCtppKSpkv4n1z1X0r8Vtp8s6c+SviHpSUn3SXpTXv6gpIWSPtjs4JvFKuko4IfA7vnYPt/LvptJ+qOkxyQtlnS+pNGF9TtJujXHfRGwai9lnJBjfFjSEYXloyR9VdIDkh6VdJak1fK6MZIuzcf7uKRrJa2U122VW6NPSrpL0oGFMqdLOlPSb/Mx/VnS6yR9M/8+/iJpp4Zz8wtJiyTNk3RsH+fxtZJ+LekpSbdIOl3SdYX1r5d0ZY73HknvbYjru5Iuy+fqJkmbldm3P5KOzH9nT0j6naSNC+tC0kcl/TWfr+9KUl43QtLX8u91nqSP5+1HSvoCsBfwnXwev1Oocr8m5W0u6U9K77PF+e+hniLCrza+gPnAzcBYYB3gbuCjed3ewII8PR54EBib57uBzfL0qcB5DeVeA5xJSnQ7AouAffK6M4A/AWsDGwJzeuopxDQb2AhYLS87JMe4EnAo8Cywfl43GVgKHAGMAE4HHgC+C4wC3g48DazR5Bz0Fetk4Lo+zt/mpK6gUUBXLuubed0qwP3AJ4CVgYOBF4HTC+d3KXBaXr8/8Bywdl7/DWBG/r2sCfwa+GJe90XgrLzfyqQkpDx9L+mf8SrAPvnYx+f9pgOLgV3y8f4RmAd8oHDursrbrgTMAv5PLmtT4D7gX5uciwvza3Vga9Lfy3V53Wvy/BHASGCnHMfWhbgeA3bN688HLiyzby9xXA18KE8flM/HVnnfzwHXF7YN4FJgNDAu/+4n5nUfBeaS/kbXBn6ftx/ZWE/J8i4ATs7ndVVgz3a//9uWd9odQN1fpCR7eGH+y8BZeXpvXk78mwMLgf2AlRvKOJVC4icl7GXAmoVlXwSm5+lXJA/gQ7w68R/ZT9yzgYPy9GTgr4V12+U34HqFZY8BO/ZSTn+xTqaPxN9Lee8GbsvTbwb+Bqiw/npemfj/3pNI8rKFwG6kJP4s+Z9rXrc7MC9PnwZcAmzeUP9ewCPASoVlFwCn5unpwA8K644B7m44d0/m6TcCDzSUfxJwTi/HPYL0T218YdnpvJz4DwWubdjn+8Aphbh+WFi3P/CXMvv2EsvVvJz4fwscVVi3Eumf68Z5PigkYOCnwNQ8/UfgI4V1+1Eu8Tcr77+AaaTrRW1/77fz5a6eFcMjhenngDUaN4iIe4HjSUl+oaQLJY1tUt5Y4PGIeLqw7H5gg8L6BwvritO9LpP0AUmz88fnJ4FtgTGFTR4tTP89x9y47FXHVSLWPklaL5+LhyQ9BZxXiGss8FDkd32h7KLHImJpYb7n/HeRWs6zCsd8eV4O8BVSS/YKpa6tqYU6H4yIl/o4nsbz0uw8bQyM7ak/x/BZYL1eTkUXqUXd7Pe6MfDGhrL+A3hdYZtmf4dl9m1mY+Bbhf0eJ/1TLZ6PZvWW+TvtTbPyTsx135y74I4sWd6w48TfQSLiJxGxJ+nNFMCXelY1bPo3YB1JaxaWjQMeytMPkz4+99iot+p6JnKf7A+AjwOvjYjRwJ2kN9Fg9Rdrf/5fjnW7iPgX4PBCXA8DG/T08RbKLmMxKQlvExGj82utiFgDICKejogTImJT4EDgk5L2zcezUU9//wCOp+hB0ieM0YXXmhGxfy/bLiJ1WzX7vT4I/KmhrDUi4uiScQxm34807LtaRFxfYt/+/k6Xa2jhiHgkIj4cEWOBjwBnqqa3fjrxdwhJ4yXtI2kU8A9SUuppVT4KdPckm4h4kNSl8UVJq0raHjiK1BqG9PH3JElrS9qAlND78hrSm2xRjuUIUot/0ErE2p81gWeAJflYPl1YdwMpGR4raWVJ/07qwy4T10ukf3bfkLQugKQNJP1rnj4gXywUsITUXfUScBOplXlirnNv4F2kvvfldTPwtNKF9tXyxc5tJb2hl3iXARcDp0paXdLrSdcNelwKbCnp/TmulSW9QdJWJeIYzL5nkf7WtgGQtJakQ0rsB+nv9Lh83kcDn2lY/yjpukcpkg6R1POP5AnS3/RLfewybDnxd45RpIuyi0kfZdcl9fcC/Cz/fEzSrXl6EukC8N+AX5L6Y3+f150GLCBdVPw98HPg+WYVR8Rc4GukRPooqR/6z0NxUCVi7c/ngZ1JyfcyUvIDICJeAP6ddJ3gcVJf9cWvLqKpz5C6c27M3Ui/J11kB9gizz9DOi9nRsRVuc53Ae8g/a7OBD4QEX9Zjnp74l8GHEC64D0vl/dDYK0mu3w8r3sE+DHp2sLzuaynSRfZDyOd50dInxhHlYhjMPv+Mm97YT6Hd5LOTRk/AK4g3XxwG/Ab0j/yZXn9t4CD891C3y5R3huAmyQ9Q7pof1xE3FcylmFFr+z+tDqSdDRwWES8pd2x2NCR9CXgdRHR9FbaTiLpHaQbHzbud2Prk1v8NSRpfUl7SFpJ0njgBFJL2zqY0r322yvZldRl1rG/19y9tX++b38D4BQ6+HhWJE789bQK6Xa8p0m3zF1C6pKwzrYmqSvrWeAiUvfcJW2NaHBE6sp7gtTVczfpOw02SO7qMTOrGbf4zcxqpt0DcJUyZsyY6O7ubncYZmYdZdasWYsjoqtxeUck/u7ubmbOnNnuMMzMOoqkxm+qA+7qMTOrHSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmqks8Us6W+k5pnc2LD9G6bmid0n6clX1m5lZ76ps8U8HJhYXSHor6RmcO0TENsBXK6zfzMx6UVnij4hrSGOgFx0NnBERPWOEL6yqfjMz612r+/i3BPaSdJOkP/X2JKEekqZImilp5qJFiwZdcffUywZdhpnZcNDqxD8SWAfYjfSIvJ82PA/1nyJiWkRMiIgJXV2vGmrCzMwGqNWJfwFwcSQ3k553OabFMZiZ1VqrE/+vgLcCSNqS9ECQxS2Owcys1iobnVPSBcDewBhJC0iPTTsbODvf4vkC8MHwk2DMzFqqssQfEZOarDq8qjrNzKx//uaumVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1UxliV/S2ZIW5qdtNa47QVJI8vN2zcxarMoW/3RgYuNCSRsBbwceqLBuMzNrorLEHxHXAI/3suobwImAn7VrZtYGLe3jl3QQ8FBE3N7Kes3M7GWVPWy9kaTVgc+SunnKbD8FmAIwbty4AdfbPfWyAe9rZjYctbLFvxmwCXC7pPnAhsCtkl7X28YRMS0iJkTEhK6urhaGaWY2vLWsxR8RdwDr9szn5D8hIha3KgYzM6v2ds4LgBuA8ZIWSDqqqrrMzKy8ylr8ETGpn/XdVdVtZmbN+Zu7ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzLRuyYUVWHMht/hnvbGMkZmbVc4vfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGqmykcvni1poaQ7C8u+IukvkuZI+qWk0VXVb2ZmvauyxT8dmNiw7Epg24jYHvhv4KQK6zczs15Ulvgj4hrg8YZlV0TE0jx7I7BhVfWbmVnv2tnHfyTw22YrJU2RNFPSzEWLFg1pxcWxeczM6qYtiV/SycBS4Pxm20TEtIiYEBETurq6Whecmdkw1/LROSVNBg4A9o2IaHX9ZmZ119LEL2kicCLwloh4rpV1m5lZUuXtnBcANwDjJS2QdBTwHWBN4EpJsyWdVVX9ZmbWu8pa/BExqZfFP6qqPjMzK6ffFr+kzSSNytN7SzrWX7wyM+tcZbp6fgEsk7Q5MA3YCPhJpVGZmVllyiT+l/KXrv4N+M+I+DSwfrVhmZlZVcok/hclTQI+CFyal61cXUhmZlalMon/CGB34AsRMU/SJsCPqw3LzMyq0u9dPRExV9JngHF5fh7wpaoDMzOzapS5q+ddwGzg8jy/o6QZFcdlZmYVKXMf/6nArsDVABExW9KmFcZUmd4GZ/OAbWZWN6Uu7kbEkoZlL1URjJmZVa9Mi/8uSe8DRkjaAjgWuL7asMzMrCplWvzHANsAzwMXAE8Bx1cYk5mZVajMXT3PASfnl5mZdbh+E7+kLYFPAd3F7SNin+rCMjOzqpTp4/8ZcBbwQ2BZteGYmVnVyiT+pRHxvcojMTOzlihzcffXkj4maX1J6/S8Ko/MzMwqUabF/8H889OFZQF05Je4zMzqrsxdPZsMpGBJZ5Meqr4wIrbNy9YBLiJdKJ4PvDcinhhI+WZmNjBlxupZXdLnJE3L81tIOqBE2dOBiQ3LpgJ/iIgtgD/keTMza6EyffznAC8Ab8rzDwGn97dTRFwDPN6w+CDg3Dx9LvDuUlGamdmQKdPHv1lEHJofxkJEPCdJA6xvvYh4OE8/AqzXbENJU4ApAOPGjRtgdc15cDYzq6syLf4XJK1GuqCLpM1IwzcMSkRET5lN1k+LiAkRMaGrq2uw1ZmZWVamxX8KaSz+jSSdD+wBTB5gfY9KWj8iHpa0PrBwgOWYmdkAlbmr50pJtwK7AQKOi4jFA6xvBun20DPyz0sGWI6ZmQ1QmbF6ds6TPX3z4yStBdwfEUv72O8CYG9gjKQFpE8OZwA/lXQUcD/w3kHEbmZmA1Cmq+dMYGdgDqnFvy1wF7CWpKMj4oredoqISU3K23cggZqZ2dAoc3H3b8BO+ULrLsBOwH3A24AvVxmcmZkNvTKJf8uIuKtnJiLmAq+PiPuqC8vMzKpS9tGL3wMuzPOHAnMljQJerCwyMzOrRJkW/2TgXtLjFo8ndfNMJiX9t1YTlpmZVaXM7Zx/B76WX42eGfKIzMysUmVa/GZmNow48ZuZ1UzTxC/px/nnca0LZ8XhQdzMbLjqq8W/i6SxwJGS1i4+dtGPXjQz61x9Xdw9i/SwlE2BWaRv7fbwoxfNzDpU0xZ/RHw7IrYCzo6ITSNik8LLSd/MrEOVuZ3zaEk7AHvlRddExJxqwzIzs6qUeebuscD5wLr5db6kY6oOzMzMqlFmyIYPAW+MiGcBJH0JuAH4zyoDMzOzapS5j1/AssL8Ml55odfMzDpImRb/OcBNkn6Z598N/KiyiMzMrFJlLu5+XdLVwJ550RERcVulUZmZWWXKtPiJiFuBW4eqUkmfIF07COAO0j+TfwxV+WZm1lzLx+qRtAFwLDAhIrYFRgCHtToOM7O6atcgbSOB1SSNBFYnPd7RzMxaoM/EL2mEpKuGssKIeAj4KvAA8DCwpLcHtkuaImmmpJmLFi0ayhCWW/fUyzxom5kNG30m/ohYBrwkaa2hqlDS2sBBwCbAWOA1kg7vpe5p+QHvE7q6uoaqejOz2itzcfcZ4A5JVwLP9iyMiGMHWOd+wLyIWAQg6WLgTcB5AyzPzMyWQ5nEf3F+DZUHgN0krQ78HdgXmDmE5ZuZWR/K3Md/rqTVgHERcc9gK4yImyT9nHR76FLgNmDaYMs1M7NyygzS9i5gNnB5nt9R0ozBVBoRp0TE6yNi24h4f0Q8P5jyzMysvDK3c54K7Ao8CRARs/FDWMzMOlaZxP9iRCxpWPZSFcGYmVn1ylzcvUvS+4ARkrYgfev2+mrDMjOzqpRp8R8DbAM8D1wAPAUcX2FMZmZWoTJ39TwHnJwfwBIR8XT1YZmZWVXK3NXzBkl3AHNIX+S6XdIu1YdmZmZVKNPH/yPgYxFxLYCkPUkPZ9m+ysDMzKwaZfr4l/UkfYCIuI70xSvLPICbmXWSpi1+STvnyT9J+j7pwm4AhwJXVx+amZlVoa+unq81zJ9SmI4KYjEzsxZomvgj4q2tDMTMzFqj34u7kkYDHwC6i9sPYlhmMzNrozJ39fwGuJH0UHQP1WBm1uHKJP5VI+KTlUdiZmYtUeZ2zh9L+rCk9SWt0/OqPDIzM6tEmRb/C8BXgJN5+W6ewEMzm5l1pDKJ/wRg84hYXHUwZmZWvTJdPfcCzw1lpZJGS/q5pL9IulvS7kNZvpmZNVemxf8sMFvSVaShmYFB3875LeDyiDhY0irA6oMoy8zMlkOZxP+r/BoSktYC3gxMBoiIF0jXEczMrAXKjMd/7hDXuQmwCDhH0g7ALOC4iHi2uJGkKcAUgHHjxg1xCGZm9VVmPP55ku5rfA2izpHAzsD3ImInUlfS1MaNImJaREyIiAldXV2DqM7MzIrKdPVMKEyvChwCDOY+/gXAgoi4Kc//nF4Sv5mZVaPfFn9EPFZ4PRQR3wTeOdAKI+IR4EFJ4/OifYG5Ay3PzMyWT5lB2nYuzK5E+gRQ5pNCX44Bzs939NwHHDHI8szMrKQyCbw4Lv9SYD7w3sFUGhGzeWUXkpmZtUiZu3o8Lr+Z2TBSpqtnFPAeXj0e/2nVhWVmZlUp09VzCbCEdL/98/1sa2ZmK7gyiX/DiJhYeSRmZtYSZQZpu17SdpVHYmZmLVGmxb8nMFnSPFJXj4CIiO0rjczMzCpRJvG/o/IozMysZcp8c/f+3l6tCG5F1j31slLLu6de1nRbM7N2KNPHb2Zmw4gTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTNsSv6QRkm6TdGm7YjAzq6N2tviPA+5uY/1mZrXUlsQvaUPgncAP21G/mVmdtavF/03gROClZhtImiJppqSZixYtallgRYMdYK3MQG4ewM3MWq3liV/SAcDCiJjV13YRMS0iJkTEhK6urhZFZ2Y2/LWjxb8HcKCk+cCFwD6SzmtDHGZmtdTyxB8RJ0XEhhHRDRwG/DEiDm91HGZmdeX7+M3MaqbMoxcrExFXA1e3MwYzs7pxi9/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGbaeh9/J2o2wFp/g60t72BsPdvPP+Ody7WfmVl/3OI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczq5mWJ35JG0m6StJcSXdJOq7VMZiZ1Vk7hmxYCpwQEbdKWhOYJenKiJjbhljMzGqn5S3+iHg4Im7N008DdwMbtDoOM7O6ausgbZK6gZ2Am3pZNwWYAjBu3LjWBtZE2YHWymzX1wBvxYHZuqde9s/5xuni9sN1ULfG4xyK8obbOTJbXm27uCtpDeAXwPER8VTj+oiYFhETImJCV1dX6wM0Mxum2pL4Ja1MSvrnR8TF7YjBzKyu2nFXj4AfAXdHxNdbXb+ZWd21o8W/B/B+YB9Js/Nr/zbEYWZWSy2/uBsR1wFqdb1mZpb4m7tmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxN+ge+plpQdja5ViTP3F17O+r+0btxloLEOx7WDOd+O+y1tOs/MyUM32XdH+nhp1atwrkk47V078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXTroetT5R0j6R7JU1tRwxmZnXVjoetjwC+C7wD2BqYJGnrVsdhZlZX7Wjx7wrcGxH3RcQLwIXAQW2Iw8yslhQRra1QOhiYGBEfyvPvB94YER9v2G4KMCXPjgfuGWCVY4DFA9y3U9XxmKGex+1jroeBHvPGEdHVuHDk4OOpRkRMA6YNthxJMyNiwhCE1DHqeMxQz+P2MdfDUB9zO7p6HgI2KsxvmJeZmVkLtCPx3wJsIWkTSasAhwEz2hCHmVkttbyrJyKWSvo48DtgBHB2RNxVYZWD7i7qQHU8ZqjncfuY62FIj7nlF3fNzKy9/M1dM7OaceI3M6uZYZP4+xsGQtIoSRfl9TdJ6m5DmEOqxDF/UtJcSXMk/UHSxu2IcyiVHe5D0nskhaSOv+2vzDFLem/+Xd8l6SetjrEKJf6+x0m6StJt+W98/3bEOVQknS1poaQ7m6yXpG/n8zFH0s4DriwiOv5Fukj8P8CmwCrA7cDWDdt8DDgrTx8GXNTuuFtwzG8FVs/TR9fhmPN2awLXADcCE9oddwt+z1sAtwFr5/l12x13i457GnB0nt4amN/uuAd5zG8GdgbubLJ+f+C3gIDdgJsGWtdwafGXGQbiIODcPP1zYF9JamGMQ63fY46IqyLiuTx7I+k7E52s7HAf/xf4EvCPVgZXkTLH/GHguxHxBEBELGxxjFUoc9wB/EueXgv4WwvjG3IRcQ3weB+bHAT8VyQ3AqMlrT+QuoZL4t8AeLAwvyAv63WbiFgKLAFe25LoqlHmmIuOIrUWOlm/x5w//m4UEZe1MrAKlfk9bwlsKenPkm6UNLFl0VWnzHGfChwuaQHwG+CY1oTWNsv7nm9qhR2ywYaOpMOBCcBb2h1LlSStBHwdmNzmUFptJKm7Z2/Sp7prJG0XEU+2M6gWmARMj4ivSdod+LGkbSPipXYHtqIbLi3+MsNA/HMbSSNJHw0fa0l01Sg19IWk/YCTgQMj4vkWxVaV/o55TWBb4GpJ80n9oDM6/AJvmd/zAmBGRLwYEfOA/yb9I+hkZY77KOCnABFxA7AqaTCz4WrIhrsZLom/zDAQM4AP5umDgT9GvmLSofo9Zkk7Ad8nJf3h0O/b5zFHxJKIGBMR3RHRTbqucWBEzGxPuEOizN/2r0itfSSNIXX93NfCGKtQ5rgfAPYFkLQVKfEvammUrTUD+EC+u2c3YElEPDyQgoZFV080GQZC0mnAzIiYAfyI9FHwXtIFlMPaF/HglTzmrwBrAD/L17EfiIgD2xb0IJU85mGl5DH/Dni7pLnAMuDTEdHJn2bLHvcJwA8kfYJ0oXdyJzfmJF1A+gc+Jl+3OAVYGSAiziJdx9gfuBd4DjhiwHV18HkyM7MBGC5dPWZmVpITv5lZzTjxm5nVjBO/mVnNOPGbma1g+huwrZftl2uQPid+swGSND/fNz+UZXZLel9hfrKk7wxlHdYRpgOlht6QtAVwErBHRGwDHN/fPk78ZiuWbuB9/W1kw1tvA7ZJ2kzS5ZJmSbpW0uvzquUepM+J34YNSf87j99+naQLJH0qL+/1DSNpeh7f/HpJ90k6uFDWpyXdksc9/3yJug+XdLOk2ZK+L2lEXv6MpC9Iuj0PoLZeIaYbJd0h6XRJz+SizgD2yuV8Ii8bm+P/q6QvD+Eps84yDTgmInYBPgWcmZcv9yB9Tvw2LEh6A/AeYAfgHaRB6Xo0e8MArA/sCRxASrpIejtprJtdgR2BXSS9uY+6twIOJX3U3pH07dn/yKtfA9wYETuQnhHw4bz8W8C3ImI70lg7PaYC10bEjhHxjbxsx1z+dsChkorjtVgNSFoDeBPpW/izSUOx9AzJXBykbxLp28yj+ypvWAzZYAbsAVwSEf8A/iHp1/CqN0zPtqMK+/0qj+Y4t6c1Drw9v27L82uQ3ljXNKl7X2AX4JZcx2pAz8ftF4BL8/Qs4G15enfg3Xn6J8BX+zi2P0TEknw8c4GNeeXwvDb8rQQ8mRsWjRaQHsryIjBPUs8gfbc0K8yJ34a7vt4wAMURS1X4+cWI+H7JOgScGxEn9bLuxcL4McsY2HuuGONAy7AOFhFPSZon6ZCI+JlSC2P7iLidNEjfJOCcsoP0uavHhos/A++StGpu5R8A6Q1DagUdAv98bukO/ZT1O+DIXA6SNpC0bh/b/wE4uGcbSeuo/+cb30jqmoJXDhj4NGl4aauxPGDbDcB4SQskHUXqPjxK0u3AXbz8RLLfAY/lT4NXUWKQPrccbFiIiFskzQDmAI8Cd5CesgbpDfM9SZ8jjXZ4IekZrs3KuiL329+Qu26eAQ7n5e6bxu3n5rKvUHoYzIvA/wLu7yPk44HzJJ0MXF6IdQ6wLL+5pwNP9H3kNhxFxKQmq1514TZ/ovxkfpXi0Tlt2JC0RkQ8I2l1Un/8lIi4td1x9SbH+PeICEmHAZMiorfnB5sNObf4bTiZJmlr0gM5zl1Rk362C/Cd3Ff7JHBke8OxOnGL38ysZnxx18ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGb+PxoH2CpMdXS9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize distribution of gene lengths\n",
    "print('adhesome gene lengths quartile = '+str(np.quantile(df_loc['txLength'].values, [0.25,0.5,0.75])))\n",
    "plt.figure()\n",
    "plt.hist(df_loc['txLength'], bins=np.arange(0,1000000,5000))\n",
    "plt.xlabel('gene length')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of adhesome gene lengths')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define gene window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define region 1kb upstream of transcripts\n",
    "df_loc['gene_window_start'] = [df_loc['txStart'].iloc[i]-1000 if (df_loc['strand'].iloc[i]=='+') \n",
    "                             else df_loc['txStart'].iloc[i] for i in range(df_loc.shape[0])]\n",
    "df_loc['gene_window_end'] = [df_loc['txEnd'].iloc[i] if (df_loc['strand'].iloc[i]=='+') \n",
    "                             else df_loc['txEnd'].iloc[i]+1000 for i in range(df_loc.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>isTF</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>uc001isx.3</td>\n",
       "      <td>ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>False</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27151016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>uc002uzz.3</td>\n",
       "      <td>ABI2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>False</td>\n",
       "      <td>204192002</td>\n",
       "      <td>204296892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>uc002iop.1</td>\n",
       "      <td>ABI3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>False</td>\n",
       "      <td>47286588</td>\n",
       "      <td>47300587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>uc004bzv.3</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>False</td>\n",
       "      <td>133588267</td>\n",
       "      <td>133763062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>uc003sot.4</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>False</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5571232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript  gene  chrom strand    txStart      txEnd  txLength   isTF  \\\n",
       "131  uc001isx.3  ABI1  chr10      -   27035524   27150016    114492  False   \n",
       "132  uc002uzz.3  ABI2   chr2      +  204193002  204296892    103890  False   \n",
       "133  uc002iop.1  ABI3  chr17      +   47287588   47300587     12999  False   \n",
       "135  uc004bzv.3  ABL1   chr9      +  133589267  133763062    173795  False   \n",
       "228  uc003sot.4  ACTB   chr7      -    5566778    5570232      3454  False   \n",
       "\n",
       "     gene_window_start  gene_window_end  \n",
       "131           27035524         27151016  \n",
       "132          204192002        204296892  \n",
       "133           47286588         47300587  \n",
       "135          133588267        133763062  \n",
       "228            5566778          5571232  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Select CAGE peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process robust CAGE peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique CAGE peaks = 201802\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>gene</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95563</th>\n",
       "      <td>p1@MTND1P23</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>129bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564571</td>\n",
       "      <td>564600</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95564</th>\n",
       "      <td>p3@MTND1P23</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>197bp_to_ENST00000416931_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564639</td>\n",
       "      <td>564649</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95565</th>\n",
       "      <td>p3@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>246bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565266</td>\n",
       "      <td>565278</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95566</th>\n",
       "      <td>p4@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>458bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565478</td>\n",
       "      <td>565483</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95567</th>\n",
       "      <td>p1@MTND2P28</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>489bp_to_ENST00000457540_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>565509</td>\n",
       "      <td>565541</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      short_description      gene    association_with_transcript chrom  \\\n",
       "95563       p1@MTND1P23  MTND1P23  129bp_to_ENST00000416931_5end  chr1   \n",
       "95564       p3@MTND1P23  MTND1P23  197bp_to_ENST00000416931_5end  chr1   \n",
       "95565       p3@MTND2P28  MTND2P28  246bp_to_ENST00000457540_5end  chr1   \n",
       "95566       p4@MTND2P28  MTND2P28  458bp_to_ENST00000457540_5end  chr1   \n",
       "95567       p1@MTND2P28  MTND2P28  489bp_to_ENST00000457540_5end  chr1   \n",
       "\n",
       "       cage_start  cage_end strand peak_id  \n",
       "95563      564571    564600      +       1  \n",
       "95564      564639    564649      +       3  \n",
       "95565      565266    565278      +       3  \n",
       "95566      565478    565483      +       4  \n",
       "95567      565509    565541      +       1  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load annotated FANTOM5 CAGE peaks\n",
    "cage_df = pd.read_csv(fantom5_dir+'hg19.cage_peak_phase1and2combined_ann.txt', sep='\\t', skiprows=7, header=0)\n",
    "# Define new columns\n",
    "cage_df['gene'] = cage_df['short_description'].str.split('@',expand=True)[1]\n",
    "cage_df['chrom'] = cage_df['00Annotation'].str.split(':', expand=True)[0]\n",
    "cage_df['cage_start'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[0].astype(int)\n",
    "cage_df['cage_end'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[2].astype(int)\n",
    "cage_df['strand'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[1]\n",
    "cage_df['peak_id'] = cage_df['short_description'].str.split('@', expand=True)[0].str.strip('p')\n",
    "# Reformat dataframe\n",
    "cage_df = cage_df[['short_description', 'gene','association_with_transcript',\n",
    "                   'chrom','cage_start','cage_end','strand','peak_id']]\n",
    "cage_df = cage_df.sort_values(by=['chrom','cage_start'])\n",
    "print('Number of unique CAGE peaks = '+str(len(np.unique(cage_df['short_description'].values))))\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>strand_gene</th>\n",
       "      <th>strand_prom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>CALR</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene strand_gene strand_prom\n",
       "279  CALR           +           -"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check which genes are located on the opposite strand of their promoter (mislabeling error)\n",
    "# The only problematic error was on PDPK1, which has been taken care of before\n",
    "merge_df = pd.merge(df_loc[['gene','strand']],cage_df[['gene','strand']], on=['gene'], suffixes=['_gene','_prom'])\n",
    "merge_df[merge_df['strand_gene'] != merge_df['strand_prom']].drop_duplicates(subset=['gene'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CAGE peaks within selected gene windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bed file for regions upstream of genes of interest\n",
    "gene_window_df = df_loc[['chrom','gene_window_start','gene_window_end','gene','strand']]\n",
    "gene_window_df['score'] = 0\n",
    "gene_window_df = gene_window_df[['chrom','gene_window_start','gene_window_end','gene','score','strand']]\n",
    "gene_window_bed = pybedtools.BedTool.from_dataframe(gene_window_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a bed file for robust CAGE peaks\n",
    "cage_peak_df = cage_df[['chrom','cage_start','cage_end','short_description','peak_id','strand']]\n",
    "cage_bed = pybedtools.BedTool.from_dataframe(cage_peak_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get CAGE peaks for each gene of interest\n",
    "out = pybedtools.bedtool.BedTool.map(gene_window_bed, cage_bed, c=4, o='distinct', s=True)\n",
    "out_df = out.to_dataframe()\n",
    "out_df.columns=['chrom','gene_window_start','gene_window_end','gene','score','strand','peaks']\n",
    "out_df = out_df[['chrom','gene_window_start','gene_window_end','gene','strand','peaks']]\n",
    "out_df['peaks'] = out_df['peaks'].str.replace(',p',';p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat matching results\n",
    "peaks_per_gene_df = pd.merge(df_loc, out_df, on=['chrom','gene_window_start','gene_window_end','gene','strand'])\n",
    "peaks_per_gene_df = peaks_per_gene_df[peaks_per_gene_df['peaks'] != '.']\n",
    "peaks_per_gene_df['peaks'] = peaks_per_gene_df['peaks'].str.split(';')\n",
    "peaks_per_gene_df['n_peaks'] = [len(peaks) for peaks in peaks_per_gene_df['peaks']]\n",
    "peaks_per_gene_df\n",
    "peaks_per_gene_df = peaks_per_gene_df[['gene', 'isTF', 'chrom', 'strand', 'txStart', 'txEnd', 'txLength',\n",
    "                                       'gene_window_start', 'gene_window_end','peaks','n_peaks']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isTF</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "      <th>peaks</th>\n",
       "      <th>n_peaks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27150016</td>\n",
       "      <td>114492</td>\n",
       "      <td>27035524</td>\n",
       "      <td>27151016</td>\n",
       "      <td>[p1@ABI1, p2@ABI1, p3@ABI1, p4@ABI1, p@chr10:2...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>204193002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>103890</td>\n",
       "      <td>204192002</td>\n",
       "      <td>204296892</td>\n",
       "      <td>[p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>47287588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>12999</td>\n",
       "      <td>47286588</td>\n",
       "      <td>47300587</td>\n",
       "      <td>[p1@ABI3, p2@ABI3, p3@ABI3, p4@ABI3, p5@ABI3, ...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>False</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>133589267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>173795</td>\n",
       "      <td>133588267</td>\n",
       "      <td>133763062</td>\n",
       "      <td>[p1@ABL1, p1@AF533988, p1@M14755, p2@ABL1, p3@...</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>False</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5570232</td>\n",
       "      <td>3454</td>\n",
       "      <td>5566778</td>\n",
       "      <td>5571232</td>\n",
       "      <td>[p10@ACTB, p11@ACTB, p12@ACTB, p13@ACTB, p14@A...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene   isTF  chrom strand    txStart      txEnd  txLength  \\\n",
       "0  ABI1  False  chr10      -   27035524   27150016    114492   \n",
       "1  ABI2  False   chr2      +  204193002  204296892    103890   \n",
       "2  ABI3  False  chr17      +   47287588   47300587     12999   \n",
       "3  ABL1  False   chr9      +  133589267  133763062    173795   \n",
       "4  ACTB  False   chr7      -    5566778    5570232      3454   \n",
       "\n",
       "   gene_window_start  gene_window_end  \\\n",
       "0           27035524         27151016   \n",
       "1          204192002        204296892   \n",
       "2           47286588         47300587   \n",
       "3          133588267        133763062   \n",
       "4            5566778          5571232   \n",
       "\n",
       "                                               peaks  n_peaks  \n",
       "0  [p1@ABI1, p2@ABI1, p3@ABI1, p4@ABI1, p@chr10:2...       10  \n",
       "1  [p1@ABI2, p2@ABI2, p3@ABI2, p4@ABI2, p5@ABI2, ...       13  \n",
       "2  [p1@ABI3, p2@ABI3, p3@ABI3, p4@ABI3, p5@ABI3, ...        9  \n",
       "3  [p1@ABL1, p1@AF533988, p1@M14755, p2@ABL1, p3@...       14  \n",
       "4  [p10@ACTB, p11@ACTB, p12@ACTB, p13@ACTB, p14@A...       28  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "peaks_per_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdNElEQVR4nO3deZgdVZnH8e+PBEnCFmIaJISmIyAKiIDNouCAwDhsM9ERBRQwgGTEYRMQYXAGZGAEAcVtwMiqYBTZBQdBBFlkkYSwJIAgCZCFfYuAQMI7f9RpKC69VLq77u176/d5nvuk1nPeunXz9rmn6p5SRGBmZtWxVKMDMDOz+nLiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgn/jqQNEfSdj2s+4SkB+sd01CizDmSnpd0R4NjmSTp5gbWv7+kJyX9TdJ761jvuZKOr1d91ljDGx1A1UXETcA6fW0n6VhgrYjYo/Sg6m9L4B+B8RHxcqODaRRJSwPfBTaPiLsbHY+1Lrf4DUmNbgCsAcxptaTfj/d1FWAEMLOEcJrKEPhMtjQn/vrZUNI9kl6U9CtJIwAkbS1pbtdGkr4haZ6khZIelLStpO2B/wB2TV0Ad6dtx0m6QtJzkh6WtF+unJGSzkvdJ/dLOqKmnjmprnuAlyUNl3SkpL+mumdJ+kxu+0mSbpH0PUkvSHpE0sfT8sclPSXpSz0dfE+xStoXOBP4WDq2b3Wz7yRJN0s6JR3PbEk71BzLdrn5YyWdn6Y7JIWkvVOcz0v6iqRN0vl4QdKP3l2lfpTO1QOSts2tWFHSWZIWpPN0vKRh3bxHzwLHdnMsy0g6TdL89DotLfsA0NXl94KkP3Szb9exTE77LpB0eG79Urlz+KykCyWNya3/taQn0nHdKGm9Hs7V8pKul/QDZXZMn4eF6ZgP72G/ruMv673r6zM9TtLFkp5On5GDcuuOTe/Hz9JxzJTUWWTflhQRfpX8AuYAdwDjgDHA/cBX0rqtgblpeh3gcWBcmu8A1kzTxwLn15R7I/C/ZK3EDYGngW3SuhOBPwIrAeOBe7rqycU0A1gdGJmWfS7FuBSwK/AysGpaNwlYBOwNDAOOBx4DfgwsA3wKWAgs18N70Fusk4Cbe3n/JgFvAPuluvcH5gPKHct2ue3feq/SexjAGanuTwF/By4DVgZWA54Ctqo5zq8BS6f34UVgTFp/KfATYNm0/x3Av9XseyBZN+rIbo7lOOC2tG8b8Cfgv2tiHd7D+9C1fmqq/8PpfdwurT84lT0+nZOfAFNz++8DLJ/WnQbMyK07N53T96ZjOj63bgHwiTS9ErBxL+epzPeux8802Wd2GvBfwHuA9wOPAP+U+0z8HdiR7DP0beC2Ivu24qvhAVThRZaY9sjNfwc4I01vnfvwrkWWhLYDlq4p41hyiZ8sYS8Gls8t+zZwbpp+xwcX+DLvTvz79BH3DGBimp4EPJRb92GyJLRKbtmzwIbdlNNXrJPoO/E/nJsflep+X+5Y+kr8q9XEuWtu/mLgkFxdb/1RScvuAPYk64p5jVxSAnYHrs/t+1gf7+lfgR1z8/9E1s2Vj7WvxP/Bms/SWWn6fmDb3LpVyf5gvqs8YHQqa8U0fy5wNnAf8PWabR8D/g1YoY9jK/u96/EzDWxWuz9wFHBO7jPx+9y6dYFXi+zbii/3o9XPE7npV8ha1u8QEQ9LOoTsQ7qepN8Bh0bE/G7KGwc8FxELc8seBTpz6x/PrctPd7tM0l7AoWQJBmA5YGxukydz06+mmGuXLdePWIt46/2LiFckdcVXVG2cvcU9L9L//uRRsmNYg6wluyDVD1lrsa/3OW9cKq+27CWRr+NRsj/CpPgulfRmbv1iYBVJTwAnkH2rawO6thlL1ioH2An4G9m3o7zPAt8ETlTWNXhkRNzaQ2xlv3c9bb8GME7SC7llw4CbcvO1/wdHKLuWUGTfluI+/iEmIn4REVuSfRgDOKlrVc2m84ExkpbPLWsH5qXpBWRfh7us3l11XROS1gB+ChwAvDciRpO1/tTNfkuqr1gH6mWybwFd3jfA8lZTLjuRxTqfLNG8BoyNiNHptUJE5PvK+xrudj7Zua0te0nkz2V+/8eBHXKxjY6IERExD/gCMJHs2+SKvP3HPX+cPwWuBn4radmuhRHx54iYSNY9cxlwYS+xlfne9faZfhyYXXPsy0fEjn2UOdB9m5IT/xAiaR1J20hahqw/8lXebpk9CXRIWgogIh4n6x/+tqQRkjYA9gXOT9tfCBwlaSVJq5El9N4sS/Yf7+kUy97A+oNxXAViHagZwG6Slk4X7HYZYHkrAwel8j4HfAj4bUQsAK4BTpW0QrqYuqakrZag7KnANyW1SRpL1q+8pO/Df0oalS7O7g38Ki0/Azgh/REn1TExrVueLPE+S/ZH8n96KPsAsovMv0kXU98j6YuSVoyIN4CXePsz2Z0y37vePtN3AAuV3bAwUtIwSetL2qRAuQPZtyk58Q8ty5BdwHqG7GvpymR9jQC/Tv8+K2l6mt6drOU2n+zC2TER8fu07jhgLjAb+D1wEdl//G5FxCzgVOBWsj8yHwZuGYyDKhDrQP0nsCbwPPAt4BcDLO92YG2y83ACsEtEPJvW7UV2AXBWqu8isr70oo4H7iS7MHkvMD0tWxJ/BB4GrgNOiYhr0vLvA1cA10haSHahd7O07mdk3S7zUuy3dVdw6qaZTPbZuZzsgviewBxJLwFfAb7YS2xlvnc9fqYjYjGwM9mNA7NT/WeSfbvp1UD2bVZdd0VYi5O0P7BbRCxJC8uGEEkdZIlp6YhY1OBw3kXSJODLqauyHvX5M91PbvG3KEmrStoifaVeBziMrKVt1pT8mR48vqundb2H7J7pCcALwC/J7qM3a1b+TA8Sd/WYmVWMu3rMzCqmKbp6xo4dGx0dHY0Ow8ysqUybNu2ZiGirXd4Uib+jo4M777yz0WGYmTUVSY92t9xdPWZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78ZmYV0xS/3K2XjiOvemt6zok7NTASM7PyuMVvZlYxTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjGlJX5JZ0t6StJ93aw7TFJIGltW/WZm1r0yW/znAtvXLpS0OvAp4LES6zYzsx6Ulvgj4kbguW5WfQ84Aoiy6jYzs57VtY9f0kRgXkTcXc96zczsbXUbnVPSKOA/yLp5imw/GZgM0N7eXmJkfcuP2pnnETzNrBnVs8W/JjABuFvSHGA8MF3S+7rbOCKmRERnRHS2tbXVMUwzs9ZWtxZ/RNwLrNw1n5J/Z0Q8U68YzMys3Ns5pwK3AutImitp37LqMjOz4kpr8UfE7n2s7yirbjMz65l/uWtmVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFlPmz9bElPSbovt+xkSQ9IukfSpZJGl1W/mZl1r8wW/7nA9jXLrgXWj4gNgL8AR5VYv5mZdaO0xB8RNwLP1Sy7JiIWpdnbgPFl1W9mZt0b3sC69wF+1dNKSZOByQDt7e31iqlfOo686q3pOSfu1MBIzMz61pCLu5KOBhYBF/S0TURMiYjOiOhsa2urX3BmZi2u7i1+SZOAnYFtIyLqXb+ZWdXVNfFL2h44AtgqIl6pZ91mZpYp83bOqcCtwDqS5kraF/gRsDxwraQZks4oq34zM+teaS3+iNi9m8VnlVWfmZkV41/umplVjBO/mVnFOPGbmVWME7+ZWcU48ZuZVYwTv5lZxTjxm5lVjBO/mVnFOPGbmVWME7+ZWcX0mfglrSlpmTS9taSD/MhEM7PmVaTFfzGwWNJawBRgdeAXpUZlZmalKZL430yPS/wM8MOI+DqwarlhmZlZWYok/jck7Q58CbgyLVu6vJDMzKxMRRL/3sDHgBMiYrakCcDPyw3LzMzK0ud4/BExS9I3gPY0Pxs4qezAzMysHEXu6vlnYAZwdZrfUNIVJcdlZmYlKdLVcyywKfACQETMAN5fWkRmZlaqQhd3I+LFmmVvlhGMmZmVr0jinynpC8AwSWtL+iHwp752knS2pKck3ZdbNkbStZIeSv+uNIDYzcysH4ok/gOB9YDXgKnAS8AhBfY7F9i+ZtmRwHURsTZwXZo3M7M6KnJXzyvA0elVWETcKKmjZvFEYOs0fR5wA/CNJSnXzMwGps/EL+kDwOFAR377iNimH/WtEhEL0vQTwCq91DsZmAzQ3t7ej6rK13HkVYO2/5wTdxpoOGZmhfSZ+IFfA2cAZwKLB6viiAhJ0cv6KWRjA9HZ2dnjdmZmtmSKJP5FEXH6INX3pKRVI2KBpFWBpwapXDMzK6jIxd3fSPqqpFXTXTljJI3pZ31XkI35Q/r38n6WY2Zm/VSkxd+VqL+eWxb08SMuSVPJLuSOlTQXOAY4EbhQ0r7Ao8DnlzRgMzMbmCJ39UzoT8ERsXsPq7btT3lmZjY4iozVM0rSNyVNSfNrS9q5/NDMzKwMRfr4zwFeBz6e5ucBx5cWkZmZlapI4l8zIr4DvAFv/aBLpUZlZmalKZL4X5c0kuyCLpLWJBu+wczMmlCRu3qOIRuLf3VJFwBbAJPKDMrMzMpT5K6eayVNBzYn6+I5OCKeKT0yMzMrRZGxejZOk11j7LRLWhF4NCIWlRaZmZmVokhXz/8CGwP3kLX41wdmAitK2j8irikxPjMzG2RFEv98YN+ImAkgaV3gOOAI4BKgKRN/XyNjDnTkzZ7K8yicZtZoRe7q+UBX0geIiFnAByPikfLCMjOzshRp8c+UdDrwyzS/KzBL0jKke/vNzKx5FGnxTwIeJnvc4iHAI2nZG8AnywnLzMzKUuR2zleBU9Or1t8GPSIzMytVkRa/mZm1ECd+M7OK6THxS/p5+vfg+oVjZmZl663F/1FJ44B9JK2Uf+ziAB69aGZmDdbbxd0zgOvIHrE4jXcOxdznoxfNzGxo6rHFHxE/iIgPAWdHxPsjYkLuNaCkL+lrkmZKuk/SVEkjBlKemZkV1+fF3YjYX9JHJB2QXhsMpEJJqwEHAZ0RsT4wDNhtIGWamVlxRZ65exBwAbByel0g6cAB1jscGClpODCKbDwgMzOrgyJDNnwZ2CwiXgaQdBJwK/DD/lQYEfMknQI8BrwKXNPdCJ+SJgOTAdrb2/tTlQ1AX4PYmVnzKnIfv4DFufnFDOCZu5JWAiYCE4BxwLKS9qjdLiKmRERnRHS2tbX1tzozM6tRpMV/DnC7pEvT/KeBswZQ53bA7Ih4GkDSJcDHgfMHUKaZmRVUZKye70q6AdgyLdo7Iu4aQJ2PAZtLGkXW1bMtcOcAyjMzsyVQpMVPREwHpg9GhRFxu6SLUnmLgLuAKYNRtpmZ9a1Q4h9sEXEMcEwj6jYzqzoP0mZmVjG9Jn5JwyRdX69gzMysfL0m/ohYDLwpacU6xWNmZiUr0sf/N+BeSdcCL3ctjIiDSovKzMxKUyTxX5JeZmbWAorcx3+epJFAe0Q8WIeYzMysREUGaftnYAZwdZrfUNIVJcdlZmYlKXI757HApsALABExAz+ExcysaRXp438jIl6U3jEu25slxdP08qNa9mf9YMfhkTXNrFaRxD9T0heAYZLWJnuIyp/KDcvMzMpSpKvnQGA94DVgKvAScEiJMZmZWYmK3NXzCnB0egBLRMTC8sMyM7OyFLmrZxNJ9wL3kP2Q625JHy0/NDMzK0ORPv6zgK9GxE0AkrYkezjLgB66bmZmjVGkj39xV9IHiIibycbRNzOzJtRji1/Sxmnyj5J+QnZhN4BdgRvKD83MzMrQW1fPqTXz+QenRAmxmJlZHfSY+CPik/UMxMzM6qPPi7uSRgN7AR357T0ss5lZcypyV89vgduAexmkoRrSH5MzgfXJuo32iYhbB6NsMzPrXZHEPyIiDh3ker8PXB0Ru0h6DzBqkMs3M7MeFEn8P5e0H3Al2bANAETEc/2pMD3G8R+ASamc14HX+1OWmZktuSKJ/3XgZOBo3r6bJ+j/0MwTgKeBcyR9BJgGHBwRL+c3kjQZmAzQ3t7ez6qaX3ejefY04ma9Rv40s+ZW5AdchwFrRURHRExIr4GMxz8c2Bg4PSI2InuO75G1G0XElIjojIjOtra2AVRnZmZ5RRL/w8Arg1jnXGBuRNye5i8i+0NgZmZ1UKSr52VghqTreWcff79u54yIJyQ9Lmmd9AzfbYFZ/SnLzMyWXJHEf1l6DaYDgQvSHT2PAHsPcvlmZtaDIuPxnzfYlabn9nYOdrlmZta3Ir/cnU03Y/MM8AKvmZk1SJGunnzLfATwOWBMOeGYmVnZ+ryrJyKezb3mRcRpQPc3kpuZ2ZBXpKsnf6vlUmTfAIp8UzAzsyGoSALPj8u/CJgDfL6UaMzMrHRF7urxuPxmZi2kSFfPMsBnefd4/MeVF5aZmZWlSFfP5cCLZIOpvdbHtmZmNsQVSfzjI2L70iNpoKEwqmU+hp5G3+xrv3rXbWbNqcggbX+S9OHSIzEzs7oo0uLfEpiUfsH7GiAgImKDUiMzM7NSFEn8O5QehZmZ1U2R2zkfrUcgZmZWH0X6+M3MrIU48ZuZVYwTv5lZxTjxm5lVjBO/mVnFNCzxSxom6S5JVzYqBjOzKmpki/9g4P4G1m9mVkkNSfySxpM9xevMRtRvZlZljWrxnwYcAbzZoPrNzCqr7o9QlLQz8FRETJO0dS/bTQYmA7S3tw9K3UNhFM7BsCTH4ZE3zaxWI1r8WwD/ImkO8EtgG0nn124UEVMiojMiOtva2uodo5lZy6p74o+IoyJifER0ALsBf4iIPeodh5lZVfk+fjOziql7H39eRNwA3NDIGMzMqsYtfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrmIb+gMu6V9ZgcvUYsM2DwpkNfW7xm5lVjBO/mVnFOPGbmVWME7+ZWcU48ZuZVYwTv5lZxTjxm5lVjBO/mVnFOPGbmVWME7+ZWcXUPfFLWl3S9ZJmSZop6eB6x2BmVmWNGKtnEXBYREyXtDwwTdK1ETGrAbGYmVVO3Vv8EbEgIqan6YXA/cBq9Y7DzKyqGjo6p6QOYCPg9m7WTQYmA7S3t9c3sApYkhFAexpx0yNxmjWnhl3clbQccDFwSES8VLs+IqZERGdEdLa1tdU/QDOzFtWQxC9pabKkf0FEXNKIGMzMqqoRd/UIOAu4PyK+W+/6zcyqrhEt/i2APYFtJM1Irx0bEIeZWSXV/eJuRNwMqN71mplZxr/cNTOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKaejonNba+hoBtMiInt2V0d+RQHuKZzDK627U0p5GMu1v3d3VN9gjpJZVnkdvHVrc4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKaUjil7S9pAclPSzpyEbEYGZWVXVP/JKGAT8GdgDWBXaXtG694zAzq6pGtPg3BR6OiEci4nXgl8DEBsRhZlZJioj6VijtAmwfEV9O83sCm0XEATXbTQYmp9l1gAf7WeVY4Jl+7tsMfHzNzcfXvJrh2NaIiLbahUN2WOaImAJMGWg5ku6MiM5BCGlI8vE1Nx9f82rmY2tEV888YPXc/Pi0zMzM6qARif/PwNqSJkh6D7AbcEUD4jAzq6S6d/VExCJJBwC/A4YBZ0fEzBKrHHB30RDn42tuPr7m1bTHVveLu2Zm1lj+5a6ZWcU48ZuZVUxLJ/5WGhpC0uqSrpc0S9JMSQen5WMkXSvpofTvSo2OdSAkDZN0l6Qr0/wESbenc/irdENAU5I0WtJFkh6QdL+kj7XS+ZP0tfTZvE/SVEkjmvn8STpb0lOS7sst6/Z8KfODdJz3SNq4cZH3rWUTfwsODbEIOCwi1gU2B/49Hc+RwHURsTZwXZpvZgcD9+fmTwK+FxFrAc8D+zYkqsHxfeDqiPgg8BGy42yJ8ydpNeAgoDMi1ie7cWM3mvv8nQtsX7Osp/O1A7B2ek0GTq9TjP3SsomfFhsaIiIWRMT0NL2QLGmsRnZM56XNzgM+3ZAAB4Gk8cBOwJlpXsA2wEVpk6Y9PkkrAv8AnAUQEa9HxAu00Pkju0twpKThwChgAU18/iLiRuC5msU9na+JwM8icxswWtKqdQm0H1o58a8GPJ6bn5uWNT1JHcBGwO3AKhGxIK16AlilUXENgtOAI4A30/x7gRciYlGab+ZzOAF4GjgndWWdKWlZWuT8RcQ84BTgMbKE/yIwjdY5f116Ol9NlW9aOfG3JEnLARcDh0TES/l1kd2b25T350raGXgqIqY1OpaSDAc2Bk6PiI2Al6np1mny87cSWat3AjAOWJZ3d5O0lGY+X62c+FtuaAhJS5Ml/Qsi4pK0+Mmur5Tp36caFd8AbQH8i6Q5ZN1y25D1iY9OXQfQ3OdwLjA3Im5P8xeR/SFolfO3HTA7Ip6OiDeAS8jOaaucvy49na+myjetnPhbamiI1N99FnB/RHw3t+oK4Etp+kvA5fWObTBExFERMT4iOsjO1R8i4ovA9cAuabNmPr4ngMclrZMWbQvMokXOH1kXz+aSRqXPatfxtcT5y+npfF0B7JXu7tkceDHXJTT0RETLvoAdgb8AfwWObnQ8AzyWLcm+Vt4DzEivHcn6wa8DHgJ+D4xpdKyDcKxbA1em6fcDdwAPA78Glml0fAM4rg2BO9M5vAxYqZXOH/At4AHgPuDnwDLNfP6AqWTXK94g+8a2b0/nCxDZXYR/Be4lu7up4cfQ08tDNpiZVUwrd/WYmVk3nPjNzCrGid/MrGKc+M3MKsaJ38ysYpz4rXIk3SCp9IdkSzoojcJ5QYl1zJE0tqzyrTXV/dGLZs1M0vB4e+yZvnwV2C4i5pYZk9mScovfhiRJHam1/NM0xvs1kkamdW+12CWNTcM8IGmSpMvSOOlzJB0g6dA0KNptksbkqthT0ow0dvymaf9l0xjsd6R9JubKvULSH8h+vFMb66GpnPskHZKWnUH246X/k/S1mu0nSbo8HcdDko7Jrdsj1T9D0k/S8OJIOl3Snem9+FY3MYyU9H+S9kvHcZWku1NMu/b/TFgrcuK3oWxt4McRsR7wAvDZAvusD/wrsAlwAvBKZIOi3QrsldtuVERsSNYqPzstO5psqIhNgU8CJ6cRNCEbV2eXiNgqX5mkjwJ7A5uRPSdhP0kbRcRXgPnAJyPie93EuWk6ng2Az0nqlPQhYFdgixTbYuCLXbFFRGfafitJG+TKWg74DTA1In5KNjja/Ij4SGRj419d4H2zCnFXjw1lsyNiRpqeBnQU2Of6yJ5XsFDSi2QJEbKf0eeT5VTIxlyXtIKk0cCnyAaKOzxtMwJoT9PXRkTt2OyQDaVxaUS8DCDpEuATwF19xHltRDyb22dLsoftfBT4czbcDSN5exCwz0uaTPZ/dlWyhwvdk9ZdDnwnIrquJdwLnCrpJLKhL27qIxarGCd+G8pey00vJkuEkCXIrm+rI3rZ583c/Ju88/NeO1ZJkI238tmIeDC/QtJmZMMoD6ae6j8vIo6qqX8CcDiwSUQ8L+lc3nnctwDbS/pFZP6i7NF/OwLHS7ouIo4b5Pitibmrx5rRHLKWMbw98uOS2hVA0pZkIym+CPwOODCNLomkjQqUcxPw6TQq5bLAZ9Kyvvyjsue3jiR7itMtZNcPdpG0cqp/jKQ1gBXI/vC8KGkVssf85f0X2WMNf5z2G0fWxXU+cDJZN5XZW9zit2Z0CnBh6vq4qp9l/F3SXcDSwD5p2X+TPQXsHklLAbOBnXsrJCKmpxb4HWnRmRHRVzcPafuLycZtPz8i7gSQ9E3gmlT/G8C/R8RtKdYHyJ7ydEs35R0MnC3pO2R/QE6W9GYqY/8C8ViFeHROszqTNIls2N4DGh2LVZO7eszMKsYtfjOzinGL38ysYpz4zcwqxonfzKxinPjNzCrGid/MrGL+HzOuAQZPq3e9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Histogram of number of peaks per gene\n",
    "plt.figure()\n",
    "plt.hist(peaks_per_gene_df['n_peaks'], bins=np.arange(0,110,1))\n",
    "plt.xlabel('number of peaks')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of number of peaks per gene')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isTF</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>txLength</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "      <th>peaks</th>\n",
       "      <th>n_peaks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, isTF, chrom, strand, txStart, txEnd, txLength, gene_window_start, gene_window_end, peaks, n_peaks]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_per_gene_df[peaks_per_gene_df['gene']=='PDPK1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Associate CAGE peaks with genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Dataframe of genes for each peak\n",
    "genes_per_peak = peaks_per_gene_df[['gene','peaks']]\n",
    "genes_per_peak = lu.unnesting(genes_per_peak,['peaks'])\n",
    "genes_per_peak = genes_per_peak.groupby('peaks')['gene'].apply(list).reset_index(name='genes')\n",
    "genes_per_peak['n_genes'] = [len(genes) for genes in genes_per_peak['genes']]\n",
    "genes_per_peak.columns = ['short_description','genes','n_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add peak location\n",
    "genes_per_peak = pd.merge(genes_per_peak, cage_df, on=['short_description'])\n",
    "genes_per_peak = genes_per_peak[['short_description','chrom','cage_start','cage_end','strand','genes']]\n",
    "genes_per_peak = genes_per_peak.sort_values(by=['chrom'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consider a window around each peak\n",
    "genes_per_peak['cage_window_start'] = [genes_per_peak['cage_start'].iloc[i]-400 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                       else genes_per_peak['cage_start'].iloc[i]-50 for i in range(genes_per_peak.shape[0])]\n",
    "genes_per_peak['cage_window_end'] = [genes_per_peak['cage_end'].iloc[i]+50 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                     else genes_per_peak['cage_end'].iloc[i]+400 for i in range(genes_per_peak.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genes</th>\n",
       "      <th>cage_window_start</th>\n",
       "      <th>cage_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1674</th>\n",
       "      <td>p7@SHC1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154943559</td>\n",
       "      <td>154943605</td>\n",
       "      <td>-</td>\n",
       "      <td>[SHC1]</td>\n",
       "      <td>154943509</td>\n",
       "      <td>154944005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>p31@NEXN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>78383264</td>\n",
       "      <td>78383274</td>\n",
       "      <td>+</td>\n",
       "      <td>[NEXN]</td>\n",
       "      <td>78382864</td>\n",
       "      <td>78383324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>p4@CAPN2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>223900286</td>\n",
       "      <td>223900299</td>\n",
       "      <td>+</td>\n",
       "      <td>[CAPN2]</td>\n",
       "      <td>223899886</td>\n",
       "      <td>223900349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1262</th>\n",
       "      <td>p4@ENAH</td>\n",
       "      <td>chr1</td>\n",
       "      <td>225742743</td>\n",
       "      <td>225742761</td>\n",
       "      <td>-</td>\n",
       "      <td>[ENAH]</td>\n",
       "      <td>225742693</td>\n",
       "      <td>225743161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>p30@ARHGEF2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155940369</td>\n",
       "      <td>155940382</td>\n",
       "      <td>-</td>\n",
       "      <td>[ARHGEF2]</td>\n",
       "      <td>155940319</td>\n",
       "      <td>155940782</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     short_description chrom  cage_start   cage_end strand      genes  \\\n",
       "1674           p7@SHC1  chr1   154943559  154943605      -     [SHC1]   \n",
       "957           p31@NEXN  chr1    78383264   78383274      +     [NEXN]   \n",
       "1247          p4@CAPN2  chr1   223900286  223900299      +    [CAPN2]   \n",
       "1262           p4@ENAH  chr1   225742743  225742761      -     [ENAH]   \n",
       "946        p30@ARHGEF2  chr1   155940369  155940382      -  [ARHGEF2]   \n",
       "\n",
       "      cage_window_start  cage_window_end  \n",
       "1674          154943509        154944005  \n",
       "957            78382864         78383324  \n",
       "1247          223899886        223900349  \n",
       "1262          225742693        225743161  \n",
       "946           155940319        155940782  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize\n",
    "genes_per_peak.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Create sequence files for all CAGE windows"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over all CAGE windows to generate sequence files\n",
    "for i in tqdm(range(genes_per_peak.shape[0])):\n",
    "    time.sleep(.01)\n",
    "    # Peak characteristics\n",
    "    peak = genes_per_peak['short_description'].iloc[i]\n",
    "    chrom = genes_per_peak['chrom'].iloc[i]\n",
    "    start = genes_per_peak['cage_window_start'].iloc[i]\n",
    "    end = genes_per_peak['cage_window_end'].iloc[i]\n",
    "    # Generate window sequence\n",
    "    result = subprocess.run(cage_fasta_dir+'twoBitToFa http://hgdownload.cse.ucsc.edu/gbdb/hg19/hg19.2bit '+\n",
    "                            cage_fasta_dir+peak+'.fa'+\n",
    "                            ' -seq='+chrom+' -start='+str(start)+' -end='+str(end)+' -noMask',\n",
    "                            shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Search for TFBS in all CAGE windows\n",
    "See process_JASPAR_with_strand.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Find TFs binding each CAGE window"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Useful data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load JASPAR TFBS metadata\n",
    "jaspar_df = pd.read_csv(tf_dir+'JASPAR-HomoSapiens.csv', \n",
    "                        header=0, usecols=['ID','Name','Species','Class','Family'])\n",
    "jaspar_tfs = np.unique(jaspar_df['Name'])\n",
    "# Mapping JASPAR ID to HGNC\n",
    "jasparid_to_hgnc = {jaspar_df.iloc[i,0]:jaspar_df.iloc[i,1] for i in range(jaspar_df.shape[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique TFs = 639\n",
      "Number of unique TF motifs = 810\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Class</th>\n",
       "      <th>Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0002.1</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Runt domain factors</td>\n",
       "      <td>Runt-related factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0003.1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0003.2</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0003.3</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID    Name       Species                                  Class  \\\n",
       "0  MA0002.1   RUNX1  Homo sapiens                    Runt domain factors   \n",
       "1  MA0003.1  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "2  MA0003.2  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "3  MA0003.3  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "4  MA0003.4  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "\n",
       "                 Family  \n",
       "0  Runt-related factors  \n",
       "1                  AP-2  \n",
       "2                  AP-2  \n",
       "3                  AP-2  \n",
       "4                  AP-2  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize data\n",
    "print('Number of unique TFs = '+str(len(np.unique(jaspar_df['Name']))))\n",
    "print('Number of unique TF motifs = '+str(len(np.unique(jaspar_df['ID']))))\n",
    "jaspar_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load MOODS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary mapping CAGE peaks to the corresponding TFs\n",
    "tfs_associated_to_cage_peaks = {}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over files in moods_out_dir and summarize results\n",
    "for peak_moods_file in tqdm(os.listdir(moods_out_dir)):\n",
    "    time.sleep(.01)\n",
    "    # load results dataframe for that gene\n",
    "    cage_peak = peak_moods_file.strip('.txt')\n",
    "    res_df = pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                         usecols=[0,1,2,3,4,5], \n",
    "                         names=['peak_id','tfbs_file','motif_loc','match_strand','match_score','motif'])\n",
    "    if (res_df.shape[0]>0):\n",
    "        # Add columns corresponding to the target peak\n",
    "        res_df['cage_peak'] = cage_peak\n",
    "        res_df['chrom'] = res_df['peak_id'].str.split(':', expand=True)[0]\n",
    "        location_cols = res_df['peak_id'].str.split(':', expand=True)[1]\n",
    "        res_df[['start','stop']] = location_cols.str.split('-', expand=True)\n",
    "        # Add columns corresponding to the TFBS\n",
    "        res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "        res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "        # Only keep TFs matching on the peak strand\n",
    "        peak_strand = cage_df[cage_df['short_description']==cage_peak]['strand'].values[0]\n",
    "        res_df = res_df[res_df['match_strand']==peak_strand]\n",
    "        # Select relevant columns\n",
    "        res_df = res_df[['cage_peak', 'chrom','start','stop','motif_loc',\n",
    "                         'jaspar_tfbs','TF','match_score','match_strand','motif']]\n",
    "        # Store list of TFs associated with the current CAGE peak in dict\n",
    "        tfs_associated_to_cage_peaks[cage_peak] = np.sort(np.unique(res_df['TF'].values))\n",
    "pickle.dump(tfs_associated_to_cage_peaks, open(saving_dir+'tfs_cage_with_strand.pkl', 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Find TFs targeting genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mapping between cage peaks and corresponding TFs\n",
    "tfs_associated_to_cage_peaks = pickle.load(open(saving_dir+'tfs_cage_with_strand.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Construct dataframe with gene-tf pairs\n",
    "tf_and_genes_df = genes_per_peak.copy()\n",
    "tf_and_genes_df['tfs'] = [tfs_associated_to_cage_peaks[cage_peak] if cage_peak in tfs_associated_to_cage_peaks.keys()\n",
    "                          else ['ToDo']\n",
    "                          for cage_peak in tf_and_genes_df['short_description'].values]\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['genes'])\n",
    "gene_to_tfs = {tf_and_genes_df.iloc[i,0]:tf_and_genes_df.iloc[i,8] for i in range(tf_and_genes_df.shape[0])}\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['tfs'])\n",
    "tf_and_genes_df = tf_and_genes_df[['genes','tfs','short_description','chrom','cage_start','cage_end','strand']]\n",
    "tf_and_genes_df = tf_and_genes_df.sort_values(by=['genes','tfs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1019</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ALX3</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1019</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ARGFX</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>BACH2</td>\n",
       "      <td>p@chr10:27149068..27149079,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149068</td>\n",
       "      <td>27149079</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1918</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>p@chr10:27132379..27132382,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27132379</td>\n",
       "      <td>27132382</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>p@chr10:27037479..27037487,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27037479</td>\n",
       "      <td>27037487</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genes    tfs             short_description  chrom  cage_start  cage_end  \\\n",
       "1019  ABI1   ALX3                       p3@ABI1  chr10    27149904  27149920   \n",
       "1019  ABI1  ARGFX                       p3@ABI1  chr10    27149904  27149920   \n",
       "1920  ABI1  BACH2  p@chr10:27149068..27149079,-  chr10    27149068  27149079   \n",
       "1918  ABI1   E2F6  p@chr10:27132379..27132382,-  chr10    27132379  27132382   \n",
       "1915  ABI1   EBF1  p@chr10:27037479..27037487,-  chr10    27037479  27037487   \n",
       "\n",
       "     strand  \n",
       "1019      -  \n",
       "1019      -  \n",
       "1920      -  \n",
       "1918      -  \n",
       "1915      -  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "tf_and_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SHC1': array(['MEIS1', 'MEIS3', 'NFATC4', 'NR2F6(var.3)', 'PBX3', 'RARA(var.2)',\n",
       "        'RARA::RXRA', 'RARA::RXRG'], dtype=object),\n",
       " 'NEXN': array(['MAZ', 'NFATC4', 'POU3F3', 'ZNF140'], dtype=object),\n",
       " 'CAPN2': array(['EWSR1-FLI1', 'NFIB', 'NFIC(var.2)', 'NFIC::TLX1', 'NFIX(var.2)',\n",
       "        'SOX18', 'SPI1', 'ZNF16', 'ZNF354C'], dtype=object),\n",
       " 'ENAH': array(['ALX3', 'ASCL1', 'BHLHA15(var.2)', 'IRF4', 'IRF8', 'LMX1A',\n",
       "        'MIXL1', 'MYOG', 'NHLH1', 'NHLH2', 'PRDM1', 'SPIC', 'STAT3',\n",
       "        'ZNF148', 'ZNF740'], dtype=object),\n",
       " 'ARHGEF2': array(['BACH1', 'ETV4', 'FOSL2', 'MAF::NFE2', 'MAFK', 'NFATC2', 'NFE2L1',\n",
       "        'TBR1'], dtype=object),\n",
       " 'MACF1': array(['CTCF', 'PRDM1', 'SPI1'], dtype=object),\n",
       " 'ITGB3BP': array(['EBF3', 'SPI1'], dtype=object),\n",
       " 'VAV3': array(['BATF', 'BATF3', 'BATF::JUN', 'FOSL1::JUN', 'HOXA4', 'JUN(var.2)',\n",
       "        'JUN::JUNB', 'NFATC2', 'RORA'], dtype=object),\n",
       " 'BCAR3': array(['ELF1', 'ERF', 'ETV1', 'ETV4', 'FOS::JUND', 'HNF1A', 'HNF1B',\n",
       "        'NFIC', 'ZNF135', 'ZNF354C', 'ZNF460'], dtype=object),\n",
       " 'TSPAN1': array(['ATF4', 'CEBPG(var.2)', 'EHF', 'ELF3', 'GABPA', 'KLF5', 'NFATC4',\n",
       "        'RXRA::VDR', 'STAT1::STAT2'], dtype=object),\n",
       " 'RHOU': array(['CTCF', 'DUX4', 'ZNF263'], dtype=object),\n",
       " 'SLC9A1': array(['ETV6', 'MAZ', 'ZNF148'], dtype=object),\n",
       " 'FBLIM1': array(['LEF1', 'MAZ', 'MZF1', 'MZF1(var.2)', 'POU1F1', 'POU2F1', 'POU3F1',\n",
       "        'POU3F2', 'POU3F3', 'POU3F4', 'SOX10', 'SOX4', 'SRF', 'ZNF148'],\n",
       "       dtype=object),\n",
       " 'ARF1': array(['EBF1'], dtype=object),\n",
       " 'CSRP1': array(['POU3F2', 'PRDM4', 'SRF', 'ZNF135'], dtype=object),\n",
       " 'ASAP3': array(['ZNF341'], dtype=object),\n",
       " 'FABP3': array(['EWSR1-FLI1', 'NR1I2', 'NR5A1', 'SPI1', 'ZKSCAN1', 'ZKSCAN5',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'ITGA10': array(['CTCF', 'CTCFL'], dtype=object),\n",
       " 'PTPRF': array([], dtype=object),\n",
       " 'HAX1': array(['EGR1', 'IRF3', 'IRF4', 'IRF7', 'IRF8', 'IRF9', 'KLF11', 'KLF14',\n",
       "        'RXRB', 'RXRG', 'SP3', 'SP9', 'ZNF317'], dtype=object),\n",
       " 'SVIL': array(['CDX1', 'CDX4', 'TFAP4'], dtype=object),\n",
       " 'ABI1': array(['MAZ', 'RREB1', 'THAP11', 'ZNF148', 'ZNF263', 'ZNF384', 'ZNF410'],\n",
       "       dtype=object),\n",
       " 'VIM': array(['FOS::JUN', 'FOS::JUNB', 'FOS::JUND', 'FOSL2::JUN', 'MZF1(var.2)',\n",
       "        'ZNF282'], dtype=object),\n",
       " 'VCL': array(['GABPA', 'ZNF140'], dtype=object),\n",
       " 'LDB3': array(['EBF1', 'EBF3', 'FOS', 'FOS::JUN', 'FOS::JUNB', 'FOS::JUND',\n",
       "        'FOSB::JUNB', 'FOSL2::JUN', 'FOSL2::JUNB', 'SPIB', 'STAT1::STAT2'],\n",
       "       dtype=object),\n",
       " 'NRP1': array(['RBPJ', 'STAT1'], dtype=object),\n",
       " 'MAPK8': array(['MYOD1', 'NHLH2', 'PLAG1'], dtype=object),\n",
       " 'ITGB1': array(['PRDM1', 'SP2', 'ZNF740'], dtype=object),\n",
       " 'PTEN': array(['ETV4', 'NKX2-2', 'SP1', 'SP2', 'SP3', 'SP4', 'SP8', 'TFE3',\n",
       "        'TFEC'], dtype=object),\n",
       " 'SORBS1': array(['CREB1', 'CREM', 'KLF6', 'NFIC'], dtype=object),\n",
       " 'ADAM12': array([], dtype=object),\n",
       " 'ITGA8': array(['CDX2', 'ZNF384'], dtype=object),\n",
       " 'DOCK1': array(['NFATC2', 'NFATC4', 'RELA', 'SPI1', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'RDX': array(['CTCF', 'NFATC3', 'STAT1::STAT2'], dtype=object),\n",
       " 'CFL1': array(['BARHL1', 'KLF4', 'THAP11'], dtype=object),\n",
       " 'CAPN1': array(['ATOH7', 'BHLHE22', 'BHLHE23', 'NEUROG1', 'NEUROG2', 'NFATC2',\n",
       "        'OLIG1', 'OLIG2', 'OLIG3', 'TAL1::TCF3', 'TBX18', 'TBX20',\n",
       "        'TFAP4(var.2)', 'ZNF140'], dtype=object),\n",
       " 'THY1': array(['ZFP42'], dtype=object),\n",
       " 'PAK1': array(['IRF2', 'NFATC2', 'SPI1', 'TBX18', 'ZNF263'], dtype=object),\n",
       " 'ARHGEF12': array([], dtype=object),\n",
       " 'CTTN': array(['EGR1', 'ELF2', 'ELF4', 'ELF5', 'ELK1', 'ELK3', 'ERF', 'ERG',\n",
       "        'ETS1', 'ETS2', 'ETV1', 'ETV2', 'ETV3', 'FEV', 'FLI1', 'HES2',\n",
       "        'IRF4', 'IRF8', 'KLF15', 'KLF5', 'SPDEF', 'SPIC'], dtype=object),\n",
       " 'SLC3A2': array(['ETV4', 'IRF6', 'NFIB', 'NFIC(var.2)', 'NFIC::TLX1', 'NFIX(var.2)'],\n",
       "       dtype=object),\n",
       " 'ILK': array([], dtype=object),\n",
       " 'INPPL1': array(['FOXF2', 'MYF5', 'SMAD2::SMAD3::SMAD4'], dtype=object),\n",
       " 'LAYN': array(['MSC', 'NR2F2', 'RORA', 'ZFP57', 'ZNF460'], dtype=object),\n",
       " 'PPFIA1': array(['EOMES', 'GFI1', 'KLF15', 'KLF5', 'NFYA', 'NFYC', 'TBR1', 'TBX2',\n",
       "        'TBX21', 'TBX6'], dtype=object),\n",
       " 'CORO1B': array(['EHF', 'ELF3', 'GABPA', 'KLF15', 'KLF5', 'SP2', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'LPXN': array(['BCL6', 'BCL6B', 'EGR1', 'ELF1', 'ETV1', 'IKZF1', 'KLF11', 'KLF15',\n",
       "        'KLF16', 'KLF5', 'SP2', 'SP9', 'SPI1', 'ZNF148'], dtype=object),\n",
       " 'CD151': array([], dtype=object),\n",
       " 'ARHGAP32': array(['CTCF', 'LHX5', 'NKX6-1', 'NR1D1', 'RORA(var.2)', 'RORB', 'RORC',\n",
       "        'TFAP2B(var.2)'], dtype=object),\n",
       " 'CBL': array(['CTCF', 'IRF3', 'KLF5', 'THRB(var.2)'], dtype=object),\n",
       " 'PARVA': array(['EWSR1-FLI1', 'NR2C2'], dtype=object),\n",
       " 'FERMT3': array(['E2F7', 'EWSR1-FLI1', 'ZBTB12', 'ZNF136'], dtype=object),\n",
       " 'HRAS': array(['ZNF449', 'ZNF460'], dtype=object),\n",
       " 'PTPN11': array(['EOMES', 'HOXA6', 'LHX5', 'MGA', 'NKX6-1', 'SPIB', 'TBX1', 'TBX15',\n",
       "        'TBX18', 'TBX2', 'TBX20', 'TBX3', 'TBX4', 'TBX5', 'TBX6', 'ZNF148',\n",
       "        'ZNF24'], dtype=object),\n",
       " 'GIT2': array(['BSX', 'DLX5', 'DLX6', 'DRGX', 'ELF1', 'ESX1', 'ETV1', 'ETV4',\n",
       "        'FOXH1', 'IKZF1', 'LHX1', 'LHX2', 'MSX1', 'MSX2', 'PHOX2A',\n",
       "        'PROP1', 'ZKSCAN5', 'ZNF354C', 'ZNF528'], dtype=object),\n",
       " 'ITGA5': array(['MZF1(var.2)', 'ZNF740'], dtype=object),\n",
       " 'PTPN6': array(['FIGLA', 'GATA1::TAL1'], dtype=object),\n",
       " 'LRP1': array(['BCL6B', 'NR2C2', 'NRF1', 'ZNF263'], dtype=object),\n",
       " 'TUBA1B': array(['FOXN3', 'IRF3', 'KLF15', 'KLF5', 'SP1', 'SP4'], dtype=object),\n",
       " 'SSH1': array(['DLX6', 'EN2', 'ESX1', 'GBX1', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D',\n",
       "        'TFAP4(var.2)'], dtype=object),\n",
       " 'PTPRO': array(['ESR2', 'ZSCAN4'], dtype=object),\n",
       " 'ITGA7': array(['ELF2', 'ELF4', 'ETV6', 'GSC', 'PITX1', 'PITX3', 'RARA', 'ZBTB7A',\n",
       "        'ZNF263'], dtype=object),\n",
       " 'ITGB7': array(['FOXH1', 'IRF2', 'RUNX2', 'SNAI3', 'SPI1', 'ZNF354C', 'ZNF460'],\n",
       "       dtype=object),\n",
       " 'TENC1': array(['ZNF354C'], dtype=object),\n",
       " 'PXN': array(['PRDM1', 'TFAP2B(var.3)', 'ZNF528'], dtype=object),\n",
       " 'AGAP2': array(['MAZ', 'ZKSCAN5'], dtype=object),\n",
       " 'ARHGEF7': array(['EBF1', 'EBF3', 'MYOD1', 'NR2F1(var.2)', 'RXRB', 'RXRG', 'SNAI2'],\n",
       "       dtype=object),\n",
       " 'STARD13': array(['GABPA', 'RREB1', 'RUNX3'], dtype=object),\n",
       " 'MMP14': array(['NR1D1', 'ZNF75D'], dtype=object),\n",
       " 'ACTN1': array(['PLAG1'], dtype=object),\n",
       " 'KTN1': array(['MEF2A', 'MEF2B', 'MEF2D', 'OSR1', 'STAT1', 'STAT3'], dtype=object),\n",
       " 'FERMT2': array(['CDX2', 'CDX4', 'HOXA10', 'HOXB13', 'HOXD9', 'IRF1', 'KLF5',\n",
       "        'NFATC2', 'NFATC3', 'NR4A1', 'POU2F2', 'SRF', 'STAT1::STAT2',\n",
       "        'ZNF16'], dtype=object),\n",
       " 'ARHGAP5': array(['HOXA7', 'HOXB8', 'HOXD8', 'MAZ', 'ZNF148'], dtype=object),\n",
       " 'HSPA2': array(['NFYB', 'POU2F2', 'RARA::RXRA', 'ZNF140'], dtype=object),\n",
       " 'AKT1': array(['MYBL1'], dtype=object),\n",
       " 'ZFYVE21': array(['NFYA', 'NFYC', 'PBX1', 'VENTX'], dtype=object),\n",
       " 'TRPM7': array(['BARHL2', 'GLIS1', 'NFATC2', 'NKX2-3', 'SREBF1', 'ZNF528'],\n",
       "       dtype=object),\n",
       " 'SYNM': array(['GFI1', 'IRF2', 'PRDM1'], dtype=object),\n",
       " 'ITGA11': array(['EHF', 'ELF3', 'GABPA', 'MAZ', 'SPIB', 'ZNF148', 'ZNF354C'],\n",
       "       dtype=object),\n",
       " 'CSK': array(['EWSR1-FLI1', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'PEAK1': array(['HOXC13', 'NFIB', 'NR2C2', 'PKNOX2', 'PPARD', 'RXRB', 'RXRG',\n",
       "        'THRB', 'ZNF652'], dtype=object),\n",
       " 'CIB1': array(['NR2C2(var.2)', 'RARA(var.2)', 'RARA::RXRG', 'RARB(var.3)',\n",
       "        'RXRB(var.2)', 'RXRG(var.2)', 'SNAI3'], dtype=object),\n",
       " 'CIB2': array(['ELF3', 'ELK4', 'ETV5', 'HNF4A', 'NEUROG2(var.2)', 'NR2C2',\n",
       "        'NR2F1(var.2)', 'NR4A1', 'PPARD', 'RXRB', 'RXRG', 'THRB',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'ITGAX': array(['ATOH1(var.2)', 'IRF1', 'KLF2', 'KLF6', 'MEF2A', 'MGA', 'TBX1',\n",
       "        'TBX15', 'TBX4', 'TBX5', 'ZNF135', 'ZNF460'], dtype=object),\n",
       " 'MAPK8IP3': array(['STAT3', 'THRB(var.2)'], dtype=object),\n",
       " 'BCAR1': array(['ARNT::HIF1A', 'POU2F2'], dtype=object),\n",
       " 'TGFB1I1': array(['EBF3', 'KLF11', 'ZNF528'], dtype=object),\n",
       " 'ITGAL': array(['FOXN3'], dtype=object),\n",
       " 'ITGAM': array(['FOS::JUND', 'FOSL2', 'FOSL2::JUN', 'JUN(var.2)', 'RELA', 'RELB'],\n",
       "       dtype=object),\n",
       " 'SH2B1': array(['ESRRA', 'ESRRB', 'NR2F1', 'NR4A2', 'NR5A1', 'POU5F1', 'POU6F2',\n",
       "        'RARA', 'ZNF354C', 'ZNF460'], dtype=object),\n",
       " 'ITGAD': array(['EWSR1-FLI1', 'KLF13', 'MZF1(var.2)', 'POU4F2', 'POU4F3', 'SP1',\n",
       "        'SP4', 'SP9', 'ZNF354C'], dtype=object),\n",
       " 'NUDT16L1': array(['MTF1', 'NFIA', 'RUNX2', 'ZBTB6'], dtype=object),\n",
       " 'PKD1': array(['ATOH1(var.2)', 'MYF5', 'RREB1', 'SREBF1'], dtype=object),\n",
       " 'SRCIN1': array(['KLF17', 'MAZ', 'SP2', 'ZNF148'], dtype=object),\n",
       " 'LASP1': array(['ETV1', 'ETV4', 'IKZF1', 'INSM1', 'NR3C1', 'NR3C2', 'ZNF740'],\n",
       "       dtype=object),\n",
       " 'ITGA3': array(['IRF1', 'STAT1::STAT2'], dtype=object),\n",
       " 'PFN1': array(['MEF2A', 'MEF2B', 'MEF2D', 'POU4F1'], dtype=object),\n",
       " 'ABI3': array(['BACH1', 'BATF', 'BATF3', 'BATF::JUN', 'FOS', 'FOS::JUNB',\n",
       "        'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'FOSL1::JUNB', 'FOSL1::JUND',\n",
       "        'FOSL2', 'FOSL2::JUN', 'FOSL2::JUNB', 'FOSL2::JUND', 'GFI1',\n",
       "        'JDP2', 'JUN::JUNB', 'JUNB', 'JUND', 'MAFK', 'NFATC4', 'NFE2',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'PRKCA': array(['ETV4', 'MZF1', 'SOX9', 'SPI1', 'ZNF384'], dtype=object),\n",
       " 'ITGAE': array(['RUNX3'], dtype=object),\n",
       " 'ITGB4': array(['GRHL2'], dtype=object),\n",
       " 'GRB2': array(['IRF7'], dtype=object),\n",
       " 'GIT1': array(['BACH2', 'BATF', 'BATF3', 'BATF::JUN', 'EHF', 'ELF3', 'FOS',\n",
       "        'FOS::JUN', 'FOS::JUNB', 'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN',\n",
       "        'FOSL1::JUNB', 'FOSL1::JUND', 'FOSL2', 'FOSL2::JUN', 'FOSL2::JUNB',\n",
       "        'FOSL2::JUND', 'GABPA', 'IRF2', 'IRF7', 'JDP2', 'JUN(var.2)',\n",
       "        'JUN::JUNB', 'JUNB', 'JUND', 'NFATC4', 'NFE2', 'PHOX2B', 'PRDM1',\n",
       "        'STAT1::STAT2', 'ZNF148'], dtype=object),\n",
       " 'STAT3': array(['BACH2(var.2)', 'NFYB', 'POU4F1', 'POU4F3', 'PRDM1', 'ZNF341'],\n",
       "       dtype=object),\n",
       " 'ITGB3': array(['POU4F1', 'POU4F2', 'POU4F3', 'PPARG', 'SCRT1', 'TBX21', 'TFCP2'],\n",
       "       dtype=object),\n",
       " 'GRB7': array(['SP4', 'SRY'], dtype=object),\n",
       " 'NDEL1': array(['ATF6', 'DMRTC2', 'KLF15', 'KLF4', 'KLF5', 'NRF1'], dtype=object),\n",
       " 'SLC16A3': array(['FOSL1::JUNB', 'FOSL1::JUND', 'FOXL1', 'HNF1B', 'MEF2A', 'MEF2B',\n",
       "        'MEF2C', 'MEF2D'], dtype=object),\n",
       " 'CRK': array(['MYB', 'NEUROD1', 'NEUROG2(var.2)', 'NFATC2', 'NR2C1', 'NRL',\n",
       "        'TGIF1'], dtype=object),\n",
       " 'PTPN2': array(['TP53'], dtype=object),\n",
       " 'MYOM1': array(['ARNT::HIF1A', 'KLF9', 'ZBTB14', 'ZNF460'], dtype=object),\n",
       " 'ROCK1': array(['ASCL1', 'ETV1', 'ETV4', 'IKZF1', 'KLF4', 'KLF5', 'MYOG', 'SP1',\n",
       "        'SP4', 'SPI1', 'THAP1'], dtype=object),\n",
       " 'RAVER1': array(['E2F6', 'GATA1', 'HOXA13', 'HOXB13', 'OSR2', 'TFDP1', 'YY1', 'YY2',\n",
       "        'ZFP42', 'ZNF354C', 'ZSCAN29'], dtype=object),\n",
       " 'KEAP1': array(['FOS', 'GLIS3', 'MZF1', 'ZBTB12'], dtype=object),\n",
       " 'CALR': array(['INSM1', 'MEIS1', 'RXRB', 'STAT1::STAT2'], dtype=object),\n",
       " 'DNM2': array(['HSF1', 'HSF4', 'IKZF1'], dtype=object),\n",
       " 'PTPRH': array(['BACH2', 'BARX2', 'FOS', 'FOS::JUNB', 'FOSB::JUNB', 'FOSL1',\n",
       "        'FOSL1::JUN', 'FOSL1::JUNB', 'FOSL1::JUND', 'FOSL2', 'FOSL2::JUNB',\n",
       "        'FOSL2::JUND', 'HNF1B', 'JUN(var.2)', 'JUN::JUNB', 'JUND',\n",
       "        'NFATC2', 'NFE2', 'ONECUT1', 'RXRB(var.2)', 'SPI1'], dtype=object),\n",
       " 'PRKACA': array(['MAZ', 'ZNF148'], dtype=object),\n",
       " 'INSR': array(['GRHL2', 'MXI1', 'NR2C2(var.2)'], dtype=object),\n",
       " 'CYTH2': array(['IRF3', 'MAFG', 'MITF', 'USF1', 'USF2'], dtype=object),\n",
       " 'PLAUR': array(['HIC2', 'ZBTB12'], dtype=object),\n",
       " 'PVR': array(['ATF6', 'MAX::MYC', 'MEIS1(var.2)', 'MYCN', 'NFATC2', 'NFKB1',\n",
       "        'PBX1', 'RELB', 'SOX10', 'SOX4', 'ZBTB26', 'ZNF354C'], dtype=object),\n",
       " 'PIP5K1C': array(['FIGLA', 'ZKSCAN5'], dtype=object),\n",
       " 'VAV1': array(['TEAD1', 'TEAD2', 'ZNF24', 'ZNF354C'], dtype=object),\n",
       " 'VASP': array(['ELF1', 'KLF4', 'NR1H4::RXRA', 'NR2F6(var.3)'], dtype=object),\n",
       " 'ITGA4': array(['SPIC'], dtype=object),\n",
       " 'ITGA6': array(['FIGLA'], dtype=object),\n",
       " 'ITGB1BP1': array(['MAX::MYC', 'MSANTD3', 'NR2F1(var.3)', 'PKNOX2', 'TGIF2',\n",
       "        'THRB(var.2)', 'ZNF317'], dtype=object),\n",
       " 'ITGAV': array(['EHF', 'ELF3', 'ELF5', 'EWSR1-FLI1', 'SPIB'], dtype=object),\n",
       " 'ILKAP': array(['KLF15', 'KLF5', 'MEIS1', 'NRF1', 'ZBTB26'], dtype=object),\n",
       " 'FHL2': array(['EWSR1-FLI1', 'FOXN3', 'SOX10', 'SOX4', 'SRF', 'ZNF460'],\n",
       "       dtype=object),\n",
       " 'INPP5D': array(['ELF1', 'ETV1', 'ETV4', 'IKZF1', 'IRF1', 'POU6F1(var.2)', 'PRDM1',\n",
       "        'SPI1', 'SPIC', 'STAT1::STAT2', 'TBXT'], dtype=object),\n",
       " 'TNS1': array(['EWSR1-FLI1', 'MAF::NFE2', 'TEAD1', 'TEAD2', 'TEAD3', 'TEAD4',\n",
       "        'ZBTB12'], dtype=object),\n",
       " 'CASP8': array(['ETV2', 'FOS::JUN', 'FOS::JUND', 'FOSL2::JUN', 'IKZF1', 'PRDM1',\n",
       "        'SMAD2::SMAD3::SMAD4', 'ZNF263'], dtype=object),\n",
       " 'LIMS2': array(['CTCFL', 'PLAG1'], dtype=object),\n",
       " 'SOS1': array(['ETV6', 'FOXG1', 'FOXI1', 'FOXL1', 'FOXN3', 'FOXO3', 'FOXO4',\n",
       "        'FOXO6', 'FOXP3', 'SRY', 'TCF7L2', 'ZKSCAN5'], dtype=object),\n",
       " 'NRP2': array(['BATF', 'BATF::JUN', 'FOS', 'MZF1', 'STAT1', 'STAT3'], dtype=object),\n",
       " 'IRS1': array(['KLF4', 'PRDM1', 'SP2'], dtype=object),\n",
       " 'ITGB6': array(['ASCL1(var.2)', 'ELF1', 'ELF4', 'ELF5', 'SPI1', 'TCF12(var.2)',\n",
       "        'TCF3', 'TCF4'], dtype=object),\n",
       " 'ABI2': array(['ELF1', 'ELF4', 'ELK1', 'ELK3', 'ERF', 'ERG', 'ETS1', 'ETV3',\n",
       "        'EWSR1-FLI1', 'FEV', 'HSF4', 'KLF15', 'KLF5', 'NR2F1(var.3)',\n",
       "        'NR2F6(var.3)', 'NR4A2::RXRA', 'PRDM4', 'SOX10', 'SOX4',\n",
       "        'THRB(var.2)', 'ZBTB7A', 'ZNF652'], dtype=object),\n",
       " 'LIMS1': array(['HES2', 'PRDM1'], dtype=object),\n",
       " 'ARPC2': array(['EWSR1-FLI1'], dtype=object),\n",
       " 'ASAP2': array(['ATF4', 'CEBPG(var.2)', 'SOX18'], dtype=object),\n",
       " 'NCK2': array(['OSR2'], dtype=object),\n",
       " 'PTPN1': array(['BATF', 'BATF3', 'BATF::JUN', 'ETV6', 'MSANTD3', 'NR1I2', 'PPARG',\n",
       "        'RFX5', 'SPIB'], dtype=object),\n",
       " 'SDC4': array(['ELF1', 'MZF1(var.2)', 'ZNF263'], dtype=object),\n",
       " 'PRNP': array(['CUX1', 'GATA1', 'NFATC2', 'SCRT1', 'SCRT2', 'VDR', 'ZNF354C',\n",
       "        'ZNF682'], dtype=object),\n",
       " 'SRC': array(['MZF1', 'NFIA', 'NKX2-5', 'ZKSCAN5'], dtype=object),\n",
       " 'FERMT1': array(['IKZF1', 'SPI1'], dtype=object),\n",
       " 'CASS4': array(['NR1H2::RXRA', 'RXRB', 'TCF7L1', 'TCF7L2'], dtype=object),\n",
       " 'SIRPA': array(['NFATC2', 'PPARD', 'ZNF652'], dtype=object),\n",
       " 'PLCG1': array(['EWSR1-FLI1', 'SPI1', 'SREBF1'], dtype=object),\n",
       " 'PTPRA': array(['GATA3', 'ONECUT1', 'ZKSCAN5'], dtype=object),\n",
       " 'TIAM1': array(['HLF', 'MNT', 'MYC', 'PBX3', 'PKNOX1', 'RXRA::VDR', 'ZNF263'],\n",
       "       dtype=object),\n",
       " 'ITGB2': array(['EWSR1-FLI1', 'GLIS3', 'ZNF460'], dtype=object),\n",
       " 'PARVB': array(['CREB3L1', 'TCF7L1', 'THRB(var.3)'], dtype=object),\n",
       " 'MYH9': array(['TFAP2A(var.2)'], dtype=object),\n",
       " 'RNF185': array(['MITF', 'NR2C2', 'THAP1', 'USF1', 'USF2', 'ZBTB7A', 'ZNF135',\n",
       "        'ZNF460'], dtype=object),\n",
       " 'CRKL': array(['ATF3', 'ATF7', 'DBP', 'FOSB::JUN', 'FOSB::JUNB(var.2)',\n",
       "        'FOSL1::JUN(var.2)', 'FOSL1::JUND(var.2)', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'GMEB2', 'GRHL2', 'NRF1', 'OSR2', 'RFX1',\n",
       "        'TFAP2A'], dtype=object),\n",
       " 'MAPK1': array(['IKZF1', 'PRDM1', 'ZNF384'], dtype=object),\n",
       " 'PPM1F': array(['PRDM4', 'SPIB'], dtype=object),\n",
       " 'NF2': array(['NFATC3', 'NKX2-2', 'NKX2-5'], dtype=object),\n",
       " 'PLD1': array(['GATA3', 'GLI2', 'GLI3', 'HSF2', 'KLF4', 'NFATC4', 'NFIA', 'NFIX',\n",
       "        'RREB1', 'TFCP2'], dtype=object),\n",
       " 'LPP': array(['KLF5'], dtype=object),\n",
       " 'PPM1M': array(['FOXN3'], dtype=object),\n",
       " 'ANKRD28': array(['IKZF1'], dtype=object),\n",
       " 'NISCH': array(['NFATC2', 'NFATC3', 'POU6F1(var.2)', 'ZIC5'], dtype=object),\n",
       " 'RHOA': array(['SNAI2'], dtype=object),\n",
       " 'ITGB5': array(['IKZF1', 'SPIC', 'TFAP2C(var.2)'], dtype=object),\n",
       " 'PIK3CA': array(['PLAG1'], dtype=object),\n",
       " 'ITGA9': array(['CTCF', 'GATA1', 'GATA2', 'GATA4', 'GATA6', 'IRF3', 'SOX10'],\n",
       "       dtype=object),\n",
       " 'CD47': array(['FOXO6', 'HES7', 'RARG(var.3)'], dtype=object),\n",
       " 'SORBS2': array(['BARX1', 'MZF1', 'PKNOX1'], dtype=object),\n",
       " 'GAB1': array(['MAF', 'MAFA', 'OSR2'], dtype=object),\n",
       " 'PALLD': array(['NFATC2', 'TCF7L2', 'TFAP2C(var.2)', 'ZNF384'], dtype=object),\n",
       " 'ARHGAP24': array(['ETV4', 'SPIB', 'STAT1::STAT2'], dtype=object),\n",
       " 'ARHGAP26': array(['ETV1', 'ETV4', 'IKZF1', 'MGA', 'OTX2', 'SPIC', 'TBX1', 'TBX15',\n",
       "        'TBX20', 'TBX21', 'TBX4', 'TBX5'], dtype=object),\n",
       " 'RASA1': array(['EWSR1-FLI1', 'NRF1', 'PBX1', 'PRDM1', 'SOX10', 'SOX4'],\n",
       "       dtype=object),\n",
       " 'ITGA2': array(['HIC2'], dtype=object),\n",
       " 'PDE4D': array(['FOS', 'FOXP2', 'IKZF1', 'NFATC3', 'ZBTB14', 'ZNF24'], dtype=object),\n",
       " 'TRIO': array(['GATA3', 'NRL'], dtype=object),\n",
       " 'ITGA1': array(['ELF1', 'NRF1', 'SPI1'], dtype=object),\n",
       " 'GNB2L1': array(['GATA3', 'NR2C1', 'ZNF384'], dtype=object),\n",
       " 'NEDD9': array(['E2F6', 'ETV4', 'IRF1', 'NFATC2', 'PRDM1', 'STAT1::STAT2'],\n",
       "       dtype=object),\n",
       " 'FYN': array(['KLF9', 'PITX1'], dtype=object),\n",
       " 'MARCKS': array(['KLF14', 'KLF15', 'KLF16', 'LMX1A', 'LMX1B', 'MAZ', 'MEIS2(var.2)',\n",
       "        'NFATC2', 'PAX3', 'POU6F1', 'SOX10', 'SOX9', 'SP1', 'SP3', 'SP4',\n",
       "        'SP8', 'SP9'], dtype=object),\n",
       " 'RNF5': array(['HNF4A', 'HNF4G', 'SPI1'], dtype=object),\n",
       " 'DEF6': array(['SIX2', 'TFCP2', 'ZBTB26', 'ZNF16'], dtype=object),\n",
       " 'EZR': array(['BACH1', 'BATF', 'BATF3', 'BATF::JUN', 'FOS', 'FOS::JUNB',\n",
       "        'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'FOSL1::JUNB', 'FOSL1::JUND',\n",
       "        'FOSL2::JUNB', 'FOSL2::JUND', 'JDP2', 'JUN(var.2)', 'JUN::JUNB',\n",
       "        'JUNB', 'JUND', 'KLF5', 'MAFK'], dtype=object),\n",
       " 'HSPB1': array(['MEF2A', 'MEF2B', 'MEF2C', 'MEF2D', 'NR1H4::RXRA', 'PHOX2B',\n",
       "        'ZNF263'], dtype=object),\n",
       " 'CAV1': array(['MAF', 'NR6A1', 'POU2F2'], dtype=object),\n",
       " 'ITGB8': array(['IRF1', 'PRDM1', 'ZNF341'], dtype=object),\n",
       " 'PTPN12': array(['BCL6B', 'KLF2', 'KLF3', 'KLF6', 'NKX2-2', 'ZNF263', 'ZNF449'],\n",
       "       dtype=object),\n",
       " 'ACTB': array(['NR2C1', 'RFX7', 'ZKSCAN1'], dtype=object),\n",
       " 'ZYX': array([], dtype=object),\n",
       " 'KCNH2': array(['ZNF684'], dtype=object),\n",
       " 'TES': array(['CTCFL', 'FOXD2', 'FOXE1', 'FOXI1', 'FOXL1', 'FOXO4', 'FOXO6',\n",
       "        'FOXP2', 'FOXP3', 'HOXA10', 'ZFP42'], dtype=object),\n",
       " 'RAC1': array(['BCL6B', 'E2F3', 'EHF', 'ELF3', 'GABPA', 'IRF1', 'KLF2', 'KLF3',\n",
       "        'KLF6', 'NFATC2', 'PRDM4', 'SPIB', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'LIMK1': array(['ASCL1', 'EHF', 'ELF3', 'GABPA', 'KLF4', 'MYOG', 'RFX3', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'TRIP6': array(['SPI1'], dtype=object),\n",
       " 'LYN': array(['ELF1', 'ELF4', 'ELF5', 'ETV1', 'MEF2B', 'POU1F1', 'POU2F1',\n",
       "        'POU2F3', 'POU3F1', 'POU3F2', 'POU3F3', 'POU3F4', 'POU5F1',\n",
       "        'POU5F1B', 'PRDM1', 'RARA', 'SPIB'], dtype=object),\n",
       " 'PLEC': array(['PKNOX2', 'TGIF2'], dtype=object),\n",
       " 'SHARPIN': array([], dtype=object),\n",
       " 'PTK2': array(['NOTO', 'PITX1', 'PITX2', 'PITX3'], dtype=object),\n",
       " 'SDCBP': array(['RFX2', 'RFX5', 'SREBF1', 'TCF7'], dtype=object),\n",
       " 'DLC1': array(['OSR2'], dtype=object),\n",
       " 'PTK2B': array(['KLF4', 'KLF9', 'RREB1', 'ZNF460', 'ZSCAN4'], dtype=object),\n",
       " 'SORBS3': array(['ESR2', 'EWSR1-FLI1', 'HIC2', 'PRDM1', 'YY1', 'ZFP42'],\n",
       "       dtype=object),\n",
       " 'PABPC1': array(['MGA', 'TBX1', 'TBX15', 'TBX4', 'TBX5'], dtype=object),\n",
       " 'RAPGEF1': array(['SREBF1(var.2)', 'SREBF2(var.2)'], dtype=object),\n",
       " 'OSTF1': array(['ALX3', 'ARGFX', 'BARX1', 'BSX', 'DLX5', 'DLX6', 'DRGX', 'EHF',\n",
       "        'ELF3', 'EMX1', 'EMX2', 'EN1', 'EN2', 'ESX1', 'EVX1', 'EVX2',\n",
       "        'EWSR1-FLI1', 'GBX1', 'GBX2', 'GSX1', 'GSX2', 'HOXB2', 'HOXB3',\n",
       "        'ISX', 'LBX2', 'LHX1', 'LHX2', 'LHX5', 'LHX6', 'LHX9', 'MIXL1',\n",
       "        'MNX1', 'MSX1', 'MSX2', 'NOTO', 'PAX4', 'PDX1', 'PRRX1', 'PRRX2',\n",
       "        'RAX', 'RAX2', 'SPIB', 'TLX2', 'UNCX', 'VSX1', 'VSX2', 'ZNF382'],\n",
       "       dtype=object),\n",
       " 'TESK1': array(['ATF3', 'FOS::JUN(var.2)', 'FOSL2::JUNB(var.2)',\n",
       "        'FOSL2::JUND(var.2)', 'JDP2(var.2)'], dtype=object),\n",
       " 'TLN1': array(['IKZF1', 'POU1F1', 'POU2F1', 'POU2F3', 'POU3F1', 'POU3F2',\n",
       "        'POU3F3', 'POU3F4', 'POU5F1', 'POU5F1B', 'RELB', 'SPI1', 'ZNF354C'],\n",
       "       dtype=object),\n",
       " 'VAV2': array(['KLF5', 'SOX15', 'SPI1'], dtype=object),\n",
       " 'SPTLC1': array(['ESR1', 'ESR2', 'FOXI1', 'FOXL1', 'FOXO4', 'FOXO6', 'FOXP3',\n",
       "        'GSX2', 'HOXA2', 'HOXB2', 'IRF4', 'IRF8', 'MSANTD3', 'NFIL3',\n",
       "        'PDX1', 'PPARG', 'ZNF528'], dtype=object),\n",
       " 'ABL1': array(['KLF4', 'NFIB', 'NFIC::TLX1', 'SOX13', 'SOX2', 'SRY'], dtype=object),\n",
       " 'CORO2A': array(['RARB(var.3)', 'YY1', 'ZKSCAN5'], dtype=object),\n",
       " 'ENG': array(['HOXA13', 'PBX2'], dtype=object)}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save dictionary mapping selected genes to their potential TFs to pickle\n",
    "pickle.dump(gene_to_tfs, open(saving_dir+'gene_to_tfs_with_strand.pkl', 'wb'))\n",
    "gene_to_tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p@chr19:44259741..44259757,-.txt\n"
     ]
    }
   ],
   "source": [
    "moods_out_dir = tf_dir+'moods_cage_outdir_without_strand/'\n",
    "peak_moods_file = os.listdir(moods_out_dir)[0]\n",
    "print(peak_moods_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_id</th>\n",
       "      <th>tfbs_file</th>\n",
       "      <th>motif_loc</th>\n",
       "      <th>match_strand</th>\n",
       "      <th>match_score</th>\n",
       "      <th>motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA0063.2.pfm</td>\n",
       "      <td>413</td>\n",
       "      <td>+</td>\n",
       "      <td>9.166249</td>\n",
       "      <td>GGCACTCAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA0616.2.pfm</td>\n",
       "      <td>161</td>\n",
       "      <td>-</td>\n",
       "      <td>10.344406</td>\n",
       "      <td>CGCACGTGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA0822.1.pfm</td>\n",
       "      <td>160</td>\n",
       "      <td>+</td>\n",
       "      <td>10.318423</td>\n",
       "      <td>CCGCACGTGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA0822.1.pfm</td>\n",
       "      <td>160</td>\n",
       "      <td>-</td>\n",
       "      <td>9.420233</td>\n",
       "      <td>CCGCACGTGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA1155.1.pfm</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>8.068437</td>\n",
       "      <td>TCTGCAGTGTGTGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA1560.1.pfm</td>\n",
       "      <td>161</td>\n",
       "      <td>+</td>\n",
       "      <td>10.152223</td>\n",
       "      <td>CGCACGTGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA1589.1.pfm</td>\n",
       "      <td>183</td>\n",
       "      <td>-</td>\n",
       "      <td>11.370343</td>\n",
       "      <td>GGCCGTGCACTTCTGCTCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA1594.1.pfm</td>\n",
       "      <td>372</td>\n",
       "      <td>+</td>\n",
       "      <td>2.384955</td>\n",
       "      <td>GGGGGCTCTCACCACATTCCCTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17:80194610-80195090</td>\n",
       "      <td>MA1645.1.pfm</td>\n",
       "      <td>411</td>\n",
       "      <td>+</td>\n",
       "      <td>7.856886</td>\n",
       "      <td>TTGGCACTCAACTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   peak_id     tfbs_file  motif_loc match_strand  match_score  \\\n",
       "0  chr17:80194610-80195090  MA0063.2.pfm        413            +     9.166249   \n",
       "1  chr17:80194610-80195090  MA0616.2.pfm        161            -    10.344406   \n",
       "2  chr17:80194610-80195090  MA0822.1.pfm        160            +    10.318423   \n",
       "3  chr17:80194610-80195090  MA0822.1.pfm        160            -     9.420233   \n",
       "4  chr17:80194610-80195090  MA1155.1.pfm          0            -     8.068437   \n",
       "5  chr17:80194610-80195090  MA1560.1.pfm        161            +    10.152223   \n",
       "6  chr17:80194610-80195090  MA1589.1.pfm        183            -    11.370343   \n",
       "7  chr17:80194610-80195090  MA1594.1.pfm        372            +     2.384955   \n",
       "8  chr17:80194610-80195090  MA1645.1.pfm        411            +     7.856886   \n",
       "\n",
       "                      motif  \n",
       "0               GGCACTCAACT  \n",
       "1                CGCACGTGCC  \n",
       "2              CCGCACGTGCCA  \n",
       "3              CCGCACGTGCCA  \n",
       "4           TCTGCAGTGTGTGCG  \n",
       "5                CGCACGTGCC  \n",
       "6     GGCCGTGCACTTCTGCTCCTG  \n",
       "7  GGGGGCTCTCACCACATTCCCTTT  \n",
       "8            TTGGCACTCAACTT  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                         usecols=[0,1,2,3,4,5], \n",
    "                         names=['peak_id','tfbs_file','motif_loc','match_strand','match_score','motif'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
