{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/anaconda3/lib/python3.7/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'process_jaspar' from '/home/louiscam/projects/gpcr/code/JASPAR_processing/process_jaspar.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import itertools\n",
    "from itertools import groupby\n",
    "import os.path\n",
    "import math\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import MOODS.parsers\n",
    "import MOODS.tools\n",
    "import MOODS.scan\n",
    "import subprocess\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import process_jaspar as pj\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(pj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "prom_hg19_seq_dir = dir_genome+'prom_hg19_seq_dir/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data_dir/'\n",
    "# Directory for storing preliminary results\n",
    "prelim_results_dir = '/home/louiscam/projects/gpcr/save/prelim_results_dir/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "# Directory of JASPAR data\n",
    "tf_dir = '/home/louiscam/projects/gpcr/data/tf_data/'\n",
    "jaspar_dir = tf_dir+'jaspar_data/'\n",
    "pfm_dir = tf_dir+'pfm_data/'\n",
    "moods_out_dir = tf_dir+'moods_out_dir/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load useful data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping UCSC to HGNC\n",
    "ucsc_to_hgnc = pj.create_ucsc_hgnc_dict(dir_genome+'chrom_hg19.name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load JASPAR TFBS metadata\n",
    "jaspar_df = pd.read_csv(tf_dir+'JASPAR-HomoSapiens.csv', \n",
    "                        header=0, usecols=['ID','Name','Species','Class','Family'])\n",
    "# Mapping JASPAR ID to HGNC\n",
    "jasparid_to_hgnc = {jaspar_df.iloc[i,0]:jaspar_df.iloc[i,1] for i in range(jaspar_df.shape[0])}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summarize MOODS run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load MOODS results"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over files in moods_out_dir and summarize results\n",
    "for gene_moods_file in os.listdir(moods_out_dir):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_moods_file = os.listdir(moods_out_dir)[0]\n",
    "# load results dataframe for that gene\n",
    "ucsc_gene = gene_moods_file.strip('.txt')\n",
    "res_df = pd.read_csv(moods_out_dir+gene_moods_file, sep=',', header=None, \n",
    "                     usecols=[0,1,2,3,4,5], \n",
    "                     names=['gene_id','tfbs_file','promoter_loc','strand','score','motif'])\n",
    "# Add columns corresponding to the target gene\n",
    "res_df['ucsc_gene'] = ucsc_gene\n",
    "res_df['hgnc_gene'] = ucsc_to_hgnc[ucsc_gene]\n",
    "res_df['loc_gene'] = res_df['gene_id'].str.split(' ', expand=True)[1].str.strip('range=')\n",
    "location_cols = res_df['loc_gene'].str.split(':', expand=True)\n",
    "res_df['chrom'] = location_cols[0]\n",
    "res_df[['start','stop']] = location_cols[1].str.split('-', expand=True)\n",
    "# Add columns corresponding to the TFBS\n",
    "res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "# Select relevant columns\n",
    "res_df = res_df[['ucsc_gene','hgnc_gene', 'chrom','start','stop','promoter_loc',\n",
    "                 'jaspar_tfbs','TF','score','strand','motif']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate MOODS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionarie smapping genes to their start and stop locations in hg19\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc_canonical'\n",
    "gene_id_filename = dir_genome+'chrom_hg19.name'\n",
    "df_loc = lu.get_gene_locations(gene_locations_filename, gene_id_filename)\n",
    "gene_to_start = {df_loc.iloc[i,1]:df_loc.iloc[i,3] for i in range(df_loc.shape[0])}\n",
    "gene_to_stop = {df_loc.iloc[i,1]:df_loc.iloc[i,4] for i in range(df_loc.shape[0])}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load FANTOM5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify directory\n",
    "fantom5_dir = '/home/louiscam/projects/gpcr/data/fantom5/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of CAGE peaks = 201802\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>00Annotation</th>\n",
       "      <th>short_description</th>\n",
       "      <th>description</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100013403..100013414,-</td>\n",
       "      <td>p@chr10:100013403..100013414,-</td>\n",
       "      <td>CAGE_peak_at_chr10:100013403..100013414,-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100027943..100027958,-</td>\n",
       "      <td>p1@LOXL4</td>\n",
       "      <td>CAGE_peak_1_at_LOXL4_5end</td>\n",
       "      <td>48bp_to_ENST00000260702,NM_032211,uc001kpa.1_5end</td>\n",
       "      <td>entrezgene:84171</td>\n",
       "      <td>HGNC:17171</td>\n",
       "      <td>uniprot:Q96JB6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100076685..100076699,+</td>\n",
       "      <td>p@chr10:100076685..100076699,+</td>\n",
       "      <td>CAGE_peak_at_chr10:100076685..100076699,+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:100150910..100150935,-</td>\n",
       "      <td>p@chr10:100150910..100150935,-</td>\n",
       "      <td>CAGE_peak_at_chr10:100150910..100150935,-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100150951..100150962,-</td>\n",
       "      <td>p@chr10:100150951..100150962,-</td>\n",
       "      <td>CAGE_peak_at_chr10:100150951..100150962,-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   00Annotation               short_description  \\\n",
       "0  chr10:100013403..100013414,-  p@chr10:100013403..100013414,-   \n",
       "1  chr10:100027943..100027958,-                        p1@LOXL4   \n",
       "2  chr10:100076685..100076699,+  p@chr10:100076685..100076699,+   \n",
       "3  chr10:100150910..100150935,-  p@chr10:100150910..100150935,-   \n",
       "4  chr10:100150951..100150962,-  p@chr10:100150951..100150962,-   \n",
       "\n",
       "                                 description  \\\n",
       "0  CAGE_peak_at_chr10:100013403..100013414,-   \n",
       "1                  CAGE_peak_1_at_LOXL4_5end   \n",
       "2  CAGE_peak_at_chr10:100076685..100076699,+   \n",
       "3  CAGE_peak_at_chr10:100150910..100150935,-   \n",
       "4  CAGE_peak_at_chr10:100150951..100150962,-   \n",
       "\n",
       "                         association_with_transcript     entrezgene_id  \\\n",
       "0                                                NaN               NaN   \n",
       "1  48bp_to_ENST00000260702,NM_032211,uc001kpa.1_5end  entrezgene:84171   \n",
       "2                                                NaN               NaN   \n",
       "3                                                NaN               NaN   \n",
       "4                                                NaN               NaN   \n",
       "\n",
       "      hgnc_id      uniprot_id  \n",
       "0         NaN             NaN  \n",
       "1  HGNC:17171  uniprot:Q96JB6  \n",
       "2         NaN             NaN  \n",
       "3         NaN             NaN  \n",
       "4         NaN             NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load annotated FANTOM5 CAGE peaks\n",
    "cage_df = pd.read_csv(fantom5_dir+'hg19.cage_peak_phase1and2combined_ann.txt', sep='\\t', skiprows=7, header=0)\n",
    "print('Number of CAGE peaks = '+str(cage_df.shape[0]))\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of CAGE peaks associated with a known gene = 89958\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>00Annotation</th>\n",
       "      <th>short_description</th>\n",
       "      <th>description</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100027943..100027958,-</td>\n",
       "      <td>p1@LOXL4</td>\n",
       "      <td>CAGE_peak_1_at_LOXL4_5end</td>\n",
       "      <td>48bp_to_ENST00000260702,NM_032211,uc001kpa.1_5end</td>\n",
       "      <td>entrezgene:84171</td>\n",
       "      <td>HGNC:17171</td>\n",
       "      <td>uniprot:Q96JB6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10:100174900..100174956,-</td>\n",
       "      <td>p1@PYROXD2</td>\n",
       "      <td>CAGE_peak_1_at_PYROXD2_5end</td>\n",
       "      <td>0bp_to_ENST00000370575,ENST00000462874_5end</td>\n",
       "      <td>entrezgene:84795</td>\n",
       "      <td>HGNC:23517</td>\n",
       "      <td>uniprot:Q8N2H3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr10:100174957..100174982,-</td>\n",
       "      <td>p2@PYROXD2</td>\n",
       "      <td>CAGE_peak_2_at_PYROXD2_5end</td>\n",
       "      <td>0bp_to_NM_032709,uc001kpc.2,uc001kpd.2,uc010qp...</td>\n",
       "      <td>entrezgene:84795</td>\n",
       "      <td>HGNC:23517</td>\n",
       "      <td>uniprot:Q8N2H3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr10:100206642..100206717,-</td>\n",
       "      <td>p1@HPS1</td>\n",
       "      <td>CAGE_peak_1_at_HPS1_5end</td>\n",
       "      <td>0bp_to_ENST00000325103,ENST00000338546,ENST000...</td>\n",
       "      <td>entrezgene:3257</td>\n",
       "      <td>HGNC:5163</td>\n",
       "      <td>uniprot:Q92902,uniprot:Q658M9,uniprot:Q8WXE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr10:100995440..100995474,-</td>\n",
       "      <td>p1@HPSE2</td>\n",
       "      <td>CAGE_peak_1_at_HPSE2_5end</td>\n",
       "      <td>84bp_to_AJ299720_5end</td>\n",
       "      <td>entrezgene:60495</td>\n",
       "      <td>HGNC:18374</td>\n",
       "      <td>uniprot:Q8WWQ2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    00Annotation short_description  \\\n",
       "1   chr10:100027943..100027958,-          p1@LOXL4   \n",
       "6   chr10:100174900..100174956,-        p1@PYROXD2   \n",
       "7   chr10:100174957..100174982,-        p2@PYROXD2   \n",
       "14  chr10:100206642..100206717,-           p1@HPS1   \n",
       "23  chr10:100995440..100995474,-          p1@HPSE2   \n",
       "\n",
       "                    description  \\\n",
       "1     CAGE_peak_1_at_LOXL4_5end   \n",
       "6   CAGE_peak_1_at_PYROXD2_5end   \n",
       "7   CAGE_peak_2_at_PYROXD2_5end   \n",
       "14     CAGE_peak_1_at_HPS1_5end   \n",
       "23    CAGE_peak_1_at_HPSE2_5end   \n",
       "\n",
       "                          association_with_transcript     entrezgene_id  \\\n",
       "1   48bp_to_ENST00000260702,NM_032211,uc001kpa.1_5end  entrezgene:84171   \n",
       "6         0bp_to_ENST00000370575,ENST00000462874_5end  entrezgene:84795   \n",
       "7   0bp_to_NM_032709,uc001kpc.2,uc001kpd.2,uc010qp...  entrezgene:84795   \n",
       "14  0bp_to_ENST00000325103,ENST00000338546,ENST000...   entrezgene:3257   \n",
       "23                              84bp_to_AJ299720_5end  entrezgene:60495   \n",
       "\n",
       "       hgnc_id                                    uniprot_id  \n",
       "1   HGNC:17171                                uniprot:Q96JB6  \n",
       "6   HGNC:23517                                uniprot:Q8N2H3  \n",
       "7   HGNC:23517                                uniprot:Q8N2H3  \n",
       "14   HGNC:5163  uniprot:Q92902,uniprot:Q658M9,uniprot:Q8WXE5  \n",
       "23  HGNC:18374                                uniprot:Q8WWQ2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select CAGE peaks mapped to a gene\n",
    "cage_df = cage_df.dropna(axis=0, how='all', subset=['entrezgene_id','hgnc_id','uniprot_id'], inplace=False)\n",
    "print('Number of CAGE peaks associated with a known gene = '+str(cage_df.shape[0]))\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define new columns\n",
    "cage_df['chrom'] = cage_df['00Annotation'].str.split(':', expand=True)[0]\n",
    "cage_df['cage_start'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[0].astype(int)\n",
    "cage_df['cage_stop'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[2].astype(int)\n",
    "cage_df['strand'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[1]\n",
    "cage_df['dist_to_gene'] = cage_df['association_with_transcript'].str.split('bp', expand=True)[0].fillna('50').astype(int)\n",
    "cage_df['peak_id'] = cage_df['short_description'].str.split('@', expand=True)[0].str.strip('p').astype(int)\n",
    "cage_df['hgnc_gene'] = cage_df['short_description'].str.split('@', expand=True)[1]\n",
    "cage_df['prom_start'] = cage_df['cage_start']-400\n",
    "cage_df['prom_stop'] = cage_df['cage_stop']+cage_df['dist_to_gene'].clip(upper=50)\n",
    "cage_df['prom_len'] = cage_df['prom_stop']-cage_df['prom_start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gene start and stop locations\n",
    "cage_df = cage_df[cage_df['hgnc_gene'].isin(df_loc['geneSymbol'].values)]\n",
    "cage_df['gene_start'] = [gene_to_start[cage_df['hgnc_gene'].iloc[i]] for i in range(cage_df.shape[0])]\n",
    "cage_df['gene_stop'] = [gene_to_stop[cage_df['hgnc_gene'].iloc[i]] for i in range(cage_df.shape[0])]\n",
    "cage_df['dist_cage_gene'] = cage_df['cage_start']-cage_df['gene_stop']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat dataframe\n",
    "cage_df = cage_df[['short_description','association_with_transcript',\n",
    "                   'chrom','cage_start','cage_stop','prom_start','prom_stop', 'prom_len',\n",
    "                   'strand','hgnc_gene','gene_start','gene_stop','dist_to_gene','dist_cage_gene','peak_id']]\n",
    "cage_df = cage_df.sort_values(by=['hgnc_gene','peak_id','dist_to_gene'])\n",
    "cage_df = cage_df.set_index('short_description')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_stop</th>\n",
       "      <th>prom_start</th>\n",
       "      <th>prom_stop</th>\n",
       "      <th>prom_len</th>\n",
       "      <th>strand</th>\n",
       "      <th>hgnc_gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>dist_to_gene</th>\n",
       "      <th>dist_cage_gene</th>\n",
       "      <th>peak_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>short_description</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p1@A1BG</th>\n",
       "      <td>0bp_to_AK056201_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58858938</td>\n",
       "      <td>58859039</td>\n",
       "      <td>58858538</td>\n",
       "      <td>58859039</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>58858171</td>\n",
       "      <td>58864865</td>\n",
       "      <td>0</td>\n",
       "      <td>-5927</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2@A1BG</th>\n",
       "      <td>17bp_to_ENST00000263100,NM_130786,uc002qsd.3_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58864822</td>\n",
       "      <td>58864847</td>\n",
       "      <td>58864422</td>\n",
       "      <td>58864864</td>\n",
       "      <td>442</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>58858171</td>\n",
       "      <td>58864865</td>\n",
       "      <td>17</td>\n",
       "      <td>-43</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p3@A1BG</th>\n",
       "      <td>0bp_to_ENST00000263100,NM_130786,uc002qsd.3_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58864848</td>\n",
       "      <td>58864868</td>\n",
       "      <td>58864448</td>\n",
       "      <td>58864868</td>\n",
       "      <td>420</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>58858171</td>\n",
       "      <td>58864865</td>\n",
       "      <td>0</td>\n",
       "      <td>-17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p4@A1BG</th>\n",
       "      <td>0bp_to_BX537419_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58858886</td>\n",
       "      <td>58858925</td>\n",
       "      <td>58858486</td>\n",
       "      <td>58858925</td>\n",
       "      <td>439</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>58858171</td>\n",
       "      <td>58864865</td>\n",
       "      <td>0</td>\n",
       "      <td>-5979</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1@A1BG-AS1</th>\n",
       "      <td>0bp_to_uc002qse.2_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58859101</td>\n",
       "      <td>58859149</td>\n",
       "      <td>58858701</td>\n",
       "      <td>58859149</td>\n",
       "      <td>448</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>58863335</td>\n",
       "      <td>58866549</td>\n",
       "      <td>0</td>\n",
       "      <td>-7448</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2@A1BG-AS1</th>\n",
       "      <td>-394bp_to_uc002qse.2_5end</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58858666</td>\n",
       "      <td>58858722</td>\n",
       "      <td>58858266</td>\n",
       "      <td>58858328</td>\n",
       "      <td>62</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>58863335</td>\n",
       "      <td>58866549</td>\n",
       "      <td>-394</td>\n",
       "      <td>-7883</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1@A1CF</th>\n",
       "      <td>0bp_to_ENST00000395489_5end</td>\n",
       "      <td>chr10</td>\n",
       "      <td>52645379</td>\n",
       "      <td>52645393</td>\n",
       "      <td>52644979</td>\n",
       "      <td>52645393</td>\n",
       "      <td>414</td>\n",
       "      <td>-</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>52559168</td>\n",
       "      <td>52645435</td>\n",
       "      <td>0</td>\n",
       "      <td>-56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2@A1CF</th>\n",
       "      <td>0bp_to_ENST00000282641,ENST00000373995,ENST000...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>52645416</td>\n",
       "      <td>52645444</td>\n",
       "      <td>52645016</td>\n",
       "      <td>52645444</td>\n",
       "      <td>428</td>\n",
       "      <td>-</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>52559168</td>\n",
       "      <td>52645435</td>\n",
       "      <td>0</td>\n",
       "      <td>-19</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1@A2M</th>\n",
       "      <td>-16bp_to_ENST00000540099_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268507</td>\n",
       "      <td>9268523</td>\n",
       "      <td>9268107</td>\n",
       "      <td>9268507</td>\n",
       "      <td>400</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>-16</td>\n",
       "      <td>-51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2@A2M</th>\n",
       "      <td>15bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268528</td>\n",
       "      <td>9268542</td>\n",
       "      <td>9268128</td>\n",
       "      <td>9268557</td>\n",
       "      <td>429</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>15</td>\n",
       "      <td>-30</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p3@A2M</th>\n",
       "      <td>0bp_to_ENST00000539638_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268445</td>\n",
       "      <td>9268461</td>\n",
       "      <td>9268045</td>\n",
       "      <td>9268461</td>\n",
       "      <td>416</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>0</td>\n",
       "      <td>-113</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p4@A2M</th>\n",
       "      <td>1bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268547</td>\n",
       "      <td>9268556</td>\n",
       "      <td>9268147</td>\n",
       "      <td>9268557</td>\n",
       "      <td>410</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>1</td>\n",
       "      <td>-11</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p5@A2M</th>\n",
       "      <td>-1bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268559</td>\n",
       "      <td>9268568</td>\n",
       "      <td>9268159</td>\n",
       "      <td>9268567</td>\n",
       "      <td>408</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p6@A2M</th>\n",
       "      <td>0bp_to_ENST00000318602_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268707</td>\n",
       "      <td>9268758</td>\n",
       "      <td>9268307</td>\n",
       "      <td>9268758</td>\n",
       "      <td>451</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>0</td>\n",
       "      <td>149</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p8@A2M</th>\n",
       "      <td>-12bp_to_ENST00000318602_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268765</td>\n",
       "      <td>9268779</td>\n",
       "      <td>9268365</td>\n",
       "      <td>9268767</td>\n",
       "      <td>402</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>-12</td>\n",
       "      <td>207</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p9@A2M</th>\n",
       "      <td>130bp_to_ENST00000462568_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9242983</td>\n",
       "      <td>9242995</td>\n",
       "      <td>9242583</td>\n",
       "      <td>9243045</td>\n",
       "      <td>462</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>130</td>\n",
       "      <td>-25575</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p10@A2M</th>\n",
       "      <td>0bp_to_ENST00000404455_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9268819</td>\n",
       "      <td>9268829</td>\n",
       "      <td>9268419</td>\n",
       "      <td>9268829</td>\n",
       "      <td>410</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p13@A2M</th>\n",
       "      <td>-96bp_to_ENST00000472360_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9254248</td>\n",
       "      <td>9254264</td>\n",
       "      <td>9253848</td>\n",
       "      <td>9254168</td>\n",
       "      <td>320</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>-96</td>\n",
       "      <td>-14310</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1@A2ML1</th>\n",
       "      <td>56bp_to_ENST00000539161,NM_144670,uc001quz.3_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975206</td>\n",
       "      <td>8975223</td>\n",
       "      <td>8974806</td>\n",
       "      <td>8975273</td>\n",
       "      <td>467</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>56</td>\n",
       "      <td>-54175</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2@A2ML1</th>\n",
       "      <td>0bp_to_ENST00000539161,NM_144670,uc001quz.3_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975144</td>\n",
       "      <td>8975169</td>\n",
       "      <td>8974744</td>\n",
       "      <td>8975169</td>\n",
       "      <td>425</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>0</td>\n",
       "      <td>-54237</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p4@A2ML1</th>\n",
       "      <td>33bp_to_ENST00000299698_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975101</td>\n",
       "      <td>8975108</td>\n",
       "      <td>8974701</td>\n",
       "      <td>8975141</td>\n",
       "      <td>440</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>33</td>\n",
       "      <td>-54280</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p5@A2ML1</th>\n",
       "      <td>0bp_to_ENST00000299698_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975061</td>\n",
       "      <td>8975072</td>\n",
       "      <td>8974661</td>\n",
       "      <td>8975072</td>\n",
       "      <td>411</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>0</td>\n",
       "      <td>-54320</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p8@A2ML1</th>\n",
       "      <td>-134bp_to_ENST00000537546_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975964</td>\n",
       "      <td>8975974</td>\n",
       "      <td>8975564</td>\n",
       "      <td>8975840</td>\n",
       "      <td>276</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>-134</td>\n",
       "      <td>-53417</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p19@A2ML1</th>\n",
       "      <td>-122bp_to_ENST00000537546_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975979</td>\n",
       "      <td>8975986</td>\n",
       "      <td>8975579</td>\n",
       "      <td>8975864</td>\n",
       "      <td>285</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>-122</td>\n",
       "      <td>-53402</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p25@A2ML1</th>\n",
       "      <td>-51bp_to_AK096448_5end</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9008112</td>\n",
       "      <td>9008120</td>\n",
       "      <td>9007712</td>\n",
       "      <td>9008069</td>\n",
       "      <td>357</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8975149</td>\n",
       "      <td>9029381</td>\n",
       "      <td>-51</td>\n",
       "      <td>-21269</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         association_with_transcript  chrom  \\\n",
       "short_description                                                             \n",
       "p1@A1BG                                         0bp_to_AK056201_5end  chr19   \n",
       "p2@A1BG            17bp_to_ENST00000263100,NM_130786,uc002qsd.3_5end  chr19   \n",
       "p3@A1BG             0bp_to_ENST00000263100,NM_130786,uc002qsd.3_5end  chr19   \n",
       "p4@A1BG                                         0bp_to_BX537419_5end  chr19   \n",
       "p1@A1BG-AS1                                   0bp_to_uc002qse.2_5end  chr19   \n",
       "p2@A1BG-AS1                                -394bp_to_uc002qse.2_5end  chr19   \n",
       "p1@A1CF                                  0bp_to_ENST00000395489_5end  chr10   \n",
       "p2@A1CF            0bp_to_ENST00000282641,ENST00000373995,ENST000...  chr10   \n",
       "p1@A2M                                 -16bp_to_ENST00000540099_5end  chr12   \n",
       "p2@A2M                  15bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end  chr12   \n",
       "p3@A2M                                   0bp_to_ENST00000539638_5end  chr12   \n",
       "p4@A2M                   1bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end  chr12   \n",
       "p5@A2M                  -1bp_to_NM_000014,uc001qvk.1,uc009zgk.1_5end  chr12   \n",
       "p6@A2M                                   0bp_to_ENST00000318602_5end  chr12   \n",
       "p8@A2M                                 -12bp_to_ENST00000318602_5end  chr12   \n",
       "p9@A2M                                 130bp_to_ENST00000462568_5end  chr12   \n",
       "p10@A2M                                  0bp_to_ENST00000404455_5end  chr12   \n",
       "p13@A2M                                -96bp_to_ENST00000472360_5end  chr12   \n",
       "p1@A2ML1           56bp_to_ENST00000539161,NM_144670,uc001quz.3_5end  chr12   \n",
       "p2@A2ML1            0bp_to_ENST00000539161,NM_144670,uc001quz.3_5end  chr12   \n",
       "p4@A2ML1                                33bp_to_ENST00000299698_5end  chr12   \n",
       "p5@A2ML1                                 0bp_to_ENST00000299698_5end  chr12   \n",
       "p8@A2ML1                              -134bp_to_ENST00000537546_5end  chr12   \n",
       "p19@A2ML1                             -122bp_to_ENST00000537546_5end  chr12   \n",
       "p25@A2ML1                                     -51bp_to_AK096448_5end  chr12   \n",
       "\n",
       "                   cage_start  cage_stop  prom_start  prom_stop  prom_len  \\\n",
       "short_description                                                           \n",
       "p1@A1BG              58858938   58859039    58858538   58859039       501   \n",
       "p2@A1BG              58864822   58864847    58864422   58864864       442   \n",
       "p3@A1BG              58864848   58864868    58864448   58864868       420   \n",
       "p4@A1BG              58858886   58858925    58858486   58858925       439   \n",
       "p1@A1BG-AS1          58859101   58859149    58858701   58859149       448   \n",
       "p2@A1BG-AS1          58858666   58858722    58858266   58858328        62   \n",
       "p1@A1CF              52645379   52645393    52644979   52645393       414   \n",
       "p2@A1CF              52645416   52645444    52645016   52645444       428   \n",
       "p1@A2M                9268507    9268523     9268107    9268507       400   \n",
       "p2@A2M                9268528    9268542     9268128    9268557       429   \n",
       "p3@A2M                9268445    9268461     9268045    9268461       416   \n",
       "p4@A2M                9268547    9268556     9268147    9268557       410   \n",
       "p5@A2M                9268559    9268568     9268159    9268567       408   \n",
       "p6@A2M                9268707    9268758     9268307    9268758       451   \n",
       "p8@A2M                9268765    9268779     9268365    9268767       402   \n",
       "p9@A2M                9242983    9242995     9242583    9243045       462   \n",
       "p10@A2M               9268819    9268829     9268419    9268829       410   \n",
       "p13@A2M               9254248    9254264     9253848    9254168       320   \n",
       "p1@A2ML1              8975206    8975223     8974806    8975273       467   \n",
       "p2@A2ML1              8975144    8975169     8974744    8975169       425   \n",
       "p4@A2ML1              8975101    8975108     8974701    8975141       440   \n",
       "p5@A2ML1              8975061    8975072     8974661    8975072       411   \n",
       "p8@A2ML1              8975964    8975974     8975564    8975840       276   \n",
       "p19@A2ML1             8975979    8975986     8975579    8975864       285   \n",
       "p25@A2ML1             9008112    9008120     9007712    9008069       357   \n",
       "\n",
       "                  strand hgnc_gene  gene_start  gene_stop  dist_to_gene  \\\n",
       "short_description                                                         \n",
       "p1@A1BG                -      A1BG    58858171   58864865             0   \n",
       "p2@A1BG                -      A1BG    58858171   58864865            17   \n",
       "p3@A1BG                -      A1BG    58858171   58864865             0   \n",
       "p4@A1BG                -      A1BG    58858171   58864865             0   \n",
       "p1@A1BG-AS1            +  A1BG-AS1    58863335   58866549             0   \n",
       "p2@A1BG-AS1            +  A1BG-AS1    58863335   58866549          -394   \n",
       "p1@A1CF                -      A1CF    52559168   52645435             0   \n",
       "p2@A1CF                -      A1CF    52559168   52645435             0   \n",
       "p1@A2M                 -       A2M     9220303    9268558           -16   \n",
       "p2@A2M                 -       A2M     9220303    9268558            15   \n",
       "p3@A2M                 -       A2M     9220303    9268558             0   \n",
       "p4@A2M                 -       A2M     9220303    9268558             1   \n",
       "p5@A2M                 -       A2M     9220303    9268558            -1   \n",
       "p6@A2M                 -       A2M     9220303    9268558             0   \n",
       "p8@A2M                 -       A2M     9220303    9268558           -12   \n",
       "p9@A2M                 -       A2M     9220303    9268558           130   \n",
       "p10@A2M                -       A2M     9220303    9268558             0   \n",
       "p13@A2M                -       A2M     9220303    9268558           -96   \n",
       "p1@A2ML1               +     A2ML1     8975149    9029381            56   \n",
       "p2@A2ML1               +     A2ML1     8975149    9029381             0   \n",
       "p4@A2ML1               +     A2ML1     8975149    9029381            33   \n",
       "p5@A2ML1               +     A2ML1     8975149    9029381             0   \n",
       "p8@A2ML1               +     A2ML1     8975149    9029381          -134   \n",
       "p19@A2ML1              +     A2ML1     8975149    9029381          -122   \n",
       "p25@A2ML1              +     A2ML1     8975149    9029381           -51   \n",
       "\n",
       "                   dist_cage_gene  peak_id  \n",
       "short_description                           \n",
       "p1@A1BG                     -5927        1  \n",
       "p2@A1BG                       -43        2  \n",
       "p3@A1BG                       -17        3  \n",
       "p4@A1BG                     -5979        4  \n",
       "p1@A1BG-AS1                 -7448        1  \n",
       "p2@A1BG-AS1                 -7883        2  \n",
       "p1@A1CF                       -56        1  \n",
       "p2@A1CF                       -19        2  \n",
       "p1@A2M                        -51        1  \n",
       "p2@A2M                        -30        2  \n",
       "p3@A2M                       -113        3  \n",
       "p4@A2M                        -11        4  \n",
       "p5@A2M                          1        5  \n",
       "p6@A2M                        149        6  \n",
       "p8@A2M                        207        8  \n",
       "p9@A2M                     -25575        9  \n",
       "p10@A2M                       261       10  \n",
       "p13@A2M                    -14310       13  \n",
       "p1@A2ML1                   -54175        1  \n",
       "p2@A2ML1                   -54237        2  \n",
       "p4@A2ML1                   -54280        4  \n",
       "p5@A2ML1                   -54320        5  \n",
       "p8@A2ML1                   -53417        8  \n",
       "p19@A2ML1                  -53402       19  \n",
       "p25@A2ML1                  -21269       25  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cage_df.head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.13231964986417144"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(cage_df['dist_cage_gene'].values>0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-98"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "12811440-12811538"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>00Annotation</th>\n",
       "      <th>short_description</th>\n",
       "      <th>description</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak_id</th>\n",
       "      <th>hgnc_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14370</th>\n",
       "      <td>chr11:45202580..45202606,+</td>\n",
       "      <td>p7@PRDM11</td>\n",
       "      <td>CAGE_peak_7_at_PRDM11_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:56981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr11</td>\n",
       "      <td>45202580</td>\n",
       "      <td>45202606</td>\n",
       "      <td>+</td>\n",
       "      <td>p7</td>\n",
       "      <td>PRDM11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14371</th>\n",
       "      <td>chr11:45202609..45202623,+</td>\n",
       "      <td>p8@PRDM11</td>\n",
       "      <td>CAGE_peak_8_at_PRDM11_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:56981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr11</td>\n",
       "      <td>45202609</td>\n",
       "      <td>45202623</td>\n",
       "      <td>+</td>\n",
       "      <td>p8</td>\n",
       "      <td>PRDM11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29497</th>\n",
       "      <td>chr12:8034651..8034657,+</td>\n",
       "      <td>p2@NANOGP1</td>\n",
       "      <td>CAGE_peak_2_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8034651</td>\n",
       "      <td>8034657</td>\n",
       "      <td>+</td>\n",
       "      <td>p2</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29503</th>\n",
       "      <td>chr12:8044956..8044963,+</td>\n",
       "      <td>p6@NANOGP1</td>\n",
       "      <td>CAGE_peak_6_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8044956</td>\n",
       "      <td>8044963</td>\n",
       "      <td>+</td>\n",
       "      <td>p6</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29504</th>\n",
       "      <td>chr12:8044970..8044999,+</td>\n",
       "      <td>p3@NANOGP1</td>\n",
       "      <td>CAGE_peak_3_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8044970</td>\n",
       "      <td>8044999</td>\n",
       "      <td>+</td>\n",
       "      <td>p3</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29505</th>\n",
       "      <td>chr12:8045003..8045004,+</td>\n",
       "      <td>p7@NANOGP1</td>\n",
       "      <td>CAGE_peak_7_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8045003</td>\n",
       "      <td>8045004</td>\n",
       "      <td>+</td>\n",
       "      <td>p7</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29506</th>\n",
       "      <td>chr12:8045012..8045024,+</td>\n",
       "      <td>p4@NANOGP1</td>\n",
       "      <td>CAGE_peak_4_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8045012</td>\n",
       "      <td>8045024</td>\n",
       "      <td>+</td>\n",
       "      <td>p4</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29507</th>\n",
       "      <td>chr12:8045030..8045044,+</td>\n",
       "      <td>p5@NANOGP1</td>\n",
       "      <td>CAGE_peak_5_at_NANOGP1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:404635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8045030</td>\n",
       "      <td>8045044</td>\n",
       "      <td>+</td>\n",
       "      <td>p5</td>\n",
       "      <td>NANOGP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29665</th>\n",
       "      <td>chr12:8309727..8309742,+</td>\n",
       "      <td>p2@ZNF705A</td>\n",
       "      <td>CAGE_peak_2_at_ZNF705A_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:440077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8309727</td>\n",
       "      <td>8309742</td>\n",
       "      <td>+</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZNF705A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43734</th>\n",
       "      <td>chr15:43663214..43663261,-</td>\n",
       "      <td>p2@ZSCAN29</td>\n",
       "      <td>CAGE_peak_2_at_ZSCAN29_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:146050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>43663214</td>\n",
       "      <td>43663261</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZSCAN29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51537</th>\n",
       "      <td>chr16:4365715..4365746,+</td>\n",
       "      <td>p23@GLIS2</td>\n",
       "      <td>CAGE_peak_23_at_GLIS2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>4365715</td>\n",
       "      <td>4365746</td>\n",
       "      <td>+</td>\n",
       "      <td>p23</td>\n",
       "      <td>GLIS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51538</th>\n",
       "      <td>chr16:4366005..4366016,+</td>\n",
       "      <td>p24@GLIS2</td>\n",
       "      <td>CAGE_peak_24_at_GLIS2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>4366005</td>\n",
       "      <td>4366016</td>\n",
       "      <td>+</td>\n",
       "      <td>p24</td>\n",
       "      <td>GLIS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51539</th>\n",
       "      <td>chr16:4366021..4366045,+</td>\n",
       "      <td>p25@GLIS2</td>\n",
       "      <td>CAGE_peak_25_at_GLIS2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>4366021</td>\n",
       "      <td>4366045</td>\n",
       "      <td>+</td>\n",
       "      <td>p25</td>\n",
       "      <td>GLIS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51540</th>\n",
       "      <td>chr16:4366053..4366076,+</td>\n",
       "      <td>p26@GLIS2</td>\n",
       "      <td>CAGE_peak_26_at_GLIS2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>4366053</td>\n",
       "      <td>4366076</td>\n",
       "      <td>+</td>\n",
       "      <td>p26</td>\n",
       "      <td>GLIS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52629</th>\n",
       "      <td>chr16:58034520..58034622,-</td>\n",
       "      <td>p6@ZNF319</td>\n",
       "      <td>CAGE_peak_6_at_ZNF319_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:57567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>58034520</td>\n",
       "      <td>58034622</td>\n",
       "      <td>-</td>\n",
       "      <td>p6</td>\n",
       "      <td>ZNF319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54971</th>\n",
       "      <td>chr16:88497018..88497035,+</td>\n",
       "      <td>p1@ZNF469</td>\n",
       "      <td>CAGE_peak_1_at_ZNF469_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>88497018</td>\n",
       "      <td>88497035</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55722</th>\n",
       "      <td>chr17:15689487..15689505,+</td>\n",
       "      <td>p2@MEIS3P1</td>\n",
       "      <td>CAGE_peak_2_at_MEIS3P1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:4213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr17</td>\n",
       "      <td>15689487</td>\n",
       "      <td>15689505</td>\n",
       "      <td>+</td>\n",
       "      <td>p2</td>\n",
       "      <td>MEIS3P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72153</th>\n",
       "      <td>chr19:22398533..22398542,-</td>\n",
       "      <td>p2@ZNF676</td>\n",
       "      <td>CAGE_peak_2_at_ZNF676_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:163223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>22398533</td>\n",
       "      <td>22398542</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZNF676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72154</th>\n",
       "      <td>chr19:22398550..22398579,-</td>\n",
       "      <td>p1@ZNF676</td>\n",
       "      <td>CAGE_peak_1_at_ZNF676_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:163223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>22398550</td>\n",
       "      <td>22398579</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72233</th>\n",
       "      <td>chr19:23493420..23493428,-</td>\n",
       "      <td>p1@ZNF725</td>\n",
       "      <td>CAGE_peak_1_at_ZNF725_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:100128853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>23493420</td>\n",
       "      <td>23493428</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75270</th>\n",
       "      <td>chr19:44860734..44860743,-</td>\n",
       "      <td>p5@ZFP112</td>\n",
       "      <td>CAGE_peak_5_at_ZFP112_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:7771</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>44860734</td>\n",
       "      <td>44860743</td>\n",
       "      <td>-</td>\n",
       "      <td>p5</td>\n",
       "      <td>ZFP112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76184</th>\n",
       "      <td>chr19:48076356..48076387,-</td>\n",
       "      <td>p3@ZNF541</td>\n",
       "      <td>CAGE_peak_3_at_ZNF541_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>48076356</td>\n",
       "      <td>48076387</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>ZNF541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76185</th>\n",
       "      <td>chr19:48076391..48076408,-</td>\n",
       "      <td>p4@ZNF541</td>\n",
       "      <td>CAGE_peak_4_at_ZNF541_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:84215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>48076391</td>\n",
       "      <td>48076408</td>\n",
       "      <td>-</td>\n",
       "      <td>p4</td>\n",
       "      <td>ZNF541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79583</th>\n",
       "      <td>chr19:9420326..9420349,-</td>\n",
       "      <td>p2@ZNF699</td>\n",
       "      <td>CAGE_peak_2_at_ZNF699_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:374879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>9420326</td>\n",
       "      <td>9420349</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZNF699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79584</th>\n",
       "      <td>chr19:9420483..9420530,-</td>\n",
       "      <td>p1@ZNF699</td>\n",
       "      <td>CAGE_peak_1_at_ZNF699_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:374879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19</td>\n",
       "      <td>9420483</td>\n",
       "      <td>9420530</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87338</th>\n",
       "      <td>chr1:197170394..197170444,-</td>\n",
       "      <td>p9@ZBTB41</td>\n",
       "      <td>CAGE_peak_9_at_ZBTB41_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:360023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>197170394</td>\n",
       "      <td>197170444</td>\n",
       "      <td>-</td>\n",
       "      <td>p9</td>\n",
       "      <td>ZBTB41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87339</th>\n",
       "      <td>chr1:197170503..197170557,-</td>\n",
       "      <td>p10@ZBTB41</td>\n",
       "      <td>CAGE_peak_10_at_ZBTB41_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:360023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>197170503</td>\n",
       "      <td>197170557</td>\n",
       "      <td>-</td>\n",
       "      <td>p10</td>\n",
       "      <td>ZBTB41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95430</th>\n",
       "      <td>chr1:54203698..54203719,-</td>\n",
       "      <td>p3@GLIS1</td>\n",
       "      <td>CAGE_peak_3_at_GLIS1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:148979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54203698</td>\n",
       "      <td>54203719</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>GLIS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95431</th>\n",
       "      <td>chr1:54203726..54203742,-</td>\n",
       "      <td>p4@GLIS1</td>\n",
       "      <td>CAGE_peak_4_at_GLIS1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:148979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54203726</td>\n",
       "      <td>54203742</td>\n",
       "      <td>-</td>\n",
       "      <td>p4</td>\n",
       "      <td>GLIS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95432</th>\n",
       "      <td>chr1:54204818..54204870,-</td>\n",
       "      <td>p5@GLIS1</td>\n",
       "      <td>CAGE_peak_5_at_GLIS1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:148979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54204818</td>\n",
       "      <td>54204870</td>\n",
       "      <td>-</td>\n",
       "      <td>p5</td>\n",
       "      <td>GLIS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102938</th>\n",
       "      <td>chr20:57757940..57757966,+</td>\n",
       "      <td>p1@ZNF831</td>\n",
       "      <td>CAGE_peak_1_at_ZNF831_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:128611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr20</td>\n",
       "      <td>57757940</td>\n",
       "      <td>57757966</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109360</th>\n",
       "      <td>chr22:42666429..42666436,-</td>\n",
       "      <td>p16@TCF20</td>\n",
       "      <td>CAGE_peak_16_at_TCF20_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:6942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr22</td>\n",
       "      <td>42666429</td>\n",
       "      <td>42666436</td>\n",
       "      <td>-</td>\n",
       "      <td>p16</td>\n",
       "      <td>TCF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109361</th>\n",
       "      <td>chr22:42666593..42666623,-</td>\n",
       "      <td>p17@TCF20</td>\n",
       "      <td>CAGE_peak_17_at_TCF20_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:6942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr22</td>\n",
       "      <td>42666593</td>\n",
       "      <td>42666623</td>\n",
       "      <td>-</td>\n",
       "      <td>p17</td>\n",
       "      <td>TCF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109362</th>\n",
       "      <td>chr22:42666624..42666673,-</td>\n",
       "      <td>p18@TCF20</td>\n",
       "      <td>CAGE_peak_18_at_TCF20_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:6942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr22</td>\n",
       "      <td>42666624</td>\n",
       "      <td>42666673</td>\n",
       "      <td>-</td>\n",
       "      <td>p18</td>\n",
       "      <td>TCF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110653</th>\n",
       "      <td>chr2:105470536..105470556,+</td>\n",
       "      <td>p4@POU3F3</td>\n",
       "      <td>CAGE_peak_4_at_POU3F3_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>105470536</td>\n",
       "      <td>105470556</td>\n",
       "      <td>+</td>\n",
       "      <td>p4</td>\n",
       "      <td>POU3F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110654</th>\n",
       "      <td>chr2:105470562..105470582,+</td>\n",
       "      <td>p5@POU3F3</td>\n",
       "      <td>CAGE_peak_5_at_POU3F3_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>105470562</td>\n",
       "      <td>105470582</td>\n",
       "      <td>+</td>\n",
       "      <td>p5</td>\n",
       "      <td>POU3F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110655</th>\n",
       "      <td>chr2:105470599..105470610,+</td>\n",
       "      <td>p6@POU3F3</td>\n",
       "      <td>CAGE_peak_6_at_POU3F3_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>105470599</td>\n",
       "      <td>105470610</td>\n",
       "      <td>+</td>\n",
       "      <td>p6</td>\n",
       "      <td>POU3F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110656</th>\n",
       "      <td>chr2:105470614..105470625,+</td>\n",
       "      <td>p7@POU3F3</td>\n",
       "      <td>CAGE_peak_7_at_POU3F3_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>105470614</td>\n",
       "      <td>105470625</td>\n",
       "      <td>+</td>\n",
       "      <td>p7</td>\n",
       "      <td>POU3F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110657</th>\n",
       "      <td>chr2:105470714..105470723,+</td>\n",
       "      <td>p8@POU3F3</td>\n",
       "      <td>CAGE_peak_8_at_POU3F3_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>105470714</td>\n",
       "      <td>105470723</td>\n",
       "      <td>+</td>\n",
       "      <td>p8</td>\n",
       "      <td>POU3F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152354</th>\n",
       "      <td>chr5:172484829..172484834,+</td>\n",
       "      <td>p14@C5orf41</td>\n",
       "      <td>CAGE_peak_14_at_C5orf41_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:153222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr5</td>\n",
       "      <td>172484829</td>\n",
       "      <td>172484834</td>\n",
       "      <td>+</td>\n",
       "      <td>p14</td>\n",
       "      <td>C5orf41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170535</th>\n",
       "      <td>chr7:127258120..127258138,-</td>\n",
       "      <td>p3@PAX4</td>\n",
       "      <td>CAGE_peak_3_at_PAX4_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr7</td>\n",
       "      <td>127258120</td>\n",
       "      <td>127258138</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>PAX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170536</th>\n",
       "      <td>chr7:127258139..127258173,-</td>\n",
       "      <td>p1@PAX4</td>\n",
       "      <td>CAGE_peak_1_at_PAX4_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr7</td>\n",
       "      <td>127258139</td>\n",
       "      <td>127258173</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>PAX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170537</th>\n",
       "      <td>chr7:127258190..127258201,-</td>\n",
       "      <td>p2@PAX4</td>\n",
       "      <td>CAGE_peak_2_at_PAX4_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:5078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr7</td>\n",
       "      <td>127258190</td>\n",
       "      <td>127258201</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>PAX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173074</th>\n",
       "      <td>chr7:155250249..155250278,+</td>\n",
       "      <td>p3@EN2</td>\n",
       "      <td>CAGE_peak_3_at_EN2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:2020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr7</td>\n",
       "      <td>155250249</td>\n",
       "      <td>155250278</td>\n",
       "      <td>+</td>\n",
       "      <td>p3</td>\n",
       "      <td>EN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191286</th>\n",
       "      <td>chr9:139294614..139294633,-</td>\n",
       "      <td>p3@SNAPC4</td>\n",
       "      <td>CAGE_peak_3_at_SNAPC4_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:6621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>139294614</td>\n",
       "      <td>139294633</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>SNAPC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191287</th>\n",
       "      <td>chr9:139294638..139294678,-</td>\n",
       "      <td>p4@SNAPC4</td>\n",
       "      <td>CAGE_peak_4_at_SNAPC4_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:6621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>139294638</td>\n",
       "      <td>139294678</td>\n",
       "      <td>-</td>\n",
       "      <td>p4</td>\n",
       "      <td>SNAPC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194158</th>\n",
       "      <td>chr9:79633621..79633632,+</td>\n",
       "      <td>p2@FOXB2</td>\n",
       "      <td>CAGE_peak_2_at_FOXB2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:442425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>79633621</td>\n",
       "      <td>79633632</td>\n",
       "      <td>+</td>\n",
       "      <td>p2</td>\n",
       "      <td>FOXB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194159</th>\n",
       "      <td>chr9:79633656..79633659,+</td>\n",
       "      <td>p1@FOXB2</td>\n",
       "      <td>CAGE_peak_1_at_FOXB2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:442425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>79633656</td>\n",
       "      <td>79633659</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>FOXB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194371</th>\n",
       "      <td>chr9:86594363..86594374,-</td>\n",
       "      <td>p23@HNRNPK</td>\n",
       "      <td>CAGE_peak_23_at_HNRNPK_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:3190</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>86594363</td>\n",
       "      <td>86594374</td>\n",
       "      <td>-</td>\n",
       "      <td>p23</td>\n",
       "      <td>HNRNPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194373</th>\n",
       "      <td>chr9:86594530..86594564,-</td>\n",
       "      <td>p24@HNRNPK</td>\n",
       "      <td>CAGE_peak_24_at_HNRNPK_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:3190</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr9</td>\n",
       "      <td>86594530</td>\n",
       "      <td>86594564</td>\n",
       "      <td>-</td>\n",
       "      <td>p24</td>\n",
       "      <td>HNRNPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197997</th>\n",
       "      <td>chrX:148841186..148841193,+</td>\n",
       "      <td>p1@HSFX1</td>\n",
       "      <td>CAGE_peak_1_at_HSFX1_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:100506164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>148841186</td>\n",
       "      <td>148841193</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>HSFX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199909</th>\n",
       "      <td>chrX:47225841..47225861,+</td>\n",
       "      <td>p1@ZNF157</td>\n",
       "      <td>CAGE_peak_1_at_ZNF157_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:7712</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>47225841</td>\n",
       "      <td>47225861</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>ZNF157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200512</th>\n",
       "      <td>chrX:55620448..55620457,+</td>\n",
       "      <td>p4@FOXR2</td>\n",
       "      <td>CAGE_peak_4_at_FOXR2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:139628</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>55620448</td>\n",
       "      <td>55620457</td>\n",
       "      <td>+</td>\n",
       "      <td>p4</td>\n",
       "      <td>FOXR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200513</th>\n",
       "      <td>chrX:55620469..55620479,+</td>\n",
       "      <td>p2@FOXR2</td>\n",
       "      <td>CAGE_peak_2_at_FOXR2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:139628</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>55620469</td>\n",
       "      <td>55620479</td>\n",
       "      <td>+</td>\n",
       "      <td>p2</td>\n",
       "      <td>FOXR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200514</th>\n",
       "      <td>chrX:55620481..55620492,+</td>\n",
       "      <td>p3@FOXR2</td>\n",
       "      <td>CAGE_peak_3_at_FOXR2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:139628</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>55620481</td>\n",
       "      <td>55620492</td>\n",
       "      <td>+</td>\n",
       "      <td>p3</td>\n",
       "      <td>FOXR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200515</th>\n",
       "      <td>chrX:55620497..55620512,+</td>\n",
       "      <td>p1@FOXR2</td>\n",
       "      <td>CAGE_peak_1_at_FOXR2_5end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>entrezgene:139628</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>55620497</td>\n",
       "      <td>55620512</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>FOXR2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       00Annotation short_description  \\\n",
       "14370    chr11:45202580..45202606,+         p7@PRDM11   \n",
       "14371    chr11:45202609..45202623,+         p8@PRDM11   \n",
       "29497      chr12:8034651..8034657,+        p2@NANOGP1   \n",
       "29503      chr12:8044956..8044963,+        p6@NANOGP1   \n",
       "29504      chr12:8044970..8044999,+        p3@NANOGP1   \n",
       "29505      chr12:8045003..8045004,+        p7@NANOGP1   \n",
       "29506      chr12:8045012..8045024,+        p4@NANOGP1   \n",
       "29507      chr12:8045030..8045044,+        p5@NANOGP1   \n",
       "29665      chr12:8309727..8309742,+        p2@ZNF705A   \n",
       "43734    chr15:43663214..43663261,-        p2@ZSCAN29   \n",
       "51537      chr16:4365715..4365746,+         p23@GLIS2   \n",
       "51538      chr16:4366005..4366016,+         p24@GLIS2   \n",
       "51539      chr16:4366021..4366045,+         p25@GLIS2   \n",
       "51540      chr16:4366053..4366076,+         p26@GLIS2   \n",
       "52629    chr16:58034520..58034622,-         p6@ZNF319   \n",
       "54971    chr16:88497018..88497035,+         p1@ZNF469   \n",
       "55722    chr17:15689487..15689505,+        p2@MEIS3P1   \n",
       "72153    chr19:22398533..22398542,-         p2@ZNF676   \n",
       "72154    chr19:22398550..22398579,-         p1@ZNF676   \n",
       "72233    chr19:23493420..23493428,-         p1@ZNF725   \n",
       "75270    chr19:44860734..44860743,-         p5@ZFP112   \n",
       "76184    chr19:48076356..48076387,-         p3@ZNF541   \n",
       "76185    chr19:48076391..48076408,-         p4@ZNF541   \n",
       "79583      chr19:9420326..9420349,-         p2@ZNF699   \n",
       "79584      chr19:9420483..9420530,-         p1@ZNF699   \n",
       "87338   chr1:197170394..197170444,-         p9@ZBTB41   \n",
       "87339   chr1:197170503..197170557,-        p10@ZBTB41   \n",
       "95430     chr1:54203698..54203719,-          p3@GLIS1   \n",
       "95431     chr1:54203726..54203742,-          p4@GLIS1   \n",
       "95432     chr1:54204818..54204870,-          p5@GLIS1   \n",
       "102938   chr20:57757940..57757966,+         p1@ZNF831   \n",
       "109360   chr22:42666429..42666436,-         p16@TCF20   \n",
       "109361   chr22:42666593..42666623,-         p17@TCF20   \n",
       "109362   chr22:42666624..42666673,-         p18@TCF20   \n",
       "110653  chr2:105470536..105470556,+         p4@POU3F3   \n",
       "110654  chr2:105470562..105470582,+         p5@POU3F3   \n",
       "110655  chr2:105470599..105470610,+         p6@POU3F3   \n",
       "110656  chr2:105470614..105470625,+         p7@POU3F3   \n",
       "110657  chr2:105470714..105470723,+         p8@POU3F3   \n",
       "152354  chr5:172484829..172484834,+       p14@C5orf41   \n",
       "170535  chr7:127258120..127258138,-           p3@PAX4   \n",
       "170536  chr7:127258139..127258173,-           p1@PAX4   \n",
       "170537  chr7:127258190..127258201,-           p2@PAX4   \n",
       "173074  chr7:155250249..155250278,+            p3@EN2   \n",
       "191286  chr9:139294614..139294633,-         p3@SNAPC4   \n",
       "191287  chr9:139294638..139294678,-         p4@SNAPC4   \n",
       "194158    chr9:79633621..79633632,+          p2@FOXB2   \n",
       "194159    chr9:79633656..79633659,+          p1@FOXB2   \n",
       "194371    chr9:86594363..86594374,-        p23@HNRNPK   \n",
       "194373    chr9:86594530..86594564,-        p24@HNRNPK   \n",
       "197997  chrX:148841186..148841193,+          p1@HSFX1   \n",
       "199909    chrX:47225841..47225861,+         p1@ZNF157   \n",
       "200512    chrX:55620448..55620457,+          p4@FOXR2   \n",
       "200513    chrX:55620469..55620479,+          p2@FOXR2   \n",
       "200514    chrX:55620481..55620492,+          p3@FOXR2   \n",
       "200515    chrX:55620497..55620512,+          p1@FOXR2   \n",
       "\n",
       "                         description association_with_transcript  \\\n",
       "14370     CAGE_peak_7_at_PRDM11_5end                         NaN   \n",
       "14371     CAGE_peak_8_at_PRDM11_5end                         NaN   \n",
       "29497    CAGE_peak_2_at_NANOGP1_5end                         NaN   \n",
       "29503    CAGE_peak_6_at_NANOGP1_5end                         NaN   \n",
       "29504    CAGE_peak_3_at_NANOGP1_5end                         NaN   \n",
       "29505    CAGE_peak_7_at_NANOGP1_5end                         NaN   \n",
       "29506    CAGE_peak_4_at_NANOGP1_5end                         NaN   \n",
       "29507    CAGE_peak_5_at_NANOGP1_5end                         NaN   \n",
       "29665    CAGE_peak_2_at_ZNF705A_5end                         NaN   \n",
       "43734    CAGE_peak_2_at_ZSCAN29_5end                         NaN   \n",
       "51537     CAGE_peak_23_at_GLIS2_5end                         NaN   \n",
       "51538     CAGE_peak_24_at_GLIS2_5end                         NaN   \n",
       "51539     CAGE_peak_25_at_GLIS2_5end                         NaN   \n",
       "51540     CAGE_peak_26_at_GLIS2_5end                         NaN   \n",
       "52629     CAGE_peak_6_at_ZNF319_5end                         NaN   \n",
       "54971     CAGE_peak_1_at_ZNF469_5end                         NaN   \n",
       "55722    CAGE_peak_2_at_MEIS3P1_5end                         NaN   \n",
       "72153     CAGE_peak_2_at_ZNF676_5end                         NaN   \n",
       "72154     CAGE_peak_1_at_ZNF676_5end                         NaN   \n",
       "72233     CAGE_peak_1_at_ZNF725_5end                         NaN   \n",
       "75270     CAGE_peak_5_at_ZFP112_5end                         NaN   \n",
       "76184     CAGE_peak_3_at_ZNF541_5end                         NaN   \n",
       "76185     CAGE_peak_4_at_ZNF541_5end                         NaN   \n",
       "79583     CAGE_peak_2_at_ZNF699_5end                         NaN   \n",
       "79584     CAGE_peak_1_at_ZNF699_5end                         NaN   \n",
       "87338     CAGE_peak_9_at_ZBTB41_5end                         NaN   \n",
       "87339    CAGE_peak_10_at_ZBTB41_5end                         NaN   \n",
       "95430      CAGE_peak_3_at_GLIS1_5end                         NaN   \n",
       "95431      CAGE_peak_4_at_GLIS1_5end                         NaN   \n",
       "95432      CAGE_peak_5_at_GLIS1_5end                         NaN   \n",
       "102938    CAGE_peak_1_at_ZNF831_5end                         NaN   \n",
       "109360    CAGE_peak_16_at_TCF20_5end                         NaN   \n",
       "109361    CAGE_peak_17_at_TCF20_5end                         NaN   \n",
       "109362    CAGE_peak_18_at_TCF20_5end                         NaN   \n",
       "110653    CAGE_peak_4_at_POU3F3_5end                         NaN   \n",
       "110654    CAGE_peak_5_at_POU3F3_5end                         NaN   \n",
       "110655    CAGE_peak_6_at_POU3F3_5end                         NaN   \n",
       "110656    CAGE_peak_7_at_POU3F3_5end                         NaN   \n",
       "110657    CAGE_peak_8_at_POU3F3_5end                         NaN   \n",
       "152354  CAGE_peak_14_at_C5orf41_5end                         NaN   \n",
       "170535      CAGE_peak_3_at_PAX4_5end                         NaN   \n",
       "170536      CAGE_peak_1_at_PAX4_5end                         NaN   \n",
       "170537      CAGE_peak_2_at_PAX4_5end                         NaN   \n",
       "173074       CAGE_peak_3_at_EN2_5end                         NaN   \n",
       "191286    CAGE_peak_3_at_SNAPC4_5end                         NaN   \n",
       "191287    CAGE_peak_4_at_SNAPC4_5end                         NaN   \n",
       "194158     CAGE_peak_2_at_FOXB2_5end                         NaN   \n",
       "194159     CAGE_peak_1_at_FOXB2_5end                         NaN   \n",
       "194371   CAGE_peak_23_at_HNRNPK_5end                         NaN   \n",
       "194373   CAGE_peak_24_at_HNRNPK_5end                         NaN   \n",
       "197997     CAGE_peak_1_at_HSFX1_5end                         NaN   \n",
       "199909    CAGE_peak_1_at_ZNF157_5end                         NaN   \n",
       "200512     CAGE_peak_4_at_FOXR2_5end                         NaN   \n",
       "200513     CAGE_peak_2_at_FOXR2_5end                         NaN   \n",
       "200514     CAGE_peak_3_at_FOXR2_5end                         NaN   \n",
       "200515     CAGE_peak_1_at_FOXR2_5end                         NaN   \n",
       "\n",
       "               entrezgene_id hgnc_id uniprot_id  chrom      start       stop  \\\n",
       "14370       entrezgene:56981     NaN        NaN  chr11   45202580   45202606   \n",
       "14371       entrezgene:56981     NaN        NaN  chr11   45202609   45202623   \n",
       "29497      entrezgene:404635     NaN        NaN  chr12    8034651    8034657   \n",
       "29503      entrezgene:404635     NaN        NaN  chr12    8044956    8044963   \n",
       "29504      entrezgene:404635     NaN        NaN  chr12    8044970    8044999   \n",
       "29505      entrezgene:404635     NaN        NaN  chr12    8045003    8045004   \n",
       "29506      entrezgene:404635     NaN        NaN  chr12    8045012    8045024   \n",
       "29507      entrezgene:404635     NaN        NaN  chr12    8045030    8045044   \n",
       "29665      entrezgene:440077     NaN        NaN  chr12    8309727    8309742   \n",
       "43734      entrezgene:146050     NaN        NaN  chr15   43663214   43663261   \n",
       "51537       entrezgene:84662     NaN        NaN  chr16    4365715    4365746   \n",
       "51538       entrezgene:84662     NaN        NaN  chr16    4366005    4366016   \n",
       "51539       entrezgene:84662     NaN        NaN  chr16    4366021    4366045   \n",
       "51540       entrezgene:84662     NaN        NaN  chr16    4366053    4366076   \n",
       "52629       entrezgene:57567     NaN        NaN  chr16   58034520   58034622   \n",
       "54971       entrezgene:84627     NaN        NaN  chr16   88497018   88497035   \n",
       "55722        entrezgene:4213     NaN        NaN  chr17   15689487   15689505   \n",
       "72153      entrezgene:163223     NaN        NaN  chr19   22398533   22398542   \n",
       "72154      entrezgene:163223     NaN        NaN  chr19   22398550   22398579   \n",
       "72233   entrezgene:100128853     NaN        NaN  chr19   23493420   23493428   \n",
       "75270        entrezgene:7771     NaN        NaN  chr19   44860734   44860743   \n",
       "76184       entrezgene:84215     NaN        NaN  chr19   48076356   48076387   \n",
       "76185       entrezgene:84215     NaN        NaN  chr19   48076391   48076408   \n",
       "79583      entrezgene:374879     NaN        NaN  chr19    9420326    9420349   \n",
       "79584      entrezgene:374879     NaN        NaN  chr19    9420483    9420530   \n",
       "87338      entrezgene:360023     NaN        NaN   chr1  197170394  197170444   \n",
       "87339      entrezgene:360023     NaN        NaN   chr1  197170503  197170557   \n",
       "95430      entrezgene:148979     NaN        NaN   chr1   54203698   54203719   \n",
       "95431      entrezgene:148979     NaN        NaN   chr1   54203726   54203742   \n",
       "95432      entrezgene:148979     NaN        NaN   chr1   54204818   54204870   \n",
       "102938     entrezgene:128611     NaN        NaN  chr20   57757940   57757966   \n",
       "109360       entrezgene:6942     NaN        NaN  chr22   42666429   42666436   \n",
       "109361       entrezgene:6942     NaN        NaN  chr22   42666593   42666623   \n",
       "109362       entrezgene:6942     NaN        NaN  chr22   42666624   42666673   \n",
       "110653       entrezgene:5455     NaN        NaN   chr2  105470536  105470556   \n",
       "110654       entrezgene:5455     NaN        NaN   chr2  105470562  105470582   \n",
       "110655       entrezgene:5455     NaN        NaN   chr2  105470599  105470610   \n",
       "110656       entrezgene:5455     NaN        NaN   chr2  105470614  105470625   \n",
       "110657       entrezgene:5455     NaN        NaN   chr2  105470714  105470723   \n",
       "152354     entrezgene:153222     NaN        NaN   chr5  172484829  172484834   \n",
       "170535       entrezgene:5078     NaN        NaN   chr7  127258120  127258138   \n",
       "170536       entrezgene:5078     NaN        NaN   chr7  127258139  127258173   \n",
       "170537       entrezgene:5078     NaN        NaN   chr7  127258190  127258201   \n",
       "173074       entrezgene:2020     NaN        NaN   chr7  155250249  155250278   \n",
       "191286       entrezgene:6621     NaN        NaN   chr9  139294614  139294633   \n",
       "191287       entrezgene:6621     NaN        NaN   chr9  139294638  139294678   \n",
       "194158     entrezgene:442425     NaN        NaN   chr9   79633621   79633632   \n",
       "194159     entrezgene:442425     NaN        NaN   chr9   79633656   79633659   \n",
       "194371       entrezgene:3190     NaN        NaN   chr9   86594363   86594374   \n",
       "194373       entrezgene:3190     NaN        NaN   chr9   86594530   86594564   \n",
       "197997  entrezgene:100506164     NaN        NaN   chrX  148841186  148841193   \n",
       "199909       entrezgene:7712     NaN        NaN   chrX   47225841   47225861   \n",
       "200512     entrezgene:139628     NaN        NaN   chrX   55620448   55620457   \n",
       "200513     entrezgene:139628     NaN        NaN   chrX   55620469   55620479   \n",
       "200514     entrezgene:139628     NaN        NaN   chrX   55620481   55620492   \n",
       "200515     entrezgene:139628     NaN        NaN   chrX   55620497   55620512   \n",
       "\n",
       "       strand peak_id hgnc_gene  \n",
       "14370       +      p7    PRDM11  \n",
       "14371       +      p8    PRDM11  \n",
       "29497       +      p2   NANOGP1  \n",
       "29503       +      p6   NANOGP1  \n",
       "29504       +      p3   NANOGP1  \n",
       "29505       +      p7   NANOGP1  \n",
       "29506       +      p4   NANOGP1  \n",
       "29507       +      p5   NANOGP1  \n",
       "29665       +      p2   ZNF705A  \n",
       "43734       -      p2   ZSCAN29  \n",
       "51537       +     p23     GLIS2  \n",
       "51538       +     p24     GLIS2  \n",
       "51539       +     p25     GLIS2  \n",
       "51540       +     p26     GLIS2  \n",
       "52629       -      p6    ZNF319  \n",
       "54971       +      p1    ZNF469  \n",
       "55722       +      p2   MEIS3P1  \n",
       "72153       -      p2    ZNF676  \n",
       "72154       -      p1    ZNF676  \n",
       "72233       -      p1    ZNF725  \n",
       "75270       -      p5    ZFP112  \n",
       "76184       -      p3    ZNF541  \n",
       "76185       -      p4    ZNF541  \n",
       "79583       -      p2    ZNF699  \n",
       "79584       -      p1    ZNF699  \n",
       "87338       -      p9    ZBTB41  \n",
       "87339       -     p10    ZBTB41  \n",
       "95430       -      p3     GLIS1  \n",
       "95431       -      p4     GLIS1  \n",
       "95432       -      p5     GLIS1  \n",
       "102938      +      p1    ZNF831  \n",
       "109360      -     p16     TCF20  \n",
       "109361      -     p17     TCF20  \n",
       "109362      -     p18     TCF20  \n",
       "110653      +      p4    POU3F3  \n",
       "110654      +      p5    POU3F3  \n",
       "110655      +      p6    POU3F3  \n",
       "110656      +      p7    POU3F3  \n",
       "110657      +      p8    POU3F3  \n",
       "152354      +     p14   C5orf41  \n",
       "170535      -      p3      PAX4  \n",
       "170536      -      p1      PAX4  \n",
       "170537      -      p2      PAX4  \n",
       "173074      +      p3       EN2  \n",
       "191286      -      p3    SNAPC4  \n",
       "191287      -      p4    SNAPC4  \n",
       "194158      +      p2     FOXB2  \n",
       "194159      +      p1     FOXB2  \n",
       "194371      -     p23    HNRNPK  \n",
       "194373      -     p24    HNRNPK  \n",
       "197997      +      p1     HSFX1  \n",
       "199909      +      p1    ZNF157  \n",
       "200512      +      p4     FOXR2  \n",
       "200513      +      p2     FOXR2  \n",
       "200514      +      p3     FOXR2  \n",
       "200515      +      p1     FOXR2  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cage_df[cage_df['association_with_transcript'].str.split('bp', expand=True)[0].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = 'hg19.cage_peak_phase1and2combined_coord.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = pybedtools.BedTool(fantom5_dir + f).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = bed.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>itemRgb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24882</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100013403</td>\n",
       "      <td>100013414</td>\n",
       "      <td>chr10:100013403..100013414,-</td>\n",
       "      <td>151</td>\n",
       "      <td>-</td>\n",
       "      <td>100013404</td>\n",
       "      <td>100013405</td>\n",
       "      <td>0,0,255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end                          name  score  \\\n",
       "24882  chr10  100013403  100013414  chr10:100013403..100013414,-    151   \n",
       "\n",
       "      strand  thickStart   thickEnd  itemRgb  \n",
       "24882      -   100013404  100013405  0,0,255  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp[(tmp['chrom']=='chr10') & (tmp['start']==100013403)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
