{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "from importlib import reload\n",
    "import sys\n",
    "import numpy as np \n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "plt.ioff()\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "import pybedtools\n",
    "from joblib import Parallel, delayed\n",
    "import utils as lu\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import pickle\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'utils' from '/home/louiscam/projects/gpcr/code/processing_regulatory_marks/GM12878_epigenome_processing/utils.py'>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload modules in case of modifications\n",
    "reload(lu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_peaks_dir = '/home/louiscam/projects/gpcr/data/regulatory_data/regulatory_data_GM12878/peaks/'\n",
    "epigenome_rnaseq_dir = '/home/louiscam/projects/gpcr/data/regulatory_data/regulatory_data_GM12878/rnaseq/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_regulatory_marks/processed_epigenome_data_GM12878/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relevant information\n",
    "cell_type = 'GM12878'\n",
    "resol_str = '250kb'\n",
    "resol = 250000\n",
    "quality = 'MAPQGE30'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load functional genomic data ids\n",
    "func_gen_table = pd.read_csv(epigenome_peaks_dir+'filenames.csv', sep=',', header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test: processing epigenomic features for all genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identify start and stop sites of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene location in hg19\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc_canonical'\n",
    "gene_id_filename = dir_genome+'chrom_hg19.name'\n",
    "df_loc = lu.get_gene_locations(gene_locations_filename, gene_id_filename)\n",
    "df_loc = df_loc[['geneSymbol','#chrom','chromStart','chromEnd']]\n",
    "df_loc['geneLength'] = df_loc['chromEnd']-df_loc['chromStart']\n",
    "df_loc.columns = ['gene','chrom','start','end','length']\n",
    "df_loc = df_loc.sort_values(by=['chrom','start']).reset_index().iloc[:,1:]\n",
    "all_genes = df_loc['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide genome into portions corresponding to all genes\n",
    "df_pos = df_loc[['chrom','start','end']]\n",
    "# Convert to bed file\n",
    "bed_all_genes = pybedtools.BedTool.from_dataframe(df_pos)\n",
    "bed_all_genes_df = bed_all_genes.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call ChIP-seq peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get bed file of feature\n",
    "f = func_gen_table[func_gen_table['name']=='H3K4me1'].iloc[0]['filename']+'.gz'\n",
    "bed = pybedtools.BedTool(epigenome_peaks_dir + f).sort()\n",
    "# Get counts for this feature in the segmented genome\n",
    "out = pybedtools.bedtool.BedTool.map(bed_all_genes, bed, c = [2,3], o = 'count_distinct')\n",
    "counts = out.to_dataframe()['name'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>H3K4me1</th>\n",
       "      <th>normH3K4me1</th>\n",
       "      <th>z_H3K4me1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>2536</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.281088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14406</td>\n",
       "      <td>29370</td>\n",
       "      <td>14964</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.281088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM138F</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34610</td>\n",
       "      <td>36081</td>\n",
       "      <td>1471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.281088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.281088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC729737</td>\n",
       "      <td>chr1</td>\n",
       "      <td>134772</td>\n",
       "      <td>140566</td>\n",
       "      <td>5794</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.281088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene chrom   start     end  length  H3K4me1  normH3K4me1  z_H3K4me1\n",
       "0    DDX11L1  chr1   11873   14409    2536        0          0.0  -1.281088\n",
       "1     WASH7P  chr1   14406   29370   14964        0          0.0  -1.281088\n",
       "2    FAM138F  chr1   34610   36081    1471        0          0.0  -1.281088\n",
       "3      OR4F5  chr1   69090   70008     918        0          0.0  -1.281088\n",
       "4  LOC729737  chr1  134772  140566    5794        0          0.0  -1.281088"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add results to df_loc\n",
    "df_loc['H3K4me1'] = counts\n",
    "# Normalize by gene length\n",
    "df_loc['normH3K4me1'] = np.log(1+1000000*df_loc['H3K4me1']/df_loc['length'])\n",
    "# z-score\n",
    "mean = df_loc['normH3K4me1'].mean()\n",
    "std = df_loc['normH3K4me1'].std()\n",
    "df_loc['z_H3K4me1'] = (df_loc['normH3K4me1']-mean)/std\n",
    "df_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process RNAseq\n",
    "Source: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90276"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mapping between Ensembl genes and HGN gene names\n",
    "map1 = pd.read_csv(dir_genome+'ensemblGene2Transcript2Protein', sep='\\t', header=0, index_col=None)\n",
    "map1 = map1.set_index('transcript')\n",
    "map2 = pd.read_csv(dir_genome+'ensemblGene2Name', sep='\\t', header=0, index_col=None)\n",
    "map2 = map2.set_index('#name')\n",
    "ensembl2name = map1.join(map2, how='inner')\n",
    "ensembl2name = ensembl2name.reset_index()\n",
    "ensembl2name.columns = ['transcript', 'gene', 'protein', 'name']\n",
    "ensembl2name = ensembl2name[['gene', 'name']].drop_duplicates(['gene', 'name'])\n",
    "ensembl2name = ensembl2name.set_index('gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get RNAseq sample 1\n",
    "sample_df = pd.read_csv(epigenome_rnaseq_dir+'GSM2400247_ENCFF383EXA_gene_quantifications_hg19.tsv', \n",
    "                         sep='\\t', header=0, index_col=None)\n",
    "sample_df = sample_df[['gene_id', 'length', 'effective_length', 'expected_count', 'TPM', 'FPKM']]\n",
    "sample_df['gene_id'] = sample_df['gene_id'].str.split('.', expand=True)[0]\n",
    "sample_df = sample_df.set_index('gene_id')\n",
    "sample_df = ensembl2name.join(sample_df, how='inner')\n",
    "sample_df = sample_df[sample_df['name'].isin(all_genes)]\n",
    "sample_df = sample_df.drop_duplicates('name')\n",
    "sample_df['logTPM'] = np.log(1+sample_df['TPM'])\n",
    "sample_df = sample_df.reset_index()[['name', 'logTPM']].set_index('name')\n",
    "mean = sample_df['logTPM'].mean()\n",
    "std = sample_df['logTPM'].std()\n",
    "sample_df['z_RNAseq'] = (sample_df['logTPM']-mean)/std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>H3K4me1</th>\n",
       "      <th>normH3K4me1</th>\n",
       "      <th>z_H3K4me1</th>\n",
       "      <th>logTPM</th>\n",
       "      <th>z_RNAseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>2536</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.955890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14406</td>\n",
       "      <td>29370</td>\n",
       "      <td>14964</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>1.682688</td>\n",
       "      <td>0.064814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM138F</td>\n",
       "      <td>chr1</td>\n",
       "      <td>34610</td>\n",
       "      <td>36081</td>\n",
       "      <td>1471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.955890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.955890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1</td>\n",
       "      <td>621095</td>\n",
       "      <td>622034</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.815365</td>\n",
       "      <td>-0.461297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20086</th>\n",
       "      <td>EHMT1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140513443</td>\n",
       "      <td>140730578</td>\n",
       "      <td>217135</td>\n",
       "      <td>44</td>\n",
       "      <td>5.316348</td>\n",
       "      <td>0.761448</td>\n",
       "      <td>4.131961</td>\n",
       "      <td>1.550522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20087</th>\n",
       "      <td>MIR602</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140732870</td>\n",
       "      <td>140732968</td>\n",
       "      <td>98</td>\n",
       "      <td>1</td>\n",
       "      <td>9.230641</td>\n",
       "      <td>2.265315</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.955890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20088</th>\n",
       "      <td>CACNA1B</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140772240</td>\n",
       "      <td>141019076</td>\n",
       "      <td>246836</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.039221</td>\n",
       "      <td>-0.932099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20089</th>\n",
       "      <td>TUBBP5</td>\n",
       "      <td>chr9</td>\n",
       "      <td>141044564</td>\n",
       "      <td>141071885</td>\n",
       "      <td>27321</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.955890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20090</th>\n",
       "      <td>FAM157B</td>\n",
       "      <td>chr9</td>\n",
       "      <td>141106636</td>\n",
       "      <td>141134172</td>\n",
       "      <td>27536</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.281088</td>\n",
       "      <td>1.523880</td>\n",
       "      <td>-0.031518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20091 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chrom      start        end  length  H3K4me1  normH3K4me1  \\\n",
       "0      DDX11L1  chr1      11873      14409    2536        0     0.000000   \n",
       "1       WASH7P  chr1      14406      29370   14964        0     0.000000   \n",
       "2      FAM138F  chr1      34610      36081    1471        0     0.000000   \n",
       "3        OR4F5  chr1      69090      70008     918        0     0.000000   \n",
       "4       OR4F29  chr1     621095     622034     939        0     0.000000   \n",
       "...        ...   ...        ...        ...     ...      ...          ...   \n",
       "20086    EHMT1  chr9  140513443  140730578  217135       44     5.316348   \n",
       "20087   MIR602  chr9  140732870  140732968      98        1     9.230641   \n",
       "20088  CACNA1B  chr9  140772240  141019076  246836        0     0.000000   \n",
       "20089   TUBBP5  chr9  141044564  141071885   27321        0     0.000000   \n",
       "20090  FAM157B  chr9  141106636  141134172   27536        0     0.000000   \n",
       "\n",
       "       z_H3K4me1    logTPM  z_RNAseq  \n",
       "0      -1.281088  0.000000 -0.955890  \n",
       "1      -1.281088  1.682688  0.064814  \n",
       "2      -1.281088  0.000000 -0.955890  \n",
       "3      -1.281088  0.000000 -0.955890  \n",
       "4      -1.281088  0.815365 -0.461297  \n",
       "...          ...       ...       ...  \n",
       "20086   0.761448  4.131961  1.550522  \n",
       "20087   2.265315  0.000000 -0.955890  \n",
       "20088  -1.281088  0.039221 -0.932099  \n",
       "20089  -1.281088  0.000000 -0.955890  \n",
       "20090  -1.281088  1.523880 -0.031518  \n",
       "\n",
       "[20091 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine ChiP-seq peaks with RNAseq\n",
    "df_loc_combined = df_loc.set_index('gene').join(sample_df, how='inner')\n",
    "df_loc_combined = df_loc_combined.rename_axis('gene').reset_index()\n",
    "df_loc_combined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load processing of all epigenomic features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load processing of all epigenomic features\n",
    "df_loc_combined_all = pd.read_csv(saving_dir+'features_matrix_all_genes_norm_GM12878.csv', \n",
    "                                  sep=',', header=0, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZUFSP</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATF2</th>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>...</td>\n",
       "      <td>1.392809</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>1.614079</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>1.526138</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>-0.549208</td>\n",
       "      <td>0.996742</td>\n",
       "      <td>1.081145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATF3</th>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "      <td>-0.197701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BATF</th>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>1.308002</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>1.610314</td>\n",
       "      <td>...</td>\n",
       "      <td>1.525879</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>1.356309</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>1.909319</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>-0.556698</td>\n",
       "      <td>1.101145</td>\n",
       "      <td>-0.556698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL11A</th>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>2.192683</td>\n",
       "      <td>...</td>\n",
       "      <td>2.090681</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>1.774518</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>-0.425190</td>\n",
       "      <td>1.116281</td>\n",
       "      <td>-0.425190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL3</th>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>1.499287</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>2.553147</td>\n",
       "      <td>-0.450960</td>\n",
       "      <td>-0.450960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF143</th>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>...</td>\n",
       "      <td>1.658517</td>\n",
       "      <td>1.180000</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>2.456494</td>\n",
       "      <td>1.580986</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>-0.602777</td>\n",
       "      <td>0.554791</td>\n",
       "      <td>-0.602777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF274</th>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "      <td>-0.068896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF384</th>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>0.727597</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>1.620714</td>\n",
       "      <td>...</td>\n",
       "      <td>1.658513</td>\n",
       "      <td>1.129702</td>\n",
       "      <td>0.765804</td>\n",
       "      <td>-0.747228</td>\n",
       "      <td>0.833490</td>\n",
       "      <td>1.667961</td>\n",
       "      <td>0.893889</td>\n",
       "      <td>1.994359</td>\n",
       "      <td>1.069178</td>\n",
       "      <td>1.058255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "      <td>-0.107422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNAseq</th>\n",
       "      <td>-0.526401</td>\n",
       "      <td>-0.380969</td>\n",
       "      <td>0.615192</td>\n",
       "      <td>-0.955890</td>\n",
       "      <td>-0.955890</td>\n",
       "      <td>0.011829</td>\n",
       "      <td>-0.955890</td>\n",
       "      <td>-0.955890</td>\n",
       "      <td>-0.955890</td>\n",
       "      <td>0.439624</td>\n",
       "      <td>...</td>\n",
       "      <td>0.930310</td>\n",
       "      <td>0.347938</td>\n",
       "      <td>0.342256</td>\n",
       "      <td>0.492556</td>\n",
       "      <td>1.185307</td>\n",
       "      <td>0.646670</td>\n",
       "      <td>0.459357</td>\n",
       "      <td>0.156708</td>\n",
       "      <td>1.660629</td>\n",
       "      <td>1.674597</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 20091 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             7SK      A1BG  A1BG-AS1      A1CF       A2M   A2M-AS1     A2ML1  \\\n",
       "ATF2   -0.549208 -0.549208 -0.549208 -0.549208 -0.549208 -0.549208 -0.549208   \n",
       "ATF3   -0.197701 -0.197701 -0.197701 -0.197701 -0.197701 -0.197701 -0.197701   \n",
       "BATF   -0.556698 -0.556698 -0.556698 -0.556698  1.308002 -0.556698 -0.556698   \n",
       "BCL11A -0.425190 -0.425190 -0.425190 -0.425190 -0.425190 -0.425190 -0.425190   \n",
       "BCL3   -0.450960 -0.450960 -0.450960 -0.450960 -0.450960 -0.450960 -0.450960   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "ZNF143 -0.602777 -0.602777 -0.602777 -0.602777 -0.602777 -0.602777 -0.602777   \n",
       "ZNF274 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896   \n",
       "ZNF384 -0.747228 -0.747228 -0.747228 -0.747228  0.727597 -0.747228 -0.747228   \n",
       "ZZZ3   -0.107422 -0.107422 -0.107422 -0.107422 -0.107422 -0.107422 -0.107422   \n",
       "RNAseq -0.526401 -0.380969  0.615192 -0.955890 -0.955890  0.011829 -0.955890   \n",
       "\n",
       "           A2MP1   A3GALT2    A4GALT  ...     ZUFSP      ZW10    ZWILCH  \\\n",
       "ATF2   -0.549208 -0.549208 -0.549208  ...  1.392809 -0.549208  1.614079   \n",
       "ATF3   -0.197701 -0.197701 -0.197701  ... -0.197701 -0.197701 -0.197701   \n",
       "BATF   -0.556698 -0.556698  1.610314  ...  1.525879 -0.556698  1.356309   \n",
       "BCL11A -0.425190 -0.425190  2.192683  ...  2.090681 -0.425190 -0.425190   \n",
       "BCL3   -0.450960 -0.450960 -0.450960  ... -0.450960 -0.450960 -0.450960   \n",
       "...          ...       ...       ...  ...       ...       ...       ...   \n",
       "ZNF143 -0.602777 -0.602777 -0.602777  ...  1.658517  1.180000 -0.602777   \n",
       "ZNF274 -0.068896 -0.068896 -0.068896  ... -0.068896 -0.068896 -0.068896   \n",
       "ZNF384 -0.747228 -0.747228  1.620714  ...  1.658513  1.129702  0.765804   \n",
       "ZZZ3   -0.107422 -0.107422 -0.107422  ... -0.107422 -0.107422 -0.107422   \n",
       "RNAseq -0.955890 -0.955890  0.439624  ...  0.930310  0.347938  0.342256   \n",
       "\n",
       "           ZWINT      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "ATF2   -0.549208 -0.549208  1.526138 -0.549208 -0.549208  0.996742  1.081145  \n",
       "ATF3   -0.197701 -0.197701 -0.197701 -0.197701 -0.197701 -0.197701 -0.197701  \n",
       "BATF   -0.556698 -0.556698  1.909319 -0.556698 -0.556698  1.101145 -0.556698  \n",
       "BCL11A -0.425190 -0.425190  1.774518 -0.425190 -0.425190  1.116281 -0.425190  \n",
       "BCL3   -0.450960 -0.450960  1.499287 -0.450960  2.553147 -0.450960 -0.450960  \n",
       "...          ...       ...       ...       ...       ...       ...       ...  \n",
       "ZNF143  2.456494  1.580986 -0.602777 -0.602777 -0.602777  0.554791 -0.602777  \n",
       "ZNF274 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896 -0.068896  \n",
       "ZNF384 -0.747228  0.833490  1.667961  0.893889  1.994359  1.069178  1.058255  \n",
       "ZZZ3   -0.107422 -0.107422 -0.107422 -0.107422 -0.107422 -0.107422 -0.107422  \n",
       "RNAseq  0.492556  1.185307  0.646670  0.459357  0.156708  1.660629  1.674597  \n",
       "\n",
       "[100 rows x 20091 columns]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_loc_combined_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
