{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'process_jaspar_with_strand' from '/home/louiscam/projects/gpcr/code/JASPAR_processing/process_jaspar_with_strand.py'>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import itertools\n",
    "from itertools import groupby\n",
    "import os.path\n",
    "import math\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import MOODS.parsers\n",
    "import MOODS.tools\n",
    "import MOODS.scan\n",
    "import subprocess\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import process_jaspar_with_strand as pj\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(pj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "prom_hg19_seq_dir = dir_genome+'prom_hg19_seq_dir/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data_dir/'\n",
    "# Directory for storing preliminary results\n",
    "prelim_results_dir = '/home/louiscam/projects/gpcr/save/prelim_results_dir/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "# Directory of tf data\n",
    "tf_dir = '/home/louiscam/projects/gpcr/data/tf_data/'\n",
    "jaspar_dir = tf_dir+'jaspar_data/'\n",
    "pfm_dir = tf_dir+'pfm_data/'\n",
    "fantom5_dir = '/home/louiscam/projects/gpcr/data/fantom5/'\n",
    "moods_out_dir = tf_dir+'moods_cage_outdir_annotated/'\n",
    "cage_fasta_dir = dir_genome+'cage_hg19_seq_dir_annotated/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Select genes of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genes of interest (active adhesome genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes of interest = 219\n"
     ]
    }
   ],
   "source": [
    "# Genes of interest and location\n",
    "selected_genes = pickle.load(open(saving_dir+'active_adhesome_genes.pkl', 'rb'))\n",
    "print('Number of genes of interest = '+str(len(selected_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping UCSC to HGNC\n",
    "ucsc_to_hgnc = pj.create_ucsc_hgnc_dict(dir_genome+'chrom_hg19.name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load human TFs list\n",
    "from http://humantfs.ccbr.utoronto.ca/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of TFs = 1639\n"
     ]
    }
   ],
   "source": [
    "# Load TFs\n",
    "tf_df = pd.read_csv(dir_genome+'human_tf.database', sep='\\t', header=0, index_col=0)\n",
    "tf_df = tf_df[tf_df['Is TF?']=='Yes'].sort_values(by=['HGNC symbol'])\n",
    "all_tfs = tf_df['HGNC symbol'].values\n",
    "print('Total number of TFs = '+str(len(all_tfs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Select CAGE peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process robust CAGE peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>gene</th>\n",
       "      <th>association_with_transcript</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84628</th>\n",
       "      <td>p5@FBLIM1</td>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>0bp_to_ENST00000496928,ENST00000508310_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16083098</td>\n",
       "      <td>16083119</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84629</th>\n",
       "      <td>p4@FBLIM1</td>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>0bp_to_ENST00000430076,ENST00000510393_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16083123</td>\n",
       "      <td>16083144</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84630</th>\n",
       "      <td>p13@FBLIM1</td>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>7bp_to_ENST00000375771,uc001axd.1_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16083161</td>\n",
       "      <td>16083168</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84631</th>\n",
       "      <td>p17@FBLIM1</td>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>16bp_to_ENST00000375771,uc001axd.1_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16083170</td>\n",
       "      <td>16083181</td>\n",
       "      <td>+</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84637</th>\n",
       "      <td>p19@FBLIM1</td>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>-328bp_to_ENST00000510929_5end</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16084107</td>\n",
       "      <td>16084112</td>\n",
       "      <td>+</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      short_description    gene                  association_with_transcript  \\\n",
       "84628         p5@FBLIM1  FBLIM1  0bp_to_ENST00000496928,ENST00000508310_5end   \n",
       "84629         p4@FBLIM1  FBLIM1  0bp_to_ENST00000430076,ENST00000510393_5end   \n",
       "84630        p13@FBLIM1  FBLIM1       7bp_to_ENST00000375771,uc001axd.1_5end   \n",
       "84631        p17@FBLIM1  FBLIM1      16bp_to_ENST00000375771,uc001axd.1_5end   \n",
       "84637        p19@FBLIM1  FBLIM1               -328bp_to_ENST00000510929_5end   \n",
       "\n",
       "      chrom  cage_start  cage_end strand peak_id  \n",
       "84628  chr1    16083098  16083119      +       5  \n",
       "84629  chr1    16083123  16083144      +       4  \n",
       "84630  chr1    16083161  16083168      +      13  \n",
       "84631  chr1    16083170  16083181      +      17  \n",
       "84637  chr1    16084107  16084112      +      19  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load annotated FANTOM5 CAGE peaks\n",
    "cage_df = pd.read_csv(fantom5_dir+'hg19.cage_peak_phase1and2combined_ann.txt', sep='\\t', skiprows=7, header=0)\n",
    "# Define new columns\n",
    "cage_df['gene'] = cage_df['short_description'].str.split('@',expand=True)[1]\n",
    "cage_df['chrom'] = cage_df['00Annotation'].str.split(':', expand=True)[0]\n",
    "cage_df['cage_start'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[0].astype(int)\n",
    "cage_df['cage_end'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[0].str.split(\".\", expand=True)[2].astype(int)\n",
    "cage_df['strand'] = cage_df['00Annotation'].str.split(':', expand=True)[1].str.split(\",\", expand=True)[1]\n",
    "cage_df['peak_id'] = cage_df['short_description'].str.split('@', expand=True)[0].str.strip('p')\n",
    "# Reformat dataframe\n",
    "cage_df = cage_df[['short_description', 'gene','association_with_transcript',\n",
    "                   'chrom','cage_start','cage_end','strand','peak_id']]\n",
    "cage_df = cage_df.sort_values(by=['chrom','cage_start'])\n",
    "# Restrict to selected genes\n",
    "cage_df = cage_df[cage_df['gene'].isin(selected_genes)]\n",
    "cage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat matching results\n",
    "peaks_per_gene_df = cage_df[['short_description','gene','chrom','strand']].copy()\n",
    "peaks_per_gene_df = peaks_per_gene_df.groupby(['gene','chrom','strand'])['short_description'].apply(list).reset_index(name='peaks')\n",
    "peaks_per_gene_df['n_peaks'] = [len(peaks) for peaks in peaks_per_gene_df['peaks']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>peaks</th>\n",
       "      <th>n_peaks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>[p4@ABI1, p2@ABI1, p1@ABI1, p3@ABI1]</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>[p1@ABI2, p3@ABI2, p2@ABI2, p4@ABI2, p5@ABI2, ...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>[p2@ABI3, p4@ABI3, p5@ABI3, p3@ABI3, p1@ABI3]</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>[p2@ABL1, p3@ABL1, p9@ABL1, p5@ABL1, p1@ABL1, ...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>[p9@ACTB, p13@ACTB, p8@ACTB, p5@ACTB, p22@ACTB...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  chrom strand                                              peaks  \\\n",
       "0  ABI1  chr10      -               [p4@ABI1, p2@ABI1, p1@ABI1, p3@ABI1]   \n",
       "1  ABI2   chr2      +  [p1@ABI2, p3@ABI2, p2@ABI2, p4@ABI2, p5@ABI2, ...   \n",
       "2  ABI3  chr17      +      [p2@ABI3, p4@ABI3, p5@ABI3, p3@ABI3, p1@ABI3]   \n",
       "3  ABL1   chr9      +  [p2@ABL1, p3@ABL1, p9@ABL1, p5@ABL1, p1@ABL1, ...   \n",
       "4  ACTB   chr7      -  [p9@ACTB, p13@ACTB, p8@ACTB, p5@ACTB, p22@ACTB...   \n",
       "\n",
       "   n_peaks  \n",
       "0        4  \n",
       "1        7  \n",
       "2        5  \n",
       "3        6  \n",
       "4       27  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "peaks_per_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAceElEQVR4nO3de7xVZZ3H8c9XJMErEsiAiseULDVDPV4abfJWo+iMOdmYlYqZlOUttYnGZqLSSc3bNNOoeC+VMq+UjmlkeSk1MEKBHE0wBRQ1QdIygd/8sZ6jy825LA5n7X32Xt/367VfZ61n3X5rr3V+e+1nPftZigjMzKw61mp0AGZmVl9O/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxF8HkuZL2q+Lae+X9Fi9Y+pPlLlS0kuSHmpwLOMl3dfA7R8n6TlJf5L09jpu9ypJZ9Rre9ZYazc6gKqLiHuBbXqaT9IkYOuI+GTpQdXfnsAHgc0i4pVGB9MokgYC5wO7R8RvGx2PtS5f8RuSGn0BsAUwv9WSfi/e1xHAIGB2CeE0lX5wTrY0J/76GStplqSlkn4gaRCApL0kPdMxk6QvSVogaZmkxyTtK2l/4F+Bw1IVwG/TvKMkTZX0R0lPSDo2t57Bkq5O1SdzJf1LzXbmp23NAl6RtLakiZJ+n7Y9R9IhufnHS7pf0gWSlkh6UtLfpvKnJS2WdFRXO99VrJKOAS4D3pf27WudLDte0n2Szk37M0/SATX7sl9ufJKka9Jwm6SQdHSK8yVJn5W0SzoeSyT996qb1H+nY/U7SfvmJmwk6XJJi9JxOkPSgE7eoxeBSZ3syzqSLpS0ML0uTGXvBDqq/JZI+lkny3bsy4S07CJJp+Wmr5U7hi9Kul7S0Nz0H0p6Nu3XPZK26+JYbSDpbknfVmZcOh+WpX0+rYvlOva/rPeup3N6lKQbJT2fzpETc9Mmpffju2k/ZktqL7JsS4oIv0p+AfOBh4BRwFBgLvDZNG0v4Jk0vA3wNDAqjbcBW6XhScA1Neu9B/gfsqvEscDzwD5p2lnAL4CNgc2AWR3bycU0E9gcGJzKPppiXAs4DHgFGJmmjQeWA0cDA4AzgD8A3wHWAT4ELAPW7+I96C7W8cB93bx/44HXgWPTto8DFgLK7ct+ufnfeK/SexjAxWnbHwL+AtwCbAJsCiwGPlCzn18ABqb3YSkwNE2/GbgEWC8t/xDwmZplTyCrRh3cyb58HXggLTsc+CXwjZpY1+7ifeiYPiVt/z3pfdwvTT8prXuzdEwuAabklv8UsEGadiEwMzftqnRM35726YzctEXA+9PwxsBO3RynMt+7Ls9psnN2BvDvwNuAdwBPAn+fOyf+AowjO4e+CTxQZNlWfDU8gCq8yBLTJ3Pj5wAXp+G9cifv1mRJaD9gYM06JpFL/GQJewWwQa7sm8BVafgtJy7waVZN/J/qIe6ZwMFpeDzweG7ae8iS0Ihc2YvA2E7W01Os4+k58T+RG183bftvcvvSU+LftCbOw3LjNwIn57b1xodKKnsIOIKsKuY1ckkJOBy4O7fsH3p4T38PjMuN/z1ZNVc+1p4S/7tqzqXL0/BcYN/ctJFkH5irrA8Ykta1URq/CrgCeBT4Ys28fwA+A2zYw76V/d51eU4Du9UuD3wZuDJ3Tvw0N21b4M9Flm3Fl+vR6ufZ3PCrZFfWbxERT0g6mewk3U7ST4BTImJhJ+sbBfwxIpblyp4C2nPTn85Nyw93WibpSOAUsgQDsD4wLDfLc7nhP6eYa8vW70WsRbzx/kXEq5I64iuqNs7u4l4Q6b8/eYpsH7Ygu5JdlLYP2dViT+9z3qi0vtp1r478Np4i+xAmxXezpJW56SuAEZKeBc4k+1Y3HOiYZxjZVTnAgcCfyL4d5X0E+ApwlrKqwYkR8asuYiv7vetq/i2AUZKW5MoGAPfmxmv/Bwcpu5dQZNmW4jr+fiYirouIPclOxgDO7phUM+tCYKikDXJlo4EFaXgR2dfhDpt3trmOAUlbAJcCxwNvj4ghZFd/6mS51dVTrGvqFbJvAR3+Zg3Xt6ly2Yks1oVkieY1YFhEDEmvDSMiX1feU3e3C8mObe26V0f+WOaXfxo4IBfbkIgYFBELgI8DB5N9m9yINz/c8/t5KXAHcLuk9ToKI+LXEXEwWfXMLcD13cRW5nvX3Tn9NDCvZt83iIhxPaxzTZdtSk78/YikbSTtI2kdsvrIP/PmldlzQJuktQAi4mmy+uFvShokaQfgGOCaNP/1wJclbSxpU7KE3p31yP7xnk+xHA1s3xf7VSDWNTUT+JikgemG3aFruL5NgBPT+j4KvBu4PSIWAXcC50naMN1M3UrSB1Zj3VOAr0gaLmkYWb3y6r4P/yZp3XRz9mjgB6n8YuDM9CFO2sbBadoGZIn3RbIPyf/oYt3Hk91k/lG6mfo2SZ+QtFFEvA68zJvnZGfKfO+6O6cfApYpa7AwWNIASdtL2qXAetdk2abkxN+/rEN2A+sFsq+lm5DVNQL8MP19UdLDafhwsiu3hWQ3zr4aET9N074OPAPMA34K3ED2j9+piJgDnAf8iuxD5j3A/X2xUwViXVP/BmwFvAR8DbhuDdf3IDCG7DicCRwaES+maUeS3QCck7Z3A1ldelFnANPJbkw+AjycylbHL4AngGnAuRFxZyr/T2AqcKekZWQ3endL075LVu2yIMX+QGcrTtU0E8jOnVvJbogfAcyX9DLwWeAT3cRW5nvX5TkdESuAg8gaDsxL27+M7NtNt9Zk2WbV0SrCWpyk44CPRcTqXGFZPyKpjSwxDYyI5Q0OZxWSxgOfTlWV9diez+le8hV/i5I0UtIe6Sv1NsCpZFfaZk3J53Tfcaue1vU2sjbTWwJLgO+TtaM3a1Y+p/uIq3rMzCrGVT1mZhXTFFU9w4YNi7a2tkaHYWbWVGbMmPFCRAyvLW+KxN/W1sb06dMbHYaZWVOR9FRn5a7qMTOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6uYpvjlbr20TbztjeH5Zx3YwEjMzMrjK34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKKS3xS9pc0t2S5kiaLemkVD5J0gJJM9NrXFkxmJnZqsrspG05cGpEPCxpA2CGpLvStAsi4twSt21mZl0oLfFHxCJgURpeJmkusGlZ2zMzs2LqUscvqQ3YEXgwFR0vaZakKyRt3MUyEyRNlzT9+eefr0eYZmaVUHril7Q+cCNwckS8DFwEbAWMJftGcF5ny0XE5Ihoj4j24cOHlx2mmVlllJr4JQ0kS/rXRsRNABHxXESsiIiVwKXArmXGYGZmb1Vmqx4BlwNzI+L8XPnI3GyHAI+WFYOZma2qzFY9ewBHAI9ImpnK/hU4XNJYIID5wGdKjMHMzGqU2arnPkCdTLq9rG2amVnP/MtdM7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKceI3M6sYJ34zs4px4jczqxgnfjOzinHiNzOrGCd+M7OKWbvRATSbtom3vTE8/6wDGxiJmVnv+IrfzKxinPjNzCrGid/MrGJKS/ySNpd0t6Q5kmZLOimVD5V0l6TH09+Ny4rBzMxWVeYV/3Lg1IjYFtgd+LykbYGJwLSIGANMS+NmZlYnpSX+iFgUEQ+n4WXAXGBT4GDg6jTb1cCHy4rBzMxWVZfmnJLagB2BB4EREbEoTXoWGNHFMhOACQCjR4+uQ5Rv5WabZtaqSr+5K2l94Ebg5Ih4OT8tIgKIzpaLiMkR0R4R7cOHDy87TDOzyig18UsaSJb0r42Im1Lxc5JGpukjgcVlxmBmZm9VZqseAZcDcyPi/NykqcBRafgo4NayYjAzs1X1mPglbSVpnTS8l6QTJQ0psO49gCOAfSTNTK9xwFnAByU9DuyXxs3MrE6K3Ny9EWiXtDUwmewK/TpgXHcLRcR9gLqYvO/qBGlmZn2nSFXPyohYDhwC/FdEfBEYWW5YZmZWliKJ/3VJh5PVx/84lQ0sLyQzMytTkcR/NPA+4MyImCdpS+B75YZlZmZl6bGOPyLmSPoSMDqNzwPOLjswMzMrR5FWPf8AzATuSONjJU0tOS4zMytJkaqeScCuwBKAiJgJvKO0iMzMrFSFbu5GxNKaspVlBGNmZuUr0o5/tqSPAwMkjQFOBH5ZblhmZlaWIlf8JwDbAa8BU4CXgZNLjMnMzEpUpFXPq8Dp6WVmZk2ux8Qv6Z3AaUBbfv6I2Ke8sMzMrCxF6vh/CFwMXAasKDccMzMrW5HEvzwiLio9EjMzq4siN3d/JOlzkkZKGtrxKj0yMzMrRZEr/o6HpnwxVxb4R1xmZk2pSKueLesRiJmZ1UeRvnrWlfQVSZPT+BhJB5UfmpmZlaFIVc+VwAzgb9P4ArKWPj/ucokm0zbxtkaHYGZWN0Vu7m4VEecAr8MbP+jq6pGKZmbWzxVJ/H+VNJjshi6StiLrvsHMzJpQkaqer5L1xb+5pGuBPYDxZQZlZmblKdKq5y5JDwO7k1XxnBQRL5QemZmZlaJIXz07pcFF6e9oSRsBT0XE8tIiMzOzUhSp6vkfYCdgFtkV//bAbGAjScdFxJ0lxmdmZn2syM3dhcCOEdEeETsDOwJPAh8EzikzODMz63tFrvjfGRGzO0YiYo6kd0XEk1Lztup0230zq6qij168CPh+Gj8MmCNpHVLbfjMzax5FqnrGA0+QPW7xZLJqnvFkSX/vcsIyM7OyFGnO+WfgvPSq9ac+j8jMzEpV5IrfzMxaSGmJX9IVkhZLejRXNknSAkkz02tcWds3M7POdZn4JX0v/T2pl+u+Cti/k/ILImJset3ey3WbmVkvdXfFv7OkUcCnJG2cf+xikUcvRsQ9wB/7LFIzM+sT3d3cvRiYRvaIxRm8tSvmNXn04vGSjgSmA6dGxEudzSRpAjABYPTo0b3cVN9wm38zayVdXvFHxLcj4t3AFRHxjojYMvfqbdK/CNgKGEvW909nLYU6tj85/Vq4ffjw4b3cnJmZ1SrSnPM4Se8F3p+K7omIWb3ZWEQ81zEs6VJa6CleZmbNosgzd08ErgU2Sa9rJZ3Qm41JGpkbPQR4tKt5zcysHEW6bPg0sFtEvAIg6WzgV8B/dbeQpCnAXsAwSc+QPdBlL0ljye4RzAc+09vAzcysd4okfgErcuMrKPDM3Yg4vJPiywvGZWZmJSmS+K8EHpR0cxr/ME7gZmZNq8jN3fMl/RzYMxUdHRG/KTUqMzMrTZErfiLiYeDhkmMxM7M6cCdtZmYV48RvZlYx3SZ+SQMk3V2vYMzMrHzdJv6IWAGslLRRneIxM7OSFbm5+yfgEUl3Aa90FEbEiaVFZWZmpSmS+G9KLzMzawFF2vFfLWkwMDoiHqtDTGZmVqIinbT9AzATuCONj5U0teS4zMysJEWac04CdgWWAETETHr/EBYzM2uwIon/9YhYWlO2soxgzMysfEVu7s6W9HFggKQxwInAL8sNy8zMylLkiv8EYDvgNWAK8DJwcokxmZlZiYq06nkVOD09gCUiYln5YZmZWVmKtOrZRdIjwCyyH3L9VtLO5YdmZmZlKFLHfznwuYi4F0DSnmQPZ9mhzMCaQdvE2wCYf9aBDY7EzKy4InX8KzqSPkBE3AcsLy8kMzMrU5dX/JJ2SoO/kHQJ2Y3dAA4Dfl5+aGZmVobuqnrOqxn/am44SojFzMzqoMvEHxF71zMQMzOrjx5v7koaAhwJtOXnd7fMZmbNqUirntuBB4BHcFcNZmZNr0jiHxQRp5QeiZmZ1UWR5pzfk3SspJGShna8So/MzMxKUeSK/6/At4DTebM1T+Cumc3MmlKRxH8qsHVEvFB2MGZmVr4iVT1PAK+WHYiZmdVHkSv+V4CZku4m65oZ6Lk5p6QrgIOAxRGxfSobCvyArGnofOCfI+KlXkVuZma9UuSK/xbgTLKHr8zIvXpyFbB/TdlEYFpEjAGmpXEzM6ujIv3xX92bFUfEPZLaaooPBvZKw1eT9fnzpd6s38zMeqfIL3fn0UnfPBHRm1Y9IyJiURp+FhjRzXYnABMARo8e3YtNraqjG+Uy5bfh7prNrD8qUsffnhseBHwUWON2/BERkrrs7C0iJgOTAdrb290pnJlZH+mxjj8iXsy9FkTEhUBvL2WfkzQSIP1d3Mv1mJlZLxWp6tkpN7oW2TeAIt8UOjMVOAo4K/29tZfrMTOzXiqSwPP98i8nNcPsaSFJU8hu5A6T9AxZf/5nAddLOgZ4qsh6zMysbxVp1dOrfvkj4vAuJu3bm/WZmVnfKFLVsw7wEVbtj//r5YVlZmZlKVLVcyuwlOxHW6/1MK+ZmfVzRRL/ZhFR+wtcy6nH7wPMzPpKkS4bfinpPaVHYmZmdVHkin9PYHz6Be9rgMh+f7VDqZGZmVkpiiT+A0qPwszM6qZIc86n6hGImZnVR5E6fjMzayFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78ZmYV09tn59pq6ui6ef5ZB65SVlvem/WuyTrMrFp8xW9mVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYxTvxmZhXjxG9mVjFux99PuD2+mdWLr/jNzCrGid/MrGKc+M3MKqYhdfyS5gPLgBXA8ohob0QcZmZV1Mibu3tHxAsN3L6ZWSW5qsfMrGIadcUfwJ2SArgkIibXziBpAjABYPTo0XUOr2/km2jWe3tuEmpmXWnUFf+eEbETcADweUl/VztDREyOiPaIaB8+fHj9IzQza1ENSfwRsSD9XQzcDOzaiDjMzKqo7olf0nqSNugYBj4EPFrvOMzMqqoRdfwjgJsldWz/uoi4owFxmJlVUt0Tf0Q8Cby33ts1M7OMm3OamVWME7+ZWcU48fdzbRNvq/vvAcystTnxm5lVjBO/mVnFOPGbmVWME7+ZWcU48ZuZVYwTv5lZxTjxm5lVTCOfwFVJvW2TX+++9t23v1nr8hW/mVnFOPGbmVWME7+ZWcU48ZuZVYwTv5lZxTjxm5lVjJtzVlRXzTXdBbRZ6/MVv5lZxTjxm5lVjBO/mVnFOPGbmVWME7+ZWcU48ZuZVYwTv5lZxbgdfz/UU1v6Im3tO5unq+6V69F2vz9389zI2Prz+2Kty1f8ZmYV48RvZlYxTvxmZhXjxG9mVjENSfyS9pf0mKQnJE1sRAxmZlVV98QvaQDwHeAAYFvgcEnb1jsOM7OqasQV/67AExHxZET8Ffg+cHAD4jAzqyRFRH03KB0K7B8Rn07jRwC7RcTxNfNNACak0W2Ax3q5yWHAC71cthl4/5qb9695NcO+bRERw2sL++0PuCJiMjB5TdcjaXpEtPdBSP2S96+5ef+aVzPvWyOqehYAm+fGN0tlZmZWB41I/L8GxkjaUtLbgI8BUxsQh5lZJdW9qicilks6HvgJMAC4IiJml7jJNa4u6ue8f83N+9e8mnbf6n5z18zMGsu/3DUzqxgnfjOzimnpxN9KXUNI2lzS3ZLmSJot6aRUPlTSXZIeT383bnSsa0LSAEm/kfTjNL6lpAfTMfxBahDQlCQNkXSDpN9Jmivpfa10/CR9IZ2bj0qaImlQMx8/SVdIWizp0VxZp8dLmW+n/ZwlaafGRd6zlk38Ldg1xHLg1IjYFtgd+Hzan4nAtIgYA0xL483sJGBubvxs4IKI2Bp4CTimIVH1jf8E7oiIdwHvJdvPljh+kjYFTgTaI2J7soYbH6O5j99VwP41ZV0drwOAMek1AbioTjH2Sssmflqsa4iIWBQRD6fhZWRJY1Oyfbo6zXY18OGGBNgHJG0GHAhclsYF7APckGZp2v2TtBHwd8DlABHx14hYQgsdP7JWgoMlrQ2sCyyiiY9fRNwD/LGmuKvjdTDw3cg8AAyRNLIugfZCKyf+TYGnc+PPpLKmJ6kN2BF4EBgREYvSpGeBEY2Kqw9cCPwLsDKNvx1YEhHL03gzH8MtgeeBK1NV1mWS1qNFjl9ELADOBf5AlvCXAjNonePXoavj1VT5ppUTf0uStD5wI3ByRLycnxZZ29ymbJ8r6SBgcUTMaHQsJVkb2Am4KCJ2BF6hplqnyY/fxmRXvVsCo4D1WLWapKU08/Fq5cTfcl1DSBpIlvSvjYibUvFzHV8p09/FjYpvDe0B/KOk+WTVcvuQ1YkPSVUH0NzH8BngmYh4MI3fQPZB0CrHbz9gXkQ8HxGvAzeRHdNWOX4dujpeTZVvWjnxt1TXEKm++3JgbkScn5s0FTgqDR8F3Frv2PpCRHw5IjaLiDayY/WziPgEcDdwaJqtmffvWeBpSdukon2BObTI8SOr4tld0rrpXO3Yv5Y4fjldHa+pwJGpdc/uwNJclVD/ExEt+wLGAf8H/B44vdHxrOG+7En2tXIWMDO9xpHVg08DHgd+CgxtdKx9sK97AT9Ow+8AHgKeAH4IrNPo+NZgv8YC09MxvAXYuJWOH/A14HfAo8D3gHWa+fgBU8juV7xO9o3tmK6OFyCyVoS/Bx4ha93U8H3o6uUuG8zMKqaVq3rMzKwTTvxmZhXjxG9mVjFO/GZmFePEb2ZWMU78VjmSfi6p9IdkSzox9cJ5bYnbmC9pWFnrt9ZU90cvmjUzSWvHm33P9ORzwH4R8UyZMZmtLl/xW78kqS1dLV+a+ni/U9LgNO2NK3ZJw1I3D0gaL+mW1E/6fEnHSzoldYr2gKShuU0cIWlm6jt+17T8eqkP9ofSMgfn1jtV0s/IfrxTG+spaT2PSjo5lV1M9uOl/5X0hZr5x0u6Ne3H45K+mpv2ybT9mZIuSd2LI+kiSdPTe/G1TmIYLOl/JR2b9uM2Sb9NMR3W+yNhrciJ3/qzMcB3ImI7YAnwkQLLbA/8E7ALcCbwamSdov0KODI337oRMZbsqvyKVHY6WVcRuwJ7A99KPWhC1q/OoRHxgfzGJO0MHA3sRvachGMl7RgRnwUWAntHxAWdxLlr2p8dgI9Kapf0buAwYI8U2wrgEx2xRUR7mv8DknbIrWt94EfAlIi4lKxztIUR8d7I+sa/o8D7ZhXiqh7rz+ZFxMw0PANoK7DM3ZE9r2CZpKVkCRGyn9Hnk+UUyPpcl7ShpCHAh8g6ijstzTMIGJ2G74qI2r7ZIetK4+aIeAVA0k3A+4Hf9BDnXRHxYm6ZPcketrMz8OusuxsG82YnYP8saQLZ/+xIsocLzUrTbgXOiYiOewmPAOdJOpus64t7e4jFKsaJ3/qz13LDK8gSIWQJsuPb6qBullmZG1/JW8/32r5Kgqy/lY9ExGP5CZJ2I+tGuS91tf2rI+LLNdvfEjgN2CUiXpJ0FW/d7/uB/SVdF5n/U/bov3HAGZKmRcTX+zh+a2Ku6rFmNJ/syhje7PlxdR0GIGlPsp4UlwI/AU5IvUsiaccC67kX+HDqlXI94JBU1pMPKnt+62CypzjdT3b/4FBJm6TtD5W0BbAh2QfPUkkjyB7zl/fvZI81/E5abhRZFdc1wLfIqqnM3uArfmtG5wLXp6qP23q5jr9I+g0wEPhUKvsG2VPAZklaC5gHHNTdSiLi4XQF/lAquiwieqrmIc1/I1m/7ddExHQASV8B7kzbfx34fEQ8kGL9HdlTnu7vZH0nAVdIOofsA+RbklamdRxXIB6rEPfOaVZnksaTddt7fKNjsWpyVY+ZWcX4it/MrGJ8xW9mVjFO/GZmFePEb2ZWMU78ZmYV48RvZlYx/w9vH7L17lk3LgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Histogram of number of peaks per gene\n",
    "plt.figure()\n",
    "plt.hist(peaks_per_gene_df['n_peaks'], bins=np.arange(0,110,1))\n",
    "plt.xlabel('number of peaks')\n",
    "plt.ylabel('number of genes')\n",
    "plt.title('histogram of number of peaks per gene')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Associate CAGE peaks with genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Dataframe of genes for each peak\n",
    "genes_per_peak = peaks_per_gene_df[['gene','peaks']]\n",
    "genes_per_peak = lu.unnesting(genes_per_peak,['peaks'])\n",
    "genes_per_peak = genes_per_peak.groupby('peaks')['gene'].apply(list).reset_index(name='genes')\n",
    "genes_per_peak['n_genes'] = [len(genes) for genes in genes_per_peak['genes']]\n",
    "genes_per_peak.columns = ['short_description','genes','n_genes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add peak location\n",
    "genes_per_peak = pd.merge(genes_per_peak, cage_df, on=['short_description'])\n",
    "genes_per_peak = genes_per_peak[['short_description','chrom','cage_start','cage_end','strand','genes']]\n",
    "genes_per_peak = genes_per_peak.sort_values(by=['chrom'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consider a window around each peak\n",
    "genes_per_peak['cage_window_start'] = [genes_per_peak['cage_start'].iloc[i]-400 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                       else genes_per_peak['cage_start'].iloc[i]-50 for i in range(genes_per_peak.shape[0])]\n",
    "genes_per_peak['cage_window_end'] = [genes_per_peak['cage_end'].iloc[i]+50 if (genes_per_peak['strand'].iloc[i]=='+') \n",
    "                                     else genes_per_peak['cage_end'].iloc[i]+400 for i in range(genes_per_peak.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genes</th>\n",
       "      <th>cage_window_start</th>\n",
       "      <th>cage_window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1843</th>\n",
       "      <td>p8@ENAH</td>\n",
       "      <td>chr1</td>\n",
       "      <td>225700609</td>\n",
       "      <td>225700622</td>\n",
       "      <td>-</td>\n",
       "      <td>[ENAH]</td>\n",
       "      <td>225700559</td>\n",
       "      <td>225701022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>p13@BCAR3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>94140243</td>\n",
       "      <td>94140275</td>\n",
       "      <td>-</td>\n",
       "      <td>[BCAR3]</td>\n",
       "      <td>94140193</td>\n",
       "      <td>94140675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>p4@SLC9A1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27481768</td>\n",
       "      <td>27481773</td>\n",
       "      <td>-</td>\n",
       "      <td>[SLC9A1]</td>\n",
       "      <td>27481718</td>\n",
       "      <td>27482173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>p13@FBLIM1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16083161</td>\n",
       "      <td>16083168</td>\n",
       "      <td>+</td>\n",
       "      <td>[FBLIM1]</td>\n",
       "      <td>16082761</td>\n",
       "      <td>16083218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1427</th>\n",
       "      <td>p4@RHOU</td>\n",
       "      <td>chr1</td>\n",
       "      <td>228871683</td>\n",
       "      <td>228871704</td>\n",
       "      <td>+</td>\n",
       "      <td>[RHOU]</td>\n",
       "      <td>228871283</td>\n",
       "      <td>228871754</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     short_description chrom  cage_start   cage_end strand     genes  \\\n",
       "1843           p8@ENAH  chr1   225700609  225700622      -    [ENAH]   \n",
       "176          p13@BCAR3  chr1    94140243   94140275      -   [BCAR3]   \n",
       "1434         p4@SLC9A1  chr1    27481768   27481773      -  [SLC9A1]   \n",
       "181         p13@FBLIM1  chr1    16083161   16083168      +  [FBLIM1]   \n",
       "1427           p4@RHOU  chr1   228871683  228871704      +    [RHOU]   \n",
       "\n",
       "      cage_window_start  cage_window_end  \n",
       "1843          225700559        225701022  \n",
       "176            94140193         94140675  \n",
       "1434           27481718         27482173  \n",
       "181            16082761         16083218  \n",
       "1427          228871283        228871754  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize\n",
    "genes_per_peak.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Create sequence files for all CAGE windows"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over all CAGE windows to generate sequence files\n",
    "for i in tqdm(range(genes_per_peak.shape[0])):\n",
    "    time.sleep(.01)\n",
    "    # Peak characteristics\n",
    "    peak = genes_per_peak['short_description'].iloc[i]\n",
    "    chrom = genes_per_peak['chrom'].iloc[i]\n",
    "    start = genes_per_peak['cage_window_start'].iloc[i]\n",
    "    end = genes_per_peak['cage_window_end'].iloc[i]\n",
    "    # Generate window sequence (sequence on the + strand)\n",
    "    result = subprocess.run(cage_fasta_dir+'twoBitToFa http://hgdownload.cse.ucsc.edu/gbdb/hg19/hg19.2bit '+\n",
    "                            cage_fasta_dir+peak+'.fa'+\n",
    "                            ' -seq='+chrom+' -start='+str(start)+' -end='+str(end)+' -noMask',\n",
    "                            shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Search for TFBS in all CAGE windows\n",
    "See process_JASPAR_annotated.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Find TFs binding each CAGE window"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Useful data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load JASPAR TFBS metadata\n",
    "jaspar_df = pd.read_csv(tf_dir+'JASPAR-HomoSapiens.csv', \n",
    "                        header=0, usecols=['ID','Name','Species','Class','Family'])\n",
    "jaspar_tfs = np.unique(jaspar_df['Name'])\n",
    "# Mapping JASPAR ID to HGNC\n",
    "jasparid_to_hgnc = {jaspar_df.iloc[i,0]:jaspar_df.iloc[i,1] for i in range(jaspar_df.shape[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique TFs = 639\n",
      "Number of unique TF motifs = 810\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Class</th>\n",
       "      <th>Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0002.1</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Runt domain factors</td>\n",
       "      <td>Runt-related factors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0003.1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0003.2</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0003.3</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0003.4</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Basic helix-span-helix factors (bHSH)</td>\n",
       "      <td>AP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID    Name       Species                                  Class  \\\n",
       "0  MA0002.1   RUNX1  Homo sapiens                    Runt domain factors   \n",
       "1  MA0003.1  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "2  MA0003.2  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "3  MA0003.3  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "4  MA0003.4  TFAP2A  Homo sapiens  Basic helix-span-helix factors (bHSH)   \n",
       "\n",
       "                 Family  \n",
       "0  Runt-related factors  \n",
       "1                  AP-2  \n",
       "2                  AP-2  \n",
       "3                  AP-2  \n",
       "4                  AP-2  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize data\n",
    "print('Number of unique TFs = '+str(len(np.unique(jaspar_df['Name']))))\n",
    "print('Number of unique TF motifs = '+str(len(np.unique(jaspar_df['ID']))))\n",
    "jaspar_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load MOODS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary mapping CAGE peaks to the corresponding TFs\n",
    "tfs_associated_to_cage_peaks = {}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Loop over files in moods_out_dir and summarize results\n",
    "for peak_moods_file in tqdm(os.listdir(moods_out_dir)):\n",
    "    time.sleep(.01)\n",
    "    # load results dataframe for that gene\n",
    "    cage_peak = peak_moods_file.strip('.txt')\n",
    "    res_df = pd.read_csv(moods_out_dir+peak_moods_file, sep=',', header=None, \n",
    "                         usecols=[0,1,2,3,4,5], \n",
    "                         names=['peak_id','tfbs_file','motif_loc','match_strand','match_score','motif'])\n",
    "    if (res_df.shape[0]>0):\n",
    "        # Add columns corresponding to the target peak\n",
    "        res_df['cage_peak'] = cage_peak\n",
    "        res_df['chrom'] = res_df['peak_id'].str.split(':', expand=True)[0]\n",
    "        location_cols = res_df['peak_id'].str.split(':', expand=True)[1]\n",
    "        res_df[['start','stop']] = location_cols.str.split('-', expand=True)\n",
    "        # Add columns corresponding to the TFBS\n",
    "        res_df['jaspar_tfbs'] = res_df['tfbs_file'].str.strip('.pfm')\n",
    "        res_df['TF'] = [jasparid_to_hgnc[jaspar_id] for jaspar_id in res_df['jaspar_tfbs']]\n",
    "        # Only keep TFs matching on the peak strand\n",
    "        peak_strand = cage_df[cage_df['short_description']==cage_peak]['strand'].values[0]\n",
    "        res_df = res_df[res_df['match_strand']==peak_strand]\n",
    "        # Select relevant columns\n",
    "        res_df = res_df[['cage_peak', 'chrom','start','stop','motif_loc',\n",
    "                         'jaspar_tfbs','TF','match_score','match_strand','motif']]\n",
    "        # Store list of TFs associated with the current CAGE peak in dict\n",
    "        tfs_associated_to_cage_peaks[cage_peak] = np.sort(np.unique(res_df['TF'].values))\n",
    "pickle.dump(tfs_associated_to_cage_peaks, open(saving_dir+'tfs_cage_annotated.pkl', 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Find TFs targeting genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mapping between cage peaks and corresponding TFs\n",
    "tfs_associated_to_cage_peaks = pickle.load(open(saving_dir+'tfs_cage_annotated.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/louiscam/projects/gpcr/code/JASPAR_processing/utils.py:37: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  return df1.join(df.drop(explode, 1), how='left')\n"
     ]
    }
   ],
   "source": [
    "# Construct dataframe with gene-tf pairs\n",
    "tf_and_genes_df = genes_per_peak.copy()\n",
    "tf_and_genes_df['tfs'] = [tfs_associated_to_cage_peaks[cage_peak] if cage_peak in tfs_associated_to_cage_peaks.keys()\n",
    "                          else ['missing']\n",
    "                          for cage_peak in tf_and_genes_df['short_description'].values]\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['genes'])\n",
    "gene_to_tfs = {tf_and_genes_df.iloc[i,0]:tf_and_genes_df.iloc[i,8] for i in range(tf_and_genes_df.shape[0])}\n",
    "tf_and_genes_df = lu.unnesting(tf_and_genes_df,['tfs'])\n",
    "tf_and_genes_df = tf_and_genes_df[['genes','tfs','short_description','chrom','cage_start','cage_end','strand']]\n",
    "tf_and_genes_df = tf_and_genes_df.sort_values(by=['genes','tfs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ALX3</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ARGFX</td>\n",
       "      <td>p3@ABI1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149904</td>\n",
       "      <td>27149920</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1639</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>BACH2</td>\n",
       "      <td>p@chr10:27149068..27149079,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27149068</td>\n",
       "      <td>27149079</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1637</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>p@chr10:27132379..27132382,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27132379</td>\n",
       "      <td>27132382</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>p@chr10:27037479..27037487,-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>27037479</td>\n",
       "      <td>27037487</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genes    tfs             short_description  chrom  cage_start  cage_end  \\\n",
       "853   ABI1   ALX3                       p3@ABI1  chr10    27149904  27149920   \n",
       "853   ABI1  ARGFX                       p3@ABI1  chr10    27149904  27149920   \n",
       "1639  ABI1  BACH2  p@chr10:27149068..27149079,-  chr10    27149068  27149079   \n",
       "1637  ABI1   E2F6  p@chr10:27132379..27132382,-  chr10    27132379  27132382   \n",
       "1634  ABI1   EBF1  p@chr10:27037479..27037487,-  chr10    27037479  27037487   \n",
       "\n",
       "     strand  \n",
       "853       -  \n",
       "853       -  \n",
       "1639      -  \n",
       "1637      -  \n",
       "1634      -  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize results\n",
    "tf_and_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SHC1': ['missing'],\n",
       " 'VAV3': array(['ESR1', 'ESR2', 'SMAD2::SMAD3::SMAD4'], dtype=object),\n",
       " 'BCAR3': ['missing'],\n",
       " 'ARHGEF2': array(['RREB1', 'ZBTB12'], dtype=object),\n",
       " 'ITGA10': array(['CTCF', 'CTCFL'], dtype=object),\n",
       " 'CSRP1': array(['ZNF263'], dtype=object),\n",
       " 'HAX1': array(['ZNF384'], dtype=object),\n",
       " 'FABP3': array(['EWSR1-FLI1', 'NR1I2', 'NR5A1', 'SPI1', 'ZKSCAN1', 'ZKSCAN5',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'FBLIM1': ['missing'],\n",
       " 'NEXN': array(['GRHL2', 'MAZ', 'NFATC4'], dtype=object),\n",
       " 'PTPRF': array(['ZNF740'], dtype=object),\n",
       " 'TSPAN1': array(['NFATC4', 'ZNF354C'], dtype=object),\n",
       " 'MACF1': ['missing'],\n",
       " 'ENAH': array(['ASCL1(var.2)', 'GLI2', 'GLI3', 'INSM1', 'NFKB1', 'TCF12(var.2)',\n",
       "        'TCF3', 'TCF4', 'TCF7L1', 'ZEB1', 'ZNF135', 'ZNF24', 'ZNF354C',\n",
       "        'ZNF460'], dtype=object),\n",
       " 'ARF1': array(['ATF6', 'KLF15', 'KLF5'], dtype=object),\n",
       " 'ASAP3': array(['ZNF341'], dtype=object),\n",
       " 'ITGB3BP': array(['EBF3', 'SPI1'], dtype=object),\n",
       " 'SLC9A1': array(['ESRRA', 'ESRRB', 'GRHL2', 'NR2F2', 'NR4A2', 'NR5A1', 'PKNOX1'],\n",
       "       dtype=object),\n",
       " 'CAPN2': ['missing'],\n",
       " 'RHOU': ['missing'],\n",
       " 'VIM': array(['FOS::JUN', 'FOS::JUNB', 'FOS::JUND', 'FOSL2::JUN', 'MZF1(var.2)',\n",
       "        'ZNF282'], dtype=object),\n",
       " 'ITGA8': array(['CDX2', 'ZNF384'], dtype=object),\n",
       " 'NRP1': array(['RBPJ', 'STAT1'], dtype=object),\n",
       " 'LDB3': array(['EBF1', 'EBF3', 'FOS', 'FOS::JUN', 'FOS::JUNB', 'FOS::JUND',\n",
       "        'FOSB::JUNB', 'FOSL2::JUN', 'FOSL2::JUNB', 'SPIB', 'STAT1::STAT2'],\n",
       "       dtype=object),\n",
       " 'PTEN': array(['TFAP2B(var.2)'], dtype=object),\n",
       " 'ADAM12': array(['EHF', 'ELF3', 'EWSR1-FLI1', 'GABPA', 'PAX1', 'PAX9', 'RFX2',\n",
       "        'RFX4', 'RFX5'], dtype=object),\n",
       " 'ITGB1': array(['EHF', 'ELF3', 'GABPA', 'PRDM4', 'SPI1', 'SPIB'], dtype=object),\n",
       " 'SVIL': ['missing'],\n",
       " 'SORBS1': array(['HOXB13', 'MAFF', 'MYBL1', 'NR2F1', 'NR2F2', 'NR4A1', 'OTX2',\n",
       "        'RORA', 'TFE3', 'TFEC', 'ZNF317'], dtype=object),\n",
       " 'VCL': array(['SMAD2::SMAD3::SMAD4', 'SRF', 'ZNF143', 'ZNF682'], dtype=object),\n",
       " 'ABI1': array(['MAZ', 'RREB1', 'THAP11', 'ZNF148', 'ZNF263', 'ZNF384', 'ZNF410'],\n",
       "       dtype=object),\n",
       " 'MAPK8': ['missing'],\n",
       " 'DOCK1': array(['NFATC2', 'NFATC4', 'RELA', 'SPI1', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'THY1': array(['IKZF1', 'MAFK', 'MAZ', 'ZNF148'], dtype=object),\n",
       " 'PARVA': array(['EWSR1-FLI1', 'NR2C2'], dtype=object),\n",
       " 'ARHGEF12': array(['CDX1', 'CDX2', 'CDX4', 'HOXA10', 'HOXD9', 'IRF3', 'IRF9'],\n",
       "       dtype=object),\n",
       " 'RDX': array(['CTCF', 'NFATC3', 'STAT1::STAT2'], dtype=object),\n",
       " 'LAYN': array(['PPARG'], dtype=object),\n",
       " 'PAK1': array(['IRF2', 'NFATC2', 'SPI1'], dtype=object),\n",
       " 'ILK': array(['RFX1'], dtype=object),\n",
       " 'CAPN1': array(['EHF', 'GATA3', 'MAZ', 'POU6F1', 'RREB1', 'ZBTB7B', 'ZNF148'],\n",
       "       dtype=object),\n",
       " 'ARHGAP32': array(['ELF1', 'ELF5', 'ETV1', 'ETV4', 'GBX2', 'SMAD2::SMAD3::SMAD4',\n",
       "        'ZBTB7A'], dtype=object),\n",
       " 'CBL': array(['CTCF', 'IRF3', 'KLF5', 'ZNF136'], dtype=object),\n",
       " 'CD151': array([], dtype=object),\n",
       " 'CFL1': array(['RREB1', 'SP3', 'SP8', 'SP9'], dtype=object),\n",
       " 'CORO1B': array(['EHF', 'ELF3', 'GABPA', 'KLF15', 'KLF5', 'SP2', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'INPPL1': array(['FOXF2', 'MYF5', 'SMAD2::SMAD3::SMAD4'], dtype=object),\n",
       " 'LPXN': array(['ELF1', 'ETV6', 'SMAD2::SMAD3::SMAD4', 'SPI1', 'ZNF140'],\n",
       "       dtype=object),\n",
       " 'CTTN': array(['EGR1', 'ELF2', 'ELF4', 'ELF5', 'ELK1', 'ELK3', 'ERF', 'ERG',\n",
       "        'ETS1', 'ETS2', 'ETV1', 'ETV2', 'ETV3', 'FEV', 'FLI1', 'HES2',\n",
       "        'IRF4', 'IRF8', 'KLF15', 'KLF5', 'SPDEF', 'SPIC'], dtype=object),\n",
       " 'HRAS': array(['ZNF449', 'ZNF460'], dtype=object),\n",
       " 'FERMT3': array(['E2F7', 'EWSR1-FLI1', 'ZBTB12', 'ZNF136'], dtype=object),\n",
       " 'SLC3A2': array([], dtype=object),\n",
       " 'PPFIA1': array(['EOMES', 'GFI1', 'KLF15', 'KLF5', 'NFYA', 'NFYC', 'TBR1', 'TBX2',\n",
       "        'TBX21', 'TBX6'], dtype=object),\n",
       " 'PTPRO': array(['ZSCAN4'], dtype=object),\n",
       " 'ITGA7': ['missing'],\n",
       " 'TUBA1B': array(['HIC2', 'IRF3', 'KLF15', 'KLF5', 'MYC', 'SP1', 'SP4', 'ZNF24'],\n",
       "       dtype=object),\n",
       " 'SSH1': array(['KLF4', 'NFATC2', 'NR1H2::RXRA', 'NR2F1(var.2)', 'NR2F2', 'NR4A1',\n",
       "        'PPARA::RXRA', 'RREB1', 'ZNF740'], dtype=object),\n",
       " 'TENC1': array(['MYOG', 'NR4A2::RXRA', 'POU1F1', 'POU2F3', 'POU5F1B', 'REST',\n",
       "        'ZNF354C', 'ZNF528'], dtype=object),\n",
       " 'ITGA5': array(['MZF1(var.2)', 'NFIC(var.2)', 'NFIX(var.2)', 'ZNF740'],\n",
       "       dtype=object),\n",
       " 'LRP1': array(['BCL6B', 'NR2C2', 'NRF1', 'ZNF263'], dtype=object),\n",
       " 'PXN': ['missing'],\n",
       " 'AGAP2': array([], dtype=object),\n",
       " 'PTPN6': array(['GABPA', 'TBX19', 'TBXT'], dtype=object),\n",
       " 'PTPN11': array(['ATF3', 'CREM', 'ELF1', 'ETV1', 'ETV4', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUND(var.2)', 'HOXD3', 'IKZF1', 'ONECUT2', 'ONECUT3',\n",
       "        'ZKSCAN1'], dtype=object),\n",
       " 'GIT2': array(['BHLHE22(var.2)', 'FERD3L', 'MITF', 'TFAP2A', 'YY1'], dtype=object),\n",
       " 'ITGB7': array(['EHF', 'ELF3', 'KLF4', 'KLF5', 'RREB1', 'SP2', 'SPIB', 'ZNF682'],\n",
       "       dtype=object),\n",
       " 'STARD13': array(['NRL'], dtype=object),\n",
       " 'ARHGEF7': array(['ATF2', 'ATF3', 'ATF7', 'CDX2', 'CREB1', 'CREM', 'FOSB::JUN',\n",
       "        'FOSB::JUNB(var.2)', 'FOSL1::JUN(var.2)', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'FOSL2::JUND(var.2)', 'FOXC1', 'FOXC2',\n",
       "        'FOXD2', 'FOXE1', 'FOXF2', 'FOXG1', 'FOXK1', 'FOXK2', 'FOXN3',\n",
       "        'FOXO3', 'FOXO6', 'FOXP1', 'HOXB13', 'IRF1', 'JUN',\n",
       "        'JUN::JUNB(var.2)', 'JUND(var.2)', 'MEF2A', 'MEF2C', 'SOX10',\n",
       "        'SOX4', 'SPIB', 'SRY', 'STAT1', 'STAT1::STAT2', 'STAT3', 'ZNF384'],\n",
       "       dtype=object),\n",
       " 'MMP14': array(['BCL6B', 'IRF1', 'NR1D1'], dtype=object),\n",
       " 'FERMT2': array(['BHLHA15(var.2)', 'FOXN3', 'ZNF140'], dtype=object),\n",
       " 'ACTN1': array(['LEF1', 'NR5A1', 'ZNF354C'], dtype=object),\n",
       " 'ZFYVE21': array(['NFYA', 'NFYC', 'PBX1', 'VENTX'], dtype=object),\n",
       " 'HSPA2': array(['FOXN3', 'NFYB', 'POU2F2'], dtype=object),\n",
       " 'KTN1': array(['ARNT::HIF1A', 'KLF5'], dtype=object),\n",
       " 'ARHGAP5': array(['MAZ', 'ZNF148'], dtype=object),\n",
       " 'AKT1': array(['MYBL1'], dtype=object),\n",
       " 'SYNM': array(['CTCFL', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D', 'NFATC3',\n",
       "        'STAT1::STAT2', 'THAP1', 'ZNF354C'], dtype=object),\n",
       " 'ITGA11': array(['ELF3', 'GABPA', 'IRF3', 'MAZ', 'SPIB', 'ZNF354C'], dtype=object),\n",
       " 'CSK': array(['CTCF', 'CTCFL', 'KLF10', 'POU4F1', 'ZIC3', 'ZIC4', 'ZIC5'],\n",
       "       dtype=object),\n",
       " 'PEAK1': array(['EWSR1-FLI1', 'KLF5', 'NFYA', 'NFYC', 'SP1', 'SP2', 'SP4'],\n",
       "       dtype=object),\n",
       " 'TRPM7': array(['BCL6B', 'MLX', 'MLXIPL', 'NFIL3', 'STAT1', 'TFE3'], dtype=object),\n",
       " 'CIB2': array(['E2F6', 'ELF3', 'ELK4', 'ETV5', 'NR2C2', 'NR2F1(var.2)', 'PPARD',\n",
       "        'RXRB', 'RXRG', 'TFDP1', 'THRB', 'ZNF354C'], dtype=object),\n",
       " 'CIB1': array(['NR2C2(var.2)', 'RARA(var.2)', 'RARA::RXRG', 'RARB(var.3)',\n",
       "        'RXRB(var.2)', 'RXRG(var.2)', 'SNAI3', 'SOX21'], dtype=object),\n",
       " 'BCAR1': array(['BCL6B', 'SREBF2', 'TFAP2C(var.2)'], dtype=object),\n",
       " 'MAPK8IP3': array(['ATF3', 'ATF7', 'CREB1', 'CREB3L4(var.2)', 'CREM', 'ELK4', 'ETS2',\n",
       "        'ETV5', 'FOSB::JUN', 'FOSB::JUNB(var.2)', 'FOSL1::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'FOSL2::JUND(var.2)', 'JDP2(var.2)',\n",
       "        'JUN::JUNB(var.2)', 'JUNB(var.2)', 'MYOG', 'NHLH1', 'NR2C1',\n",
       "        'NR4A2', 'PBX3', 'PKNOX1', 'RORA', 'RORB', 'SIX1'], dtype=object),\n",
       " 'TGFB1I1': array([], dtype=object),\n",
       " 'PDPK1': ['missing'],\n",
       " 'ITGAL': array(['ELF1', 'ETV1', 'HNF4A(var.2)', 'IKZF1', 'NR1H2::RXRA',\n",
       "        'NR2F1(var.2)', 'PLAG1', 'PPARA::RXRA', 'PPARD', 'RXRB', 'RXRG',\n",
       "        'THRB'], dtype=object),\n",
       " 'ITGAX': array(['ESRRA', 'ESRRB', 'FOS', 'NR2F1', 'NR2F1(var.2)', 'NR2F2', 'NR5A1',\n",
       "        'NR6A1', 'RORA', 'SPDEF', 'THRB'], dtype=object),\n",
       " 'SH2B1': ['missing'],\n",
       " 'PKD1': array(['ATF2', 'ATF3', 'CREB1', 'FOSB::JUN', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'JDP2(var.2)', 'JUN', 'JUNB(var.2)',\n",
       "        'JUND(var.2)'], dtype=object),\n",
       " 'ITGAM': array(['ARGFX', 'EWSR1-FLI1', 'ISX', 'LHX9', 'MEF2A', 'MIXL1', 'MNX1',\n",
       "        'NKX6-2', 'POU6F2', 'PRRX1', 'PRRX2', 'RAX', 'UNCX', 'ZNF135',\n",
       "        'ZNF263', 'ZNF460'], dtype=object),\n",
       " 'NUDT16L1': array(['MTF1', 'NFIA', 'RUNX2', 'ZBTB6'], dtype=object),\n",
       " 'ITGAD': array(['EWSR1-FLI1', 'KLF13', 'POU4F2', 'POU4F3', 'SP1', 'SP4', 'SP9',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'SRCIN1': array(['CDX4', 'HOXA13', 'HOXC13'], dtype=object),\n",
       " 'STAT3': array(['BACH2(var.2)', 'MAX', 'MAX::MYC', 'POU4F1', 'POU4F3', 'PRDM1',\n",
       "        'ZNF341'], dtype=object),\n",
       " 'GIT1': array(['BACH2', 'BATF', 'BATF3', 'BATF::JUN', 'ELF3', 'FOS', 'FOS::JUN',\n",
       "        'FOS::JUNB', 'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'FOSL1::JUNB',\n",
       "        'FOSL1::JUND', 'FOSL2', 'FOSL2::JUN', 'FOSL2::JUNB', 'FOSL2::JUND',\n",
       "        'GSC', 'GSC2', 'IRF2', 'IRF7', 'JDP2', 'JUN(var.2)', 'JUN::JUNB',\n",
       "        'JUNB', 'JUND', 'MZF1', 'NFE2', 'OTX1', 'PITX1', 'PITX2', 'PITX3',\n",
       "        'PRDM1'], dtype=object),\n",
       " 'ITGB4': array(['BACH1', 'BATF', 'BATF3', 'BATF::JUN', 'CDX4', 'FOS', 'FOS::JUNB',\n",
       "        'FOSB::JUNB', 'FOSL1', 'FOSL1::JUN', 'HOXA13', 'HOXD13', 'JUNB',\n",
       "        'JUND', 'MAFK', 'SNAI3'], dtype=object),\n",
       " 'PRKCA': array(['ETV4', 'MZF1', 'SOX9', 'SPI1', 'ZNF384'], dtype=object),\n",
       " 'PFN1': array(['ELF1', 'ELF5', 'ETV1', 'ETV4', 'HOXB13', 'IKZF1', 'KLF15', 'KLF5',\n",
       "        'SP1', 'SP2', 'SP4', 'SPI1', 'ZBTB7A', 'ZNF263'], dtype=object),\n",
       " 'NDEL1': array(['ATF6', 'KLF15', 'KLF16', 'KLF4', 'KLF5', 'NR1I3', 'NRF1'],\n",
       "       dtype=object),\n",
       " 'GRB2': array(['EWSR1-FLI1', 'NFYB', 'PKNOX2', 'SPI1', 'SPIC', 'TGIF2'],\n",
       "       dtype=object),\n",
       " 'GRB7': array(['SP4', 'SRY'], dtype=object),\n",
       " 'ITGA3': array(['IRF1', 'STAT1::STAT2'], dtype=object),\n",
       " 'CRK': array(['MYB', 'NEUROD1', 'NEUROG2(var.2)', 'NFATC2', 'NR2C1', 'NRL',\n",
       "        'TGIF1'], dtype=object),\n",
       " 'ITGAE': array(['ASCL1', 'FERD3L', 'MYOG', 'SP2', 'ZSCAN4'], dtype=object),\n",
       " 'ABI3': array(['EHF', 'EWSR1-FLI1', 'GFI1', 'GRHL2', 'IRF2', 'NFATC4', 'SPIB',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'SLC16A3': array(['GRHL2', 'NFATC3', 'POU1F1'], dtype=object),\n",
       " 'ITGB3': array(['EOMES', 'MAZ', 'NRL', 'PAX3', 'PAX7', 'PPARG', 'PROP1', 'TBR1',\n",
       "        'TBX2', 'TBX21', 'ZNF148'], dtype=object),\n",
       " 'LASP1': array(['TWIST1', 'ZBTB18', 'ZNF75D'], dtype=object),\n",
       " 'PTPN2': array(['ARNT::HIF1A', 'EWSR1-FLI1'], dtype=object),\n",
       " 'MYOM1': array(['ARNT::HIF1A', 'ZBTB14', 'ZNF460'], dtype=object),\n",
       " 'ROCK1': array(['ASCL1', 'ETV1', 'ETV4', 'IKZF1', 'KLF4', 'KLF5', 'MYOG', 'SP1',\n",
       "        'SP4', 'SPI1', 'THAP1'], dtype=object),\n",
       " 'INSR': array(['FOXA1', 'FOXB1', 'FOXC1', 'FOXC2', 'FOXD2', 'FOXE1', 'FOXF2',\n",
       "        'IRF3', 'IRF4', 'IRF9', 'ZNF384'], dtype=object),\n",
       " 'CALR': array(['INSM1', 'MEIS1', 'RXRB', 'STAT1::STAT2'], dtype=object),\n",
       " 'PLAUR': array(['E2F6', 'MEF2A', 'MEF2D', 'RUNX3', 'RXRB(var.2)', 'RXRG(var.2)',\n",
       "        'ZBTB12'], dtype=object),\n",
       " 'PRKACA': array(['GLIS3', 'IRF1', 'STAT1::STAT2'], dtype=object),\n",
       " 'DNM2': array(['CEBPA', 'CEBPD', 'EBF1', 'MEIS1', 'MZF1', 'PBX3', 'PKNOX1',\n",
       "        'SPIB'], dtype=object),\n",
       " 'KEAP1': array(['FOS', 'GLIS3', 'MZF1', 'ZBTB12'], dtype=object),\n",
       " 'PVR': array(['IKZF1', 'NFKB2', 'TFDP1'], dtype=object),\n",
       " 'VASP': array(['MEIS1', 'ZBTB7B', 'ZNF143'], dtype=object),\n",
       " 'CYTH2': array(['SP1', 'SP2', 'SP4', 'ZNF263'], dtype=object),\n",
       " 'PTPRH': array(['EHF', 'ELF3', 'GABPA', 'IRF1', 'PRDM1', 'SPIB'], dtype=object),\n",
       " 'VAV1': array(['ELF1', 'RREB1', 'SPI1'], dtype=object),\n",
       " 'PIP5K1C': array(['FOS::JUN', 'FOS::JUND', 'FOSL2::JUN', 'HES7', 'KLF15', 'KLF16',\n",
       "        'KLF5', 'NFE2', 'SOX21', 'SP1', 'SP4'], dtype=object),\n",
       " 'RAVER1': array(['E2F6', 'GATA1', 'HOXA13', 'HOXB13', 'OSR2', 'TFDP1', 'YY1', 'YY2',\n",
       "        'ZFP42', 'ZNF354C', 'ZSCAN29'], dtype=object),\n",
       " 'TNS1': array(['SOHLH2'], dtype=object),\n",
       " 'SOS1': array(['ASCL1', 'MYOG', 'THAP11'], dtype=object),\n",
       " 'LIMS2': array(['ASCL1', 'BHLHE22(var.2)', 'IRF1', 'NHLH1', 'NR5A1', 'PRDM1',\n",
       "        'ZNF354C'], dtype=object),\n",
       " 'FHL2': array(['LEF1', 'TCF7L1', 'TCF7L2', 'ZNF384'], dtype=object),\n",
       " 'IRS1': array(['CTCF', 'EBF1'], dtype=object),\n",
       " 'INPP5D': array(['ELF1', 'ETV1', 'ETV4', 'IKZF1', 'IRF1', 'POU6F1(var.2)', 'PRDM1',\n",
       "        'SPI1', 'SPIC', 'STAT1::STAT2', 'TBXT'], dtype=object),\n",
       " 'ITGA4': array(['GATA1::TAL1', 'GATA6', 'HES6', 'SOHLH2'], dtype=object),\n",
       " 'NRP2': array(['LMX1B', 'PAX6', 'SPI1'], dtype=object),\n",
       " 'ITGA6': array(['FIGLA', 'FOXE1', 'HNF1A'], dtype=object),\n",
       " 'ITGAV': array(['EHF', 'ELF3', 'ELF5', 'EWSR1-FLI1', 'SPIB'], dtype=object),\n",
       " 'ITGB6': ['missing'],\n",
       " 'ITGB1BP1': array(['MAX::MYC', 'MSANTD3', 'NR2F1(var.3)', 'PKNOX2', 'TGIF2',\n",
       "        'THRB(var.2)', 'ZNF317'], dtype=object),\n",
       " 'ASAP2': array(['KLF5', 'SOX13', 'SOX18', 'SOX2', 'SOX4', 'SOX8', 'SRY'],\n",
       "       dtype=object),\n",
       " 'ARPC2': array(['ETV6'], dtype=object),\n",
       " 'LIMS1': ['missing'],\n",
       " 'ILKAP': array(['FOXC1', 'ZNF384'], dtype=object),\n",
       " 'NCK2': ['missing'],\n",
       " 'ABI2': array(['ELF1', 'ELF4', 'ELK1', 'ELK3', 'ERF', 'ERG', 'ETS1', 'ETV1',\n",
       "        'ETV3', 'ETV4', 'ETV6', 'EWSR1-FLI1', 'FEV', 'FLI1', 'HSF4',\n",
       "        'KLF15', 'KLF5', 'MYBL1', 'PRDM4', 'SOX10', 'SOX4', 'SOX8', 'SRY',\n",
       "        'ZBTB7A', 'ZNF652'], dtype=object),\n",
       " 'CASP8': ['missing'],\n",
       " 'SDC4': array(['ASCL1(var.2)', 'MEF2B', 'MEF2D', 'NHLH1', 'TCF12(var.2)', 'TCF4',\n",
       "        'ZNF341'], dtype=object),\n",
       " 'FERMT1': array(['SPI1'], dtype=object),\n",
       " 'SIRPA': array(['FOXC1', 'FOXC2', 'FOXD2', 'FOXE1', 'GRHL1', 'GRHL2', 'ZNF652'],\n",
       "       dtype=object),\n",
       " 'PLCG1': array(['EGR1', 'HNF4A', 'HNF4G', 'KLF15', 'KLF5', 'PROP1'], dtype=object),\n",
       " 'PTPN1': array(['BATF', 'BATF3', 'BATF::JUN', 'ETV6', 'MSANTD3', 'NR1I2', 'RFX5',\n",
       "        'SPIB'], dtype=object),\n",
       " 'PRNP': array(['NR3C2'], dtype=object),\n",
       " 'CASS4': array(['EWSR1-FLI1', 'NR1H2::RXRA', 'PBX3', 'PKNOX1', 'RXRB', 'TCF7L1',\n",
       "        'TCF7L2'], dtype=object),\n",
       " 'SRC': ['missing'],\n",
       " 'PTPRA': ['missing'],\n",
       " 'ITGB2': array(['ELF1', 'ELF3', 'ELF5', 'ERF', 'ETV1', 'ETV2', 'ETV4', 'IKZF1',\n",
       "        'NR2C2', 'OSR1', 'SPI1', 'TBR1'], dtype=object),\n",
       " 'TIAM1': array(['E2F8', 'NFATC2', 'PRDM1'], dtype=object),\n",
       " 'MYH9': array(['RARA::RXRA'], dtype=object),\n",
       " 'NF2': array(['NFATC3', 'NKX2-2', 'NKX2-5', 'THRB(var.2)'], dtype=object),\n",
       " 'CRKL': array(['ATF3', 'ATF7', 'DBP', 'FOSB::JUN', 'FOSB::JUNB(var.2)',\n",
       "        'FOSL1::JUN(var.2)', 'FOSL1::JUND(var.2)', 'FOSL2::JUN(var.2)',\n",
       "        'FOSL2::JUNB(var.2)', 'GMEB2', 'GRHL2', 'MZF1(var.2)', 'OSR2',\n",
       "        'TEF', 'TFAP2A'], dtype=object),\n",
       " 'PPM1F': array(['FOXC1', 'FOXC2', 'GABPA', 'MEF2A', 'MEF2B', 'MEF2C', 'MEF2D',\n",
       "        'MEIS1', 'MYF5', 'ONECUT1', 'SOX21', 'SPIB'], dtype=object),\n",
       " 'MAPK1': array(['EWSR1-FLI1', 'KLF15', 'KLF5', 'YY1', 'YY2', 'ZFP42'], dtype=object),\n",
       " 'RNF185': array(['MITF', 'NR2C2', 'THAP1', 'USF1', 'USF2', 'ZBTB7A', 'ZNF135',\n",
       "        'ZNF460'], dtype=object),\n",
       " 'ANKRD28': array(['GLIS2', 'RFX2', 'RFX4', 'RFX5'], dtype=object),\n",
       " 'CD47': array(['ETV6', 'NEUROG1', 'NEUROG2', 'NFATC2', 'NFATC4', 'OTX2', 'TGIF1',\n",
       "        'TGIF2', 'TGIF2LX', 'TGIF2LY'], dtype=object),\n",
       " 'LPP': ['missing'],\n",
       " 'PLD1': array(['PKNOX1', 'ZNF354C'], dtype=object),\n",
       " 'PPM1M': array(['PBX3', 'PKNOX1', 'ZKSCAN5'], dtype=object),\n",
       " 'ITGB5': array(['KLF5', 'SPI1', 'ZNF460'], dtype=object),\n",
       " 'RHOA': array(['SNAI2'], dtype=object),\n",
       " 'NISCH': array(['CEBPG', 'DBP', 'HLF', 'IRF1', 'IRF7', 'NR4A2::RXRA', 'SP1', 'SP4',\n",
       "        'SP8', 'SP9', 'YY1', 'YY2', 'ZFP42'], dtype=object),\n",
       " 'PIK3CA': array(['FOXN3', 'HOXA13', 'HOXB13', 'IRF1', 'SPIB', 'ZNF384'],\n",
       "       dtype=object),\n",
       " 'ITGA9': array(['CTCF', 'GATA1', 'GATA2', 'GATA4', 'GATA6', 'IRF3', 'SOX10'],\n",
       "       dtype=object),\n",
       " 'SORBS2': array(['GATA3', 'MEF2C', 'NFYB', 'TFAP2A(var.3)', 'TFAP2B(var.3)',\n",
       "        'TFAP2C(var.3)'], dtype=object),\n",
       " 'GAB1': array(['OSR2'], dtype=object),\n",
       " 'PALLD': array(['SIX2'], dtype=object),\n",
       " 'ARHGAP24': array(['NKX2-8', 'SIX2'], dtype=object),\n",
       " 'PDE4D': array(['FOXC2', 'FOXE1', 'KLF14'], dtype=object),\n",
       " 'ARHGAP26': array(['DUX4', 'FOXC2', 'FOXE1'], dtype=object),\n",
       " 'GNB2L1': array(['ATF4', 'CEBPG(var.2)', 'MZF1', 'NEUROD1', 'NEUROG2(var.2)',\n",
       "        'PKNOX2', 'TGIF1', 'TGIF2'], dtype=object),\n",
       " 'ITGA1': array(['TBX18', 'TBX20'], dtype=object),\n",
       " 'RASA1': array(['EMX1', 'EMX2', 'EN1', 'HOXB3', 'KLF4', 'KLF5', 'MEIS1', 'MSGN1',\n",
       "        'NEUROD2', 'NOTO', 'PAX4', 'SREBF2', 'TCF21(var.2)', 'TLX2',\n",
       "        'VSX2', 'ZNF135', 'ZNF460'], dtype=object),\n",
       " 'TRIO': ['missing'],\n",
       " 'ITGA2': array(['HIC2', 'MZF1'], dtype=object),\n",
       " 'EZR': array(['HOXC4', 'SPIB', 'ZNF354C'], dtype=object),\n",
       " 'MARCKS': array(['KLF14', 'KLF15', 'KLF16', 'LMX1A', 'LMX1B', 'MAZ', 'MEIS2(var.2)',\n",
       "        'PAX3', 'SOX9', 'SP1', 'SP3', 'SP4', 'SP8', 'SP9', 'ZNF148'],\n",
       "       dtype=object),\n",
       " 'FYN': array(['HNF4A(var.2)', 'SNAI1', 'SNAI3', 'TFAP2A(var.3)', 'TFAP2B(var.3)',\n",
       "        'TFAP2C(var.3)'], dtype=object),\n",
       " 'NEDD9': ['missing'],\n",
       " 'DEF6': array(['SIX2', 'TFCP2', 'ZBTB26', 'ZNF16'], dtype=object),\n",
       " 'KCNH2': array(['MAFK', 'MEIS1(var.2)', 'NKX6-3', 'PBX2'], dtype=object),\n",
       " 'ITGB8': array(['RARA'], dtype=object),\n",
       " 'ACTB': array(['ZNF354C'], dtype=object),\n",
       " 'RAC1': array(['BCL6B', 'E2F3', 'EHF', 'ELF3', 'GABPA', 'IRF1', 'KLF2', 'KLF3',\n",
       "        'KLF6', 'NFATC2', 'PRDM4', 'SPIB', 'YY1', 'ZFP42'], dtype=object),\n",
       " 'HSPB1': array(['FOXC2', 'FOXN3', 'HSF1', 'HSF2', 'HSF4', 'NFIB', 'NFIC(var.2)',\n",
       "        'NFIX(var.2)', 'ZNF384'], dtype=object),\n",
       " 'TES': array(['HOXB13', 'NFATC4'], dtype=object),\n",
       " 'LIMK1': array(['EWSR1-FLI1', 'KLF15', 'KLF5', 'ONECUT1', 'POU4F1', 'POU4F2',\n",
       "        'POU4F3'], dtype=object),\n",
       " 'PTPN12': array(['SOX15', 'SOX9', 'ZBTB7B'], dtype=object),\n",
       " 'TRIP6': array(['SPI1'], dtype=object),\n",
       " 'PLEC': array(['ATF2', 'CREB1', 'JUN'], dtype=object),\n",
       " 'LYN': array(['ELF1', 'ELF5', 'POU1F1', 'POU2F1', 'POU2F3', 'POU3F1', 'POU3F2',\n",
       "        'POU3F3', 'POU3F4', 'POU5F1', 'POU5F1B', 'PRDM1', 'RARA', 'SPIB'],\n",
       "       dtype=object),\n",
       " 'PABPC1': array(['ELF2', 'ELF4', 'ELF5', 'EWSR1-FLI1', 'IRF1', 'KLF5', 'NRF1',\n",
       "        'SP2', 'SP4', 'SPIB'], dtype=object),\n",
       " 'SHARPIN': array(['SMAD2::SMAD3::SMAD4', 'ZNF460'], dtype=object),\n",
       " 'DLC1': array(['ARGFX', 'ISX', 'LHX9', 'NKX6-2', 'PRRX1', 'PRRX2', 'UNCX'],\n",
       "       dtype=object),\n",
       " 'PTK2': array(['FOXI1', 'FOXL1', 'FOXO4', 'FOXO6', 'FOXP3', 'NFATC2', 'PRDM1',\n",
       "        'ZBTB26'], dtype=object),\n",
       " 'SORBS3': array(['IRF3', 'IRF4', 'IRF5', 'IRF8', 'SIX1', 'STAT3'], dtype=object),\n",
       " 'PTK2B': ['missing'],\n",
       " 'SDCBP': array(['RFX2', 'RFX5', 'SREBF1', 'TCF7'], dtype=object),\n",
       " 'TLN1': array(['IKZF1', 'MEF2C', 'POU1F1', 'POU2F1', 'POU2F3', 'POU3F1', 'POU3F2',\n",
       "        'POU3F3', 'POU3F4', 'POU5F1', 'POU5F1B', 'SPI1'], dtype=object),\n",
       " 'RAPGEF1': array(['ASCL1', 'EWSR1-FLI1', 'MYOG', 'NFIC'], dtype=object),\n",
       " 'ABL1': array(['MEIS1(var.2)', 'THAP11', 'ZFP42'], dtype=object),\n",
       " 'VAV2': array(['ELF1', 'ETV1', 'ETV4', 'HNF4A', 'HNF4G', 'HOXA9', 'KLF5', 'SP2'],\n",
       "       dtype=object),\n",
       " 'ENG': ['missing'],\n",
       " 'TESK1': array(['ATF3', 'FOS::JUN(var.2)', 'FOSL2::JUNB(var.2)',\n",
       "        'FOSL2::JUND(var.2)', 'JDP2(var.2)'], dtype=object),\n",
       " 'OSTF1': array(['ELF1', 'ETV1', 'HOXC4', 'NR1D1', 'NR1D2', 'PAX5', 'RXRB(var.2)',\n",
       "        'RXRG(var.2)', 'ZNF384'], dtype=object),\n",
       " 'SPTLC1': array(['ESR1', 'ESR2', 'FOXI1', 'FOXL1', 'FOXO4', 'FOXO6', 'FOXP3',\n",
       "        'GSX2', 'HOXA2', 'HOXB2', 'IRF4', 'IRF8', 'MSANTD3', 'NFIL3',\n",
       "        'PDX1', 'PPARG', 'ZNF528'], dtype=object),\n",
       " 'CORO2A': array(['FOXE1', 'NFATC2', 'NKX2-8', 'NR2F1(var.3)', 'THRB(var.3)',\n",
       "        'ZBTB12', 'ZNF135', 'ZNF460'], dtype=object)}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save dictionary mapping selected genes to their potential TFs to pickle\n",
    "pickle.dump(gene_to_tfs, open(saving_dir+'gene_to_tfs_annotated.pkl', 'wb'))\n",
    "gene_to_tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>tfs</th>\n",
       "      <th>short_description</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cage_start</th>\n",
       "      <th>cage_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p1@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2588012</td>\n",
       "      <td>2588031</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p2@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2587980</td>\n",
       "      <td>2587994</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p3@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2587999</td>\n",
       "      <td>2588011</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p4@PDPK1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2588131</td>\n",
       "      <td>2588143</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876</th>\n",
       "      <td>PDPK1</td>\n",
       "      <td>missing</td>\n",
       "      <td>p@chr16:2589098..2589102,+</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2589098</td>\n",
       "      <td>2589102</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes      tfs           short_description  chrom  cage_start  cage_end  \\\n",
       "430   PDPK1  missing                    p1@PDPK1  chr16     2588012   2588031   \n",
       "736   PDPK1  missing                    p2@PDPK1  chr16     2587980   2587994   \n",
       "952   PDPK1  missing                    p3@PDPK1  chr16     2587999   2588011   \n",
       "1123  PDPK1  missing                    p4@PDPK1  chr16     2588131   2588143   \n",
       "1876  PDPK1  missing  p@chr16:2589098..2589102,+  chr16     2589098   2589102   \n",
       "\n",
       "     strand  \n",
       "430       +  \n",
       "736       +  \n",
       "952       +  \n",
       "1123      +  \n",
       "1876      +  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_and_genes_df[tf_and_genes_df['genes']=='PDPK1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Confirm findings using ChIP-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
