{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure FISH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'utils' from '/home/louiscam/projects/gpcr/code/processing_FISH/utils.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, fcluster, cut_tree\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from sklearn.metrics import adjusted_mutual_info_score\n",
    "import seaborn as sns\n",
    "from scipy import sparse\n",
    "import scipy.stats as ss\n",
    "import csv\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import community\n",
    "import communities as com\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "import operator\n",
    "from scipy.sparse import csr_matrix\n",
    "import itertools\n",
    "import os.path\n",
    "import math\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.metrics import adjusted_mutual_info_score\n",
    "from networkx.algorithms.community.kclique import k_clique_communities\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import OmicsIntegrator as oi\n",
    "import gseapy\n",
    "from gseapy.plot import barplot, dotplot\n",
    "from ortools.linear_solver import pywraplp\n",
    "import pyvis as pv\n",
    "from pyvis.network import Network\n",
    "from joblib import Parallel, delayed\n",
    "\n",
    "import utils as lu\n",
    "reload(lu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = '/home/louiscam/projects/gpcr/save/processed_hic_data/processed_hic_data_IMR90/final_BP250000_intraKR_interINTERKR/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/regulatory_data/regulatory_data_IMR90/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_regulatory_marks/processed_epigenome_data_IMR90/'\n",
    "# Directory of FISH data\n",
    "dir_fish = '/home/louiscam/projects/gpcr/data/fish_data/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'\n",
    "saving_main_dir = '/home/louiscam/projects/gpcr/save/figures_main/'\n",
    "saving_supp_dir = '/home/louiscam/projects/gpcr/save/figures_supplement/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene to chromosome directory\n",
    "with open(saving_dir+'gene2chrom.pkl', 'rb') as f:\n",
    "    gene2chrom = pickle.load(f)\n",
    "# Load gene to locus directory\n",
    "with open(saving_dir+'gene2locus.pkl', 'rb') as f:\n",
    "    gene2locus = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load FISH data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genomic scale FISH data\n",
    "Including for each locus whether the genes measured within it are actively transcribing and the distances from the locus to the nuclear lamina, the nearest nuclear speckle and the nearest nucleolus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "merfish_data = pd.read_csv(dir_fish+'dna_merfish_data.tsv', sep='\\t', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3668/3668 [00:43<00:00, 84.29it/s]\n"
     ]
    }
   ],
   "source": [
    "# For each cell, determine its nucleus size (maximum probe-probe distance)\n",
    "merfish_data_reduced = merfish_data[['cell number', 'genomic coordinate', 'x(nm)', 'y(nm)', 'z(nm)']]\n",
    "merfish_data_reduced = merfish_data_reduced.dropna(how='any', axis=0)\n",
    "cell_ids = merfish_data_reduced['cell number'].unique()\n",
    "max_dist_dict = {}\n",
    "for cell_id in tqdm(cell_ids):\n",
    "    coords = merfish_data_reduced[merfish_data_reduced['cell number']==cell_id][['x(nm)', 'y(nm)', 'z(nm)']]\n",
    "    dist = pdist(coords, metric='euclidean')\n",
    "    max_dist = dist.max()\n",
    "    max_dist_dict[cell_id] = max_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build dictionary from Ensembl gene name to HGNC\n",
    "enst2hgnc_df = pd.read_csv(dir_genome+'ensemblGene2Name', sep='\\t', header=0)\n",
    "enst2hgnc_df.columns = ['ENST', 'HGNC']\n",
    "enst2hgnc_df = enst2hgnc_df.set_index('ENST')\n",
    "ensg2enst_df = pd.read_csv(dir_genome+'ensemblGene2Transcript2Protein', sep='\\t', header=0)\n",
    "ensg2enst_df.columns = ['ENSG', 'ENST', 'ENSP']\n",
    "ensg2enst_df = ensg2enst_df[['ENSG', 'ENST']]\n",
    "ensg2enst_df = ensg2enst_df.set_index('ENST')\n",
    "ensg2hgnc_df = ensg2enst_df.join(enst2hgnc_df, how='outer').reset_index()\n",
    "ensg2hgnc_df = ensg2hgnc_df[['ENSG', 'HGNC']].drop_duplicates(['ENSG', 'HGNC'])\n",
    "ensg2hgnc_df['HGNC'] = ensg2hgnc_df['HGNC'].str.upper()\n",
    "ensg2hgnc_df = ensg2hgnc_df.reset_index()[['ENSG', 'HGNC']]\n",
    "ensg2hgnc = {ensg2hgnc_df.loc[i,'ENSG']: ensg2hgnc_df.loc[i,'HGNC'] for i in range(len(ensg2hgnc_df))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gene information to MERFISH data\n",
    "merfish_data['gene names'] = merfish_data['gene names'].astype(str)\n",
    "merfish_data['gene names'] = merfish_data['gene names'].apply(lambda x: x.split(',') if x!='nan' else [])\n",
    "merfish_data['gene_hgnc'] = merfish_data['gene names'].apply(lambda x: [ensg2hgnc[g] if (g in ensg2hgnc.keys()) \n",
    "                                                                    else 'unknown' for g in x] if len(x)>0 else [])\n",
    "merfish_data['n_genes'] = merfish_data['gene_hgnc'].apply(lambda x: len(x))\n",
    "merfish_data['transciption'] = merfish_data['transciption'].astype(str)\n",
    "merfish_data['transciption'] = merfish_data['transciption'].apply(lambda x: x.split(',') if x!='nan' else [])\n",
    "merfish_data['chrom'] = merfish_data['genomic coordinate'].str.split(':', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reduce data (drop NAs, chrX and gene-less loci)\n",
    "merfish_data = merfish_data.dropna(how='any')\n",
    "merfish_data = merfish_data[merfish_data['chrom']!='chrX']\n",
    "merfish_data = merfish_data[['cell number',\n",
    "                             'chrom',\n",
    "                             'homolog number',\n",
    "                             'genomic coordinate',\n",
    "                             'gene_hgnc',\n",
    "                             'transciption',\n",
    "                             'x(nm)', 'y(nm)', 'z(nm)']]\n",
    "merfish_data.columns = [ 'cell_ID',\n",
    "                         'chrom',\n",
    "                         'homolog',\n",
    "                         'probe_locus',\n",
    "                         'gene_ID',\n",
    "                         'transcription',\n",
    "                         'x(nm)', 'y(nm)', 'z(nm)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of cells = 3533\n",
      "Total number of loci = 990\n"
     ]
    }
   ],
   "source": [
    "# Print basic statistics\n",
    "print(f'Total number of cells = {len(np.unique(merfish_data[\"cell_ID\"].values))}')\n",
    "print(f'Total number of loci = {len(np.unique(merfish_data[\"probe_locus\"].values))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes that we want to validate\n",
    "# Load all adhesome genes\n",
    "with open(saving_dir+'all_adhesome_genes.pkl', 'rb') as f:\n",
    "    all_adhesome_genes = pickle.load(f)\n",
    "# Load active adhesome genes\n",
    "with open(saving_dir+'active_adhesome_genes.pkl', 'rb') as f:\n",
    "    active_adhesome_genes = pickle.load(f)\n",
    "# Load adhesome TF genes and loci\n",
    "with open(saving_dir+'active_lung_adhesome_tf_genes.pkl', 'rb') as f:\n",
    "    active_lung_adhesome_tf_genes = pickle.load(f)\n",
    "active_adhesome_genes = np.concatenate([active_adhesome_genes, active_lung_adhesome_tf_genes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample random non-adhesome active genes\n",
    "with open(saving_dir+'all_active_genes.pkl', 'rb') as f:\n",
    "    all_active_genes = pickle.load(f)\n",
    "with open(saving_dir+'all_inactive_genes.pkl', 'rb') as f:\n",
    "    all_inactive_genes = pickle.load(f)\n",
    "active_gene_pool = list(set(all_active_genes)-set(active_adhesome_genes))\n",
    "active_gene_pool = np.array(active_gene_pool)\n",
    "random_active_genes = active_gene_pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of selected genes = 22595\n"
     ]
    }
   ],
   "source": [
    "# Genes for downstream analysis\n",
    "selected_genes = np.concatenate([active_adhesome_genes, random_active_genes])\n",
    "print(f'Number of selected genes = {len(selected_genes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Location of genes in hg38\n",
    "df_loc = pd.read_csv(dir_genome+'chrom_hg38.loc', sep='\\t', header=0)\n",
    "df_loc = df_loc[df_loc['#chrom'].isin(['chr'+str(i) for i in range(1,23)])]\n",
    "df_loc = df_loc[['#chrom', 'chromStart', 'chromEnd', 'geneName']]\n",
    "df_loc.columns = ['chrom', 'start', 'end', 'geneName']\n",
    "df_loc = df_loc.sort_values(['chrom','start','end'])\n",
    "df_loc = df_loc.groupby(['chrom','geneName']).agg({'start':'min', 'end':'max'})\n",
    "df_loc = df_loc.reset_index().sort_values(['chrom','start','end'])\n",
    "df_loc_adhesome = df_loc[df_loc['geneName'].isin(selected_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify adhesome genes present within 500kb of available MERFISH probes\n",
    "resol = 500000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a bed object for adhesome genes\n",
    "df_adhesome_pos = df_loc_adhesome[['chrom','start','end', 'geneName']]\n",
    "bed_adhesome = pybedtools.BedTool.from_dataframe(df_adhesome_pos).sort()\n",
    "bed_adhesome_df = bed_adhesome.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a bed object for MERFISH probes\n",
    "df_merfish_pos = merfish_data.loc[:,'probe_locus'].to_frame().drop_duplicates()\n",
    "df_merfish_pos['chrom'] = df_merfish_pos['probe_locus'].str.split(':', expand=True)[0]\n",
    "df_merfish_pos['extent'] = df_merfish_pos['probe_locus'].str.split(':', expand=True)[1]\n",
    "df_merfish_pos['start'] = df_merfish_pos['extent'].str.split('-', expand=True)[0].astype(int)\n",
    "df_merfish_pos['end'] = df_merfish_pos['extent'].str.split('-', expand=True)[1].astype(int)\n",
    "df_merfish_pos = df_merfish_pos[['chrom', 'start', 'end', 'probe_locus']]\n",
    "df_merfish_pos['start'] = (df_merfish_pos['start']-resol).clip(lower=0)\n",
    "df_merfish_pos['end'] = (df_merfish_pos['end']+resol)\n",
    "bed_merfish = pybedtools.BedTool.from_dataframe(df_merfish_pos).sort()\n",
    "bed_merfish_df = bed_merfish.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts for this feature in the segmented genome\n",
    "out = pybedtools.bedtool.BedTool.map(bed_merfish, bed_adhesome, c=[4], o='distinct', f=1e-9, F=1e-9)\n",
    "counts = out.to_dataframe()\n",
    "counts['score'] = counts['score'].str.split(',')\n",
    "counts['score'] = counts['score'].apply(lambda x: x if x[0]!='.' else [])\n",
    "merfishloc2gene = {counts.loc[i,'name']: counts.loc[i,'score'] for i in range(len(counts))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build dictionary mapping selected genes to the corresponding MERFISH locus\n",
    "counts = counts.loc[:,['name', 'score']]\n",
    "counts = counts.set_index('name').apply(pd.Series.explode).reset_index().dropna().reset_index().iloc[:,1:]\n",
    "gene2merfishloc = {counts.loc[i,'score']: counts[counts['score']==counts.loc[i,'score']]['name'].values\n",
    "                   for i in range(len(counts))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coverage of adhesome genes by MERFISH probes = 97\n",
      "Coverage of active adhesome genes by MERFISH probes = 105\n",
      "Coverage of random genes by MERFISH probes = 7425\n",
      "Number of MERFISH probes covering adhesome loci = 88\n",
      "Number of MERFISH probes covering active adhesome loci = 92\n"
     ]
    }
   ],
   "source": [
    "# Basic stats\n",
    "print(f'Coverage of adhesome genes by MERFISH probes = {len(np.unique(list(set(counts[\"score\"].values).intersection(set(all_adhesome_genes)))))}')\n",
    "print(f'Coverage of active adhesome genes by MERFISH probes = {len(np.unique(list(set(counts[\"score\"].values).intersection(set(active_adhesome_genes)))))}')\n",
    "print(f'Coverage of random genes by MERFISH probes = {len(np.unique(list(set(counts[\"score\"].values).intersection(set(random_active_genes)))))}')\n",
    "print(f'Number of MERFISH probes covering adhesome loci = {counts[counts[\"score\"].isin(all_adhesome_genes)][\"name\"].nunique()}')\n",
    "print(f'Number of MERFISH probes covering active adhesome loci = {counts[counts[\"score\"].isin(active_adhesome_genes)][\"name\"].nunique()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>chrom</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr_10_loc_128500000</td>\n",
       "      <td>10</td>\n",
       "      <td>514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr_10_loc_128750000</td>\n",
       "      <td>10</td>\n",
       "      <td>515.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr_10_loc_129000000</td>\n",
       "      <td>10</td>\n",
       "      <td>516.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr_10_loc_129250000</td>\n",
       "      <td>10</td>\n",
       "      <td>517.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr_10_loc_17250000</td>\n",
       "      <td>10</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chr_9_loc_133750000</td>\n",
       "      <td>9</td>\n",
       "      <td>535.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>chr_9_loc_136500000</td>\n",
       "      <td>9</td>\n",
       "      <td>546.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>chr_9_loc_136750000</td>\n",
       "      <td>9</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>chr_9_loc_93500000</td>\n",
       "      <td>9</td>\n",
       "      <td>374.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>chr_9_loc_94750000</td>\n",
       "      <td>9</td>\n",
       "      <td>379.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>158 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    locus  chrom  location\n",
       "0    chr_10_loc_128500000     10     514.0\n",
       "1    chr_10_loc_128750000     10     515.0\n",
       "2    chr_10_loc_129000000     10     516.0\n",
       "3    chr_10_loc_129250000     10     517.0\n",
       "4     chr_10_loc_17250000     10      69.0\n",
       "..                    ...    ...       ...\n",
       "153   chr_9_loc_133750000      9     535.0\n",
       "154   chr_9_loc_136500000      9     546.0\n",
       "155   chr_9_loc_136750000      9     547.0\n",
       "156    chr_9_loc_93500000      9     374.0\n",
       "157    chr_9_loc_94750000      9     379.0\n",
       "\n",
       "[158 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Where are uncovered adhesome genes\n",
    "uncovered_adh_genes = np.unique(list(set(all_adhesome_genes)-set(counts[\"score\"].values)))\n",
    "uncovered_adh_loc = np.unique(list(itertools.chain.from_iterable([gene2locus[g] for g in uncovered_adh_genes])))\n",
    "uncovered_adh_loc_df = pd.DataFrame({'locus': uncovered_adh_loc})\n",
    "uncovered_adh_loc_df['chrom'] = uncovered_adh_loc_df['locus'].str.split('_', expand=True)[1].astype(int)\n",
    "uncovered_adh_loc_df['location'] = uncovered_adh_loc_df['locus'].str.split('_', expand=True)[3].astype(int)/250000\n",
    "uncovered_adh_loc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify MERFISH loci of adhesome genes\n",
    "adh_merfish_loc = np.unique(list((itertools.chain.from_iterable([gene2merfishloc[g] \n",
    "                             for g in active_adhesome_genes if (g in gene2merfishloc.keys())]))))\n",
    "# Identiy MERFISH loci of non adhesome genes\n",
    "rand_merfish_loc = np.unique(list((itertools.chain.from_iterable([gene2merfishloc[g] \n",
    "                             for g in random_active_genes if (g in gene2merfishloc.keys())]))))\n",
    "rand_merfish_loc = np.array(list(set(rand_merfish_loc)-set(adh_merfish_loc)))\n",
    "\n",
    "# Combine adh_merfish_loc and rand_merfish_loc\n",
    "selected_merfish_loc = np.concatenate([adh_merfish_loc, rand_merfish_loc])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare MERFISH distances between adhesome loci and random loci all non-adhesome active loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select MERFISH data corresponding to active adhesome genes and active random genes\n",
    "selected_merfish_data = merfish_data.copy()\n",
    "selected_merfish_data = selected_merfish_data.loc[selected_merfish_data['probe_locus'].isin(selected_merfish_loc),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique loci = 969\n",
      "Number of unique cells = 3533\n"
     ]
    }
   ],
   "source": [
    "# A few statistics\n",
    "all_merfish_loci = np.unique(selected_merfish_data[\"probe_locus\"])\n",
    "all_merfish_cells = np.unique(selected_merfish_data[\"cell_ID\"])\n",
    "print(f'Number of unique loci = {len(all_merfish_loci)}')\n",
    "print(f'Number of unique cells = {len(all_merfish_cells)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3533/3533 [01:19<00:00, 44.63it/s]\n"
     ]
    }
   ],
   "source": [
    "# Build a dictionary where for each cell, we report a gene by gene distance matrix between adhesome genes\n",
    "cell2distmat = {}\n",
    "for cell in tqdm(all_merfish_cells):\n",
    "    cell_data = selected_merfish_data.loc[selected_merfish_data['cell_ID']==cell,\n",
    "                                          ['x(nm)','y(nm)','z(nm)','probe_locus']]\n",
    "    cell_data = cell_data.groupby('probe_locus').agg({'x(nm)': 'mean', 'y(nm)': 'mean', 'z(nm)': 'mean'})\n",
    "    dist_mat = squareform(pdist(cell_data, metric='euclidean'))/max_dist_dict[cell_id]\n",
    "    cell2distmat[cell] = pd.DataFrame(dist_mat,columns = cell_data.index.values, index = cell_data.index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3533/3533 [02:07<00:00, 27.74it/s]\n"
     ]
    }
   ],
   "source": [
    "# Combine information across all cells into one dataframe\n",
    "distances_df = pd.DataFrame(0, columns=all_merfish_loci, index=all_merfish_loci)\n",
    "n_samples_df = pd.DataFrame(0, columns=all_merfish_loci, index=all_merfish_loci)\n",
    "for cell in tqdm(all_merfish_cells):\n",
    "    cell_dist_mat = cell2distmat[cell]\n",
    "    cell_idx = cell_dist_mat.index.values\n",
    "    distances_df.loc[cell_idx, cell_idx] += cell_dist_mat\n",
    "    n_samples_df.loc[cell_idx, cell_idx] += 1 \n",
    "n_samples_df = n_samples_df.replace(0, np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximum average distance = 8326.957249187171\n"
     ]
    }
   ],
   "source": [
    "# Average distance dataframe\n",
    "avg_dist_df = distances_df/n_samples_df\n",
    "print(f'Maximum average distance = {avg_dist_df.max().max()}')\n",
    "avg_scaled_dist_df = avg_dist_df/avg_dist_df.max().max()\n",
    "avg_scaled_prox_df = 1-avg_scaled_dist_df\n",
    "final_adh_loc = np.sort([g for g in all_merfish_loci if (g in adh_merfish_loc)])\n",
    "final_rand_loc = np.sort([g for g in all_merfish_loci if (g in rand_merfish_loc)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Procedure to sort final_adh_loc and final_rand_loc in genomic order\n",
    "sort_adh = pd.DataFrame({'locus': final_adh_loc})\n",
    "sort_adh['chrom'] = sort_adh['locus'].str.split(':', expand=True)[0]\n",
    "sort_adh['chrom'] = sort_adh['chrom'].str.split('chr', expand=True)[1].astype(int)\n",
    "sort_adh['ext'] = sort_adh['locus'].str.split(':', expand=True)[1]\n",
    "sort_adh['start'] = sort_adh['ext'].str.split('-', expand=True)[0].astype(int)\n",
    "sort_adh['end'] = sort_adh['ext'].str.split('-', expand=True)[1].astype(int)\n",
    "sort_adh = sort_adh[['locus','chrom','start','end']]\n",
    "sort_adh = sort_adh.sort_values(['chrom','start','end'])\n",
    "final_adh_loc =  sort_adh['locus'].values\n",
    "sort_rand = pd.DataFrame({'locus': final_rand_loc})\n",
    "sort_rand['chrom'] = sort_rand['locus'].str.split(':', expand=True)[0]\n",
    "sort_rand['chrom'] = sort_rand['chrom'].str.split('chr', expand=True)[1].astype(int)\n",
    "sort_rand['ext'] = sort_rand['locus'].str.split(':', expand=True)[1]\n",
    "sort_rand['start'] = sort_rand['ext'].str.split('-', expand=True)[0].astype(int)\n",
    "sort_rand['end'] = sort_rand['ext'].str.split('-', expand=True)[1].astype(int)\n",
    "sort_rand = sort_rand[['locus','chrom','start','end']]\n",
    "sort_rand = sort_rand.sort_values(['chrom','start','end'])\n",
    "final_rand_loc =  sort_rand['locus'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXgAAAEGCAYAAABvtY4XAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAkuklEQVR4nO3deXwV9b3/8dcHDFLKIkhULG3BXleIQQiL7LJfFwRF0Z8bWOVqvdhqRe3FXxvt8lPx6q3oFWlVtFeRC7K472JQ9kCI4EZrUalUA1gKgsry+f0xk3gM55ADyZyTTN7PxyOPzJkzZ76fmUk+53u+M+cz5u6IiEj8NMh2ACIiEg0leBGRmFKCFxGJKSV4EZGYUoIXEYmpg7IdQKLWrVt7u3btsh2GiEidUVxcvNHdc5M9V6sSfLt27Vi+fHm2wxARqTPM7MNUz2mIRkQkppTgRURiSgleRCSmatUYfDI7d+5k/fr1fPnll9kOReSANW7cmLZt25KTk5PtUKQeqfUJfv369TRr1ox27dphZtkOR2S/uTubNm1i/fr1tG/fPtvhSD0SaYI3s3XAVmA3sMvdC/Z3HV9++aWSu9RpZsahhx5KWVlZtkOReiYTPfhT3H1jdVag5C51nf6GJRt0klVEJKai7sE78KKZOXC/u0+tvICZjQPGAfzgBz+ocoV3La7ZAK/pUbPrmz9/Po0aNaJnz54ATJkyhSZNmnDxxRfXbEMplH9ZbNu2bZx++umsXr06I+1Wx7Rp01i+fDn33HPPfr+2Z8+eLFy4MIKoROq+qBN8b3f/m5kdBrxkZu+6e1HiAmHSnwpQUFBQ5+8+Mn/+fJo2bVqR4K+44orsBrT9k+TzmxyZ2TgiouQuklqkQzTu/rfw92fAHKBblO1FZcSIEXTp0oUOHTowdeo3H0Kef/55OnfuTH5+PgMHDmTdunVMmTKFu+66i06dOrFgwQIKCwu54447ePfdd+nW7ZvNX7duHXl5eQAUFxfTr18/unTpwtChQ9mwYcNeMTz11FN0796dk046iUGDBvHpp58CsGnTJoYMGUKHDh247LLLSLxD1+7du7n8qgl0KDiFIWecz44dOwD4ywfrGDZsGF26dKFPnz68++67AMycOZOOHTuSn59P3759geAk99ixY8nLy+Okk07itddeA4Je94gRIxg8eDDt2rXjnnvu4c477+Skk06iR48ebN68OWjrL39J2lYq69atY8CAAZx44okMHDiQjz76CIBPP/2UkSNHkp+fT35+fkVib9q0abqHUaTeiSzBm9l3zaxZ+TQwBKj94wVJPPjggxQXF7N8+XLuvvtuNm3aRFlZGZdffjlPPPEEq1atYubMmbRr144rrriCa665hpKSEvr06VOxjuOOO46vv/6av/71rwDMmDGD0aNHs3PnTsaPH8+sWbMoLi7m0ksvZeLEiXvF0Lt3bxYvXszKlSs577zzuP322wG4+eab6d27N2vWrGHkyJEVCRFg7dq1XDXuEtYsf41DDmnOE3OfBWDcv1/P5MmTKS4u5o477uAnP/kJALfccgsvvPACq1at4sknnwTg3nvvxcx46623mD59OpdccknFdxJWr17N7NmzWbZsGRMnTqRJkyasXLmSk08+mUceeSRoa9y4pG2lMn78eC655BJKS0u54IILuPrqqwG4+uqr6devH6tWrWLFihV06NBh/w+kSD0T5RDN4cCc8OqBg4DH3P35CNuLzN13382cOXMA+Pjjj1m7di1lZWX07du34rrmVq1aVbmec889lxkzZnDjjTcyY8YMZsyYwXvvvcfq1asZPHgwEPS627Rps9dr169fz+jRo9mwYQNff/11RbtFRUXMnj0bgNNOO42WLVtWvKZ9+/Z0yu8IQJdOJ7Luo4/Ztu0LFi4p5pxzzqlY7quvvgKgV69ejBkzhnPPPZezzjoLgDfeeIPx48cDwZvUD3/4Q95//30ATjnlFJo1a0azZs1o0aIFZ5xxBgB5eXmUlpaybds2Fi5cmLStVBYtWlSxPRdddBHXX389AK+++mrFm0bDhg1p0aJFFXtbRCJL8O7+AZAf1fozZf78+bz88sssWrSIJk2a0L9//wP+Vu3o0aM555xzOOusszAzjj76aN566y06dOjAokWL9vna8ePHc+211zJ8+HDmz59PYWFhle0dfPDBFdMNGzZkx5dfsmfPHg5p0ZySkpK9lp8yZQpLlizhmWeeoUuXLhQXF6e9/gYNGlQ8btCgAbt27QraOuSQpG2JSPR0mWQVtmzZQsuWLWnSpAnvvvsuixcHl/H06NGDoqKiiiGX8jHnZs2asXXr1qTr+tGPfkTDhg359a9/zejRowE49thjKSsrq0jwO3fuZM2aNUnj+N73vgfAww8/XDG/b9++PPbYYwA899xzfP755/vcnubNm9G+3feZOXMmEHzLctWqVUAwXt69e3duueUWcnNz+fjjj+nTpw+PPvooAO+//z4fffQRxx57bFW7LWyrOe3bt0/aVio9e/bk8ccfB+DRRx+tGOYaOHAg9913HxB8ytmyZUtaMYjUZ7W+VEFlNX1ZY1WGDRvGlClTOP744zn22GPp0SMIIDc3l6lTp3LWWWexZ88eDjvsMF566SXOOOMMRo0axbx585g8efJe6xs9ejQTJkyoeGNo1KgRs2bN4uqrr2bLli3s2rWLn/3sZ3uNMRcWFnLOOefQsmVLBgwYUPH6X/3qV5x//vl06NCBnj17pnWp6aMP3sOV1xbym9/8hp07d3LeeeeRn5/PhAkTWLt2Le7OwIEDyc/P57jjjuPKK68kLy+Pgw46iGnTpn2r515lW48+ypVXXrlXW6lMnjyZsWPHMmnSJHJzc3nooYcA+P3vf8+4ceN44IEHaNiwIffddx8nn3xy2nGI1EeWeNVFthUUFHjlG3688847HH/88VmKKAZifplkXaK/ZYmCmRWnKgNT53rwkh2fbtt73uG6QlGkVtMYvIhITKkHHxephmL2d3kN3YjEhnrwIiIxpQQvIhJTSvAiIjFV98bgSwtrdn0nHvj6ykvztm7dmqZNm7JtW5JLTfZDtksNR6WwsJCmTZty3XXX0b9/f+644w4KCvb75l4Zd6DH9Je//CV9+/Zl0KBBEUQlkr66l+BjrNaVGq5CsksnQZdP3nLLLdkOQQTQEE1aUpULrs5ra7zUcK9hDB3+f9iw4dO9Yhgz7mdcfd3/peeA4RzV4WRmzXkaCEoHTPiPX9OxYAB5XQcyY9Y8IHij6d+/P6NGjeK4447jggsuINkX4r7Yto1Rpw9kcO/O9O+ex/NPz6t47re//S3HHHMMvXv35r333vvW62bOnEm3bt045phjWLBgARCUH5gwYQJdu3blxBNP5P777wdgw4YN9O3bl06dOtGxY8eK5adPn05eXh4dO3bkhhtuqFh306ZNmTBhAh06dGDQoEEsXbqU/v37c9RRR1VUyEzVViruzoQJE+jYsSN5eXnMmDGj4rnbbruNvLw88vPzufHGG4P9PWYMs2bN2uc6RTJBPfg0PPjgg7Rq1YodO3bQtWtXzj77bA499NADfu2ePXu4/PLLKSoqon379mzevJlWrVpxxRVXVAxlALzyyivAt0sNt2/ffq9Sw/PmzSP3uzuZMWseE2++jQen3LlXHBv+/ilvvDyXd9/7M8PPHcuokacze96zlJSuYdWSl9i4cTNd+55K38EjAFi5ciVr1qzhyCOPpFevXixd9Cbde/b+1joPbtyYhx6bQ7Pmzdm0cSOnDezB0NOGU1y8gscff5ySkhJ27dpF586d6dKlS8Xrdu3axdKlS3n22We5+eabefnll3nggQdo0aIFy5Yt46uvvqJXr14MGTKE2bNnM3ToUCZOnMju3bvZvn07n3zyCTfccAPFxcW0bNmSIUOGMHfuXEaMGMEXX3zBgAEDmDRpEiNHjuSmm27ipZde4u233+aSSy5h+PDhKdsqr9BZ2ezZsykpKWHVqlVs3LiRrl270rdvX0pKSpg3bx5LliyhSZMmFfWIRGoLJfg0JCsXnG6Cz1ip4T072b17D22OOCzp60ecPowGDRpwwvHH8OlnZQC8sXAp558zgoYNG3L44bn0692DZcuW0bx5c7p160bbtm0B6NSpEx9/tG6vBO/u/O7m/2Dxm0U0aNCAv3/yN8o++5QFCxYwcuRImjRpAsDw4cO/9bryUsRdunRh3bp1ALz44ouUlpZW9Hy3bNnC2rVr6dq1K5deeik7d+5kxIgRdOrUiVdffZX+/fuTm5sLwAUXXEBRUREjRoygUaNGDBs2DAjKFh988MHk5OSQl5dXZVupEvwbb7zB+eefH+6nw+nXrx/Lli3j9ddfZ+zYsRXbmc5xFMkkJfgqVKdccEZLDVfxRaeDD25UMZ1O/aHKpYZ37drFimVLmPDTfwPg+om38Pnnm9m0sYwXFxSTk5NDQYd2aW1f+brL11se0+TJkxk6dOheyxcVFfHMM88wZswYrr322n3Wgs/JySG8B0HSEsZVtSUSJxqDr0KqcsHVeW2kpYbffi/p65Pp06s7M554kt27d1NWtomiN5d8a6y/ss5du/PKwhJeWVjC0NOGs/WfW2idexg5OTm8UfQa6z/6EAhKGM+dO5cdO3awdetWnnrqqSpjGTp0KPfddx87d+4EgtLEX3zxBR9++CGHH344l19+OZdddhkrVqygW7duvP7662zcuJHdu3czffp0+vXrl/Z2p2or5X7q04cZM2aE+6mMoqIiunXrxuDBg3nooYfYvn07gIZopNapez34alzWeCBSlQuuzmsjKTX8+UZ27d7Nz666jA4npFevfeTwf2XRkmLyuw/GzLj9NxM54ogjqrxvarmzzr2Ai889g/7d88g/qYCjjzkOgM6dOzN69Gjy8/M57LDD6Nq1a5Xruuyyy1i3bh2dO3fG3cnNzWXu3LnMnz+fSZMmkZOTQ9OmTXnkkUdo06YNt956K6eccgruzmmnncaZZ56ZVsz7aiuVkSNHsmjRIvLz84P9dPvtHHHEEQwbNoySkhIKCgpo1KgRp556Kr/73e/SjkMkaioXHBf7W4smlRS1aFJdEplMfb9MMhX9LUsU9lUuWEM0IiIxVfeGaKTW0xegRGqHOtGDr03DSCIHQn/Dkg21vgffuHFjNm3axKGHHlpx+ZtEKOVYvurEHyh3Z9OmTTRu3DjboUg9U+sTfNu2bVm/fj1lZWXZDqV2+/ofka7+n76l2uvYnP69umOncePGFV8cE8mUWp/gc3JyUn7DUBKUFka6+rt2VH/913Sq9ipEZD/UiTF4ERHZf0rwIiIxVeuHaKSS0sJsRyAidYR68CIiMaUELyISU0rwIiIxpQQvIhJTSvAiIjEVeYI3s4ZmttLMno66LRER+UYmLpP8KfAO0DwDbUk1LVqf4okauN3oXenfDAuAa9K/t4qIJBFpD97M2gKnAX+Msh0REdlb1EM0/wVcD+xJtYCZjTOz5Wa2XAXFRERqTmQJ3sxOBz5z9+J9LefuU929wN0LcnNzowpHRKTeibIH3wsYbmbrgMeBAWb2PxG2JyIiCSJL8O7+C3dv6+7tgPOAV939wqjaExGRb9N18CIiMZWRapLuPh+Yn4m2REQkoB68iEhMKcGLiMSUEryISEwpwYuIxJRu2Sdp6bG5cK95i1vtPU9Eag/14EVEYkoJXkQkppTgRURiSmPw9VTKuu8iEhvqwYuIxJQSvIhITCnBi4jElBK8iEhMKcGLiMSUEryISEwpwYuIxJQSvIhITCnBi4jElBK8iEhMKcGLiMSUEryISEyp2FhtVVqY7QhEpI5TD15EJKaU4EVEYkoJXkQkppTgRURiSgleRCSmlOBFRGJKCV5EJKZ0HXzMRXlz7R6bC5POX9wq+XwRySz14EVEYkoJXkQkppTgRURiKrIEb2aNzWypma0yszVmdnNUbYmIyN6iPMn6FTDA3beZWQ7whpk95+6LI2xTRERCkSV4d3dgW/gwJ/zxqNoTEZFvS2uIxszOMLP9Hs4xs4ZmVgJ8Brzk7kuSLDPOzJab2fKysrL9bUJERFJItwc/GvgvM3sCeNDd303nRe6+G+hkZocAc8yso7uvrrTMVGAqQEFBgXr4UuGuFIN51/TIbBwidVVavXJ3vxA4CfgLMM3MFoU972Zpvv4fwGvAsAMNVERE9k/awy7u/k9gFvA40AYYCawws/HJljez3LDnjpl9BxgMpNXzFxGR6ktriMbMzgTGAP8CPAJ0c/fPzKwJ8DYwOcnL2gAPm1lDgjeS/3X3p2skahERqVK6Y/BnAXe5e1HiTHffbmY/TvYCdy8lGNYREZEsSHeI5u+Vk7uZ3Qbg7q/UeFQiIlJt6Sb4wUnm/WtNBiIiIjVrn0M0ZnYl8BPgR2ZWmvBUM+DNKAMTEZHqqWoM/jHgOeD/ATcmzN/q7psji0pERKqtqgTv7r7OzK6q/ISZtVKSFxGpvdLpwZ8OFBPUkbGE5xw4KqK4RESkmvaZ4N399PB3+8yEIyIiNSXdYmO9zOy74fSFZnanmf0g2tBERKQ60r1M8j5gu5nlAz8nqEnzp8iiEhGRaks3we8K67ufCdzj7vcSXCopIiK1VLqlCraa2S+AC4G+YW34nOjCEhGR6kq3Bz+a4BZ8P3b3vwNtgUmRRSUiItWWVg8+TOp3Jjz+iKCqpIiI1FLpXkVzlpmtNbMtZvZPM9tqZv+MOjgRETlw6Y7B3w6c4e7vRBmMxEOPzYVJ5y9ulXy+iEQj3TH4T5XcRUTqlnR78MvNbAYwl+BkKwDuPjuKoEREpPrSTfDNge3AkIR5DijB14TSwmxHICIxlO5VNGOjDkRERGpWulfRHGNmr5jZ6vDxiWZ2U7ShiYhIdaR7kvUPwC+AnVBxQ+3zogpKRESqL90E38Tdl1aat6umgxERkZqTboLfaGY/IjixipmNAjZEFpWIiFRbulfRXAVMBY4zs78BfwUuiCwqERGptn0meDO7NuHhs8BrBL3+L4CzSahPIyIitUtVPfjymu/HAl2BeQT3Zb0IqDwmLyIitUhV92S9GcDMioDO7r41fFwIPBN5dCIicsDSPcl6OPB1wuOvw3kiIlJLpXuS9RFgqZnNCR+PAKZFEZCIiNSMdEsV/NbMngP6hLPGuvvK6MKS/bVofbYjEJHaJt0ePO6+AlgRYSwiIlKD0h2DFxGROiayBG9m3zez18zsbTNbY2Y/jaotERHZW9pDNAdgF/Bzd19hZs2AYjN7yd3fjrBNEREJRZbg3X0DYb0ad99qZu8A3wOU4KVa7lqcfP41PTIbh0htl5ExeDNrB5wELMlEeyIikoEEb2ZNgSeAn7n7P5M8P87MlpvZ8rKysqjDERGpNyJN8GaWQ5DcH011g253n+ruBe5ekJubG2U4IiL1SpRX0RjwAPCOu6vqpIhIhkXZg+9FUHVygJmVhD+nRtieiIgkiPIqmjcISguLANBjc2HS+YtbJZ8vItWjb7KKiMSUEryISEwpwYuIxJQSvIhITCnBi4jElBK8iEhMKcGLiMSUEryISEwpwYuIxJQSvIhITEV5RyeRjEp2IxDdBETqMyX4TCotzHYEIlKPaIhGRCSmlOBFRGJKCV5EJKaU4EVEYkoJXkQkppTgRURiSgleRCSmlOBFRGJKCV5EJKb0TVbJuh6bC5POX9wq+XwRSY968CIiMaUELyISU0rwIiIxpQQvIhJTSvAiIjGlq2jqmEXrsx2BiNQV6sGLiMSUEryISEwpwYuIxJQSvIhITCnBi4jEVGRX0ZjZg8DpwGfu3jGqdkT25a7Fyedf0yOzcYhkQ5Q9+GnAsAjXLyIi+xBZD97di8ysXVTrjztd7y4i1ZX1MXgzG2dmy81seVlZWbbDERGJjawneHef6u4F7l6Qm5ub7XBERGJDpQqk1tKNQESqJ+s9eBERiUaUl0lOB/oDrc1sPfArd38gqvZqldLCbEcgIhLpVTTnR7VuERGpmoZoRERiSgleRCSmlOBFRGJKCV5EJKaU4EVEYkoJXkQkppTgRURiSqUKpM5RCQOR9KgHLyISU+rBS72kOz1JfaAevIhITCnBi4jElBK8iEhMaQy+FtD9V0UkCurBi4jElBK8iEhMaYimOkoLsx2BJEj2BSh9+UnqM/XgRURiSgleRCSmNESTQbpaRkQySQleRA6YSj7UbkrwIlKlVIlcajcleBGpcerZ1w46ySoiElPqwUdAJ1Nrj/29OYh6nhInSvAidU1pYfL5Jyafrzet+ksJXiTbSgsjXk/y+Ur88acEn47SwmxHIHFRWpjxJmtqmErqHiV4qZciv3F3aQ2tJ0K6eXn8KcFXg06mSjZE/Xenom3xoQQvUh2lhdmOQCQlJXiRBKmGLSiNrs268ElQwzl1kxJ8GurCP6BINijx126RJngzGwb8HmgI/NHdb42yvWorLcx2BFJL7e+b/Mlto4mjrtvfK3R0yWb1RJbgzawhcC8wGFgPLDOzJ9397ajaFKnt6sunQfXsa4coe/DdgD+7+wcAZvY4cCaQ0QSfrMdwTZPCpMvWl38+iZ7+lpJLeY4jhbsW7728evXpizLBfw/4OOHxeqB75YXMbBwwLny4zczeO8D2WgMb01nw2gNsoJZIeztjoL5sa33ZTtjvbb15rzl15P83k8f0h6meyPpJVnefCkyt7nrMbLm7F9RASLVafdlOqD/bWl+2E+rPttaW7YyyXPDfgO8nPG4bzhMRkQyIMsEvA442s/Zm1gg4D3gywvZERCRBZEM07r7LzP4deIHgMskH3X1NVO1RA8M8dUR92U6oP9taX7YT6s+21ortNHfPdgwiIhIB3bJPRCSmlOBFRGKqziV4MxtmZu+Z2Z/N7MYkzx9sZjPC55eYWbsshFltaWzntWb2tpmVmtkrZpbyWtjarqptTVjubDNzM8v65WcHIp3tNLNzw+O6xswey3SMNSWNv98fmNlrZrYy/Bs+NRtxVpeZPWhmn5nZ6hTPm5ndHe6HUjPrnNEA3b3O/BCcrP0LcBTQCFgFnFBpmZ8AU8Lp84AZ2Y47ou08BWgSTl9ZF7cz3W0Nl2sGFAGLgYJsxx3RMT0aWAm0DB8flu24I9zWqcCV4fQJwLpsx32A29oX6AysTvH8qcBzgAE9gCWZjK+u9eAryh+4+9dAefmDRGcCD4fTs4CBZmYZjLEmVLmd7v6au28PHy4m+J5BXZTOMQX4NXAb8GUmg6tB6Wzn5cC97v45gLt/luEYa0o62+pA83C6BfBJBuOrMe5eBGzexyJnAo94YDFwiJm1yUx0dW+IJln5g++lWsbddwFbgEMzEl3NSWc7E/2YoJdQF1W5reHH2u+7+zOZDKyGpXNMjwGOMbM3zWxxWI21LkpnWwuBC81sPfAsMD4zoWXc/v4v16islyqQ6jGzC4ECoF+2Y4mCmTUA7gTGZDmUTDiIYJimP8EnsiIzy3P3f2QzqIicD0xz9/80s5OBP5lZR3ffk+3A4qSu9eDTKX9QsYyZHUTw8W9TRqKrOWmVeTCzQcBEYLi7f5Wh2GpaVdvaDOgIzDezdQTjmE/WwROt6RzT9cCT7r7T3f8KvE+Q8OuadLb1x8D/Arj7IqAxQYGuuMlqyZa6luDTKX/wJHBJOD0KeNXDsx11SJXbaWYnAfcTJPe6OlYLVWyru29x99bu3s7d2xGcbxju7suzE+4BS+dvdy5B7x0za00wZPNBBmOsKels60fAQAAzO54gwZdlNMrMeBK4OLyapgewxd03ZKrxOjVE4ynKH5jZLcByd38SeIDg496fCU5+nJe9iA9Mmts5CWgKzAzPIX/k7sOzFvQBSnNb67w0t/MFYIiZvQ3sBia4e1379Jnutv4c+IOZXUNwwnVMHeyIYWbTCd6UW4fnE34F5AC4+xSC8wunAn8GtgNjMxpfHdynIiKShro2RCMiImlSghcRiSkleBGRmFKCFxGJKSV4EZGYUoKXyJhZu1RV9vbxmmlmNqoG2i4ws7sP8LVHmtmsGohhflhRsST8OaB1Ju4TM/ujmZ2QZJkxZnZPdWPOBjNbuJ/LX2FmF4fTY8zsyGgiq/vq1HXwsjcza+juu7MdR20TfhFqry9DmdlBYY2ifb32E4IvydWEC2ryS1nufllNrSsK6ezfyty9534uPyXh4RhgNXW0WFnU1IPPIDOba2bFYa3vceG8K8xsUsIyFT0xM7vQzJaGvb/7zaxhOH+bmf2nma0CTjazX5rZMjNbbWZTy6tnmlnXsAZ1iZlNKu9Nm1nD8PGy8Pl/SxLrd83sGTNbFa53dMI6F4bzl5pZs7CnvsDMVoQ/e/3Dpmoz/IbfPWFP92XgsBT7br6Z/T7cltVm1i2c383MFllQV3yhmR0bzu9vZk+H04Vm9icze5PgS3DPmNmJ4XMrzeyX4fQtZnZ54icPM+uQcAxKzezofR2bNP8OvvUpxcy2JUzfYGZvhfv31hT7oSCcHmtm75vZUqBXwjK5ZvZEuK+XmVmvKvbVGDObbWbPm9laM7s9RdzrzOz2ML6lZvYvCdszxcyWALebWScLiqWVmtkcM2tpZj8M193azBqEfy9DErc/PGavm9k8M/vAzG41swvCtt4ysx8lHM/rwn1YADwaHofTzGxuQryDzWxOusclljJZm7i+/wCtwt/fIeh1HArkEpRWLV/mOaA3cDzwFJATzv9v4OJw2oFzK683nP4TcEY4vRo4OZy+lbBmNTAOuCmcPpigp9u+UqxnA39IeNyCoLb3B0DXcF5zgk+BTYDG4byjCb6tCNCuqjaBs4CXCL7xeCTwD2BUkn03vzweghrcqxNjCKcHAU+E0/2Bp8PpQqAY+E74+EbgqnCblgEvhPNfA46tFPdkgl444fZ/Z1/HJknM7wEl4c+kcP60xG0EtoW//xVYyDd1/ltVXj5cZwHQhuDr/rlhXG8C94TLPAb0Dqd/ALxTxb4aEx7XFgQlAz4kqN5ZeXvWARPD6YsT9u804GmgYfi4FOgXTt8C/Fc4fRkwE5gA3J9k+/sTHP82BH8jfwNuDp/7acJ6CoHrEvdHOG3Au0Buwn44I9v/99n80RBNZl1tZiPD6e8DR7v74rC30gNYCxxH8M96FdAFWGZBh/w7QHnNmd3AEwnrPcXMridItK2ANWa2AGjmQSEnCP7YTw+nhwAnJvQiWxAk5r8mrPMt4D/N7DaCf+QFZpYHbHD3ZQDu/k8IevvAPWbWKYztmCTbnqrNvsB0D4aZPjGzV/ex/6aH7RaZWXMzO4SgGNnDYc/aCb8mnsST7r4jnF4AXB1u7zPAYDNrQvAm9559+y5gi4CJZtYWmO3ua81sIKmPTWX7M0QzCHjIwzr/7r6vOuPdgfnuXgZgZjP4Zr8PAk6wb26D0NzMmhLs81T76hV33xKu623gh3y7zG256Qm/70qYP9Pdd5tZC+AQd389nP8wQVLH3f9oZucAVwCdUmzXMg9rtZjZX4AXw/lvEdzkJiV3dzP7E0EZ4oeAkwneiOotJfgMMbP+BP94J7v7djObT9BbguCGCOcS9D7mhH+oBjzs7r9Isrovw4SImTUm6EEWuPvHZlaYsN6U4QDj3f2FVAu4+/sW1GE/FfiNmb0CpPq4ew3wKZBPMOyX7KYcSdu0/btVW+W6Gk5wI5DX3H1kmJjnp3jtFwnTywh6wR8QfHpoTXCzjeK9GnR/LBx6OA141oKhpX0dm3TsIhwetaAccqMDXE8qDYAe7v6t42DB0F+qfZVYjXQ3qXODp5j+ovKClYVvouU3pmkKbE2yWGIcexIe79lHTIkeIvh09SXBm85+nQ+IG43BZ04L4PMwuR9HUPa23ByCO7+cT5DsAV4BRpnZYQBm1sqS33e1PJlvDHtpowA8qCG+1cy6h88nFl17AbjSzHLCdR8T9sIrWHBlwnZ3/x+CwmadCYYb2phZ13CZZvZNSeYNHtTyvohguKWyVG0WAaMtGKNvw757aeXnAXoTVOXbErZdXn51zD5eW8GDuwx9DJxD0ENfAFwXxvItZnYU8IG73w3MA04k/WOTyjqCTwAAw/mmJ/0SMDZMhJhZq32sYwnQz8wODffpOQnPvUjCDTTCT1ZwAPsqidEJvxdVfjI8Jp+bWZ9w1kVAeW/+NuBR4JfAHw6w/cq2EnyKK2//E4ITrjcRJPt6TT34zHkeuMLM3iFIlIvLn3D3z8P5J7j70nDe22Z2E/Bi2MvbSTBs82HiSt39H2b2B4Lx9r8T9E7L/ZigYt8egn+yLeH8PxKMM68IPymUASMqxZsHTApfu5Pg/plfW3CydbKZfQfYQfCp5L+BJyy4dO15kvfmUrU5BxgAvE0wprxX0kjwpZmtJEiIl4bzbicYdriJYLglXQuAge6+IxzOahvOq+xc4CIz20mwf3/n7pvTOTahR82sfGhoo7sPIkhu8yw4SV6xv9z9+TAZLzezrwkqEf5HsuDdfUP4aW0Rwbh1ScLTVwP3mlkpwf94EcGwyIHuq0Qtw/V+RdAhSeYSYEr4RvUBwZtWP6Ar0CscyjnbzMa6e3WT8LSwrR0En453ELyJ5Lr7O9Vcd52napIxZmZN3b38CoUbgTbu/tMsh3VAwiGt6/ZjPFtqmAU3XClw943ZjmVfwqGole7+QLZjyTb14OPtNDP7BcFx/pD6cds7qcfMrJjgE9HPsx1LbaAevIhITOkkq4hITCnBi4jElBK8iEhMKcGLiMSUEryISEz9f/2S7fnZ5bBwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wilcoxon Rank-Sums test (blue vs orange):\n",
      "RanksumsResult(statistic=43.367122968322356, pvalue=0.0)\n"
     ]
    }
   ],
   "source": [
    "# Plot histogram of MERFISH distances\n",
    "plt.figure()\n",
    "plt.hist(avg_scaled_prox_df.loc[final_adh_loc,final_adh_loc].values[np.triu_indices(len(final_adh_loc),1)], \n",
    "         color='dodgerblue', alpha=0.5, density=True, bins=np.arange(0, 1.05, 0.02),\n",
    "         label='active adhesome loci')\n",
    "plt.hist(avg_scaled_prox_df.loc[final_rand_loc,final_rand_loc].values[np.triu_indices(len(final_rand_loc),1)], \n",
    "         color='orange', alpha=0.5, density=True, bins=np.arange(0, 1.05, 0.02),\n",
    "         label='all active non-adhesome loci')\n",
    "plt.xlabel('average scaled pairwise Euclidean proximity')\n",
    "plt.ylabel('density')\n",
    "plt.legend(loc='upper left')\n",
    "plt.show()\n",
    "# Wilcoxon Rank-Sums test\n",
    "print('Wilcoxon Rank-Sums test (blue vs orange):')\n",
    "print(ss.ranksums(avg_scaled_prox_df.loc[final_adh_loc,final_adh_loc].values[np.triu_indices(len(final_adh_loc),1)],\n",
    "                  avg_scaled_prox_df.loc[final_rand_loc,final_rand_loc].values[np.triu_indices(len(final_rand_loc),1)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
