{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Goal: Save any interesting quantity to pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'correlation_clustering' from '/home/louiscam/projects/gpcr/code/figures4/correlation_clustering.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import standard libraries\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, fcluster\n",
    "from sklearn.metrics import adjusted_mutual_info_score\n",
    "import seaborn as sns\n",
    "from scipy import sparse\n",
    "import scipy.stats as ss\n",
    "import csv\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import community\n",
    "import communities as com\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "import operator\n",
    "from scipy.sparse import csr_matrix\n",
    "import itertools\n",
    "import os.path\n",
    "import math\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.metrics import adjusted_mutual_info_score\n",
    "from networkx.algorithms.community.kclique import k_clique_communities\n",
    "import pybedtools\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import OmicsIntegrator as oi\n",
    "import gseapy\n",
    "from gseapy.plot import barplot, dotplot\n",
    "from ortools.linear_solver import pywraplp\n",
    "from matplotlib_venn import venn2, venn3\n",
    "# Custom libraries\n",
    "import utils as lu\n",
    "import correlation_clustering as cc\n",
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(cc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HiC information\n",
    "cell_type = 'IMR90'\n",
    "resol_str = '250kb'\n",
    "resol = 250000\n",
    "quality = 'MAPQGE30'\n",
    "norm = 'GWKR'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory of genome data\n",
    "dir_genome = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "# Directory of adhesome data\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "# Directory of processed HiC\n",
    "dir_processed_hic = f'/home/louiscam/projects/gpcr/save/processed_hic_data/processed_hic_data_IMR90/final_BP250000_intraKR_inter{norm}/'\n",
    "# Directory of epigenomic data\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/regulatory_data/regulatory_data_IMR90/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_regulatory_marks/processed_epigenome_data_IMR90/'\n",
    "# Directory of TF data\n",
    "dir_htftarget = '/home/louiscam/projects/gpcr/data/tf_data/hTFtarget/'\n",
    "# Directory of TCR/BCR data\n",
    "dir_tcr = '/home/louiscam/projects/gpcr/data/tcr_data/'\n",
    "dir_bcr = '/home/louiscam/projects/gpcr/data/bcr_data/'\n",
    "# Directory of proteins of interest\n",
    "dir_proteins = '/home/louiscam/projects/gpcr/data/protein_families/'\n",
    "# Saving directory\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load  activity data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active/inactive loci\n",
    "with open(saving_dir+'all_active_loci.pkl', 'rb') as f:\n",
    "    all_active_loci = pickle.load(f)\n",
    "with open(saving_dir+'all_inactive_loci.pkl', 'rb') as f:\n",
    "    all_inactive_loci = pickle.load(f)\n",
    "with open(saving_dir+'active_adhesome_loci.pkl', 'rb') as f:\n",
    "    active_adhesome_loci = pickle.load(f)\n",
    "with open(saving_dir+'inactive_adhesome_loci.pkl', 'rb') as f:\n",
    "    inactive_adhesome_loci = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active/inactive genes\n",
    "with open(saving_dir+'active_genes.pkl', 'rb') as f:\n",
    "    active_genes = pickle.load(f)\n",
    "with open(saving_dir+'inactive_genes.pkl', 'rb') as f:\n",
    "    inactive_genes = pickle.load(f)\n",
    "with open(saving_dir+'active_adhesome_genes.pkl', 'rb') as f:\n",
    "    active_adhesome_genes = pickle.load(f)\n",
    "with open(saving_dir+'inactive_adhesome_genes.pkl', 'rb') as f:\n",
    "    inactive_adhesome_genes = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load location data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene location in hg19\n",
    "gene_locations_filename = dir_genome+'chrom_hg19.loc_canonical'\n",
    "gene_id_filename = dir_genome+'chrom_hg19.name'\n",
    "df_loc = lu.get_all_gene_locations(gene_locations_filename, gene_id_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load adhesome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adhesome data\n",
    "adhesome_components_filename = dir_adhesome+'components.csv'\n",
    "df_components = lu.load_adhesome_data(adhesome_components_filename)\n",
    "adhesome_chr_loci, missing_adhesome = lu.annotate_genes(df_components[['geneSymbol']], df_loc, resol)\n",
    "adhesome_chr_loci['active'] = adhesome_chr_loci['gene'].isin(active_adhesome_genes)\n",
    "all_adhesome_genes = df_components['geneSymbol'].unique()\n",
    "with open(saving_dir+'adhesome_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(adhesome_chr_loci, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load TF data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All TF data\n",
    "tftarget_df = pd.read_csv(dir_htftarget+'tftarget_full_processed.csv', \n",
    "                          sep='\\t', index_col=0, header=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5896536"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tftarget_df.drop_duplicates(['tf', 'target_gene', 'cell_line', 'cell_type', 'tissue']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "333"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tftarget_df['tf'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55714"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tftarget_df['target_gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tf</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_id</th>\n",
       "      <th>target_synonym</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>fibroblast</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>condition</th>\n",
       "      <th>project</th>\n",
       "      <th>samples</th>\n",
       "      <th>control</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EP300</td>\n",
       "      <td>NADK</td>\n",
       "      <td>ENSG00000008130</td>\n",
       "      <td>['dJ283E3.1']</td>\n",
       "      <td>913</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>lymphoblastoid cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>epstein-barr virus transformed</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE31477']</td>\n",
       "      <td>['GSM935294']</td>\n",
       "      <td>['GSM935390']</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EP300</td>\n",
       "      <td>SLC2A5</td>\n",
       "      <td>ENSG00000142583</td>\n",
       "      <td>['GLUT-5', 'GLUT5']</td>\n",
       "      <td>913</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>lymphoblastoid cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>epstein-barr virus transformed</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE31477']</td>\n",
       "      <td>['GSM935294']</td>\n",
       "      <td>['GSM935390']</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EP300</td>\n",
       "      <td>SCARNA16</td>\n",
       "      <td>ENSG00000252404</td>\n",
       "      <td>[]</td>\n",
       "      <td>913</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>lymphoblastoid cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>epstein-barr virus transformed</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE31477']</td>\n",
       "      <td>['GSM935294']</td>\n",
       "      <td>['GSM935390']</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EP300</td>\n",
       "      <td>SZRD1</td>\n",
       "      <td>ENSG00000055070</td>\n",
       "      <td>['C1orf144']</td>\n",
       "      <td>913</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>lymphoblastoid cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>epstein-barr virus transformed</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE31477']</td>\n",
       "      <td>['GSM935294']</td>\n",
       "      <td>['GSM935390']</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EP300</td>\n",
       "      <td>RP11-430L17.1</td>\n",
       "      <td>ENSG00000226457</td>\n",
       "      <td>[]</td>\n",
       "      <td>913</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>lymphoblastoid cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>epstein-barr virus transformed</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE31477']</td>\n",
       "      <td>['GSM935294']</td>\n",
       "      <td>['GSM935390']</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022598</th>\n",
       "      <td>YY1</td>\n",
       "      <td>RP11-298I3.3</td>\n",
       "      <td>ENSG00000280129</td>\n",
       "      <td>[]</td>\n",
       "      <td>5278</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>acute t cell leukemia cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>acute t cell leukemia</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE99518']</td>\n",
       "      <td>['GSM2773998']</td>\n",
       "      <td>['GSM2773999']</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022599</th>\n",
       "      <td>YY1</td>\n",
       "      <td>RP3-523C21.3</td>\n",
       "      <td>ENSG00000280155</td>\n",
       "      <td>[]</td>\n",
       "      <td>5278</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>acute t cell leukemia cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>acute t cell leukemia</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE99518']</td>\n",
       "      <td>['GSM2773998']</td>\n",
       "      <td>['GSM2773999']</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022600</th>\n",
       "      <td>YY1</td>\n",
       "      <td>RP5-1180D12.1</td>\n",
       "      <td>ENSG00000280202</td>\n",
       "      <td>[]</td>\n",
       "      <td>5278</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>acute t cell leukemia cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>acute t cell leukemia</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE99518']</td>\n",
       "      <td>['GSM2773998']</td>\n",
       "      <td>['GSM2773999']</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022601</th>\n",
       "      <td>YY1</td>\n",
       "      <td>RP11-972P1.7</td>\n",
       "      <td>ENSG00000280328</td>\n",
       "      <td>[]</td>\n",
       "      <td>5278</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>acute t cell leukemia cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>acute t cell leukemia</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE99518']</td>\n",
       "      <td>['GSM2773998']</td>\n",
       "      <td>['GSM2773999']</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022602</th>\n",
       "      <td>YY1</td>\n",
       "      <td>AC005355.3</td>\n",
       "      <td>ENSG00000280420</td>\n",
       "      <td>[]</td>\n",
       "      <td>5278</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>acute t cell leukemia cell</td>\n",
       "      <td>False</td>\n",
       "      <td>blood</td>\n",
       "      <td>acute t cell leukemia</td>\n",
       "      <td>other</td>\n",
       "      <td>['GSE99518']</td>\n",
       "      <td>['GSM2773998']</td>\n",
       "      <td>['GSM2773999']</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9022603 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            tf    target_gene        target_id       target_synonym  \\\n",
       "0        EP300           NADK  ENSG00000008130        ['dJ283E3.1']   \n",
       "1        EP300         SLC2A5  ENSG00000142583  ['GLUT-5', 'GLUT5']   \n",
       "2        EP300       SCARNA16  ENSG00000252404                   []   \n",
       "3        EP300          SZRD1  ENSG00000055070         ['C1orf144']   \n",
       "4        EP300  RP11-430L17.1  ENSG00000226457                   []   \n",
       "...        ...            ...              ...                  ...   \n",
       "9022598    YY1   RP11-298I3.3  ENSG00000280129                   []   \n",
       "9022599    YY1   RP3-523C21.3  ENSG00000280155                   []   \n",
       "9022600    YY1  RP5-1180D12.1  ENSG00000280202                   []   \n",
       "9022601    YY1   RP11-972P1.7  ENSG00000280328                   []   \n",
       "9022602    YY1     AC005355.3  ENSG00000280420                   []   \n",
       "\n",
       "         dataset_id cell_line                   cell_type fibroblast tissue  \\\n",
       "0               913   GM12878         lymphoblastoid cell      False  blood   \n",
       "1               913   GM12878         lymphoblastoid cell      False  blood   \n",
       "2               913   GM12878         lymphoblastoid cell      False  blood   \n",
       "3               913   GM12878         lymphoblastoid cell      False  blood   \n",
       "4               913   GM12878         lymphoblastoid cell      False  blood   \n",
       "...             ...       ...                         ...        ...    ...   \n",
       "9022598        5278    Jurkat  acute t cell leukemia cell      False  blood   \n",
       "9022599        5278    Jurkat  acute t cell leukemia cell      False  blood   \n",
       "9022600        5278    Jurkat  acute t cell leukemia cell      False  blood   \n",
       "9022601        5278    Jurkat  acute t cell leukemia cell      False  blood   \n",
       "9022602        5278    Jurkat  acute t cell leukemia cell      False  blood   \n",
       "\n",
       "                                disease condition       project  \\\n",
       "0        epstein-barr virus transformed     other  ['GSE31477']   \n",
       "1        epstein-barr virus transformed     other  ['GSE31477']   \n",
       "2        epstein-barr virus transformed     other  ['GSE31477']   \n",
       "3        epstein-barr virus transformed     other  ['GSE31477']   \n",
       "4        epstein-barr virus transformed     other  ['GSE31477']   \n",
       "...                                 ...       ...           ...   \n",
       "9022598           acute t cell leukemia     other  ['GSE99518']   \n",
       "9022599           acute t cell leukemia     other  ['GSE99518']   \n",
       "9022600           acute t cell leukemia     other  ['GSE99518']   \n",
       "9022601           acute t cell leukemia     other  ['GSE99518']   \n",
       "9022602           acute t cell leukemia     other  ['GSE99518']   \n",
       "\n",
       "                samples         control  source  \n",
       "0         ['GSM935294']   ['GSM935390']  ENCODE  \n",
       "1         ['GSM935294']   ['GSM935390']  ENCODE  \n",
       "2         ['GSM935294']   ['GSM935390']  ENCODE  \n",
       "3         ['GSM935294']   ['GSM935390']  ENCODE  \n",
       "4         ['GSM935294']   ['GSM935390']  ENCODE  \n",
       "...                 ...             ...     ...  \n",
       "9022598  ['GSM2773998']  ['GSM2773999']    NCBI  \n",
       "9022599  ['GSM2773998']  ['GSM2773999']    NCBI  \n",
       "9022600  ['GSM2773998']  ['GSM2773999']    NCBI  \n",
       "9022601  ['GSM2773998']  ['GSM2773999']    NCBI  \n",
       "9022602  ['GSM2773998']  ['GSM2773999']    NCBI  \n",
       "\n",
       "[9022603 rows x 15 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tftarget_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only TF-target pairs in lung fibroblasts\n",
    "all_lung_cell_lines = tftarget_df[tftarget_df['tissue']=='lung']['cell_line'].unique()\n",
    "lung_fibroblasts = ['IMR90', 'WI38', 'AG04450', 'HPF', 'NHLF', ]\n",
    "tftarget_df_lungfibro = tftarget_df[tftarget_df['cell_line'].isin(lung_fibroblasts)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Include loci information\n",
    "htargettf_df = tftarget_df_lungfibro[['tf','target_gene']]\n",
    "htargettf_df.columns = ['geneSymbol', 'target']\n",
    "lungTF_chr_loci, missing_TF = lu.annotate_genes(htargettf_df[['geneSymbol']].drop_duplicates(), df_loc, resol)\n",
    "lungTF_chr_loci['active'] = lungTF_chr_loci['gene'].isin(active_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load TCR data\n",
    "- https://www.wikipathways.org/index.php/Pathway:WP69\n",
    "- http://www.netpath.org/netslim/tcr_pathway.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of TCR intrinsic and associated components = 92\n"
     ]
    }
   ],
   "source": [
    "# TCR data\n",
    "tcr_df = pd.read_csv(dir_tcr+'TCR.csv', header=0, index_col=None)\n",
    "print('Total number of TCR intrinsic and associated components = '+str(len(tcr_df)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load BCR data\n",
    "- https://www.wikipathways.org/index.php/Pathway:WP23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of BCR intrinsic and associated components = 103\n"
     ]
    }
   ],
   "source": [
    "# BCR data\n",
    "bcr_df = pd.read_csv(dir_bcr+'BCR.csv', header=0, index_col=None, sep='\\t')\n",
    "print('Total number of BCR intrinsic and associated components = '+str(len(bcr_df)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load genes related to mechanoregulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proteins of interest\n",
    "with open(dir_proteins+'mechanoreg_proteins.pkl', 'rb') as f:\n",
    "    proteins_df = pickle.load(f)\n",
    "proteins_chr_loci, missing_prot = lu.annotate_genes(proteins_df, df_loc, resol)\n",
    "proteins_chr_loci['active'] = proteins_chr_loci['gene'].isin(active_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mechanoreg protein to family\n",
    "prot2fam = {proteins_df.iloc[i,0]: proteins_df.iloc[i,1] for i in range(len(proteins_df))}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define useful quantities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene to chromosome dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25959/25959 [00:41<00:00, 621.35it/s]\n"
     ]
    }
   ],
   "source": [
    "# Build dictionary matching each gene to its corresponding chromosome\n",
    "gene2chrom = {g: df_loc[df_loc['geneSymbol']==g]['#chrom'].values[0] \n",
    "              for g in tqdm(df_loc['geneSymbol'].values)}\n",
    "with open(saving_dir+'gene2chrom.pkl', 'wb') as f:\n",
    "    pickle.dump(gene2chrom, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map from genes to locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25959/25959 [00:10<00:00, 2418.07it/s]\n"
     ]
    }
   ],
   "source": [
    "# Build dictionary matching each gene to its corresponding locus\n",
    "gene2locus = {df_loc.iloc[i]['geneSymbol']: \n",
    "              ['chr_'+df_loc.iloc[i]['#chrom'].strip('chr')+'_loc_'+str(resol*x)\n",
    "               for x in range(df_loc.iloc[i]['chromStart']//resol, df_loc.iloc[i]['chromEnd']//resol+1, 1)]\n",
    "              for i in tqdm(range(df_loc.shape[0]))}\n",
    "with open(saving_dir+'gene2locus.pkl', 'wb') as f:\n",
    "    pickle.dump(gene2locus, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25959/25959 [00:00<00:00, 1112313.69it/s]\n"
     ]
    }
   ],
   "source": [
    "# Build dictionary matching each locus to its corresponding genes\n",
    "locus2gene = {}\n",
    "for k, v in tqdm(gene2locus.items()):\n",
    "    for vals in v:\n",
    "        locus2gene[vals] = locus2gene.get(vals, []) + [k]\n",
    "with open(saving_dir+'locus2gene.pkl', 'wb') as f:\n",
    "    pickle.dump(locus2gene, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adhesome genes/loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All adhesome genes\n",
    "all_adhesome_genes = np.unique(adhesome_chr_loci['gene'])\n",
    "with open(saving_dir+'all_adhesome_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(all_adhesome_genes, f)\n",
    "\n",
    "all_adhesome_loci = np.unique(adhesome_chr_loci['locus_id'])\n",
    "with open(saving_dir+'all_adhesome_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(all_adhesome_loci, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adhesome gene2locus and locus2gene dictionaries\n",
    "adhesome_gene2locus = {gene: adhesome_chr_loci[adhesome_chr_loci['gene']==gene]['locus_id'].values.astype(str)  \n",
    "              for gene in np.unique(adhesome_chr_loci['gene'])}\n",
    "with open(saving_dir+'adhesome_gene2locus.pkl', 'wb') as f:\n",
    "    pickle.dump(adhesome_gene2locus, f)\n",
    "\n",
    "adhesome_locus2gene = {locus: adhesome_chr_loci[adhesome_chr_loci['locus_id']==locus]['gene'].values.astype(str) \n",
    "              for locus in np.unique(adhesome_chr_loci['locus_id'])}\n",
    "with open(saving_dir+'adhesome_locus2gene.pkl', 'wb') as f:\n",
    "    pickle.dump(adhesome_locus2gene, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active adhesome gene2locus and locus2gene dictionaries\n",
    "active_adhesome_gene2locus = {gene: adhesome_chr_loci[(adhesome_chr_loci['gene']==gene) &\n",
    "                                                      (adhesome_chr_loci['active']==True)]['locus_id'].values.astype(str)  \n",
    "              for gene in np.unique(adhesome_chr_loci['gene'])}\n",
    "with open(saving_dir+'active_adhesome_gene2locus.pkl', 'wb') as f:\n",
    "    pickle.dump(active_adhesome_gene2locus, f)\n",
    "\n",
    "active_adhesome_locus2gene = {locus: adhesome_chr_loci[(adhesome_chr_loci['locus_id']==locus) &\n",
    "                                                      (adhesome_chr_loci['active']==True)]['gene'].values.astype(str) \n",
    "              for locus in np.unique(adhesome_chr_loci['locus_id'])}\n",
    "with open(saving_dir+'active_adhesome_locus2gene.pkl', 'wb') as f:\n",
    "    pickle.dump(active_adhesome_locus2gene, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All active lung TF genes/loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_int</th>\n",
       "      <th>loci</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>active</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LMNB1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>5</td>\n",
       "      <td>504</td>\n",
       "      <td>chr_5_loc_126000000</td>\n",
       "      <td>True</td>\n",
       "      <td>AGBL5-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNB1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>5</td>\n",
       "      <td>504</td>\n",
       "      <td>chr_5_loc_126000000</td>\n",
       "      <td>True</td>\n",
       "      <td>AGBL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNB1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>5</td>\n",
       "      <td>504</td>\n",
       "      <td>chr_5_loc_126000000</td>\n",
       "      <td>True</td>\n",
       "      <td>RP11-204L24.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNB1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>5</td>\n",
       "      <td>504</td>\n",
       "      <td>chr_5_loc_126000000</td>\n",
       "      <td>True</td>\n",
       "      <td>CLASP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LMNB1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>5</td>\n",
       "      <td>504</td>\n",
       "      <td>chr_5_loc_126000000</td>\n",
       "      <td>True</td>\n",
       "      <td>RNU4ATAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  geneSymbol chrom  chrom_int  loci             locus_id  active  \\\n",
       "0      LMNB1  chr5          5   504  chr_5_loc_126000000    True   \n",
       "1      LMNB1  chr5          5   504  chr_5_loc_126000000    True   \n",
       "2      LMNB1  chr5          5   504  chr_5_loc_126000000    True   \n",
       "3      LMNB1  chr5          5   504  chr_5_loc_126000000    True   \n",
       "4      LMNB1  chr5          5   504  chr_5_loc_126000000    True   \n",
       "\n",
       "          target  \n",
       "0      AGBL5-AS1  \n",
       "1          AGBL5  \n",
       "2  RP11-204L24.2  \n",
       "3         CLASP1  \n",
       "4       RNU4ATAC  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge TF and target information\n",
    "lungTF_chr_loci1 = lungTF_chr_loci.copy()\n",
    "lungTF_chr_loci1.columns = ['chrom', 'chrom_int', 'loci', 'tf', 'locus_id', 'active']\n",
    "tf_target_df = tftarget_df_lungfibro.merge(lungTF_chr_loci1, on='tf')\n",
    "tf_target_df = tf_target_df[['tf', 'chrom', 'chrom_int', 'loci', 'locus_id', 'active', 'target_gene']]\n",
    "tf_target_df.columns = ['geneSymbol', 'chrom', 'chrom_int', 'loci', 'locus_id', 'active', 'target']\n",
    "tf_target_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 28/28 [00:00<00:00, 87.46it/s]\n"
     ]
    }
   ],
   "source": [
    "# Select active TFs targeting active genes\n",
    "active_tf_target_df = tf_target_df[(tf_target_df['geneSymbol'].isin(active_genes)) & \n",
    "                                   (tf_target_df['target'].isin(active_genes))]\n",
    "\n",
    "active_lung_tf2target = {tf: set(active_tf_target_df[active_tf_target_df['geneSymbol']==tf]['target'].unique())\n",
    "                        for tf in tqdm(active_tf_target_df['geneSymbol'].unique())}\n",
    "with open(saving_dir+'active_lung_tf2target.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_tf2target, f)\n",
    "\n",
    "active_lung_target2tf = {target: set(active_tf_target_df[active_tf_target_df['target']==target]['geneSymbol'].unique())\n",
    "                        for target in active_tf_target_df['target'].unique()}\n",
    "with open(saving_dir+'active_lung_target2tf.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_target2tf, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store active TFs targeting active genes\n",
    "active_lung_tf_genes = tf_target_df[(tf_target_df['geneSymbol'].isin(active_genes)) & \n",
    "                                    (tf_target_df['target'].isin(active_genes))]['geneSymbol'].unique()\n",
    "with open(saving_dir+'active_lung_tf_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_tf_genes, f)\n",
    "\n",
    "active_lung_tf_loci = tf_target_df[(tf_target_df['geneSymbol'].isin(active_genes)) & \n",
    "                                   (tf_target_df['target'].isin(active_genes))]['locus_id'].unique()\n",
    "with open(saving_dir+'active_lung_tf_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_tf_loci, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TFs in normal lung fibroblasts cell lines = 31\n",
      "['LMNB1' 'RBL2' 'MAFK' 'BRF2' 'CTCF' 'RB1' 'E2F7' 'CHD1' 'EP300' 'SUMO2'\n",
      " 'UBE2I' 'PIAS4' 'SNAPC2' 'SUMO1' 'GTF2B' 'BRCA1' 'RFX5' 'BDP1' 'RCOR1'\n",
      " 'SNAPC1' 'TP53' 'CEBPB' 'RAD21' 'POLR2A' 'RBL1' 'SNAPC5' 'SNAPC4' 'BRF1'\n",
      " 'MXI1' 'MAZ' 'SPI1']\n",
      "TFs in the different cell lines\n",
      "  cell_line                                                 tf  n_tfs  \\\n",
      "0   AG04450                                             [CTCF]      1   \n",
      "1       HPF                                             [CTCF]      1   \n",
      "2     IMR90  [LMNB1, RBL2, MAFK, BRF2, RB1, E2F7, CHD1, EP3...     27   \n",
      "3      NHLF                                             [CTCF]      1   \n",
      "4      WI38                 [SUMO2, UBE2I, PIAS4, SUMO1, CTCF]      5   \n",
      "\n",
      "   n_active_tfs  \n",
      "0             1  \n",
      "1             1  \n",
      "2            24  \n",
      "3             1  \n",
      "4             5  \n",
      "TFs in WI38 that are not in IMR90\n",
      "{'PIAS4', 'SUMO1', 'SUMO2', 'UBE2I'}\n"
     ]
    }
   ],
   "source": [
    "# Breakdown of TFs in normal lung fibroblasts\n",
    "print('Number of TFs in normal lung fibroblasts cell lines = '+str(tftarget_df_lungfibro['tf'].nunique()))\n",
    "print(tftarget_df_lungfibro['tf'].unique())\n",
    "print('TFs in the different cell lines')\n",
    "lungfibro_breadkdown_df = tftarget_df_lungfibro.drop_duplicates(['tf', 'cell_line'])\n",
    "lungfibro_breadkdown_df = lungfibro_breadkdown_df.groupby(['cell_line'])['tf'].agg(list).reset_index()\n",
    "lungfibro_breadkdown_df['n_tfs'] = [len(ls) \n",
    "                                    for ls in lungfibro_breadkdown_df['tf']]\n",
    "lungfibro_breadkdown_df['n_active_tfs'] = [len(set(ls).intersection(set(active_genes))) \n",
    "                                           for ls in lungfibro_breadkdown_df['tf']]\n",
    "print(lungfibro_breadkdown_df)\n",
    "print('TFs in WI38 that are not in IMR90')\n",
    "print(set(lungfibro_breadkdown_df[lungfibro_breadkdown_df['cell_line']=='WI38']['tf'].values[0])-\n",
    "      set(lungfibro_breadkdown_df[lungfibro_breadkdown_df['cell_line']=='IMR90']['tf'].values[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAE9CAYAAAD6c07jAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAA6IElEQVR4nO3deXxV1b3//9eHQQZFRVCLIgIKKINECA5UbCqIqCBUEfCKEmsvFkeuw61aW2fFqq3D/RbkV1EUBRxKRcQpKFULooAoolJwJBQHBpEYkMHP74+9Eg+QcM7Z5CQn5P18PPbj7LX2/qy1zqnlkz2tbe6OiIjIjtSq6gGIiEj2U7IQEZGklCxERCQpJQsREUlKyUJERJJSshARkaTqVPUAMqFp06besmXLqh6GiEi1Mm/evJXuvm9Z23bJZNGyZUvmzp1b1cMQEalWzOzz8rbpNJSIiCSlZCEiIkkpWYiISFK75DULEdm1bdq0icLCQjZs2FDVQ6mW6tevT/Pmzalbt27KMUoWIlLtFBYW0qhRI1q2bImZVfVwqhV3Z9WqVRQWFtKqVauU43QaSkSqnQ0bNtCkSRMlihjMjCZNmqR9VKZkISLVkhJFfHF+u4wmCzP7HzNbZGbvm9lEM6tvZq3MbI6ZLTWzyWa2W9i3XigvDdtbJrRzTahfbGYnZXLMIiI7o2XLlqxcuRKAPfbYo4pHU3Eyds3CzA4ELgXau/t6M3sCGAKcAvzF3SeZ2RjgfGB0+Fzj7oea2RDgDmCwmbUPcR2AA4ACM2vr7lsyNXYRqV5aXv1chbb32ahTK7S9XUGmT0PVARqYWR2gIbACOAF4KmwfDwwI6/1DmbC9p0XHSv2BSe7+g7t/CiwFjsrwuEVEdmjAgAF07dqVDh06MHbs2LRib775Ztq1a8dxxx3HWWedxV133QXAxx9/TJ8+fejatSs9evTgo48+AiA/P59LL72U7t2707p1a5566qnStu688066devGEUccwfXXXw/A999/z6mnnkrnzp3p2LEjkydP3unvm7EjC3dfbmZ3AV8A64GXgHnAt+6+OexWCBwY1g8EloXYzWa2FmgS6t9MaDoxRkSkSowbN4599tmH9evX061bN8444wyaNGmSNO7tt9/m6aef5t1332XTpk106dKFrl27AjB8+HDGjBlDmzZtmDNnDhdeeCGvvPIKACtWrOCNN97go48+4rTTTmPgwIG89NJLLFmyhLfeegt357TTTuO1117jm2++4YADDuC556IjrrVr1+70983kaajGREcFrYBvgSeBPhnsbzgwHKBFixaZ6kak2hoc/hidPBDIy4sKM2em3U7ew1HszPw0YgtCf73S7y9b3XfffUyZMgWAZcuWsWTJkpSSxb/+9S/69+9P/fr1qV+/Pv369QOgqKiIWbNmceaZZ5bu+8MPP5SuDxgwgFq1atG+fXu++uorAF566SVeeukljjzyyNI2lixZQo8ePbjiiiv43e9+R9++fenRo8dOf99MPmfRC/jU3b8BMLO/Az8H9jazOuHoojmwPOy/HDgIKAynrfYCViXUl0iMKeXuY4GxALm5uZ6RbyQiAsycOZOCggJmz55Nw4YNycvL2+kHBH/88Uf23ntvFixYUOb2evXqla67e+nnNddcwwUXXLDd/vPnz2f69Olcd9119OzZkz/+8Y87Nb5MXrP4AjjGzBqGaw89gQ+AV4GBYZ9hwDNhfWooE7a/4tEvMhUYEu6WagW0Ad7K4LhFRHZo7dq1NG7cmIYNG/LRRx/x5ptvJg8Kfv7zn/Pss8+yYcMGioqKmDZtGgB77rknrVq14sknnwSiRPDuu+/usK2TTjqJcePGUVRUBMDy5cv5+uuv+c9//kPDhg0ZOnQoV111FfPnz4/5TX+SyWsWc8zsKWA+sBl4h+gv/+eASWZ2S6h7MIQ8CDxqZkuB1UR3QOHui8KdVB+Edi7SnVAiUpX69OnDmDFjOPzww2nXrh3HHHNMyrHdunXjtNNO44gjjmD//fenU6dO7LXXXgA89thjjBgxgltuuYVNmzYxZMgQOnfuXG5bvXv35sMPP+TYY48Folt1J0yYwNKlS7nqqquoVasWdevWZfTo0Tv3hQErOZzZleTm5rreZyGytQfmRZ8XdAXC3TdceWXa7dw1K4q9snsasR+G/g5Pv78ym/vwQw4//PAKaasqFBUVsccee1BcXMzxxx/P2LFj6dKlS6WOoazf0MzmuXtuWftrbiiRGuKCrgmFGEmiNDSdJFGigpLErmL48OF88MEHbNiwgWHDhlV6oohDySJNyR7+0cM8IpLM448/XtVDSJvmhhKpIQY/9dPts+Tl/XT7bJryHs4rvX02ZQV5P90+K9WSkoWIiCSlZCEiIkkpWYiISFJKFiIiMdx666106NCBI444gpycHObMmbPV9OQQPendt29fAB5++GHMjIKCgtLt//jHPzCz0okBN27cyMiRIzn00ENp06YN/fv3p7CwEIimFPnlL39J+/bt6dChA/fee28lflvdDSVSY5zaJqEwaFDsdgZ1iBHbIn5/2Wj27NlMmzaN+fPnU69ePVauXMnGjRuTxnXq1IlJkybRq1cvACZOnLjVQ3fXXnst69atY/HixdSuXZuHHnqI008/nTlz5lCnTh3uvvtuunTpwrp16+jatSsnnngi7du3z9j3TKRkIVJDnJv4IPCFF8Zu58JuMWLbxu8vG61YsYKmTZuWztfUtGnTlOJ69OjB66+/zqZNm/jhhx9YunQpOTk5ABQXF/PQQw/x6aefUrt2bQDOO+88xo0bxyuvvELPnj1p1qwZAI0aNeLwww9n+fLllZYsdBpKpIZYvylaACgujpYYijcVU7wpzdjNxdGyi+jduzfLli2jbdu2XHjhhfzzn/9MKc7M6NWrFy+++CLPPPMMp512Wum2pUuX0qJFC/bcc8+tYnJzc1m0aNFWdZ999hnvvPMORx999M5/mRTpyEKkhsgPU3ZOHgicckpUiDFF+SmPRbFpTVE+M/SXoSnKy3ruY1CHQVzY7UKKNxWXjjlRfk4++Tn5rCxeycAnBm61Ldl322OPPZg3bx6vv/46r776KoMHD2bUqFFlvtt627ohQ4Zw3333sXbtWu6++25uu+225F8wQVFREWeccQb33HPPdoklk5QsRERiqF27Nnl5eeTl5dGpUyfGjx9PkyZNWLNmTelpqdWrV293iuqoo45i4cKFNGzYkLZt25bWH3LIIXzxxResW7eORo0aldbPmzev9CL5pk2bOOOMMzj77LM5/fTTK+Fb/kTJQkSqvR0dCTSs23CH25s2bJreURKwePFiatWqRZs20V0DCxYs4OCDD6Zdu3Y8+uij3HTTTWzZsoUJEyYwYMCA7eJHjRpF/fr1t6rbfffdGTZsGJdffjljxoyhdu3aPPLIIxQXF3PCCSfg7px//vkcfvjhXH755WmNtyIoWYiIpKmoqIhLLrmEb7/9ljp16nDooYcyduxY6taty4gRI+jcuTPuTp8+fRg6dOh28SeffHKZ7d5+++1ceeWVtG3bllq1anHYYYcxZcoUzIw33niDRx99lE6dOpVeFL/ttts45ZTtT7FlgpKFiEiaunbtyqxZs8rcVt4kgfn5+eTn529X//DDD5eu16tXj/vvv5/7779/u/2OO+44qvKVEkoWIjXEwMQ7LMv4RytV+TkxYlvH70+yg5KFSA1xppKF7AQ9ZyFSQ6xeHy0ArFwZLTGsLF7JyuI0YzesjBaptnRkIVJDjAjv7Zo8EBgYniuI8ZxFyTMJad1B9EboL0PPWUjm6chCRESSUrIQEZGklCxERGL47LPP6Nix41Z1N9xwA3fddRf5+fm0atWKnJwcDjvsMG688cbSffLy8mjXrh05OTnk5OQwMJwS/POf/0z79u054ogj6NmzJ59//nm5/VQFXbMQEcmAO++8k4EDB7Jhwwbat2/PueeeS6tWrQB47LHHyM3N3Wr/I488krlz59KwYUNGjx7N//7v/zJ58uSqGHqZlCxEaoihRyQURoyI3c6I3BixbeL3V91t2LABiKbz2JFf/vKXpevHHHMMEyZMyOi40pWxZGFm7YDEtNga+CPwSKhvCXwGDHL3NRZNzXgvcApQDOS7+/zQ1jDgutDOLe4+PlPjFtlV9WubUBg8OHY7gzvGiD04fn/V1VVXXcUtt9zC0qVLufTSS9lvv/1Kt5199tk0aNAAgBNPPJE777xzq9gHH3yw3ClBqkrGkoW7LwZyAMysNrAcmAJcDcxw91FmdnUo/w44GWgTlqOB0cDRZrYPcD2QCzgwz8ymuvuaTI1dZFf0n3XR5wGNgGXLosJBB6XdzrK1UexBe6UR+33ob/f0+0tJQd72dS0GRS9d2lz80xTpiVrnR8uGlT/d2lsihVt8y5qOPLG+5DRUUVERPXv2ZNasWXTv3h0o+zRUiQkTJjB37tyU35FRWSrrAndP4GN3/xzoD5QcGYwHBoT1/sAjHnkT2NvMmgEnAS+7++qQIF4G+lTSuEV2Gf/zYrQAcM450RLDOVPO4ZwpacbOPidadiEl05EnKmtK8j322IO8vDzeeOONpG0WFBRw6623MnXq1NK38GWLyrpmMQSYGNb3d/cVYf1LYP+wfiCwLCGmMNSVV78VMxsODAdo0aJFhQ1cRKqBHR0J1Gm44+31m8Z6WHCPPfagWbNmvPLKK5xwwgmsXr2aF154gcsuu4xXX321dL/NmzczZ84cLrnkkh22984773DBBRfwwgsvbHXKKltkPFmY2W7AacA1225zdzezCplG0d3HAmMBcnNzq25qRhGpMR555BEuuuii0vdLXH/99RxyyCHAT9csNm7cSM+ePbd6WVHiNYumTZtSUFDAVVddRVFREWeeeSYQ/dE7derUSv5G5auMI4uTgfnu/lUof2Vmzdx9RTjN9HWoXw4kntBsHuqWA3nb1M/M6IhFRFLQvn37rY4iSiROO76tmeVMsVJQUFBmfcuWLXn//ffjDK9CVcY1i7P46RQUwFRgWFgfBjyTUH+uRY4B1obTVS8Cvc2ssZk1BnqHOhERqSQZPbIws92BE4ELEqpHAU+Y2fnA58CgUD+d6LbZpUS3zp4H4O6rzexm4O2w303uvjqT4xbZFf13l4TCFVfEbueKY2PEHha/P8kOGU0W7v490GSbulVEd0dtu68DF5XTzjhgXCbGKFJT9GqdUOjXL3Y7/drFiG0evz/JDpobSqSG+HhNtACweHG0xLB45WIWr0wz9rvF0SLVlqb7EKkhrp0RfU4eCFwQzgzHeJ/FBdOi2LTeZ/FW6E/vs6i2dGQhIiJJKVmIiMRw66230qFDB4444ghycnKYM2cOLVu2ZGXC62pnzpxJ3759geh2WjPb6hbZf/zjH5gZTz31FAAbN25k5MiRHHroobRp04b+/ftTWFhYuv+vf/1r9ttvvyqZslzJQkQkTbNnz2batGnMnz+f9957j4KCAg5KYZ6tTp06MWnSpNLyxIkT6dy5c2n52muvZd26dSxevJglS5YwYMAATj/9dKL7fyA/P58XXnih4r9QCpQsRETStGLFCpo2bVo6f1PTpk054IADksb16NGDt956i02bNlFUVMTSpUvJyckBoLi4mIceeoi//OUv1K5dG4DzzjuPevXq8corrwBw/PHHs88++2TmSyWhC9wiNcQlRyUUrruu3P2Sue74GLEd4/eXjXr37s1NN91E27Zt6dWrF4MHD+YXv/hF0jgzo1evXrz44ousXbuW0047jU8//RSApUuX0qJFC/bcc8+tYnJzc1m0aBE9e273xEGlUrIQqSGOS5xfs1ev2O30ah0j9mfx+0tJXt72dYMGwYUXQnExnFLGFOX5+dGyciUM3GaK8iR3ie2xxx7MmzeP119/nVdffZXBgwczatSoMqct37ZuyJAh3Hfffaxdu5a7776b2267bYd9ZQslC5EaYtE30WeHfYEFC6JCOAWSjgVfRrE5P0sjdk3or3H6/WWr2rVrk5eXR15eHp06dWL8+PGl05aXTFNe1pTlRx11FAsXLqRhw4a0bfvTG6kOOeQQvvjiC9atW0ejRo1K6+fNm1d6kbwqKVmI1BA3hXfpTB4IjBwZFWI8ZzHyhSg2recs5oX+MvWcxY6+R8OGO97etGnav8PixYupVasWbdq0AWDBggUcfPDBtGvXjkcffZSbbrqJLVu2MGHCBAYMGLBd/KhRo6hfv/5WdbvvvjvDhg3j8ssvZ8yYMdSuXZtHHnmE4uJiTjjhhLTGlwlKFiIiaSoqKuKSSy7h22+/pU6dOhx66KGMHTuWunXrMmLECDp37oy706dPH4YOHbpdfHmvTL399tu58soradu2LbVq1eKwww5jypQppaeyzjrrLGbOnMnKlStp3rw5N954I+eff35Gv2sJJQsRkTR17dqVWbNmlbnt8ccfL7M+Pz+f/Pz87eoTpzOvV68e999/P/fff3+ZbUycOLHM+sqgW2dFRCQpJQsREUlKp6FEaoiruicUduJ2zdt6xojtXD1uD5XyKVmI1BC5iQ8Yd+9e7n7JdD8oRuy+8fuT7KDTUCI1xNz/RAsAs2ZFSwyzls1i1rI0Y7+ZFS1SbenIQqSGuDP8Wz15IHDttVEhxnMW186IYtN6zuLd0J/eZ1Ft6chCRCSG2rVrk5OTQ8eOHTnzzDMpLi4GoqlAEt1zzz3Ur1+ftWvXltYVFxdz9tln06lTJzp27Mhxxx1HUVFR6fYtW7Zw5JFHZsWT2yWULEREYmjQoAELFizg/fffZ7fddmPMmDFl7jdx4kS6devG3//+99K6e++9l/3335+FCxfy/vvv8+CDD1K3bt2tth9++OEZ/w7pULIQEdlJPXr0YOnSpdvVf/zxxxQVFXHLLbds9UDdihUrOPDAA0vL7dq1K53uvLCwkOeee47f/OY3mR94GpQsRER2wubNm3n++efp1KnTdtsmTZrEkCFD6NGjB4sXL+arr74Cojfe3XHHHRx77LFcd911LFmypDRm5MiR/OlPf6JWrez651kXuEVqiD8mvm7hnntit3NPnxixXeP3l4rBT21fd2obOLczrN8E+c9sv31gezizPaxeDyOe23rb5IHb77+t9evXl764qEePHmXO0TRx4kSmTJlCrVq1OOOMM3jyySe5+OKLycnJ4ZNPPuGll16ioKCAbt26MXv2bD7++GP2228/unbtyswYNx9kUkaThZntDfwN6Ag48GtgMTAZaAl8Bgxy9zUWzZR1L3AKUAzku/v80M4woOTtKbe4+/hMjltkV9Rh34RCjKnJS0PTmZq8xC40NXmJkmsW5Vm4cCFLlizhxBNPBKL3a7dq1YqLL74YiC6En3766Zx++unUqlWL6dOns3LlSqZOncr06dPZsGED3333HUOHDmXChAmV8ZV2KNNHFvcCL7j7QDPbDWgIXAvMcPdRZnY1cDXwO+BkoE1YjgZGA0eb2T7A9UAuUcKZZ2ZT3X1Nhscuskt544vo87gWQEFBVIjxEqSCT6LYtF6C9GXoL0MvQdrRkUCDujvevk+D1I4k0jVx4kRuuOEGrrnmmtK6Vq1a8fnnn1NYWEj79u1p3LgxGzdu5IMPPiAvL48rrriC22+/HYCZM2dy1113ZUWigAwmCzPbCzgeyAdw943ARjPrD+SF3cYDM4mSRX/gEY/eTP6mme1tZs3Cvi+7++rQ7stAH6Dqpl8UqYbufyv6PK4FcMstUSFGsrjltSg2rWTxfugv02/MyyKTJk1i+vTpW9X96le/YtKkSTRr1owRI0bg7vz444+ceuqpnHHGGVU00tRk8siiFfAN8JCZdQbmAZcB+7v7irDPl8D+Yf1AYFlCfGGoK69eRKTKJD4XUVb9J598st22P//5z6Xr55577g7bL3kLX7bI5OX2OkAXYLS7Hwl8T3TKqVQ4ivCK6MzMhpvZXDOb+80331REkyIiEmQyWRQChe4+J5SfIkoeX4XTS4TPr8P25cBBCfHNQ1159Vtx97Hunuvuufvuu++2m0VEZCdkLFm4+5fAMjNrF6p6Ah8AU4FhoW4YUHJT21TgXIscA6wNp6teBHqbWWMzawz0DnUiIlJJMn031CXAY+FOqE+A84gS1BNmdj7wOTAo7Dud6LbZpUS3zp4H4O6rzexm4O2w300lF7tFJHW39UwoPPBA7HYe6Bsj9qj4/ZXH3UvfTS3pia4ApCejycLdFxDd8rqtnttWhOsXF5XTzjhgXIUOTqSGOaRxQqFdu3L3S6Zd0xixe8bvryz169dn1apVNGnSRAkjTe7OqlWrqF+/flpxeoJbpIYoCDfn9GoNPPtsVOjXL+12nl0cxfZrl0ZsYeivefr9laV58+YUFhaim1niqV+/Ps2bN08rRslCpIb4/+ZHn71aA3ffHRViJIu7Z0exaSWLj0J/FZQs6tatS6tWrSqkLUlNds1UJSIiWUnJQkREklKyEBGRpJQsREQkKV3gFqkh/nJSQuHRR2O38+ivYsQeG78/yQ5KFiI1xAGNEgoHHVTufskctFeM2N3j9yfZQaehRGqIZ/8dLQBMnhwtMUx+fzKT308z9vPJ0SLVlo4sRGqICe9Fn/3aAqNHR4XBg9NuZ/TcKHZwxzRil4T+Dk6/P8kOOrIQEZGklCxERCSppMnCzA4xs3phPc/MLjWzvTM+MhERyRqpHFk8DWwxs0OBsUQvIno8o6MSEZGsksoF7h/dfbOZ/Qq4393vN7N3Mj0wEalYo09NKDz1VOx2nhoUI/a4+P1JdkglWWwys7OI3mpXMmVk3cwNSUQyYZ8GCYWmTWO307RhjNj68fuT7JDKaajzgGOBW939UzNrBehxTJFq5skPogWAhx+OlhgeXvAwDy9IM/aTh6NFqq2kRxbu/oGZ/Q5oEcqfAndkemAiUrGeConizPb8lCjy89NupyRR5OekEVuSKFqn359kh1TuhuoHLABeCOUcM5ua4XGJiEgWSeU01A3AUcC3UPpe7dYZG5GIiGSdVJLFJndfu03dj5kYjIiIZKdU7oZaZGb/BdQ2szbApcCszA5LRESySSrJ4hLg98APwETgReDmTA5KRCrew/0TCtOnx25n+tkxYvPi9yfZIelpKHcvdvffu3s3d88N6xtSadzMPjOzhWa2wMzmhrp9zOxlM1sSPhuHejOz+8xsqZm9Z2ZdEtoZFvZfYmbD4n5ZkZqsQd1oAaBhw2iJoWHdhjSsm2ZsnYbRItVW0iMLM2sLXAm0TNzf3U9IsY9fuvvKhPLVwAx3H2VmV4fy74CTgTZhORoYDRxtZvsA1wO5gAPzzGyqu69JsX8RAR55N/o8tzPw179GhQsvTLudv74dxV7YLY3Yf4f+2qbfn2SHVE5DPQmMAf4GbKmAPvsDeWF9PDCTKFn0Bx5xdwfeNLO9zaxZ2Pdld18NYGYvA32ITomJSIqeWxJ9ntsZeOKJqBAjWTyxKIpNK1l8EfpTsqi2UkkWm919dMz2HXjJzBx4wN3HAvu7+4qw/Utg/7B+ILAsIbYw1JVXLyIilSSVZPGsmV0ITCG6yA1AyV/6SRzn7svNbD/gZTP7KHGju3tIJDvNzIYDwwFatGhREU2KiEiQynMWw4CriG6XnReWuak07u7Lw+fXRMnmKOCrcHqJ8Pl12H050fTnJZqHuvLqt+1rbLgAn7vvvvumMjwREUlRKndDtSpjSfoEt5ntbmaNStaB3sD7wFSiBET4fCasTwXODXdFHQOsDaerXgR6m1njcOdU71AnIiKVJJW7oRoClwMt3H14eDCvnbtPSxK6PzDFzEr6edzdXzCzt4EnzOx84HNgUNh/OnAKsBQoJprtFndfbWY3A2+H/W5K8RSYiCSYPDChMHNm7HZm5seI7RW/P8kOqVyzeIjo1FP3UF5OdIfUDpOFu38CdC6jfhXQs4x6By4qp61xwLgUxioiIhmQyjWLQ9z9T8AmiB7SAyyjoxKRCvfAvGgB4K67oiWGu2bdxV2z0oz98K5okWorlWSx0cwaEN0Gi5kdQsJdUSJSPbzyabQAMG1atMQw7d/TmPbvNGOXT4sWqbZSOQ11PdG7LA4ys8eAnwP5mRyUiIhkl1TelPeymc0HjiE6/XTZNtN3iIjILi6Vu6FKJvQreeq6hZntBXzu7pszNjIREckaqZyG+ivQBXiP6MiiI7AI2MvMRrj7Sxkcn4hUkHqJ/29v0CB2Ow3qxoitHb8/yQ6pJIv/AOe7+yIAM2sP3AT8L/B3QMlCpBp4ZEBC4fnnY7fz/NkxYn8Zvz/JDqncDdW2JFEAuPsHwGHhOQoREakBUn2t6mhgUigPBj4ws3qEZy9EJPvdOyf6vOxo4Obwsss//CHtdm7+ZxT7h1+kEbsw9Ncp/f4kO6RyZJFPNAXHyLB8Euo2Ab/MzLBEpKLNWhYtAMyYES0xzPh0BjM+TTP2qxnRItVWKrfOrgfuDsu2iip8RCIiknVSObIQEZEaTslCRESSKjdZmNmj4fOyyhuOiGTK3vWjBYAmTaIlhiYNm9CkYZqx9ZpEi1RbO7pm0dXMDgB+bWaPsM1Ms3qnhEj18kDfhMLTT8du5+lBMWJ7xO9PssOOksUYYAbQmuh9FonJwkO9iIjUAOWehnL3+9z9cGCcu7dO97WqIpJd7vhXtABwzTXREsM1BddwTUGasQuuiRaptlK5dXaEmXUGeoSq19z9vcwOS0Qq2vwVCYXZs2O3M7swRuzK+P1Jdkh6N5SZXQo8BuwXlsfM7JJMD0xERLJHKtN9/AY42t2/BzCzO4DZwP2ZHJiIiGSPVJ6zMGBLQnkLege3iEiNksqRxUPAHDObEsoDgAczNiIRyYif7ZFQaN48djvN94wR2zB+f5IdzN2T7xS9Le+4UHzd3d/J6Kh2Um5urs+dOzcjbbe8+rkdbv9s1KkZ6VdEJNPMbJ6755a1LZUjC9x9PjC/QkclIiLVRsbnhjKz2mb2jplNC+VWZjbHzJaa2WQz2y3U1wvlpWF7y4Q2rgn1i83spEyPWWRXdOM/owWAkSOjJYaRL4xk5Atpxs4bGS1SbaV0ZLGTLgM+BPYM5TuAv7j7JDMbA5wPjA6fa9z9UDMbEvYbHF7jOgToABwAFJhZW3ffsm1HIlK+D75JKCxYELudBV/GiF0Tvz/JDjs8sghHBa/GbdzMmgOnAn8LZQNOAJ4Ku4wnumAO0D+UCdt7hv37A5Pc/Qd3/5ToRUxHxR2TiIikb4fJIvz1/qOZ7RWz/XuA/wV+DOUmwLfuvjmUC4EDw/qBwLLQ72Zgbdi/tL6MmFJmNtzM5prZ3G+++WbbzSIishNSOQ1VBCw0s5eB70sq3f3SHQWZWV/ga3efZ2Z5OzPIVLj7WGAsRHdDZbo/EZGaJJVk8fewpOvnwGlmdgpQn+iaxb3A3mZWJxw9NAeWh/2XAwcBhWZWB9gLWJVQXyIxRkRS1KpxQqFt29jttG0SI7ZR/P4kO6T6nEUDoIW7L47VSXRkcaW79zWzJ4GnEy5wv+fufzWzi4BO7v7bcIH7dHcfZGYdgMeJrlMcQDRtepsdXeDWcxYiIunb0XMWqUwk2A9YALwQyjlmNnUnxvM74HIzW0p0TaLkafAHgSah/nLgagB3XwQ8AXwQxnCR7oQSEalcqZyGuoHor/qZAO6+wMzSep+Fu89MiP+EMu5mcvcNwJnlxN8K3JpOnyKytatnRJ+jegLDh0eFsWPTbmf4s1Hs2H5pxM4J/R2dfn+SHVJJFpvcfW10F2upH8vbWUSy06drEgr//nfsdv69Kkbsuvj9SXZIJVksMrP/AmqbWRvgUmBWZoclIiLZJJXpPi4henr6B2Ai8B0wMoNjEhGRLJPKa1WLgd+Hlx65u6/L/LBERCSbJE0WZtYNGAc0CuW1wK/dfV6GxyYiFaj9vgmFnJzY7eT8LEZs4/j9SXZI+pyFmb1HdLvq66F8HPBXdz+iEsYXi56zEBFJ3049ZwFsKUkUAO7+BrB5B/uLiMguptzTUOHteAD/NLMHiC5uOzCY8MyEiFQfl70Qfd7bBxg6NCpMmJB2O0P/HsVOOD2N2Fmhv+7p9yfZYUfXLO7epnx9wrom6hOpZr4sSigUFsZup/C7GLHF8fuT7FBusnD3X1bmQEREJHulcjfU3sC5QMvE/ZNNUS4iIruOVJ7gng68CSxE03yIiNRIqSSL+u5+ecZHIiIZ1aVZQuHYY2O3c2zzGLFN4/cn2SGV5yz+h+htedOIpvwAwN1XZ3Zo8ek5CxGR9O3oOYtUjiw2AncCv+enu6AcSGuachERqb5SSRZXAIe6+8pMD0ZEMueCadHnA32BM86ICk8/nXY7ZzwRxT49KI3Y10N/PdLvT7JDKsliKVCc6YGISGZ9uyGhsGpV7HZWFceI/SF+f5IdUkkW3wMLzOxVtr5moVtnRURqiFSSxT/CIiIiNVQq77MYXxkDERGR7JXKE9yfUsZcUO6uu6FEqpHuByUUevaM3U7PVjFi94/fn2SHVE5DJd5zWx84E9gnM8MRkUy57OiEwh/+ELudP/wiRmyn+P1Jdkj6Pgt3X5WwLHf3ewA9eSYiUoMkTRZm1iVhyTWz35La6av6ZvaWmb1rZovM7MZQ38rM5pjZUjObbGa7hfp6obw0bG+Z0NY1oX6xmZ0U/+uK1Fzn/iNaADj55GiJ4eTHTubkx9KMffXkaJFqK5XTUInvtdgMfAYMSiHuB+AEdy8ys7rAG2b2PHA58Bd3n2RmY4DzgdHhc427H2pmQ4A7gMFm1h4YAnQADgAKzKytu29J7SuKCMAPie+3XL8+djvrN8WI3RK/P8kOqdwNFeu9Fh5NOlXyupW6YXHgBOC/Qv144AaiZNE/rAM8BfyfmVmon+TuPwCfmtlS4ChgdpxxiYhI+lI5nVQPOIPt32dxUwqxtYF5wKHA/wM+Br5195K/cQqBA8P6gcCy0PZmM1sLNAn1byY0mxgjIiKVIJXTUM8Aa4n+0f8hyb5bCaeKcsILlKYAh6U7wFSZ2XBgOECLFi0y1Y2ISI2USrJo7u59dqYTd/82TBdyLLC3mdUJRxfNgeVht+XAQUChmdUB9gJWJdSXjichJrGPscBYiKYo35nxiuyKTmiVUOjbN3Y7fdvGiD0wfn+SHVJJFrPMrJO7L0ynYTPbF9gUEkUD4ESii9avAgOBScAwoiMXgKmhPDtsf8Xd3cymAo+b2Z+JLnC3Ad5KZywiAhd0TShceWXsdq7sHiP28Pj9SXZIJVkcB+SHJ7l/AIzo+vURSeKaAePDdYtawBPuPs3MPgAmmdktwDvAg2H/B4FHwwXs1UR3QOHui8zsCeADoruxLtKdUCIilSuVZBHr5mh3fw84soz6T4juZtq2fgPR0+FltXUrcGuccYhIZPBT0efkgUBeXlSYOTPtdvIejmJn5qcRWxD665V+f5IdUrl19vPKGIiIiGSvpE9wi4iIKFmIiEhSShYiIpJUKhe4RWQXcGqbhMKgVKZ3K9ugDjFiW8TvT7KDkoVIDXFu54TChRfGbufCbjFi28bvT7KDTkOJ1BDrN0ULAMXF0RJD8aZiijelGbu5OFqk2tKRhUgNkR/mSpg8EDjllKgQ4zmLUx6LYtN6zmJm6E/PWVRbOrIQEZGklCxERCQpJQsREUlKyUJERJLSBW6RGmJg+4RCfn7sdvJzYsS2jt+fZAclC5Ea4kwlC9kJOg0lUkOsXh8tAKxcGS0xrCxeycriNGM3rIwWqbZ0ZCFSQ4x4LvqcPBAYODAqxHjOYuATUWxaz1m8EfrTcxbVlo4sREQkKSULERFJSslCRESSUrIQEZGkdIFbpIYYekRCYcSI2O2MyI0R2yZ+f5IdlCxEaoh+bRMKgwfHbmdwxxixB8fvT7KDTkOJ1BD/WRctACxbFi0xLFu7jGVr04z9flm0SLWVsWRhZgeZ2atm9oGZLTKzy0L9Pmb2spktCZ+NQ72Z2X1mttTM3jOzLgltDQv7LzGzYZkas8iu7H9ejBYAzjknWmI4Z8o5nDMlzdjZ50SLVFuZPLLYDFzh7u2BY4CLzKw9cDUww93bADNCGeBkoE1YhgOjIUouwPXA0cBRwPUlCUZERCpHxpKFu69w9/lhfR3wIXAg0B8YH3YbDwwI6/2BRzzyJrC3mTUDTgJedvfV7r4GeBnok6lxi4jI9irlmoWZtQSOBOYA+7v7irDpS2D/sH4gkHhSszDUlVcvIiKVJOPJwsz2AJ4GRrr7d4nb3N0Br6B+hpvZXDOb+80331REkyIiEmT01lkzq0uUKB5z97+H6q/MrJm7rwinmb4O9cuBgxLCm4e65UDeNvUzt+3L3ccCYwFyc3MrJAGJ7Er+u0tC4YorYrdzxbExYg+L359kh4wlCzMz4EHgQ3f/c8KmqcAwYFT4fCah/mIzm0R0MXttSCgvArclXNTuDVyTqXGL7Kp6tU4o9OsXu51+7WLENo/fn2SHTB5Z/Bw4B1hoZgtC3bVESeIJMzsf+BwYFLZNB04BlgLFwHkA7r7azG4G3g773eTuqzM4bpFd0sdros9DGgOLF0eFdu3Sbmfxyii2XdM0Yr8L/e2Zfn+SHTKWLNz9DcDK2dyzjP0duKictsYB4ypudCI1z7Uzos/JA4ELLogKMd5nccG0KDat91m8FfrT+yyqLT3BLSIiSSlZiIhIUkoWIiKSlJKFiIgkpSnKRWqIS45KKFx3Xex2rjs+RmzH+P1JdlCyEKkhjmuRUOjVK3Y7vVrHiP1Z/P4kO+g0lEgNseibaAFgwYJoiWHBlwtY8GWasWsWRItUWzqyEKkhbvpn9Dl5IDByZFSI8ZzFyBei2LSes5gX+tNzFtWWjixERCQpJQsREUlKyUJERJJSshARkaR0gVski7W8+rkdbv9s1Kkpt3VV94TCbbfFHBHc1jNGbOf4/Ul2ULIQqSFyD0godO9e7n7JdD8oRuy+8fuT7KDTUCI1xNz/RAsAs2ZFSwyzls1i1rI0Y7+ZFS1SbenIQmQXlngay/c5BgBb/SaTHr8agGO+WJh2m9fOuBZI8zmLd6MYPWdRfSlZiFRjya5piFQUnYYSEZGklCxERCQpJQsREUlK1yxEqlilXXf47oPS1Zt6Dgdgeoxm7ulzT/pBXWPESFZRshCpIWzzd6XrH+zfGoj30F/Oz3LS77xxjBjJKjoNJVJD+G5N8N2aAPDzzxbw888WxGqn4JMCCj4pSC/oy4JokWpLRxYiGZY1t7fu0Sb6XL2KS2ZNAuBfLXPSbuaW124B0nxj3vtRjN6YV31l7MjCzMaZ2ddm9n5C3T5m9rKZLQmfjUO9mdl9ZrbUzN4zsy4JMcPC/kvMbFimxisiIuXL5Gmoh4E+29RdDcxw9zbAjFAGOBloE5bhwGiIkgtwPXA0cBRwfUmCERGRypOxZOHurwGrt6nuD4wP6+OBAQn1j3jkTWBvM2sGnAS87O6r3X0N8DLbJyAREcmwyr5msb+7rwjrXwL7h/UDgWUJ+xWGuvLqt2Nmw4mOSmjRokUFDlmk5irresuXu62qgpFIVauyC9zu7mbmFdjeWGAsQG5uboW1K7LLWPvTpIHXnnRx7GaabIoRe9QDsfuT7FDZyeIrM2vm7ivCaaavQ/1y4KCE/ZqHuuVA3jb1MythnCK7HNvyfen6J02ax26nrseI3bNd7P4kO1R2spgKDANGhc9nEuovNrNJRBez14aE8iJwW8JF7d7ANZU8ZpEdyppbY5PwevsBYD98Tc+lcwCYcejRabdTXGtOWEv9LX0UPht9Nu+Xdn+SHTKWLMxsItFRQVMzKyS6q2kU8ISZnQ98DgwKu08HTgGWAsXAeQDuvtrMbgbeDvvd5O7bXjQXkVTsHj21zQ9f899vTQHiJYvv6kwJazelHvTR3dGnkkW1lbFk4e5nlbOpZxn7OnBROe2MA8ZV4NBE0lJdjhwqW0W+H1yyn6b7EBGRpJQsREQkKSULERFJShMJSo1XY65JfLugdPV/+l4Ru5mmm2LEHvto7P4kOyhZiNQQ9uOG0vUVe+4bu506nlps2Un4vdI1XQCvXnQaSqSG8PrN8PrNAOj74Wv0/fC1WO18X/s1vq+dXmzfvV6j717x+pPsoCMLkZqi4cHR54YVDH0neqHqtMOPT7uZdbWj2N23pB47tEnob236/Ul2ULKQaqHGXFcQyVI6DSUiIknpyEKygo4cRLKbkoWIVAlNF1K9KFlIpdCRQxZYM690dcSA+JM377sx/dgRn2uy6OpOyUKkhjDfVLq+puFesdupTfqxa7bE70+ygy5wi9QQ3qA53iB6cdHAhQUMXFgQq52i2gUU1U4vdmDjAgY2jtefZAclC5GaokHzaEHJQtKn01BSYXRdQmTXpWQhIllJd0tlF52GEhGRpHRkISnTaSaRmkvJQqSmWP1W6Wr+mTfEbma/jenH5n8avz/JDkoWIjWE8WPp+oa69WO3U4v0Yzd4/P7Ks6MjXV3PqHhKFiI1hIcpyq34c4bOj/6hndAl/X9U19WOYhttST12aJPQ36rK+UdcF8crnpKFlNI1iV1cePERxZ/T96PXgXjJ4vvaUWw6yaLvXqG/SkoWUvGqTbIwsz7AvUBt4G/uPqqKhyQi1ZSOPNJXLZKFmdUG/h9wIlAIvG1mU939g6odWfWiIweR1CiZbK9aJAvgKGCpu38CYGaTgP6AkoWIVLqamEyqS7I4EFiWUC4Ejq6isWQtHTmIZIdM/n+xqhKRuXuVdJwOMxsI9HH334TyOcDR7n5xwj7DgeGh2A5YvBNdNgVW7kR8Jmls8Whs8Whs8VTXsR3s7vuWtaG6HFksBw5KKDcPdaXcfSwwtiI6M7O57p5bEW1VNI0tHo0tHo0tnl1xbNVlbqi3gTZm1srMdgOGAFOreEwiIjVGtTiycPfNZnYx8CLRrbPj3H1RFQ9LRKTGqBbJAsDdpwPTK6m7CjmdlSEaWzwaWzwaWzy73NiqxQVuERGpWtXlmoWIiFShGp8szOxMM1tkZj+aWbl3CJjZZ2a20MwWmNncLBxfHzNbbGZLzezqShrbPmb2spktCZ+Ny9lvS/jdFphZxm5MSPYbmFk9M5scts8xs5aZGkuMseWb2TcJv9NvKnFs48zsazN7v5ztZmb3hbG/Z2ZdsmhseWa2NuF3+2Mlju0gM3vVzD4I/x+9rIx9Kv23S3Fc6f9u7l6jF+BwoucyZgK5O9jvM6BpNo6P6KL/x0BrYDfgXaB9JYztT8DVYf1q4I5y9iuqhLEk/Q2AC4ExYX0IMLmS/jdMZWz5wP9V9n9foe/jgS7A++VsPwV4HjDgGGBOFo0tD5hWRb9bM6BLWG8E/LuM/10r/bdLcVxp/241/sjC3T909515gC+jUhxf6XQo7r4RKJkOJdP6A+PD+nhgQCX0WZ5UfoPE8T4F9DQzy5KxVRl3fw1YvYNd+gOPeORNYG8za5YlY6sy7r7C3eeH9XXAh0SzTSSq9N8uxXGlrcYnizQ48JKZzQtPi2eTsqZD2en/OFKwv7uvCOtfAvuXs199M5trZm+a2YAMjSWV36B0H3ffDKwFmmRoPOmODeCMcKriKTM7qIztVaWq/vtK1bFm9q6ZPW9mHapiAOGU5pHAnG02Velvt4NxQZq/W7W5dXZnmFkB8LMyNv3e3Z9JsZnj3H25me0HvGxmH4W/erJlfBmxo7ElFtzdzay8W+sODr9da+AVM1vo7h9X9FiruWeBie7+g5ldQHQEdEIVj6k6mE/031eRmZ0C/ANoU5kDMLM9gKeBke7+XWX2vSNJxpX271YjkoW796qANpaHz6/NbArRqYUKSRYVML6k06HEtaOxmdlXZtbM3VeEQ+uvy2mj5Lf7xMxmEv2lU9HJIpXfoGSfQjOrA+wFrKrgccQam7snjuNvRNeDskXG/vvaWYn/CLr7dDP7q5k1dfdKmZfJzOoS/YP8mLv/vYxdquS3SzauOL+bTkOlwMx2N7NGJetAb6DMuzOqSFVNhzIVGBbWhwHbHQWZWWMzqxfWmwI/JzNTy6fyGySOdyDwioerfRmWdGzbnMc+jeg8c7aYCpwb7uw5BlibcPqxSpnZz0quO5nZUUT/plXGHwCEfh8EPnT3P5ezW6X/dqmMK9bvlukr89m+AL8iOo/4A/AV8GKoPwCYHtZbE93B8i6wiOj0UNaML5RPIbrr4ePKGh/R+f4ZwBKgANgn1OcSvc0QoDuwMPx2C4HzMzie7X4D4CbgtLBeH3gSWAq8BbSuxP8dk43t9vDf1rvAq8BhlTi2icAKYFP4b+184LfAb8N2I3r52Mfhf8Ny7xqsgrFdnPC7vQl0r8SxHUd0LfM9YEFYTqnq3y7FcaX9u+kJbhERSUqnoUREJCklCxERSUrJQkREklKyEBGRpJQsREQkKSULkRSYWVE59SUz6r5vZs+a2d6hvqWZuZldkrDv/5lZfkK5jkUzzY7K9Ph3hpndYGZXVvU4pGopWUiNFZ7i3lnr3T3H3TsSTXh3UcK2r4HLwoN4ZTmR6NmLMytpQkOR2JQsJCuY2W8T5tb/1MxeLWOfDmb2VtjnPTNrE+rPDeV3zezRUNfSzF4J9TPMrEWof9jMxpjZHOBPZnaImb0QJoh83cwOC/u1MrPZFr3D5JYUv8Zstp4k7huihxaHlb07ZwH3Al8Ax5bzu/wi4Xd5J2EmgavM7O3w/W5M2P/3ZvZvM3vDzCZue0RgZnuZ2edmViuUdzezZWZW18z+O7T5rpk9bWYNyxjPTAvvVTGzpmb2WVivbWZ3JozpgpR+Mak2lCwkK7j7GHfPAboRPalb1jQFvwXuDfvlEs3x1AG4DjjB3TsDJS96uR8Y7+5HAI8B9yW005zoidXLid5HfIm7dwWuBP4a9rkXGO3unYieIN4hM6sN9GT7KUbuAK4M2xP3rw/0IkwgSJQ4ynIlcFH4zj2A9WbWm2jSt6OAHKCrmR1vZl2JphLJIXpit9u2jbn7WqInen8RqvoSzQqwCfi7u3cLv+OHRE9Lp+p8oqksuoV+/9vMWqURL1lOyUKyzb1EczY9W8a22cC1ZvY7ohkz1xPNzPqkhwnQ3L3k3QfHAo+H9UeJpkAo8aS7b7FoVs7uwJNmtgB4gOjFMRDNYTUxIb48DUJsyRTtLydudPdPiKaH/q9t4voCr4bv8DQwYNuEEvwL+LOZXQrs7dHU6r3D8g7R7KGHESWPHsAUdy/2aKK48uYHmwwMDutDQhmgYzi6WgicDaQz3XdvojmQFoTv24RKnv1VMkvJQrJGuPh7MHBjKP8q4RRMrrs/TjTJ3npgupnFncL7+/BZC/g2XHMoWQ5P2C+VuXDWh7/6DyaaB+iiMva5Dfhd2F7iLKBXOI0zj+gf1xPM7KKE73yAu48CfgM0AP4VTpMZcHvCmA919wdT/fJESaSPme0DdAVeCfUPAxeHo6kbiebS2tZmfvp3I3G7ER2hlYyplbu/lMaYJMspWUhWCKdQrgSGuvuPAO4+JeEfn7kWvQ/jE3e/j2iG2yOI/qE708yahHb2CU3OIvqrGaK/kl/fts/w1/enZnZmiDUz6xw2/2ub+B1y92LgUuCKbS+cu/tHRDPt9gv97El0FNDC3Vu6e0uiJHOWu/+/hO/8HzM7xN0XuvsdRLPXHga8CPw6HBlhZgda9J6V14iOUBqEaxv9yhlrUWjrXqJXa24JmxoBKyya3rq87/wZUYKBaObeEi8CI0IsZtbWohmaZRehZCHZ4mJgH+DV8Ff138rYZxDwfjjV0ZHodZWLgFuBf5rZu/x0reMS4Dwzew84h5+uZWzrbOD8ELuIn153ehlwUTglk9Kbzdz9HaKZPsu6/nAr0bUSiGYSfsXdf0jY/gzQz8J07glGWnRb7ntEM68+H/5ifxyYHcb3FNDIo1dpTiaaSfR5ooRQnsnAUH46BQXwB6JTSP8CPion7i6ipPAO0DSh/m9ECXG+mb1PdEqvRrwvp6bQrLMiuygzuwEocve7qnosUv3pyEJERJLSkYWIiCSlIwsREUlKyUJERJJSshARkaSULEREJCklCxERSUrJQkREkvr/Aa5VL7trYGYdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check whether 'SUMO2', 'UBE2I', 'SUMO1', 'PIAS4' are expressed in IMR90\n",
    "df_all_norm = pd.read_csv(saving_dir+'features_matrix_all_genes_norm.csv', header=0, index_col=0)\n",
    "df_select = df_all_norm.loc[['RNAseq'], :]\n",
    "# Plot expression of genes compared to histogram of RNAseq of all genes\n",
    "col_dict = {'SUMO2':'green', 'UBE2I':'orange', 'SUMO1':'red', 'PIAS4':'dodgerblue'}\n",
    "plt.figure(figsize=(6,5))\n",
    "plt.hist(df_all_norm.loc['RNAseq', :].values, bins=np.arange(-1.5, 2.5, 0.1),\n",
    "         label='all genes')\n",
    "for g in ['SUMO2', 'UBE2I', 'SUMO1', 'PIAS4']:\n",
    "    plt.vlines(x=df_all_norm.loc['RNAseq', g], ymin=0, ymax=8000, \n",
    "               color=col_dict[g], linestyle='dashed', label=g)\n",
    "plt.xlabel('z-scored RNA-seq value')\n",
    "plt.ylabel('number of genes')\n",
    "plt.legend(loc='upper right')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung TF genes/loci targeting adhesome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select all lung fibroblast TFs targeting active genes\n",
    "all_lung_adhesome_tf_genes = tftarget_df_lungfibro[tftarget_df_lungfibro['target_gene'].isin(active_adhesome_genes)]\n",
    "all_lung_adhesome_tf_genes = all_lung_adhesome_tf_genes['tf'].unique()\n",
    "with open(saving_dir+'all_lung_adhesome_tf_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(all_lung_adhesome_tf_genes, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active TF genes targeting active adhesome\n",
    "active_lung_adhesome_tf_genes = tf_target_df[(tf_target_df['target'].isin(active_adhesome_genes)) &\n",
    "                                             (tf_target_df['geneSymbol'].isin(active_genes))]['geneSymbol'].unique()\n",
    "with open(saving_dir+'active_lung_adhesome_tf_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_adhesome_tf_genes, f)\n",
    "\n",
    "active_lung_adhesome_tf_loci = tf_target_df[(tf_target_df['target'].isin(active_adhesome_genes)) &\n",
    "                                            (tf_target_df['geneSymbol'].isin(active_genes))]['locus_id'].unique()\n",
    "with open(saving_dir+'active_lung_adhesome_tf_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(active_lung_adhesome_tf_loci, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CHD1', 'SNAPC1', 'SNAPC4', 'SNAPC5', 'SPI1'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lung fibroblast TFs that are not active or do not target active adhesome genes\n",
    "set(np.sort(tf_target_df['geneSymbol'].unique()))-set(np.sort(active_lung_adhesome_tf_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary tf2target and target2tf\n",
    "active_adhesome_tf_target_df = tf_target_df[(tf_target_df['geneSymbol'].isin(active_lung_adhesome_tf_genes)) &\n",
    "                                            (tf_target_df['target'].isin(active_adhesome_genes))]\n",
    "\n",
    "adhesome_active_lung_tf2target = {tf: active_adhesome_tf_target_df[active_adhesome_tf_target_df['geneSymbol']==tf]['target'].unique() \n",
    "                                  for tf in active_lung_adhesome_tf_genes}\n",
    "with open(saving_dir+'adhesome_active_lung_tf2target.pkl', 'wb') as f:\n",
    "    pickle.dump(adhesome_active_lung_tf2target, f)\n",
    "\n",
    "adhesome_active_lung_target2tf = {target: active_adhesome_tf_target_df[active_adhesome_tf_target_df['target']==target]['geneSymbol'].unique() \n",
    "                                  for target in active_adhesome_tf_target_df['target'].unique()}\n",
    "with open(saving_dir+'adhesome_active_lung_target2tf.pkl', 'wb') as f:\n",
    "    pickle.dump(adhesome_active_lung_target2tf, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TCR genes/loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCR genes missing from location data = ['IKBKG', 'WAS']\n",
      "Total number of TCR genes = 90\n",
      "Total number of TCR loci = 113\n",
      "Total number of active TCR genes = 63\n",
      "Total number of active TCR loci = 103\n"
     ]
    }
   ],
   "source": [
    "# Select active/inactive TCR genes and loci\n",
    "tcr_chr_loci, missing_tcr = lu.annotate_genes(tcr_df[['geneSymbol']], df_loc, resol)\n",
    "tcr_chr_loci['gene_active'] = tcr_chr_loci['gene'].isin(active_genes)\n",
    "tcr_chr_loci['locus_active'] = tcr_chr_loci['locus_id'].isin(all_active_loci)\n",
    "tcr_genes = tcr_chr_loci['gene'].unique()\n",
    "tcr_loci = tcr_chr_loci['locus_id'].unique()\n",
    "active_tcr_genes = tcr_chr_loci[tcr_chr_loci['gene_active']==True]['gene'].unique()\n",
    "active_tcr_loci = tcr_chr_loci[tcr_chr_loci['locus_active']==True]['locus_id'].unique()\n",
    "inactive_tcr_genes = tcr_chr_loci[tcr_chr_loci['gene_active']==False]['gene'].unique()\n",
    "inactive_tcr_loci = tcr_chr_loci[tcr_chr_loci['locus_active']==False]['locus_id'].unique()\n",
    "print('TCR genes missing from location data = '+str(missing_tcr))\n",
    "print('Total number of TCR genes = '+str(len(tcr_genes)))\n",
    "print('Total number of TCR loci = '+str(len(tcr_loci)))\n",
    "print('Total number of active TCR genes = '+str(len(active_tcr_genes)))\n",
    "print('Total number of active TCR loci = '+str(len(active_tcr_loci)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to pickle\n",
    "with open(saving_dir+'tcr_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(tcr_chr_loci, f)\n",
    "with open(saving_dir+'tcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(tcr_genes, f)\n",
    "with open(saving_dir+'tcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(tcr_loci, f)\n",
    "with open(saving_dir+'active_tcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(active_tcr_genes, f)\n",
    "with open(saving_dir+'inactive_tcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(inactive_tcr_genes, f)\n",
    "with open(saving_dir+'active_tcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(active_tcr_loci, f)\n",
    "with open(saving_dir+'inactive_tcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(inactive_tcr_loci, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BCR genes/loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BCR genes missing from location data = ['PI-4,5-P2', 'PI-4-P', 'ELK1', 'DAG', 'IKBKG', 'IP3', 'CD45', 'PDPK2', 'PRKCB1', 'BTK']\n",
      "Total number of BCR genes = 93\n",
      "Total number of BCR loci = 118\n",
      "Total number of active BCR genes = 72\n",
      "Total number of active BCR loci = 112\n"
     ]
    }
   ],
   "source": [
    "# Select active/inactive BCR genes and loci\n",
    "bcr_chr_loci, missing_bcr = lu.annotate_genes(bcr_df[['geneSymbol']], df_loc, resol)\n",
    "bcr_chr_loci['gene_active'] = bcr_chr_loci['gene'].isin(active_genes)\n",
    "bcr_chr_loci['locus_active'] = bcr_chr_loci['locus_id'].isin(all_active_loci)\n",
    "bcr_genes = bcr_chr_loci['gene'].unique()\n",
    "bcr_loci = bcr_chr_loci['locus_id'].unique()\n",
    "active_bcr_genes = bcr_chr_loci[bcr_chr_loci['gene_active']==True]['gene'].unique()\n",
    "active_bcr_loci = bcr_chr_loci[bcr_chr_loci['locus_active']==True]['locus_id'].unique()\n",
    "inactive_bcr_genes = bcr_chr_loci[bcr_chr_loci['gene_active']==False]['gene'].unique()\n",
    "inactive_bcr_loci = bcr_chr_loci[bcr_chr_loci['locus_active']==False]['locus_id'].unique()\n",
    "print('BCR genes missing from location data = '+str(missing_bcr))\n",
    "print('Total number of BCR genes = '+str(len(bcr_genes)))\n",
    "print('Total number of BCR loci = '+str(len(bcr_loci)))\n",
    "print('Total number of active BCR genes = '+str(len(active_bcr_genes)))\n",
    "print('Total number of active BCR loci = '+str(len(active_bcr_loci)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to pickle\n",
    "with open(saving_dir+'bcr_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(bcr_chr_loci, f)\n",
    "with open(saving_dir+'bcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(bcr_genes, f)\n",
    "with open(saving_dir+'bcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(bcr_loci, f)\n",
    "with open(saving_dir+'active_bcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(active_bcr_genes, f)\n",
    "with open(saving_dir+'inactive_bcr_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(inactive_bcr_genes, f)\n",
    "with open(saving_dir+'active_bcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(active_bcr_loci, f)\n",
    "with open(saving_dir+'inactive_bcr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(inactive_bcr_loci, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jointly consider active adhesome genes, and active adhesome lung TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene and locus dataframe for acive adhesome genes and adhesome TFs\n",
    "adh_and_tf_chr_loci = pd.concat([adhesome_chr_loci[adhesome_chr_loci['active']],\n",
    "                                 lungTF_chr_loci[lungTF_chr_loci['gene'].isin(active_lung_adhesome_tf_genes)]], \n",
    "                                 axis=0)\n",
    "with open(saving_dir+'adh_and_tf_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(adh_and_tf_chr_loci, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene2locus and locus2gene dictionaries for adhesome+adhesome TFs\n",
    "adh_and_tf_gene2locus = {gene: adh_and_tf_chr_loci[adh_and_tf_chr_loci['gene']==gene]['locus_id'].values.astype(str)  \n",
    "              for gene in np.unique(adh_and_tf_chr_loci['gene'])}\n",
    "with open(saving_dir+'adh_and_tf_gene2locus.pkl', 'wb') as f:\n",
    "    pickle.dump(adh_and_tf_gene2locus, f)\n",
    "\n",
    "adh_and_tf_locus2gene = {locus: adh_and_tf_chr_loci[adh_and_tf_chr_loci['locus_id']==locus]['gene'].values.astype(str) \n",
    "              for locus in np.unique(adh_and_tf_chr_loci['locus_id'])}\n",
    "with open(saving_dir+'adh_and_tf_locus2gene.pkl', 'wb') as f:\n",
    "    pickle.dump(adh_and_tf_locus2gene, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jointly consider active adhesome genes, lung TFs and mechanoreg proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active adhesome genes\n",
    "act_adhesome_chr_loci = adhesome_chr_loci[adhesome_chr_loci['active']==True]\n",
    "with open(saving_dir+'act_adhesome_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(act_adhesome_chr_loci, f)\n",
    "\n",
    "# Active lung TFs\n",
    "act_lung_TF_chr_loci = lungTF_chr_loci1[lungTF_chr_loci1['tf'].isin(active_lung_tf_genes)]\n",
    "act_lung_TF_chr_loci.columns = ['chrom', 'chrom_int', 'loci', 'gene', 'locus_id', 'active']\n",
    "with open(saving_dir+'act_lung_TF_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(act_lung_TF_chr_loci, f)\n",
    "\n",
    "# Active mechanoreg proteins\n",
    "act_proteins_chr_loci = proteins_chr_loci[proteins_chr_loci['active']==True]\n",
    "with open(saving_dir+'act_proteins_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(act_proteins_chr_loci, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Joint annotated dataframe\n",
    "joint_chr_loci = pd.concat([act_adhesome_chr_loci,act_lung_TF_chr_loci,act_proteins_chr_loci], axis=0)\n",
    "with open(saving_dir+'joint_chr_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(joint_chr_loci, f)\n",
    "\n",
    "joint_loci = np.unique(joint_chr_loci['locus_id'])\n",
    "with open(saving_dir+'joint_loci.pkl', 'wb') as f:\n",
    "    pickle.dump(joint_loci, f)\n",
    "\n",
    "joint_genes = np.unique(joint_chr_loci['gene'])\n",
    "with open(saving_dir+'joint_genes.pkl', 'wb') as f:\n",
    "    pickle.dump(joint_genes, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene2locus and locus2gene dictionaries\n",
    "selected_gene2locus = {gene: joint_chr_loci[joint_chr_loci['gene']==gene]['locus_id'].values.astype(str)  \n",
    "                       for gene in np.unique(joint_chr_loci['gene'])}\n",
    "with open(saving_dir+'selected_gene2locus.pkl', 'wb') as f:\n",
    "    pickle.dump(selected_gene2locus, f)\n",
    "\n",
    "selected_locus2gene = {locus: joint_chr_loci[joint_chr_loci['locus_id']==locus]['gene'].values.astype(str) \n",
    "                       for locus in np.unique(joint_chr_loci['locus_id'])}\n",
    "with open(saving_dir+'selected_locus2gene.pkl', 'wb') as f:\n",
    "    pickle.dump(selected_locus2gene, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 231/231 [00:18<00:00, 12.79it/s]\n"
     ]
    }
   ],
   "source": [
    "# Get HiC contact map\n",
    "contacts_df = pd.DataFrame(0, index=joint_loci, columns=joint_loci)\n",
    "# Loop over all chromosome pairs and fill gene_contacts_df\n",
    "chr_list = np.arange(1,23,1)\n",
    "chr_pairs = list(itertools.combinations(chr_list, 2))\n",
    "for pair in tqdm(chr_pairs):\n",
    "    time.sleep(.01)\n",
    "    chr1, chr2 = pair\n",
    "\n",
    "    # Select loci on chr1 and chr2\n",
    "    loci_chr1 = [locus for locus in joint_loci if ('chr_'+str(chr1)+'_') in locus]\n",
    "    loci_chr2 = [locus for locus in joint_loci if ('chr_'+str(chr2)+'_') in locus]\n",
    "\n",
    "    # Load HiC data for this chromosome pair\n",
    "    processed_hic_filename = 'hic_'+'chr'+str(chr1)+'_'+'chr'+str(chr2)+'_norm1_filter3'+'.pkl'\n",
    "    with open(dir_processed_hic+processed_hic_filename, 'rb') as f:\n",
    "        hic_chpair_df = pickle.load(f)\n",
    "\n",
    "    # Fill in corresponding submatrix of gene_contacts_df by sum of Hi-C contacts across all gene loci\n",
    "    locus_id_pairs = itertools.product(loci_chr1,loci_chr2)\n",
    "    for locus1_id, locus2_id in locus_id_pairs:\n",
    "        locus1 = int(locus1_id.split('_')[3])\n",
    "        locus2 = int(locus2_id.split('_')[3])\n",
    "        contacts_df.loc[locus1_id, locus2_id] = hic_chpair_df.loc[locus1,locus2]\n",
    "# Symmetrize data frame\n",
    "contacts_df = contacts_df+contacts_df.T\n",
    "# Save to pickle\n",
    "with open(saving_dir+f'contacts_df_{norm}.pkl', 'wb') as f:\n",
    "    pickle.dump(contacts_df, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
