{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload\n",
    "import sys\n",
    "import numpy as np \n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "plt.ioff()\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "import pybedtools\n",
    "from joblib import Parallel, delayed\n",
    "import get_feature_matrix_adhesomegenes as gfma\n",
    "import utils as lu\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import pickle\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'get_feature_matrix_adhesomegenes' from '/home/louiscam/projects/gpcr/code/epigenome_processing/get_feature_matrix_adhesomegenes.py'>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload modules in case of modifications\n",
    "reload(lu)\n",
    "reload(gfma)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_dir = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/epigenome_data/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'\n",
    "dir_adhesome = '/home/louiscam/projects/gpcr/data/adhesome_data/'\n",
    "saving_dir = '/home/louiscam/projects/gpcr/save/figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'IMR90'\n",
    "resol_str = '250kb'\n",
    "resol = 250000\n",
    "quality = 'MAPQGE30'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "code_folding": [
     0,
     42,
     84,
     126
    ]
   },
   "outputs": [],
   "source": [
    "features = [\"CEBPB\",\n",
    "            \"CHD1\",\n",
    "            \"CTCF\",\n",
    "            \"DNase-seq\",\n",
    "            \"EP300\",\n",
    "            \"H2AK5ac\",\n",
    "            \"H2AK9ac\",\n",
    "            \"MacroH2A1.1\",\n",
    "            \"H2AZ\",\n",
    "            \"H2BK120ac\",\n",
    "            \"H2BK12ac\",\n",
    "            \"H2BK15ac\",\n",
    "            \"H2BK20ac\",\n",
    "            \"H2BK5ac\",\n",
    "            \"H3K14ac\",\n",
    "            \"H3K18ac\",\n",
    "            \"H3K23ac\",\n",
    "            \"H3K27ac\",\n",
    "            \"H3K27me3\",\n",
    "            \"H3K36me3\",\n",
    "            \"H3K4ac\",\n",
    "            \"H3K4me1\",\n",
    "            \"H3K4me2\",\n",
    "            \"H3K4me3\",\n",
    "            \"H3K56ac\",\n",
    "            \"H3K79me1\",\n",
    "            \"H3K79me2\",\n",
    "            \"H3K9ac\",\n",
    "            \"H3K9me1\",\n",
    "            \"H3K9me3\",\n",
    "            \"H4K20me1\",\n",
    "            \"H4K5ac\",\n",
    "            \"H4K8ac\",\n",
    "            \"H4K91ac\",\n",
    "            \"MAFK\",\n",
    "            \"MAZ\",\n",
    "            \"MXI1\",\n",
    "            \"POLR2A\",\n",
    "            \"RAD21\",\n",
    "            \"RCOR1\",\n",
    "            \"RFX5\"]\n",
    "\n",
    "filenames = [   \"wgEncodeAwgTfbsSydhImr90CebpbIggrabUniPk.narrowPeak\",\n",
    "                \"wgEncodeSydhTfbsImr90Chd1nb10060411IggrabPk.narrowPeak\",\n",
    "                \"wgEncodeAwgTfbsSydhImr90CtcfbIggrabUniPk.narrowPeak\",\n",
    "                \"E017-DNase.hotspot.fdr0.01.broad.bed\",\n",
    "                \"GSM1055812_p300.IMR90.mono.bed\",\n",
    "                \"E017-H2AK5ac.broadPeak\",\n",
    "                \"E017-H2AK9ac.broadPeak\",\n",
    "                \"GSM1325070_m1_c_.In.101118.P0.01.G.1000.gappedPeak\",\n",
    "                \"E017-H2A.Z.broadPeak\",\n",
    "                \"E017-H2BK120ac.broadPeak\",\n",
    "                \"E017-H2BK12ac.broadPeak\",\n",
    "                \"E017-H2BK15ac.broadPeak\",\n",
    "                \"E017-H2BK20ac.broadPeak\",\n",
    "                \"E017-H2BK5ac.broadPeak\",\n",
    "                \"E017-H3K14ac.broadPeak\",\n",
    "                \"E017-H3K18ac.broadPeak\",\n",
    "                \"E017-H3K23ac.broadPeak\",\n",
    "                \"E017-H3K27ac.broadPeak\",\n",
    "                \"E017-H3K27me3.broadPeak\",\n",
    "                \"E017-H3K36me3.broadPeak\",\n",
    "                \"E017-H3K4ac.broadPeak\",\n",
    "                \"E017-H3K4me1.broadPeak\",\n",
    "                \"E017-H3K4me2.broadPeak\",\n",
    "                \"E017-H3K4me3.broadPeak\",\n",
    "                \"E017-H3K56ac.broadPeak\",\n",
    "                \"E017-H3K79me1.broadPeak\",\n",
    "                \"E017-H3K79me2.broadPeak\",\n",
    "                \"E017-H3K9ac.broadPeak\",\n",
    "                \"E017-H3K9me1.broadPeak\",\n",
    "                \"E017-H3K9me3.broadPeak\",\n",
    "                \"E017-H4K20me1.broadPeak\",\n",
    "                \"E017-H4K5ac.broadPeak\",\n",
    "                \"E017-H4K8ac.broadPeak\",\n",
    "                \"E017-H4K91ac.broadPeak\",\n",
    "                \"wgEncodeAwgTfbsSydhImr90MafkIggrabUniPk.narrowPeak\",\n",
    "                \"wgEncodeSydhTfbsImr90Mazab85725IggrabPk.narrowPeak\",\n",
    "                \"wgEncodeSydhTfbsImr90Mxi1IggrabPk.narrowPeak\",\n",
    "                \"wgEncodeAwgTfbsSydhImr90Pol2IggrabUniPk.narrowPeak\",\n",
    "                \"wgEncodeAwgTfbsSydhImr90Rad21IggrabUniPk.narrowPeak\",\n",
    "                \"wgEncodeSydhTfbsImr90Corestsc30189IggrabPk.narrowPeak\",\n",
    "                \"wgEncodeSydhTfbsImr90Rfx5IggrabPk.narrowPeak\"]\n",
    "\n",
    "sources = [ \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Jin_et_al.\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Chen_et_al.\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"Roadmap_Epigenomics\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\",\n",
    "            \"ENCODE\"]\n",
    "\n",
    "accessions = [  \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE18927\",\n",
    "                \"GSE43070\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE54847\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE16256\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\",\n",
    "                \"GSE31477\"]\n",
    "\n",
    "identifier_col = np.repeat(4, len(features))\n",
    "\n",
    "df = pd.DataFrame({'feature': features,\n",
    "                   'filename': filenames,\n",
    "                   'identifier_col': identifier_col,\n",
    "                   'source': sources, \n",
    "                   'accession':accessions})\n",
    "df.to_csv(epigenome_dir + 'features_extended.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test: processing epigenomic features for all adhesome genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identify start and stop sites of each adhesome gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16085254</td>\n",
       "      <td>16113084</td>\n",
       "      <td>27830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>ASAP3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23755055</td>\n",
       "      <td>23810750</td>\n",
       "      <td>55695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>SLC9A1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27425299</td>\n",
       "      <td>27481621</td>\n",
       "      <td>56322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1618</th>\n",
       "      <td>FABP3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31838099</td>\n",
       "      <td>31845923</td>\n",
       "      <td>7824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>MACF1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39549838</td>\n",
       "      <td>39952810</td>\n",
       "      <td>402972</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene chrom     start       end  length\n",
       "146   FBLIM1  chr1  16085254  16113084   27830\n",
       "1548   ASAP3  chr1  23755055  23810750   55695\n",
       "1588  SLC9A1  chr1  27425299  27481621   56322\n",
       "1618   FABP3  chr1  31838099  31845923    7824\n",
       "335    MACF1  chr1  39549838  39952810  402972"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load active adhesome genes\n",
    "active_adhesome_genes = pickle.load(open(saving_dir+'active_adhesome_genes.pkl', 'rb'))\n",
    "# Load gene location in hg19\n",
    "gene_locations_filename = genome_dir+'chrom_hg19.loc_canonical'\n",
    "gene_id_filename = genome_dir+'chrom_hg19.name'\n",
    "df_loc = lu.get_gene_locations(gene_locations_filename, gene_id_filename)\n",
    "# Find location of adhesome genes\n",
    "adhesome_loc = df_loc[df_loc['geneSymbol'].isin(active_adhesome_genes)][['geneSymbol','#chrom','chromStart','chromEnd']]\n",
    "adhesome_loc['geneLength'] = adhesome_loc['chromEnd']-adhesome_loc['chromStart']\n",
    "adhesome_loc.columns = ['gene','chrom','start','end','length']\n",
    "adhesome_loc = adhesome_loc.sort_values(by=['chrom','start'])\n",
    "adhesome_loc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide genome into portions corresponding to adhesome genes of interest \n",
    "df_adhesome_pos = adhesome_loc[['chrom','start','end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to bed file\n",
    "bed_adhesome = pybedtools.BedTool.from_dataframe(df_adhesome_pos)\n",
    "bed_adhesome_df = bed_adhesome.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call RNAseq peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get bed file of RNAseq\n",
    "f = 'GSM438363_UCSD.IMR90.mRNA-Seq.mRNA-seq_imr90_r1.bed'\n",
    "bed = pybedtools.BedTool(epigenome_dir + f).sort()\n",
    "# Get counts for this feature in the segmented genome\n",
    "out = pybedtools.bedtool.BedTool.map(bed_adhesome, bed, c = 4, o = 'count_distinct')\n",
    "counts = out.to_dataframe()['name'].values\n",
    "# Add results to tf_loc\n",
    "adhesome_loc['RNAseq'] = counts\n",
    "# Normalize by gene length\n",
    "adhesome_loc['normRNAseq'] = np.log(1+1000000*adhesome_loc['RNAseq']/adhesome_loc['length'])\n",
    "# z-score\n",
    "mean = adhesome_loc['normRNAseq'].mean()\n",
    "std = adhesome_loc['normRNAseq'].std()\n",
    "adhesome_loc['z_RNAseq'] = (adhesome_loc['normRNAseq']-mean)/std\n",
    "adhesome_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call ChIP-seq peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>RNAseq</th>\n",
       "      <th>normRNAseq</th>\n",
       "      <th>H3K4me1</th>\n",
       "      <th>normH3K4me1</th>\n",
       "      <th>z_H3K4me1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>FBLIM1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16085254</td>\n",
       "      <td>16113084</td>\n",
       "      <td>27830</td>\n",
       "      <td>853</td>\n",
       "      <td>10.330433</td>\n",
       "      <td>11</td>\n",
       "      <td>5.982063</td>\n",
       "      <td>1.077013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>ASAP3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23755055</td>\n",
       "      <td>23810750</td>\n",
       "      <td>55695</td>\n",
       "      <td>677</td>\n",
       "      <td>9.405618</td>\n",
       "      <td>12</td>\n",
       "      <td>5.377402</td>\n",
       "      <td>0.228899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>SLC9A1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27425299</td>\n",
       "      <td>27481621</td>\n",
       "      <td>56322</td>\n",
       "      <td>202</td>\n",
       "      <td>8.185217</td>\n",
       "      <td>13</td>\n",
       "      <td>5.445943</td>\n",
       "      <td>0.325036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1618</th>\n",
       "      <td>FABP3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31838099</td>\n",
       "      <td>31845923</td>\n",
       "      <td>7824</td>\n",
       "      <td>42</td>\n",
       "      <td>8.588415</td>\n",
       "      <td>2</td>\n",
       "      <td>5.547611</td>\n",
       "      <td>0.467639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>MACF1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39549838</td>\n",
       "      <td>39952810</td>\n",
       "      <td>402972</td>\n",
       "      <td>10845</td>\n",
       "      <td>10.200385</td>\n",
       "      <td>74</td>\n",
       "      <td>5.218384</td>\n",
       "      <td>0.005856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene chrom     start       end  length  RNAseq  normRNAseq  H3K4me1  \\\n",
       "146   FBLIM1  chr1  16085254  16113084   27830     853   10.330433       11   \n",
       "1548   ASAP3  chr1  23755055  23810750   55695     677    9.405618       12   \n",
       "1588  SLC9A1  chr1  27425299  27481621   56322     202    8.185217       13   \n",
       "1618   FABP3  chr1  31838099  31845923    7824      42    8.588415        2   \n",
       "335    MACF1  chr1  39549838  39952810  402972   10845   10.200385       74   \n",
       "\n",
       "      normH3K4me1  z_H3K4me1  \n",
       "146      5.982063   1.077013  \n",
       "1548     5.377402   0.228899  \n",
       "1588     5.445943   0.325036  \n",
       "1618     5.547611   0.467639  \n",
       "335      5.218384   0.005856  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get bed file of RNAseq\n",
    "f = 'E017-H3K4me1.broadPeak'\n",
    "bed = pybedtools.BedTool(epigenome_dir + f).sort()\n",
    "# Get counts for this feature in the segmented genome\n",
    "out = pybedtools.bedtool.BedTool.map(bed_adhesome, bed, c = 4, o = 'count_distinct')\n",
    "counts = out.to_dataframe()['name'].values\n",
    "# Add results to tf_loc\n",
    "adhesome_loc['H3K4me1'] = counts\n",
    "# Normalize by gene length\n",
    "adhesome_loc['normH3K4me1'] = np.log(1+1000000*adhesome_loc['H3K4me1']/adhesome_loc['length'])\n",
    "# z-score\n",
    "mean = adhesome_loc['normH3K4me1'].mean()\n",
    "std = adhesome_loc['normH3K4me1'].std()\n",
    "adhesome_loc['z_H3K4me1'] = (adhesome_loc['normH3K4me1']-mean)/std\n",
    "adhesome_loc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load processing of all epigenomic features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>FBLIM1</th>\n",
       "      <th>ASAP3</th>\n",
       "      <th>SLC9A1</th>\n",
       "      <th>FABP3</th>\n",
       "      <th>MACF1</th>\n",
       "      <th>PTPRF</th>\n",
       "      <th>TSPAN1</th>\n",
       "      <th>ITGB3BP</th>\n",
       "      <th>NEXN</th>\n",
       "      <th>BCAR3</th>\n",
       "      <th>...</th>\n",
       "      <th>SHARPIN</th>\n",
       "      <th>TESK1</th>\n",
       "      <th>TLN1</th>\n",
       "      <th>OSTF1</th>\n",
       "      <th>SPTLC1</th>\n",
       "      <th>CORO2A</th>\n",
       "      <th>ENG</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>RAPGEF1</th>\n",
       "      <th>VAV2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>z_CEBPB</th>\n",
       "      <td>0.935898</td>\n",
       "      <td>0.489748</td>\n",
       "      <td>0.482657</td>\n",
       "      <td>-1.478095</td>\n",
       "      <td>-0.643699</td>\n",
       "      <td>0.171379</td>\n",
       "      <td>1.552601</td>\n",
       "      <td>0.243686</td>\n",
       "      <td>0.493353</td>\n",
       "      <td>0.017950</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.478095</td>\n",
       "      <td>2.102120</td>\n",
       "      <td>0.786133</td>\n",
       "      <td>0.456317</td>\n",
       "      <td>0.230655</td>\n",
       "      <td>0.330856</td>\n",
       "      <td>-1.478095</td>\n",
       "      <td>-0.200483</td>\n",
       "      <td>-0.045527</td>\n",
       "      <td>-0.357634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z_CHD1</th>\n",
       "      <td>1.498630</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>1.089608</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>0.763480</td>\n",
       "      <td>0.808699</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>0.670238</td>\n",
       "      <td>...</td>\n",
       "      <td>2.452693</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>-0.679852</td>\n",
       "      <td>0.845080</td>\n",
       "      <td>0.612954</td>\n",
       "      <td>-0.679852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z_CTCF</th>\n",
       "      <td>-1.516834</td>\n",
       "      <td>0.377904</td>\n",
       "      <td>0.371077</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>-0.713435</td>\n",
       "      <td>0.071364</td>\n",
       "      <td>1.401273</td>\n",
       "      <td>0.140984</td>\n",
       "      <td>0.381376</td>\n",
       "      <td>-0.076366</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>0.663279</td>\n",
       "      <td>0.345716</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>-1.516834</td>\n",
       "      <td>-0.286684</td>\n",
       "      <td>-0.137485</td>\n",
       "      <td>-0.437998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z_DNase-seq</th>\n",
       "      <td>0.424802</td>\n",
       "      <td>-0.105118</td>\n",
       "      <td>-0.113540</td>\n",
       "      <td>1.417280</td>\n",
       "      <td>-1.451383</td>\n",
       "      <td>-0.483264</td>\n",
       "      <td>1.157297</td>\n",
       "      <td>-0.397380</td>\n",
       "      <td>-0.100836</td>\n",
       "      <td>-0.665501</td>\n",
       "      <td>...</td>\n",
       "      <td>1.680508</td>\n",
       "      <td>1.809995</td>\n",
       "      <td>0.246917</td>\n",
       "      <td>-0.144825</td>\n",
       "      <td>-0.412859</td>\n",
       "      <td>-0.293843</td>\n",
       "      <td>-2.442446</td>\n",
       "      <td>-0.924947</td>\n",
       "      <td>-0.740897</td>\n",
       "      <td>-1.111606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z_EP300</th>\n",
       "      <td>0.073656</td>\n",
       "      <td>1.418606</td>\n",
       "      <td>-0.365324</td>\n",
       "      <td>-0.595259</td>\n",
       "      <td>0.347610</td>\n",
       "      <td>0.409860</td>\n",
       "      <td>0.586623</td>\n",
       "      <td>0.494332</td>\n",
       "      <td>-0.314952</td>\n",
       "      <td>0.317697</td>\n",
       "      <td>...</td>\n",
       "      <td>0.450341</td>\n",
       "      <td>0.923858</td>\n",
       "      <td>1.427625</td>\n",
       "      <td>-0.528200</td>\n",
       "      <td>0.098285</td>\n",
       "      <td>0.378541</td>\n",
       "      <td>-1.489268</td>\n",
       "      <td>0.488919</td>\n",
       "      <td>0.392668</td>\n",
       "      <td>1.191726</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 219 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene           FBLIM1     ASAP3    SLC9A1     FABP3     MACF1     PTPRF  \\\n",
       "z_CEBPB      0.935898  0.489748  0.482657 -1.478095 -0.643699  0.171379   \n",
       "z_CHD1       1.498630 -0.679852  1.089608 -0.679852  0.763480  0.808699   \n",
       "z_CTCF      -1.516834  0.377904  0.371077 -1.516834 -0.713435  0.071364   \n",
       "z_DNase-seq  0.424802 -0.105118 -0.113540  1.417280 -1.451383 -0.483264   \n",
       "z_EP300      0.073656  1.418606 -0.365324 -0.595259  0.347610  0.409860   \n",
       "\n",
       "gene           TSPAN1   ITGB3BP      NEXN     BCAR3    ...      SHARPIN  \\\n",
       "z_CEBPB      1.552601  0.243686  0.493353  0.017950    ...    -1.478095   \n",
       "z_CHD1      -0.679852 -0.679852 -0.679852  0.670238    ...     2.452693   \n",
       "z_CTCF       1.401273  0.140984  0.381376 -0.076366    ...    -1.516834   \n",
       "z_DNase-seq  1.157297 -0.397380 -0.100836 -0.665501    ...     1.680508   \n",
       "z_EP300      0.586623  0.494332 -0.314952  0.317697    ...     0.450341   \n",
       "\n",
       "gene            TESK1      TLN1     OSTF1    SPTLC1    CORO2A       ENG  \\\n",
       "z_CEBPB      2.102120  0.786133  0.456317  0.230655  0.330856 -1.478095   \n",
       "z_CHD1      -0.679852 -0.679852 -0.679852 -0.679852 -0.679852 -0.679852   \n",
       "z_CTCF      -1.516834  0.663279  0.345716 -1.516834 -1.516834 -1.516834   \n",
       "z_DNase-seq  1.809995  0.246917 -0.144825 -0.412859 -0.293843 -2.442446   \n",
       "z_EP300      0.923858  1.427625 -0.528200  0.098285  0.378541 -1.489268   \n",
       "\n",
       "gene             ABL1   RAPGEF1      VAV2  \n",
       "z_CEBPB     -0.200483 -0.045527 -0.357634  \n",
       "z_CHD1       0.845080  0.612954 -0.679852  \n",
       "z_CTCF      -0.286684 -0.137485 -0.437998  \n",
       "z_DNase-seq -0.924947 -0.740897 -1.111606  \n",
       "z_EP300      0.488919  0.392668  1.191726  \n",
       "\n",
       "[5 rows x 219 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load processing of all epigenomic features\n",
    "adhesome_loc = pickle.load(open(saving_dir+'adhesome_with_epigenomics.pkl', 'rb'))\n",
    "adhesome_loc = adhesome_loc[[col for col in adhesome_loc.columns if ('z_' in col) or (col == 'gene')]]\n",
    "adhesome_loc = adhesome_loc.set_index('gene').transpose()\n",
    "adhesome_loc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Spearman correlation between adhesome genes\n",
    "adhesome_loc_corr = adhesome_loc.corr(method='spearman')\n",
    "pickle.dump(adhesome_loc_corr, open(saving_dir+'adhesome_loc_corr.pkl','wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
